// codestral/src/mentions/MentionResolver.js
import path from 'path';
import fastGlob from 'fast-glob';

/**
 * MentionResolver handles @-mention resolution and file autocomplete
 */
class MentionResolver {
  /**
   * @param {FileSystemManager} fileSystemManager - File system manager instance
   * @param {Object} options - Configuration options
   */
  constructor(fileSystemManager, options = {}) {
    this.fsManager = fileSystemManager;
    this.maxFileSize = options.maxFileSize || 1024 * 1024; // 1MB default
    this.warnFileSize = options.warnFileSize || 512 * 1024; // 512KB warning threshold
  }

  /**
   * Resolve @-mentions to actual file paths
   * @param {Array<string>} mentions - Array of mentioned file patterns
   * @returns {Promise<Array<{mention: string, resolvedPaths: Array<string>, error?: string}>>}
   */
  async resolve(mentions) {
    const results = [];

    for (const mention of mentions) {
      try {
        // Check if it's a glob pattern
        const isGlob = mention.includes('*') || mention.includes('?') || mention.includes('[');
        
        if (isGlob) {
          // Handle glob patterns
          const matches = await this._resolveGlob(mention);
          results.push({
            mention,
            resolvedPaths: matches,
            isGlob: true
          });
        } else {
          // Handle single file
          const resolved = await this._resolveSingleFile(mention);
          results.push(resolved);
        }
      } catch (error) {
        results.push({
          mention,
          resolvedPaths: [],
          error: error.message
        });
      }
    }

    return results;
  }

  /**
   * Resolve a single file mention
   * @param {string} mention - File path mention
   * @returns {Promise<Object>} Resolution result
   * @private
   */
  async _resolveSingleFile(mention) {
    // Try exact match first
    const exists = await this.fsManager.exists(mention);
    
    if (exists) {
      return {
        mention,
        resolvedPaths: [mention],
        isGlob: false
      };
    }

    // File doesn't exist, try to find similar files
    const suggestions = await this.getAutocomplete(mention);
    
    return {
      mention,
      resolvedPaths: [],
      error: `File not found: ${mention}`,
      suggestions: suggestions.slice(0, 5) // Top 5 suggestions
    };
  }

  /**
   * Resolve glob pattern to matching files
   * @param {string} pattern - Glob pattern
   * @returns {Promise<Array<string>>} Matching file paths
   * @private
   */
  async _resolveGlob(pattern) {
    await this.fsManager.loadGitignore();
    
    const matches = await fastGlob(pattern, {
      cwd: this.fsManager.rootPath,
      dot: false,
      ignore: ['node_modules/**', '.git/**', ...this.fsManager.ignorePatterns],
      onlyFiles: true
    });

    return matches;
  }

  /**
   * Read contents of mentioned files
   * @param {Array<string>} filePaths - Array of file paths to read
   * @returns {Promise<Array<{path: string, content: string, size: number, error?: string, warning?: string}>>}
   */
  async readMentionedFiles(filePaths) {
    const results = [];

    for (const filePath of filePaths) {
      try {
        // Validate file size first
        const sizeValidation = await this.validateFileSize(filePath);
        
        if (!sizeValidation.allowed) {
          results.push({
            path: filePath,
            content: null,
            size: sizeValidation.size,
            error: sizeValidation.error
          });
          continue;
        }

        // Check if file is binary
        const isBinary = await this.fsManager.isBinary(filePath);
        
        if (isBinary) {
          results.push({
            path: filePath,
            content: null,
            size: sizeValidation.size,
            error: 'File is binary and cannot be read as text'
          });
          continue;
        }

        // Read file content
        const content = await this.fsManager.readFile(filePath);
        
        const result = {
          path: filePath,
          content,
          size: sizeValidation.size
        };

        // Add warning if file is large
        if (sizeValidation.warning) {
          result.warning = sizeValidation.warning;
        }

        results.push(result);
      } catch (error) {
        results.push({
          path: filePath,
          content: null,
          size: 0,
          error: error.message
        });
      }
    }

    return results;
  }

  /**
   * Validate file size against limits
   * @param {string} filePath - Path to file
   * @returns {Promise<{allowed: boolean, size: number, warning?: string, error?: string}>}
   */
  async validateFileSize(filePath) {
    try {
      const stats = await this.fsManager.getStats(filePath);
      const size = stats.size;

      if (size > this.maxFileSize) {
        return {
          allowed: false,
          size,
          error: `File is too large (${this._formatSize(size)}). Maximum allowed size is ${this._formatSize(this.maxFileSize)}`
        };
      }

      if (size > this.warnFileSize) {
        return {
          allowed: true,
          size,
          warning: `File is large (${this._formatSize(size)}). This may impact performance.`
        };
      }

      return {
        allowed: true,
        size
      };
    } catch (error) {
      return {
        allowed: false,
        size: 0,
        error: error.message
      };
    }
  }

  /**
   * Get autocomplete suggestions for partial filename
   * @param {string} partial - Partial filename or path
   * @param {Object} options - Autocomplete options
   * @returns {Promise<Array<{path: string, score: number, isDirectory: boolean}>>}
   */
  async getAutocomplete(partial, options = {}) {
    const maxResults = options.maxResults || 10;
    const includeDirectories = options.includeDirectories !== false;
    
    try {
      await this.fsManager.loadGitignore();
      
      // Get all files in the project
      const allFiles = await fastGlob('**/*', {
        cwd: this.fsManager.rootPath,
        dot: false,
        ignore: ['node_modules/**', '.git/**', ...this.fsManager.ignorePatterns],
        onlyFiles: !includeDirectories,
        markDirectories: true
      });

      // Score and filter files based on fuzzy matching
      const scored = allFiles
        .map(file => ({
          path: file,
          score: this._fuzzyMatch(partial, file),
          isDirectory: file.endsWith('/')
        }))
        .filter(item => item.score > 0)
        .sort((a, b) => {
          // Sort by score (descending), then by path length (ascending)
          if (b.score !== a.score) {
            return b.score - a.score;
          }
          return a.path.length - b.path.length;
        })
        .slice(0, maxResults);

      return scored;
    } catch (error) {
      // If autocomplete fails, return empty array
      return [];
    }
  }

  /**
   * Fuzzy match a partial string against a full string
   * @param {string} partial - Partial string to match
   * @param {string} full - Full string to match against
   * @returns {number} Match score (0 = no match, higher = better match)
   * @private
   */
  _fuzzyMatch(partial, full) {
    if (!partial) return 0;
    
    const partialLower = partial.toLowerCase();
    const fullLower = full.toLowerCase();
    
    // Exact match gets highest score
    if (fullLower === partialLower) {
      return 1000;
    }
    
    // Starts with match gets high score
    if (fullLower.startsWith(partialLower)) {
      return 900;
    }
    
    // Contains match gets medium score
    if (fullLower.includes(partialLower)) {
      return 800;
    }
    
    // Fuzzy match: check if all characters in partial appear in order in full
    let partialIndex = 0;
    let fullIndex = 0;
    let consecutiveMatches = 0;
    let maxConsecutive = 0;
    let matchCount = 0;
    
    while (partialIndex < partialLower.length && fullIndex < fullLower.length) {
      if (partialLower[partialIndex] === fullLower[fullIndex]) {
        partialIndex++;
        matchCount++;
        consecutiveMatches++;
        maxConsecutive = Math.max(maxConsecutive, consecutiveMatches);
      } else {
        consecutiveMatches = 0;
      }
      fullIndex++;
    }
    
    // If not all characters matched, no match
    if (partialIndex < partialLower.length) {
      return 0;
    }
    
    // Score based on match quality
    // Higher score for more consecutive matches and shorter full string
    const matchRatio = matchCount / partialLower.length;
    const lengthPenalty = full.length - partial.length;
    const consecutiveBonus = maxConsecutive * 10;
    
    return Math.max(0, 500 * matchRatio + consecutiveBonus - lengthPenalty);
  }

  /**
   * Format file size in human-readable format
   * @param {number} bytes - Size in bytes
   * @returns {string} Formatted size
   * @private
   */
  _formatSize(bytes) {
    if (bytes < 1024) return `${bytes} B`;
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(2)} KB`;
    return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
  }
}

export default MentionResolver;
