// codestral/tests/__mocks__/FileSystemManager.js
/**
 * Mock FileSystemManager for testing
 */
class FileSystemManager {
  constructor(rootPath = process.cwd(), ignorePatterns = []) {
    this.rootPath = rootPath;
    this.ignorePatterns = ignorePatterns;
    this.files = new Map(); // Mock file system
  }

  async loadGitignore() {
    // Mock implementation
  }

  validatePath(filePath) {
    return filePath;
  }

  async readFile(filePath, encoding = 'utf8') {
    if (this.files.has(filePath)) {
      return this.files.get(filePath);
    }
    throw new Error(`File not found: ${filePath}`);
  }

  async writeFile(filePath, content, options = {}) {
    this.files.set(filePath, content);
  }

  async listDirectory(dirPath = '.', recursive = false, respectIgnore = true) {
    // Mock directory listing
    return [
      { name: 'file1.txt', path: 'file1.txt', isDirectory: false, isFile: true },
      { name: 'file2.js', path: 'file2.js', isDirectory: false, isFile: true },
      { name: 'subdir', path: 'subdir', isDirectory: true, isFile: false }
    ];
  }

  async exists(filePath) {
    return this.files.has(filePath);
  }

  async getStats(filePath) {
    if (!this.files.has(filePath)) {
      throw new Error(`File not found: ${filePath}`);
    }
    return {
      size: this.files.get(filePath).length,
      isDirectory: false,
      isFile: true,
      modified: new Date(),
      created: new Date()
    };
  }

  async isBinary(filePath) {
    // Mock binary detection - return false for text files
    return filePath.endsWith('.jpg') || filePath.endsWith('.png') || filePath.endsWith('.pdf');
  }

  shouldIgnore(filePath, isDirectory = false) {
    // Mock ignore logic
    return filePath.includes('node_modules') || filePath.startsWith('.');
  }

  filterIgnored(files) {
    return files.filter(file => !this.shouldIgnore(file));
  }
}

export default FileSystemManager;
