// codestral/tests/__mocks__/SearchEngine.js
/**
 * Mock SearchEngine for testing
 */
class SearchEngine {
  constructor(fileSystemManager, codebaseIndexer) {
    this.fileSystemManager = fileSystemManager;
    this.codebaseIndexer = codebaseIndexer;
  }

  async search(query, options = {}) {
    // Mock search results
    return [
      {
        filePath: 'src/app.js',
        line: 'function startServer() {',
        lineNumber: 10,
        context: {
          before: ['// Server setup'],
          after: ['  console.log("Server starting");']
        }
      }
    ];
  }

  async searchRegex(pattern, options = {}) {
    // Mock regex search results
    return [
      {
        filePath: 'src/utils.js',
        line: 'function validateEmail(email) {',
        lineNumber: 5,
        context: {
          before: ['// Utility functions'],
          after: ['  return email.includes("@");']
        }
      }
    ];
  }

  filterByFileType(results, extensions) {
    return results.filter(r => extensions.some(ext => r.filePath.endsWith(ext)));
  }

  filterByDirectory(results, directories) {
    return results.filter(r => directories.some(dir => r.filePath.includes(dir)));
  }

  rankResults(results) {
    return results; // Mock ranking - return as is
  }
}

export default SearchEngine;
