// codestral/tests/__mocks__/templates.js
/**
 * Mock templates for testing
 */

export const mockTemplates = {
  'static-website': {
    name: 'Static Website',
    description: 'A basic HTML/CSS/JS website',
    generate: (projectName) => [
      {
        path: 'index.html',
        content: `<!DOCTYPE html><html><head><title>${projectName}</title></head><body><h1>${projectName}</h1></body></html>`
      },
      {
        path: 'css/style.css',
        content: 'body { font-family: Arial, sans-serif; }'
      },
      {
        path: 'js/main.js',
        content: `console.log('${projectName} loaded');`
      },
      {
        path: 'README.md',
        content: `# ${projectName}\n\nA static website.`
      }
    ]
  },
  'react-app': {
    name: 'React App',
    description: 'A React application',
    generate: (projectName) => [
      {
        path: 'package.json',
        content: `{"name": "${projectName.toLowerCase()}", "version": "1.0.0"}`
      },
      {
        path: 'src/App.js',
        content: `function App() { return <h1>${projectName}</h1>; }`
      },
      {
        path: 'README.md',
        content: `# ${projectName}\n\nA React application.`
      }
    ]
  }
};

export function getTemplateNames() {
  return Object.keys(mockTemplates);
}

export function getTemplate(templateName) {
  return mockTemplates[templateName] || null;
}

export function generateTemplate(templateName, projectName, options = {}) {
  const template = getTemplate(templateName);
  if (!template) {
    throw new Error(`Template "${templateName}" not found`);
  }
  return template.generate(projectName, options);
}

export default {
  getTemplateNames,
  getTemplate,
  generateTemplate,
  templates: mockTemplates
};
