// codestral/src/tools/ToolRegistry.js
import logger from '../utils/logger.js';

/**
 * ToolRegistry manages tool definitions and execution for the AI agent
 */
class ToolRegistry {
  /**
   * @param {FileSystemManager} fileSystemManager - File system manager instance
   * @param {SearchEngine} searchEngine - Search engine instance
   * @param {Object} commandExecutor - Command executor instance (optional)
   */
  constructor(fileSystemManager, searchEngine, commandExecutor = null) {
    this.fileSystemManager = fileSystemManager;
    this.searchEngine = searchEngine;
    this.commandExecutor = commandExecutor;
    
    // Map of tool name -> { schema, handler }
    this.tools = new Map();
  }

  /**
   * Register a tool with its schema and handler function
   * @param {string} name - Tool name
   * @param {Object} schema - Tool schema (JSON Schema format)
   * @param {Function} handler - Tool handler function
   * @returns {void}
   */
  registerTool(name, schema, handler) {
    if (!name || typeof name !== 'string') {
      throw new Error('Tool name must be a non-empty string');
    }

    if (!schema || typeof schema !== 'object') {
      throw new Error('Tool schema must be an object');
    }

    if (!handler || typeof handler !== 'function') {
      throw new Error('Tool handler must be a function');
    }

    // Validate schema has required fields
    if (!schema.description) {
      throw new Error(`Tool schema for "${name}" must have a description`);
    }

    if (!schema.parameters || typeof schema.parameters !== 'object') {
      throw new Error(`Tool schema for "${name}" must have parameters object`);
    }

    this.tools.set(name, {
      schema: {
        name,
        description: schema.description,
        parameters: schema.parameters
      },
      handler
    });

    logger.debug(`Registered tool: ${name}`);
  }

  /**
   * Get all tool definitions in the format expected by Mistral AI
   * @returns {Array<Object>} Array of tool schemas
   */
  getToolDefinitions() {
    const definitions = [];
    
    for (const [name, { schema }] of this.tools.entries()) {
      definitions.push({
        type: 'function',
        function: schema
      });
    }
    
    return definitions;
  }

  /**
   * Execute a tool by name with given parameters
   * @param {string} toolName - Name of the tool to execute
   * @param {Object} parameters - Tool parameters
   * @returns {Promise<Object>} Tool execution result
   */
  async executeTool(toolName, parameters = {}) {
    const tool = this.tools.get(toolName);
    
    if (!tool) {
      const error = `Tool not found: ${toolName}`;
      logger.error(error);
      throw new Error(error);
    }

    logger.info(`Executing tool: ${toolName}`, { parameters });

    try {
      // Validate parameters against schema
      this._validateParameters(toolName, parameters, tool.schema.parameters);
      
      // Execute the tool handler
      const result = await tool.handler(parameters);
      
      logger.info(`Tool execution successful: ${toolName}`);
      
      return {
        success: true,
        result
      };
    } catch (error) {
      logger.error(`Tool execution failed: ${toolName}`, { error: error.message });
      
      return {
        success: false,
        error: error.message
      };
    }
  }

  /**
   * Get a tool by name
   * @param {string} name - Tool name
   * @returns {Object|null} Tool object with schema and handler, or null if not found
   */
  getToolByName(name) {
    return this.tools.get(name) || null;
  }

  /**
   * Get list of all registered tool names
   * @returns {Array<string>} Array of tool names
   */
  getToolNames() {
    return Array.from(this.tools.keys());
  }

  /**
   * Check if a tool is registered
   * @param {string} name - Tool name
   * @returns {boolean} True if tool is registered
   */
  hasTool(name) {
    return this.tools.has(name);
  }

  /**
   * Validate parameters against schema
   * @param {string} toolName - Tool name (for error messages)
   * @param {Object} parameters - Parameters to validate
   * @param {Object} schema - Parameter schema
   * @returns {void}
   * @throws {Error} If validation fails
   * @private
   */
  _validateParameters(toolName, parameters, schema) {
    // Check required parameters
    if (schema.required && Array.isArray(schema.required)) {
      for (const requiredParam of schema.required) {
        if (!(requiredParam in parameters)) {
          throw new Error(
            `Missing required parameter "${requiredParam}" for tool "${toolName}"`
          );
        }
      }
    }

    // Basic type checking for provided parameters
    if (schema.properties) {
      for (const [paramName, paramValue] of Object.entries(parameters)) {
        const paramSchema = schema.properties[paramName];
        
        if (!paramSchema) {
          // Parameter not in schema - log warning but don't fail
          logger.warn(
            `Unknown parameter "${paramName}" for tool "${toolName}"`
          );
          continue;
        }

        // Check type if specified
        if (paramSchema.type) {
          const actualType = Array.isArray(paramValue) ? 'array' : typeof paramValue;
          const expectedType = paramSchema.type;
          
          if (actualType !== expectedType && paramValue !== null && paramValue !== undefined) {
            throw new Error(
              `Invalid type for parameter "${paramName}" in tool "${toolName}": ` +
              `expected ${expectedType}, got ${actualType}`
            );
          }
        }
      }
    }
  }
}

export default ToolRegistry;
