// codestral/src/cli/AsciiLogo.js
import chalk from 'chalk';

/**
 * ASCII Art Logo for Codestral
 */
export function displayLogo() {
    const logo = `
 ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó  ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó ‚Ėą‚Ėą‚ēó     
‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēź‚ēź‚ēĚ‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēź‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēź‚ēź‚ēĚ‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēź‚ēź‚ēĚ‚ēö‚ēź‚ēź‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēĚ‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚ēĎ     
‚Ėą‚Ėą‚ēĎ     ‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĎ  ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó  ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó   ‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēĒ‚ēĚ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĎ     
‚Ėą‚Ėą‚ēĎ     ‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĎ  ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚ēĚ  ‚ēö‚ēź‚ēź‚ēź‚ēź‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚ēĒ‚ēź‚ēź‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĎ     
‚ēö‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó‚ēö‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēĒ‚ēĚ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēĒ‚ēĚ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĎ   ‚Ėą‚Ėą‚ēĎ  ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚ēĎ  ‚Ėą‚Ėą‚ēĎ‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚Ėą‚ēó
 ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ   ‚ēö‚ēź‚ēĚ   ‚ēö‚ēź‚ēĚ  ‚ēö‚ēź‚ēĚ‚ēö‚ēź‚ēĚ  ‚ēö‚ēź‚ēĚ‚ēö‚ēź‚ēź‚ēź‚ēź‚ēź‚ēź‚ēĚ
`;

    // Create gradient effect from cyan to blue to magenta to pink
    const lines = logo.split('\n');
    const colors = [
        chalk.hex('#00D9FF'), // Cyan
        chalk.hex('#00B8FF'), // Light Blue
        chalk.hex('#0088FF'), // Blue
        chalk.hex('#6B5FFF'), // Purple
        chalk.hex('#A855F7'), // Violet
        chalk.hex('#E879F9'), // Magenta
        chalk.hex('#F0ABFC'), // Pink
    ];

    lines.forEach((line, index) => {
        if (line.trim()) {
            const colorIndex = Math.min(index, colors.length - 1);
            console.log(colors[colorIndex](line));
        } else {
            console.log(line);
        }
    });
}

/**
 * Display welcome tips
 */
export function displayWelcomeTips() {
    console.log(chalk.gray('\nTips for getting started:'));
    console.log(chalk.gray('1. Ask questions, edit files, or run commands.'));
    console.log(chalk.gray('2. Be specific for the best results.'));
    console.log(chalk.gray('3. Create ') + chalk.white('CODESTRAL.md') + chalk.gray(' files to customize your interactions with Codestral.'));
    console.log(chalk.gray('4. ') + chalk.white('/help') + chalk.gray(' for more information.'));
    console.log();
}

/**
 * Display status bar at bottom
 */
export function displayStatusBar(options = {}) {
    const {
        workingDir = process.cwd().split('/').pop(),
        mode = 'safe',
        model = 'mistral-large-latest',
        contextUsage = 0
    } = options;

    const modeText = mode === 'safe' ? 'safe mode' : mode === 'autonomous' ? 'autonomous mode' : 'full autonomy';
    const contextText = `(${contextUsage}% context used)`;

    const leftPart = chalk.gray(`~/${workingDir}`);
    const middlePart = chalk.gray(`${modeText} (see /docs)`);
    const rightPart = chalk.gray(`${model} ${contextText}`);

    // Calculate spacing to distribute evenly
    const terminalWidth = process.stdout.columns || 100;
    const leftText = `~/${workingDir}`;
    const middleText = `${modeText} (see /docs)`;
    const rightText = `${model} ${contextText}`;

    const totalTextLen = leftText.length + middleText.length + rightText.length;
    const availableSpace = terminalWidth - totalTextLen;
    const spacing1 = Math.max(2, Math.floor(availableSpace / 2));
    const spacing2 = Math.max(2, availableSpace - spacing1);

    console.log(leftPart + ' '.repeat(spacing1) + middlePart + ' '.repeat(spacing2) + rightPart);
}

/**
 * Display input box border (top)
 */
export function displayInputBoxTop() {
    const terminalWidth = process.stdout.columns || 100;
    const border = '‚ĒÄ'.repeat(terminalWidth - 2);
    console.log(chalk.gray(`‚ē≠${border}‚ēģ`));
}

/**
 * Display input box border (bottom)
 */
export function displayInputBoxBottom() {
    const terminalWidth = process.stdout.columns || 100;
    const border = '‚ĒÄ'.repeat(terminalWidth - 2);
    console.log(chalk.gray(`‚ēį${border}‚ēĮ`));
}

/**
 * Get custom prompt with green cursor
 */
export function getPrompt() {
    return chalk.gray('‚Ēā ') + chalk.green('‚Ėą ') + chalk.gray('Type your message or @path/to/file');
}

export default {
    displayLogo,
    displayWelcomeTips,
    displayStatusBar,
    displayInputBoxTop,
    displayInputBoxBottom,
    getPrompt
};
