// codestral/src/cli/CommandHandler.js
import chalk from 'chalk';
import logger from '../utils/logger.js';

/**
 * CommandHandler - Handles special commands in the terminal interface
 * 
 * Responsibilities:
 * - Process special commands (starting with /)
 * - Provide help and configuration commands
 * - Manage context and search operations
 * - Handle autonomy mode changes
 * - Display tool and configuration information
 */
class CommandHandler {
  /**
   * Create a new CommandHandler
   * @param {ConfigManager} configManager - Configuration manager instance
   * @param {ContextManager} contextManager - Context manager instance
   * @param {SearchEngine} searchEngine - Search engine instance
   * @param {ToolRegistry} toolRegistry - Tool registry instance
   * @param {ApprovalManager} approvalManager - Approval manager instance
   * @param {TerminalInterface} terminalInterface - Terminal interface instance
   */
  constructor(configManager, contextManager, searchEngine, toolRegistry, approvalManager, terminalInterface) {
    if (!configManager) {
      throw new Error('ConfigManager is required');
    }
    if (!contextManager) {
      throw new Error('ContextManager is required');
    }
    if (!searchEngine) {
      throw new Error('SearchEngine is required');
    }
    if (!toolRegistry) {
      throw new Error('ToolRegistry is required');
    }
    if (!approvalManager) {
      throw new Error('ApprovalManager is required');
    }
    if (!terminalInterface) {
      throw new Error('TerminalInterface is required');
    }

    this.configManager = configManager;
    this.contextManager = contextManager;
    this.searchEngine = searchEngine;
    this.toolRegistry = toolRegistry;
    this.approvalManager = approvalManager;
    this.terminal = terminalInterface;

    // Define available commands
    this.commands = new Map([
      ['help', this.handleHelp.bind(this)],
      ['clear', this.handleClear.bind(this)],
      ['search', this.handleSearch.bind(this)],
      ['files', this.handleFiles.bind(this)],
      ['config', this.handleConfig.bind(this)],
      ['setup', this.handleSetup.bind(this)],
      ['set', this.handleSet.bind(this)],
      ['mode', this.handleMode.bind(this)],
      ['tools', this.handleTools.bind(this)]
    ]);
  }

  /**
   * Check if a message is a command (starts with /)
   * @param {string} message - Message to check
   * @returns {boolean} True if message is a command
   */
  isCommand(message) {
    return typeof message === 'string' && message.trim().startsWith('/');
  }

  /**
   * Execute a command
   * @param {string} message - Command message
   * @returns {Promise<boolean>} True if command was handled, false otherwise
   */
  async executeCommand(message) {
    if (!this.isCommand(message)) {
      return false;
    }

    const trimmed = message.trim();
    const parts = trimmed.slice(1).split(/\s+/); // Remove '/' and split by whitespace
    const commandName = parts[0].toLowerCase();
    const args = parts.slice(1);

    logger.debug('Executing command', { commandName, args });

    const handler = this.commands.get(commandName);
    if (!handler) {
      this.terminal.displayError(`Unknown command: /${commandName}`);
      this.terminal.displayInfo('Type /help to see available commands');
      return true; // Command was handled (even if unknown)
    }

    try {
      await handler(args);
      return true;
    } catch (error) {
      logger.error('Command execution failed', { commandName, error: error.message });
      this.terminal.displayError(`Command failed: ${error.message}`);
      return true;
    }
  }

  /**
   * Handle /help command - show available commands
   * @param {Array<string>} args - Command arguments
   */
  async handleHelp(args) {
    const specificCommand = args[0];

    if (specificCommand) {
      // Show help for specific command
      this._showCommandHelp(specificCommand);
    } else {
      // Show general help
      this._showGeneralHelp();
    }
  }

  /**
   * Handle /clear command - reset context
   * @param {Array<string>} args - Command arguments
   */
  async handleClear(args) {
    const stats = this.contextManager.getStats();
    this.contextManager.clear();
    
    this.terminal.displaySuccess('Context cleared successfully');
    this.terminal.displayInfo(`Cleared ${stats.messageCount} messages and ${stats.fileCount} file contexts`);
    
    logger.info('Context cleared via command', stats);
  }

  /**
   * Handle /search command - search codebase
   * @param {Array<string>} args - Command arguments
   */
  async handleSearch(args) {
    if (args.length === 0) {
      this.terminal.displayError('Usage: /search <query> [options]');
      this.terminal.displayInfo('Options: --case-sensitive, --whole-word, --context=N');
      return;
    }

    // Parse arguments
    const query = args[0];
    const options = {
      caseSensitive: args.includes('--case-sensitive'),
      wholeWord: args.includes('--whole-word'),
      contextLines: 2
    };

    // Parse context option
    const contextArg = args.find(arg => arg.startsWith('--context='));
    if (contextArg) {
      const contextValue = parseInt(contextArg.split('=')[1]);
      if (!isNaN(contextValue) && contextValue >= 0) {
        options.contextLines = contextValue;
      }
    }

    try {
      this.terminal.displayInfo(`Searching for: "${query}"`);
      
      const results = await this.searchEngine.search(query, options);
      
      if (results.length === 0) {
        this.terminal.displayWarning('No results found');
        return;
      }

      // Rank and limit results
      const rankedResults = this.searchEngine.rankResults(results, query);
      const limitedResults = rankedResults.slice(0, 20); // Show top 20 results

      this.terminal.displaySuccess(`Found ${results.length} results (showing top ${limitedResults.length}):`);
      
      limitedResults.forEach((result, index) => {
        console.log(chalk.cyan(`\n${index + 1}. ${result.filePath}:${result.lineNumber}`));
        console.log(chalk.gray(`   ${result.line.trim()}`));
        
        if (result.context.before.length > 0 || result.context.after.length > 0) {
          console.log(chalk.dim('   Context:'));
          result.context.before.forEach(line => {
            console.log(chalk.dim(`   - ${line.trim()}`));
          });
          result.context.after.forEach(line => {
            console.log(chalk.dim(`   + ${line.trim()}`));
          });
        }
      });

      logger.info('Search completed', { query, resultCount: results.length });

    } catch (error) {
      this.terminal.displayError(`Search failed: ${error.message}`);
      logger.error('Search command failed', { query, error: error.message });
    }
  }

  /**
   * Handle /files command - list files in context
   * @param {Array<string>} args - Command arguments
   */
  async handleFiles(args) {
    const fileContexts = this.contextManager.getAllFileContexts();
    
    if (fileContexts.length === 0) {
      this.terminal.displayInfo('No files currently in context');
      return;
    }

    this.terminal.displaySuccess(`Files in context (${fileContexts.length}):`);
    
    fileContexts.forEach((fileContext, index) => {
      const sizeKB = (fileContext.size / 1024).toFixed(1);
      console.log(chalk.cyan(`${index + 1}. ${fileContext.path}`));
      console.log(chalk.gray(`   Language: ${fileContext.language}, Size: ${sizeKB} KB`));
      console.log(chalk.gray(`   Added: ${fileContext.addedAt.toLocaleString()}`));
    });

    // Show context statistics
    const stats = this.contextManager.getStats();
    console.log(chalk.bold(`\nContext Statistics:`));
    console.log(chalk.gray(`Messages: ${stats.messageCount}`));
    console.log(chalk.gray(`Total tokens: ${stats.totalTokens}/${stats.maxTokens} (${stats.tokenUsagePercent.toFixed(1)}%)`));
    
    if (stats.isNearLimit) {
      console.log(chalk.yellow('⚠️  Context is near token limit'));
    }
  }

  /**
   * Handle /config command - view current configuration
   * @param {Array<string>} args - Command arguments
   */
  async handleConfig(args) {
    this.configManager.displayConfig();
  }

  /**
   * Handle /setup command - run configuration wizard
   * @param {Array<string>} args - Command arguments
   */
  async handleSetup(args) {
    try {
      this.terminal.displayInfo('Starting configuration wizard...');
      
      const config = await this.configManager.runSetupWizard();
      
      this.terminal.displaySuccess('Configuration wizard completed successfully!');
      
      // Update approval manager with new config
      if (config.agent && config.agent.autonomyMode) {
        this.approvalManager.setAutonomyMode(config.agent.autonomyMode);
      }
      
    } catch (error) {
      if (error.message === 'Setup cancelled by user') {
        this.terminal.displayWarning('Configuration wizard cancelled');
      } else {
        this.terminal.displayError(`Setup failed: ${error.message}`);
        logger.error('Setup command failed', { error: error.message });
      }
    }
  }

  /**
   * Handle /set command - update specific config values
   * @param {Array<string>} args - Command arguments
   */
  async handleSet(args) {
    if (args.length < 2) {
      this.terminal.displayError('Usage: /set <key> <value>');
      this.terminal.displayInfo('Examples:');
      this.terminal.displayInfo('  /set apiKey YOUR_API_KEY');
      this.terminal.displayInfo('  /set mistral.model mistral-large-latest');
      this.terminal.displayInfo('  /set agent.autonomyMode autonomous');
      return;
    }

    const key = args[0];
    const value = args.slice(1).join(' '); // Join remaining args as value

    try {
      await this.configManager.updateConfig(key, value);
      
      // Special handling for autonomy mode changes
      if (key === 'agent.autonomyMode' || key === 'autonomyMode') {
        this.approvalManager.setAutonomyMode(value);
      }
      
    } catch (error) {
      this.terminal.displayError(`Failed to update configuration: ${error.message}`);
      logger.error('Set command failed', { key, value, error: error.message });
    }
  }

  /**
   * Handle /mode command - change autonomy mode
   * @param {Array<string>} args - Command arguments
   */
  async handleMode(args) {
    const mode = args[0];
    
    if (!mode) {
      // Display current mode and available modes
      const currentMode = this.approvalManager.getAutonomyMode();
      console.log(chalk.bold.cyan(`\n🤖 Current Autonomy Mode: ${currentMode.toUpperCase()}\n`));
      
      console.log(chalk.bold('Available modes:'));
      console.log(chalk.green('  safe       - Requires approval for file writes and commands'));
      console.log(chalk.yellow('  autonomous - Auto-approve reads and writes, require approval for commands'));
      console.log(chalk.red('  full       - Auto-approve all except dangerous operations'));
      
      console.log(chalk.gray('\nUsage: /mode <safe|autonomous|full>'));
      return;
    }

    try {
      this.approvalManager.setAutonomyMode(mode);
      
      // Update configuration
      await this.configManager.updateConfig('agent.autonomyMode', mode);
      
    } catch (error) {
      this.terminal.displayError(`Failed to change mode: ${error.message}`);
      logger.error('Mode command failed', { mode, error: error.message });
    }
  }

  /**
   * Handle /tools command - list available tools
   * @param {Array<string>} args - Command arguments
   */
  async handleTools(args) {
    const toolNames = this.toolRegistry.getToolNames();
    
    if (toolNames.length === 0) {
      this.terminal.displayWarning('No tools are currently registered');
      return;
    }

    const specificTool = args[0];
    
    if (specificTool) {
      // Show details for specific tool
      this._showToolDetails(specificTool);
    } else {
      // Show all tools
      this._showAllTools(toolNames);
    }
  }

  /**
   * Start the terminal (codestral command implementation)
   * This method would be called by the main CLI entry point
   */
  startTerminal() {
    this.terminal.displayWelcome();
    this.terminal.displayInfo('Type /help for available commands');
    logger.info('Terminal started via codestral command');
  }

  /**
   * Show general help information
   * @private
   */
  _showGeneralHelp() {
    console.log(chalk.bold.cyan('\n🤖 AI Terminal Assistant - Available Commands\n'));
    
    const commandHelp = [
      { cmd: '/help [command]', desc: 'Show this help or help for specific command' },
      { cmd: '/clear', desc: 'Reset conversation context and history' },
      { cmd: '/search <query>', desc: 'Search codebase for text or patterns' },
      { cmd: '/files', desc: 'List files currently in context' },
      { cmd: '/config', desc: 'View current configuration' },
      { cmd: '/setup', desc: 'Run interactive configuration wizard' },
      { cmd: '/set <key> <value>', desc: 'Update specific configuration value' },
      { cmd: '/mode [mode]', desc: 'View or change autonomy mode (safe/autonomous/full)' },
      { cmd: '/tools [tool]', desc: 'List available tools or show tool details' }
    ];

    commandHelp.forEach(({ cmd, desc }) => {
      console.log(chalk.cyan(`${cmd.padEnd(20)} ${chalk.gray(desc)}`));
    });

    console.log(chalk.bold('\nExamples:'));
    console.log(chalk.gray('  /search "function handleError"'));
    console.log(chalk.gray('  /set mistral.model mistral-large-latest'));
    console.log(chalk.gray('  /mode autonomous'));
    console.log(chalk.gray('  /tools write_file'));
    console.log('');
  }

  /**
   * Show help for a specific command
   * @param {string} commandName - Name of the command
   * @private
   */
  _showCommandHelp(commandName) {
    const helpText = {
      help: {
        usage: '/help [command]',
        description: 'Show available commands or detailed help for a specific command',
        examples: ['/help', '/help search', '/help mode']
      },
      clear: {
        usage: '/clear',
        description: 'Reset the conversation context, clearing all messages and file contexts',
        examples: ['/clear']
      },
      search: {
        usage: '/search <query> [options]',
        description: 'Search through the codebase for text patterns',
        options: [
          '--case-sensitive: Make search case sensitive',
          '--whole-word: Match whole words only',
          '--context=N: Show N lines of context around matches'
        ],
        examples: [
          '/search "function handleError"',
          '/search TODO --case-sensitive',
          '/search "class.*Manager" --context=5'
        ]
      },
      files: {
        usage: '/files',
        description: 'List all files currently loaded in the conversation context',
        examples: ['/files']
      },
      config: {
        usage: '/config',
        description: 'Display the current configuration settings',
        examples: ['/config']
      },
      setup: {
        usage: '/setup',
        description: 'Run the interactive configuration wizard to set up the assistant',
        examples: ['/setup']
      },
      set: {
        usage: '/set <key> <value>',
        description: 'Update a specific configuration value',
        examples: [
          '/set apiKey YOUR_API_KEY',
          '/set mistral.model mistral-large-latest',
          '/set agent.autonomyMode autonomous',
          '/set mistral.temperature 0.8'
        ]
      },
      mode: {
        usage: '/mode [mode]',
        description: 'View current autonomy mode or change to a new mode',
        options: [
          'safe: Requires approval for file writes and commands',
          'autonomous: Auto-approve reads and writes, require approval for commands',
          'full: Auto-approve all except dangerous operations'
        ],
        examples: ['/mode', '/mode safe', '/mode autonomous', '/mode full']
      },
      tools: {
        usage: '/tools [tool_name]',
        description: 'List all available tools or show details for a specific tool',
        examples: ['/tools', '/tools write_file', '/tools search_code']
      }
    };

    const help = helpText[commandName];
    if (!help) {
      this.terminal.displayError(`No help available for command: /${commandName}`);
      return;
    }

    console.log(chalk.bold.cyan(`\n📖 Help: /${commandName}\n`));
    console.log(chalk.bold('Usage:'));
    console.log(chalk.cyan(`  ${help.usage}`));
    
    console.log(chalk.bold('\nDescription:'));
    console.log(chalk.gray(`  ${help.description}`));

    if (help.options) {
      console.log(chalk.bold('\nOptions:'));
      help.options.forEach(option => {
        console.log(chalk.yellow(`  ${option}`));
      });
    }

    if (help.examples) {
      console.log(chalk.bold('\nExamples:'));
      help.examples.forEach(example => {
        console.log(chalk.gray(`  ${example}`));
      });
    }

    console.log('');
  }

  /**
   * Show all available tools
   * @param {Array<string>} toolNames - Array of tool names
   * @private
   */
  _showAllTools(toolNames) {
    console.log(chalk.bold.cyan(`\n🔧 Available Tools (${toolNames.length})\n`));

    // Group tools by category for better organization
    const categories = {
      'File Operations': ['read_file', 'write_file', 'replace_in_file', 'create_directory', 'list_directory', 'create_multiple_files'],
      'Search & Analysis': ['search_code', 'analyze_code'],
      'Execution': ['execute_command'],
      'Other': []
    };

    // Categorize tools
    const categorized = { ...categories };
    toolNames.forEach(toolName => {
      let found = false;
      for (const [category, tools] of Object.entries(categories)) {
        if (tools.includes(toolName)) {
          found = true;
          break;
        }
      }
      if (!found) {
        categorized['Other'].push(toolName);
      }
    });

    // Display categorized tools
    Object.entries(categorized).forEach(([category, tools]) => {
      if (tools.length > 0) {
        console.log(chalk.bold(category + ':'));
        tools.forEach(toolName => {
          const tool = this.toolRegistry.getToolByName(toolName);
          const description = tool ? tool.schema.description : 'No description available';
          console.log(chalk.cyan(`  ${toolName.padEnd(20)} ${chalk.gray(description.substring(0, 60))}${description.length > 60 ? '...' : ''}`));
        });
        console.log('');
      }
    });

    console.log(chalk.gray('Use /tools <tool_name> to see detailed information about a specific tool'));
    console.log('');
  }

  /**
   * Show details for a specific tool
   * @param {string} toolName - Name of the tool
   * @private
   */
  _showToolDetails(toolName) {
    const tool = this.toolRegistry.getToolByName(toolName);
    
    if (!tool) {
      this.terminal.displayError(`Tool not found: ${toolName}`);
      this.terminal.displayInfo('Use /tools to see all available tools');
      return;
    }

    console.log(chalk.bold.cyan(`\n🔧 Tool Details: ${toolName}\n`));
    
    console.log(chalk.bold('Description:'));
    console.log(chalk.gray(`  ${tool.schema.description}`));

    if (tool.schema.parameters && tool.schema.parameters.properties) {
      console.log(chalk.bold('\nParameters:'));
      
      const required = tool.schema.parameters.required || [];
      
      Object.entries(tool.schema.parameters.properties).forEach(([paramName, paramSchema]) => {
        const isRequired = required.includes(paramName);
        const requiredText = isRequired ? chalk.red(' (required)') : chalk.gray(' (optional)');
        
        console.log(chalk.cyan(`  ${paramName}${requiredText}`));
        console.log(chalk.gray(`    Type: ${paramSchema.type || 'any'}`));
        
        if (paramSchema.description) {
          console.log(chalk.gray(`    Description: ${paramSchema.description}`));
        }
        
        if (paramSchema.enum) {
          console.log(chalk.gray(`    Allowed values: ${paramSchema.enum.join(', ')}`));
        }
      });
    }

    // Show approval status
    const isAutoApproved = this.approvalManager.isAutoApproved(toolName);
    const approvalStatus = isAutoApproved ? chalk.green('Auto-approved') : chalk.yellow('Requires approval');
    console.log(chalk.bold(`\nApproval Status: ${approvalStatus}`));

    console.log('');
  }
}

export default CommandHandler;
