// codestral/src/cli/SetApiKeyCommand.js
import prompts from 'prompts';
import chalk from 'chalk';
import ApiKeyManager from '../config/ApiKeyManager.js';

/**
 * Command to set or update the hardcoded API key
 */
class SetApiKeyCommand {
  constructor() {
    this.apiKeyManager = new ApiKeyManager();
  }

  /**
   * Execute the set API key command
   * @param {string} apiKey - Optional API key to set directly
   * @returns {Promise<void>}
   */
  async execute(apiKey = null) {
    try {
      console.log(chalk.bold.cyan('\n🔑 Set Mistral AI API Key\n'));

      let keyToSet = apiKey;

      // If no API key provided, prompt for it
      if (!keyToSet) {
        const response = await prompts({
          type: 'password',
          name: 'apiKey',
          message: 'Enter your Mistral AI API key:',
          validate: value => {
            if (!value || value.length === 0) {
              return 'API key cannot be empty';
            }
            if (value.length < 10) {
              return 'API key seems too short. Please check and try again.';
            }
            return true;
          }
        });

        if (!response.apiKey) {
          console.log(chalk.yellow('\n⚠ Operation cancelled'));
          return;
        }

        keyToSet = response.apiKey;
      }

      // Validate the API key format
      if (!keyToSet || keyToSet.length < 10) {
        throw new Error('Invalid API key format');
      }

      // Create/update the API key file
      await this.apiKeyManager.createApiKeyFile(keyToSet);

      console.log(chalk.green('\n✓ API key has been hardcoded successfully!'));
      console.log(chalk.cyan('\n📝 Your API key is now stored in:'));
      console.log(chalk.gray('   src/config/HardcodedApiKey.js'));
      console.log(chalk.yellow('\n⚠ Security Warning:'));
      console.log(chalk.gray('   The API key is now hardcoded in your source code.'));
      console.log(chalk.gray('   Do NOT commit this file to public repositories!'));
      console.log(chalk.gray('   Add it to .gitignore if needed.\n'));

      // Show next steps
      console.log(chalk.cyan('Next steps:'));
      console.log(chalk.gray('   1. Run "codestral" to start using the assistant'));
      console.log(chalk.gray('   2. Your API key will be automatically loaded\n'));

    } catch (error) {
      console.error(chalk.red(`\n✗ Failed to set API key: ${error.message}\n`));
      throw error;
    }
  }

  /**
   * Show current API key status
   * @returns {Promise<void>}
   */
  async showStatus() {
    try {
      const hasValid = await this.apiKeyManager.hasValidApiKey();
      const apiKey = await this.apiKeyManager.readApiKey();

      console.log(chalk.bold.cyan('\n🔑 API Key Status\n'));

      if (hasValid && apiKey) {
        const maskedKey = this.maskApiKey(apiKey);
        console.log(chalk.green('✓ API key is configured'));
        console.log(chalk.gray(`  Key: ${maskedKey}`));
        console.log(chalk.gray(`  File: src/config/HardcodedApiKey.js\n`));
      } else {
        console.log(chalk.yellow('⚠ No valid API key found'));
        console.log(chalk.gray('  Run "codestral set-api-key" to configure\n'));
      }
    } catch (error) {
      console.error(chalk.red(`\n✗ Error checking API key status: ${error.message}\n`));
    }
  }

  /**
   * Mask API key for display (show first 8 and last 4 characters)
   * @param {string} apiKey - The API key to mask
   * @returns {string} Masked API key
   */
  maskApiKey(apiKey) {
    if (!apiKey || apiKey.length < 12) {
      return '***';
    }
    const start = apiKey.substring(0, 8);
    const end = apiKey.substring(apiKey.length - 4);
    const middle = '*'.repeat(Math.min(apiKey.length - 12, 20));
    return `${start}${middle}${end}`;
  }
}

export default SetApiKeyCommand;
