// codestral/src/cli/TerminalInterface.js
import prompts from 'prompts';
import chalk from 'chalk';
import ora from 'ora';
import { highlight } from 'cli-highlight';
import { displayLogo, displayWelcomeTips, displayStatusBar, displayInputBoxTop, displayInputBoxBottom, getPrompt } from './AsciiLogo.js';

/**
 * Enhanced TerminalInterface manages terminal I/O and user interaction
 * Handles display, prompts, user input with command history, progress indicators,
 * color themes, and improved user experience
 */
class TerminalInterface {
  constructor(config = {}) {
    this.config = config;
    this.commandHistory = [];
    this.historyIndex = -1;
    this.spinner = null;
    this.isRunning = false;
    this.syntaxHighlight = config.syntaxHighlight !== false; // Default to true
    this.interruptHandler = null;
    this.onInterruptCallback = null;

    // Enhanced UI options
    this.theme = config.theme || 'default';
    this.showProgress = config.showProgress !== false;
    this.animateOutput = config.animateOutput !== false;
    this.verboseMode = config.verboseMode || false;

    // Progress tracking
    this.currentProgress = null;
    this.progressSteps = [];
    this.completedSteps = 0;

    // Color themes
    this.themes = {
      default: {
        primary: chalk.cyan,
        secondary: chalk.blue,
        success: chalk.green,
        warning: chalk.yellow,
        error: chalk.red,
        info: chalk.blue,
        muted: chalk.gray,
        accent: chalk.magenta,
        highlight: chalk.white.bold
      },
      dark: {
        primary: chalk.white,
        secondary: chalk.gray,
        success: chalk.green.bright,
        warning: chalk.yellow.bright,
        error: chalk.red.bright,
        info: chalk.cyan.bright,
        muted: chalk.gray.dim,
        accent: chalk.magenta.bright,
        highlight: chalk.white.bold
      },
      minimal: {
        primary: chalk.white,
        secondary: chalk.white,
        success: chalk.white,
        warning: chalk.white,
        error: chalk.white,
        info: chalk.white,
        muted: chalk.gray,
        accent: chalk.white,
        highlight: chalk.white.bold
      },
      colorful: {
        primary: chalk.cyan.bold,
        secondary: chalk.blue.bold,
        success: chalk.green.bold,
        warning: chalk.yellow.bold,
        error: chalk.red.bold,
        info: chalk.blue.bold,
        muted: chalk.gray,
        accent: chalk.magenta.bold,
        highlight: chalk.rainbow
      }
    };

    // Get current theme colors
    this.colors = this.themes[this.theme] || this.themes.default;

    // Handle terminal resize events
    if (typeof process !== 'undefined' && process.stdout) {
      process.stdout.on('resize', () => this.handleResize());
    }
  }

  /**
   * Start the terminal interface with Gemini-style welcome
   */
  async start() {
    this.isRunning = true;

    // Clear screen for clean start
    console.clear();

    // Display ASCII art logo
    displayLogo();

    // Display welcome tips
    displayWelcomeTips();

    // Display accepting edits status
    console.log(chalk.gray('                                accepting edits ') + chalk.gray('(shift + tab to toggle)'));
    console.log();
  }

  /**
   * Display quick tips for new users
   */
  displayQuickTips() {
    console.log(this.colors.info('💡 Quick Tips:'));
    console.log(this.colors.muted('   • Ask questions about your code: "Explain this function"'));
    console.log(this.colors.muted('   • Get help with errors: "Fix this bug in @app.js"'));
    console.log(this.colors.muted('   • Search your codebase: /search "function name"'));
    console.log(this.colors.muted('   • Change settings: /config or /setup'));
    console.log(this.colors.muted('   • Switch themes: /set theme dark\n'));
  }

  /**
   * Prompt user for input with Gemini-style prompt
   * @param {string} message - Prompt message to display (optional, uses custom prompt)
   * @returns {Promise<string>} User input
   */
  async prompt(message = null) {
    try {
      // Display status bar before prompt
      const workingDir = process.cwd().split('/').pop();
      const mode = this.config.autonomyMode || 'safe';
      const model = this.config.model || 'mistral-large-latest';
      const contextUsage = 0; // TODO: Calculate actual context usage

      displayStatusBar({ workingDir, mode, model, contextUsage });

      // Display input box top border
      displayInputBoxTop();

      const response = await prompts({
        type: 'text',
        name: 'input',
        message: getPrompt(),
        initial: '',
        onState: (state) => {
          // Handle state changes if needed
          if (state.aborted) {
            process.nextTick(() => process.exit(0));
          }
        }
      });

      // Display input box bottom border
      displayInputBoxBottom();
      console.log();

      // Handle Ctrl+C or empty input
      if (response.input === undefined) {
        return null;
      }

      const input = response.input.trim();

      // Add to command history if not empty and different from last command
      if (input && (this.commandHistory.length === 0 || this.commandHistory[this.commandHistory.length - 1] !== input)) {
        this.commandHistory.push(input);
      }

      // Reset history index
      this.historyIndex = this.commandHistory.length;

      return input;
    } catch (error) {
      // Handle prompts cancellation
      if (error.message === 'User force closed the prompt') {
        return null;
      }
      throw error;
    }
  }

  /**
   * Get previous command from history
   * @returns {string|null} Previous command or null
   */
  getPreviousCommand() {
    if (this.commandHistory.length === 0) {
      return null;
    }

    if (this.historyIndex > 0) {
      this.historyIndex--;
    }

    return this.commandHistory[this.historyIndex];
  }

  /**
   * Get next command from history
   * @returns {string|null} Next command or null
   */
  getNextCommand() {
    if (this.commandHistory.length === 0) {
      return null;
    }

    if (this.historyIndex < this.commandHistory.length - 1) {
      this.historyIndex++;
      return this.commandHistory[this.historyIndex];
    }

    // At the end of history, return empty string
    this.historyIndex = this.commandHistory.length;
    return '';
  }

  /**
   * Get command history
   * @returns {string[]} Array of previous commands
   */
  getHistory() {
    return [...this.commandHistory];
  }

  /**
   * Clear command history
   */
  clearHistory() {
    this.commandHistory = [];
    this.historyIndex = -1;
  }

  /**
   * Display content with enhanced formatting, syntax highlighting, and theme support
   * @param {string} content - Content to display
   * @param {Object} options - Display options
   * @param {string} options.type - Type of content (text, code, error, warning, success, info)
   * @param {string} options.language - Language for syntax highlighting (js, python, etc.)
   * @param {boolean} options.highlight - Whether to apply syntax highlighting
   * @param {boolean} options.animate - Whether to animate the output
   * @param {string} options.prefix - Prefix icon or text
   */
  display(content, options = {}) {
    const {
      type = 'text',
      language = 'javascript',
      highlight: shouldHighlight = false,
      animate = false,
      prefix = null
    } = options;

    let output = content;

    // Apply syntax highlighting for code blocks
    if (shouldHighlight && this.syntaxHighlight && language) {
      try {
        output = highlight(content, {
          language,
          ignoreIllegals: true,
          theme: this.theme === 'dark' ? 'monokai' : 'default'
        });
      } catch (error) {
        // If highlighting fails, just display the content as-is
        output = content;
      }
    }

    // Add prefix if provided
    if (prefix) {
      output = `${prefix} ${output}`;
    }

    // Apply color formatting based on type using theme colors
    switch (type) {
      case 'error':
        console.log(this.colors.error(output));
        break;
      case 'warning':
        console.log(this.colors.warning(output));
        break;
      case 'success':
        console.log(this.colors.success(output));
        break;
      case 'info':
        console.log(this.colors.info(output));
        break;
      case 'code':
        console.log(this.colors.muted(output));
        break;
      case 'primary':
        console.log(this.colors.primary(output));
        break;
      case 'secondary':
        console.log(this.colors.secondary(output));
        break;
      case 'accent':
        console.log(this.colors.accent(output));
        break;
      case 'highlight':
        console.log(this.colors.highlight(output));
        break;
      default:
        console.log(output);
    }

    // Animate output if requested
    if (animate && this.animateOutput) {
      // Simple animation delay for better UX
      return new Promise(resolve => setTimeout(resolve, 50));
    }
  }

  /**
   * Display formatted code block with syntax highlighting
   * @param {string} code - Code to display
   * @param {string} language - Programming language
   */
  displayCode(code, language = 'javascript') {
    console.log(chalk.gray('\n─────────────────────────────────────────────────────'));
    this.display(code, { type: 'code', language, highlight: true });
    console.log(chalk.gray('─────────────────────────────────────────────────────\n'));
  }

  /**
   * Display enhanced error message with better formatting
   * @param {string|Object} message - Error message or error info object
   * @param {Error} error - Optional error object
   */
  displayError(message, error = null) {
    if (typeof message === 'object' && message.title) {
      // Enhanced error object from ErrorHandler
      this.displayEnhancedError(message);
    } else {
      // Simple error message
      console.log(this.colors.error.bold('\n🚨 Error: ') + this.colors.error(message));
      if (error && error.stack && this.config.debug) {
        console.log(this.colors.muted(error.stack));
      }
      console.log();
    }
  }

  /**
   * Display enhanced error with suggestions and help
   * @param {Object} errorInfo - Enhanced error information
   */
  displayEnhancedError(errorInfo) {
    console.log();

    // Error title and message
    console.log(this.colors.error.bold(`🚨 ${errorInfo.title}`));
    console.log(this.colors.error(`💬 ${errorInfo.message}`));

    // Suggestions
    if (errorInfo.suggestions && errorInfo.suggestions.length > 0) {
      console.log(this.colors.info('\n💡 Suggestions:'));
      errorInfo.suggestions.forEach((suggestion, index) => {
        console.log(this.colors.muted(`   ${index + 1}. ${suggestion}`));
      });
    }

    // Contextual help
    if (errorInfo.contextualHelp && errorInfo.contextualHelp.length > 0) {
      console.log(this.colors.accent('\n🔧 Quick Help:'));
      errorInfo.contextualHelp.forEach(help => {
        console.log(this.colors.muted(`   • ${help}`));
      });
    }

    // Recovery information
    if (errorInfo.recoverable && errorInfo.recovery) {
      console.log(this.colors.warning('\n🔄 Recovery: '), end = '');
      if (errorInfo.recovery.action === 'retry') {
        console.log(this.colors.muted(`Will retry automatically in ${errorInfo.recovery.delay}ms`));
      } else if (errorInfo.recovery.action === 'prompt') {
        console.log(this.colors.muted(errorInfo.recovery.message));
      }
    }

    console.log();
  }

  /**
   * Display enhanced success message
   * @param {string} message - Success message
   * @param {Object} options - Display options
   */
  displaySuccess(message, options = {}) {
    const { icon = '✅', animate = false } = options;
    const output = this.colors.success.bold(`\n${icon} `) + this.colors.success(message) + '\n';

    if (animate && this.animateOutput) {
      this.typewriterEffect(output, 30);
    } else {
      console.log(output);
    }
  }

  /**
   * Display enhanced warning message
   * @param {string} message - Warning message
   * @param {Object} options - Display options
   */
  displayWarning(message, options = {}) {
    const { icon = '⚠️', showHelp = false } = options;
    console.log(this.colors.warning.bold(`\n${icon} Warning: `) + this.colors.warning(message));

    if (showHelp && options.help) {
      console.log(this.colors.muted(`💡 Tip: ${options.help}`));
    }
    console.log();
  }

  /**
   * Display enhanced info message
   * @param {string} message - Info message
   * @param {Object} options - Display options
   */
  displayInfo(message, options = {}) {
    const { icon = 'ℹ️', type = 'info' } = options;
    const colorFn = type === 'tip' ? this.colors.accent : this.colors.info;
    console.log(colorFn.bold(`\n${icon} `) + colorFn(message) + '\n');
  }

  /**
   * Display a tip or hint message
   * @param {string} message - Tip message
   */
  displayTip(message) {
    this.displayInfo(message, { icon: '💡', type: 'tip' });
  }

  /**
   * Show enhanced loading spinner with message and options
   * @param {string} message - Loading message
   * @param {Object} options - Spinner options
   */
  showSpinner(message = 'Loading...', options = {}) {
    if (this.spinner) {
      this.spinner.stop();
    }

    const {
      type = 'dots',
      color = 'cyan',
      prefixText = '',
      showElapsed = false
    } = options;

    this.spinner = ora({
      text: message,
      color: color,
      spinner: type,
      prefixText: prefixText
    }).start();

    // Track elapsed time if requested
    if (showElapsed) {
      this.spinnerStartTime = Date.now();
      this.spinnerInterval = setInterval(() => {
        const elapsed = ((Date.now() - this.spinnerStartTime) / 1000).toFixed(1);
        this.spinner.text = `${message} (${elapsed}s)`;
      }, 100);
    }
  }

  /**
   * Hide loading spinner with enhanced feedback
   * @param {string} message - Optional final message
   * @param {boolean} success - Whether operation was successful
   * @param {Object} options - Additional options
   */
  hideSpinner(message = null, success = true, options = {}) {
    if (this.spinner) {
      // Clear elapsed time interval
      if (this.spinnerInterval) {
        clearInterval(this.spinnerInterval);
        this.spinnerInterval = null;
      }

      if (message) {
        const { showDuration = false } = options;
        let finalMessage = message;

        // Add duration if tracking elapsed time
        if (showDuration && this.spinnerStartTime) {
          const duration = ((Date.now() - this.spinnerStartTime) / 1000).toFixed(1);
          finalMessage += ` (${duration}s)`;
        }

        if (success) {
          this.spinner.succeed(finalMessage);
        } else {
          this.spinner.fail(finalMessage);
        }
      } else {
        this.spinner.stop();
      }
      this.spinner = null;
      this.spinnerStartTime = null;
    }
  }

  /**
   * Update spinner text and options
   * @param {string} message - New message
   * @param {Object} options - Update options
   */
  updateSpinner(message, options = {}) {
    if (this.spinner) {
      this.spinner.text = message;

      if (options.color) {
        this.spinner.color = options.color;
      }

      if (options.spinner) {
        this.spinner.spinner = options.spinner;
      }
    }
  }

  /**
   * Show progress bar for multi-step operations
   * @param {Array} steps - Array of step descriptions
   * @param {Object} options - Progress options
   */
  startProgress(steps, options = {}) {
    if (!this.showProgress) return;

    this.progressSteps = steps;
    this.completedSteps = 0;
    this.currentProgress = {
      total: steps.length,
      current: 0,
      startTime: Date.now(),
      ...options
    };

    this.displayProgressHeader();
  }

  /**
   * Update progress to next step
   * @param {string} message - Optional custom message for current step
   */
  nextProgressStep(message = null) {
    if (!this.currentProgress) return;

    this.completedSteps++;
    this.currentProgress.current = this.completedSteps;

    const stepMessage = message || this.progressSteps[this.completedSteps - 1] || 'Processing...';
    this.displayProgressStep(stepMessage);
  }

  /**
   * Complete progress tracking
   * @param {string} message - Final completion message
   */
  completeProgress(message = 'Completed successfully') {
    if (!this.currentProgress) return;

    const duration = ((Date.now() - this.currentProgress.startTime) / 1000).toFixed(1);
    console.log(this.colors.success(`\n✅ ${message} (${duration}s)\n`));

    this.currentProgress = null;
    this.progressSteps = [];
    this.completedSteps = 0;
  }

  /**
   * Display progress header
   * @private
   */
  displayProgressHeader() {
    if (!this.currentProgress) return;

    console.log(this.colors.primary('\n📋 Progress:'));
    console.log(this.colors.muted(`   ${this.currentProgress.total} steps to complete\n`));
  }

  /**
   * Display current progress step
   * @param {string} stepMessage - Message for current step
   * @private
   */
  displayProgressStep(stepMessage) {
    if (!this.currentProgress) return;

    const { current, total } = this.currentProgress;
    const percentage = Math.round((current / total) * 100);
    const progressBar = this.createProgressBar(percentage);

    console.log(this.colors.info(`[${current}/${total}] ${stepMessage}`));
    console.log(this.colors.muted(`${progressBar} ${percentage}%\n`));
  }

  /**
   * Create a visual progress bar
   * @param {number} percentage - Completion percentage
   * @returns {string} Progress bar string
   * @private
   */
  createProgressBar(percentage) {
    const width = 20;
    const filled = Math.round((percentage / 100) * width);
    const empty = width - filled;

    return this.colors.success('█'.repeat(filled)) + this.colors.muted('░'.repeat(empty));
  }

  /**
   * Display tool execution feedback
   * @param {string} toolName - Name of the tool being executed
   * @param {Object} parameters - Tool parameters
   * @param {string} status - Execution status (starting, success, error)
   */
  displayToolExecution(toolName, parameters, status) {
    const icons = {
      starting: '🔧',
      success: '✅',
      error: '❌',
      waiting: '⏳'
    };

    const colors = {
      starting: this.colors.info,
      success: this.colors.success,
      error: this.colors.error,
      waiting: this.colors.warning
    };

    const icon = icons[status] || '🔧';
    const colorFn = colors[status] || this.colors.info;

    let message = `${icon} Tool: ${toolName}`;

    if (status === 'starting') {
      message += ` (${Object.keys(parameters).length} parameters)`;
    }

    console.log(colorFn(message));

    // Show parameters in verbose mode
    if (this.verboseMode && status === 'starting') {
      console.log(this.colors.muted(`   Parameters: ${JSON.stringify(parameters, null, 2)}`));
    }
  }

  /**
   * Typewriter effect for animated text
   * @param {string} text - Text to animate
   * @param {number} speed - Speed in milliseconds per character
   */
  async typewriterEffect(text, speed = 50) {
    for (let i = 0; i < text.length; i++) {
      process.stdout.write(text[i]);
      await new Promise(resolve => setTimeout(resolve, speed));
    }
  }

  /**
   * Set color theme
   * @param {string} themeName - Theme name (default, dark, minimal, colorful)
   */
  setTheme(themeName) {
    if (this.themes[themeName]) {
      this.theme = themeName;
      this.colors = this.themes[themeName];
      this.displayInfo(`Theme changed to: ${themeName}`, { icon: '🎨' });
    } else {
      this.displayWarning(`Unknown theme: ${themeName}. Available themes: ${Object.keys(this.themes).join(', ')}`);
    }
  }

  /**
   * Toggle verbose mode
   */
  toggleVerboseMode() {
    this.verboseMode = !this.verboseMode;
    this.displayInfo(`Verbose mode: ${this.verboseMode ? 'enabled' : 'disabled'}`, { icon: '🔍' });
  }

  /**
   * Display available themes
   */
  displayThemes() {
    console.log(this.colors.primary('\n🎨 Available Themes:'));
    Object.keys(this.themes).forEach(theme => {
      const indicator = theme === this.theme ? '→' : ' ';
      console.log(this.colors.muted(`  ${indicator} ${theme}`));
    });
    console.log();
  }

  /**
   * Handle terminal resize events
   */
  handleResize() {
    // Get new terminal dimensions
    const width = process.stdout.columns;
    const height = process.stdout.rows;

    // Store dimensions for potential use
    this.terminalWidth = width;
    this.terminalHeight = height;

    // Emit resize event if needed for other components
    // This is a placeholder for future functionality
  }

  /**
   * Get terminal dimensions
   * @returns {Object} Terminal width and height
   */
  getTerminalSize() {
    return {
      width: process.stdout.columns || 80,
      height: process.stdout.rows || 24
    };
  }

  /**
   * Handle interrupt signal (Ctrl+C)
   * @param {Function} callback - Optional callback to execute before shutdown
   */
  handleInterrupt(callback = null) {
    // Store callback for later use
    if (callback) {
      this.onInterruptCallback = callback;
    }

    // Remove existing handler if present
    if (this.interruptHandler) {
      process.removeListener('SIGINT', this.interruptHandler);
    }

    // Create new interrupt handler
    this.interruptHandler = async () => {
      console.log(chalk.yellow('\n\n⚠ Interrupt received (Ctrl+C)'));

      // Stop any running spinner
      if (this.spinner) {
        this.spinner.stop();
        this.spinner = null;
      }

      // Execute callback if provided
      if (this.onInterruptCallback) {
        try {
          await this.onInterruptCallback();
        } catch (error) {
          console.error(chalk.red('Error during cleanup:'), error.message);
        }
      }

      // Perform graceful shutdown
      await this.gracefulShutdown();
    };

    // Register the interrupt handler
    process.on('SIGINT', this.interruptHandler);
  }

  /**
   * Perform graceful shutdown
   */
  async gracefulShutdown() {
    console.log(chalk.cyan('\nShutting down gracefully...'));

    // Stop the terminal interface
    this.stop();

    // Display goodbye message
    console.log(chalk.cyan('Goodbye! 👋\n'));

    // Exit the process
    process.exit(0);
  }

  /**
   * Remove interrupt handler
   */
  removeInterruptHandler() {
    if (this.interruptHandler) {
      process.removeListener('SIGINT', this.interruptHandler);
      this.interruptHandler = null;
      this.onInterruptCallback = null;
    }
  }

  /**
   * Stop the terminal interface
   */
  stop() {
    this.isRunning = false;
    if (this.spinner) {
      this.spinner.stop();
    }
    // Clean up interrupt handler
    this.removeInterruptHandler();
  }
}

export default TerminalInterface;
