// codestral/src/modifier/FileModifier.js
import * as diff from 'diff';
import path from 'path';
import fs from 'fs/promises';

/**
 * FileModifier handles code modifications with diff generation, preview, and validation
 */
class FileModifier {
  /**
   * @param {FileSystemManager} fileSystemManager - File system manager instance
   */
  constructor(fileSystemManager) {
    this.fileSystemManager = fileSystemManager;
    this.backupDir = path.join(this.fileSystemManager.rootPath, '.backups');
    this.backups = new Map(); // Track backups: filePath -> { backupPath, timestamp, originalContent }
  }

  /**
   * Generate a diff between original and modified content
   * @param {string} original - Original content
   * @param {string} modified - Modified content
   * @param {Object} options - Diff options
   * @returns {Object} Diff result with patches and statistics
   */
  generateDiff(original, modified, options = {}) {
    const {
      context = 3,
      ignoreWhitespace = false
    } = options;

    // Generate unified diff
    const patches = diff.createPatch(
      'file',
      original,
      modified,
      'original',
      'modified',
      { context, ignoreWhitespace }
    );

    // Generate structured diff for programmatic use
    const structuredDiff = diff.diffLines(original, modified, { ignoreWhitespace });

    // Calculate statistics
    let additions = 0;
    let deletions = 0;
    let changes = 0;

    structuredDiff.forEach(part => {
      if (part.added) {
        additions += part.count || 0;
      } else if (part.removed) {
        deletions += part.count || 0;
      } else {
        changes += part.count || 0;
      }
    });

    return {
      patches,
      structuredDiff,
      statistics: {
        additions,
        deletions,
        unchanged: changes,
        total: additions + deletions + changes
      }
    };
  }

  /**
   * Preview changes for a single file or multiple files
   * @param {Object|Array<Object>} modifications - Single modification or array of modifications
   * @returns {Promise<Object>} Preview result with diffs and statistics
   */
  async previewChanges(modifications) {
    // Normalize to array
    const mods = Array.isArray(modifications) ? modifications : [modifications];
    
    const previews = [];
    let totalAdditions = 0;
    let totalDeletions = 0;

    for (const mod of mods) {
      const { filePath, newContent, options = {} } = mod;

      // Validate modification object
      if (!filePath || newContent === undefined) {
        throw new Error('Modification must include filePath and newContent');
      }

      // Read original content
      let originalContent = '';
      let fileExists = false;

      try {
        originalContent = await this.fileSystemManager.readFile(filePath);
        fileExists = true;
      } catch (error) {
        if (error.message.includes('File not found')) {
          // File doesn't exist, will be created
          fileExists = false;
          originalContent = '';
        } else {
          throw error;
        }
      }

      // Generate diff
      const diffResult = this.generateDiff(originalContent, newContent, options);

      previews.push({
        filePath,
        fileExists,
        diff: diffResult,
        originalSize: originalContent.length,
        newSize: newContent.length
      });

      totalAdditions += diffResult.statistics.additions;
      totalDeletions += diffResult.statistics.deletions;
    }

    return {
      previews,
      summary: {
        filesModified: previews.length,
        totalAdditions,
        totalDeletions,
        filesCreated: previews.filter(p => !p.fileExists).length,
        filesUpdated: previews.filter(p => p.fileExists).length
      }
    };
  }

  /**
   * Validate changes before applying
   * @param {Object|Array<Object>} modifications - Modifications to validate
   * @returns {Promise<Object>} Validation result
   */
  async validateChanges(modifications) {
    const mods = Array.isArray(modifications) ? modifications : [modifications];
    const errors = [];
    const warnings = [];

    for (const mod of mods) {
      const { filePath, newContent } = mod;

      // Check if filePath is provided
      if (!filePath) {
        errors.push({ filePath: 'unknown', error: 'Missing filePath' });
        continue;
      }

      // Check if newContent is provided
      if (newContent === undefined || newContent === null) {
        errors.push({ filePath, error: 'Missing newContent' });
        continue;
      }

      // Validate path
      try {
        this.fileSystemManager.validatePath(filePath);
      } catch (error) {
        errors.push({ filePath, error: error.message });
        continue;
      }

      // Check if file exists and is writable
      try {
        const exists = await this.fileSystemManager.exists(filePath);
        
        if (exists) {
          // Check if it's a directory
          const stats = await this.fileSystemManager.getStats(filePath);
          if (stats.isDirectory) {
            errors.push({ filePath, error: 'Path is a directory, not a file' });
            continue;
          }

          // Check if file is binary
          const isBinary = await this.fileSystemManager.isBinary(filePath);
          if (isBinary) {
            warnings.push({ filePath, warning: 'File appears to be binary' });
          }
        }
      } catch (error) {
        errors.push({ filePath, error: error.message });
      }

      // Check content size
      if (newContent.length > 10 * 1024 * 1024) { // 10MB
        warnings.push({ filePath, warning: 'File size exceeds 10MB' });
      }

      // Check for potentially dangerous patterns
      const dangerousPatterns = [
        /rm\s+-rf\s+\//,
        /sudo\s+/,
        /eval\s*\(/,
        /__import__\s*\(\s*['"]os['"]\s*\)/
      ];

      for (const pattern of dangerousPatterns) {
        if (pattern.test(newContent)) {
          warnings.push({ 
            filePath, 
            warning: `Content contains potentially dangerous pattern: ${pattern.source}` 
          });
        }
      }
    }

    return {
      valid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * Create a backup of a file before modification
   * @param {string} filePath - Path to file to backup
   * @returns {Promise<string>} Path to backup file
   */
  async createBackup(filePath) {
    const exists = await this.fileSystemManager.exists(filePath);
    
    if (!exists) {
      // No backup needed for new files
      return null;
    }

    // Ensure backup directory exists
    await fs.mkdir(this.backupDir, { recursive: true });

    // Read original content
    const originalContent = await this.fileSystemManager.readFile(filePath);

    // Generate backup filename with timestamp
    const timestamp = Date.now();
    const fileName = path.basename(filePath);
    const backupFileName = `${fileName}.${timestamp}.backup`;
    const backupPath = path.join(this.backupDir, backupFileName);

    // Write backup
    await fs.writeFile(backupPath, originalContent, 'utf8');

    // Store backup info
    this.backups.set(filePath, {
      backupPath,
      timestamp,
      originalContent
    });

    return backupPath;
  }

  /**
   * Apply changes to files with optional confirmation
   * @param {Object|Array<Object>} modifications - Modifications to apply
   * @param {boolean} confirmed - Whether changes are already confirmed
   * @param {Function} confirmCallback - Optional callback for confirmation (async function)
   * @returns {Promise<Object>} Result of applying changes
   */
  async applyChanges(modifications, confirmed = false, confirmCallback = null) {
    const mods = Array.isArray(modifications) ? modifications : [modifications];

    // Validate changes first
    const validation = await this.validateChanges(mods);
    if (!validation.valid) {
      throw new Error(`Validation failed: ${validation.errors.map(e => e.error).join(', ')}`);
    }

    // Request confirmation if not already confirmed
    if (!confirmed && confirmCallback) {
      const shouldProceed = await confirmCallback(mods, validation);
      if (!shouldProceed) {
        return {
          success: false,
          message: 'Changes cancelled by user',
          filesModified: 0
        };
      }
    }

    const results = [];
    const backupPaths = [];
    let successCount = 0;

    try {
      // Create backups and apply changes
      for (const mod of mods) {
        const { filePath, newContent } = mod;

        try {
          // Create backup if file exists
          const backupPath = await this.createBackup(filePath);
          if (backupPath) {
            backupPaths.push({ filePath, backupPath });
          }

          // Write new content
          await this.fileSystemManager.writeFile(filePath, newContent);

          results.push({
            filePath,
            success: true,
            backupPath
          });

          successCount++;
        } catch (error) {
          results.push({
            filePath,
            success: false,
            error: error.message
          });

          // Rollback on failure
          throw new Error(`Failed to modify ${filePath}: ${error.message}`);
        }
      }

      return {
        success: true,
        message: `Successfully modified ${successCount} file(s)`,
        filesModified: successCount,
        results,
        backups: backupPaths
      };
    } catch (error) {
      // Rollback all changes on any failure
      await this.rollback(backupPaths.map(b => b.filePath));

      return {
        success: false,
        message: `Failed to apply changes: ${error.message}`,
        filesModified: 0,
        results,
        error: error.message
      };
    }
  }

  /**
   * Rollback changes for specified files
   * @param {string|Array<string>} filePaths - File path(s) to rollback
   * @returns {Promise<Object>} Rollback result
   */
  async rollback(filePaths) {
    const paths = Array.isArray(filePaths) ? filePaths : [filePaths];
    const results = [];
    let successCount = 0;

    for (const filePath of paths) {
      const backup = this.backups.get(filePath);

      if (!backup) {
        results.push({
          filePath,
          success: false,
          error: 'No backup found for this file'
        });
        continue;
      }

      try {
        // Restore from backup
        await this.fileSystemManager.writeFile(filePath, backup.originalContent);

        // Clean up backup file
        try {
          await fs.unlink(backup.backupPath);
        } catch (unlinkError) {
          // Backup file cleanup failed, but restoration succeeded
        }

        // Remove from backups map
        this.backups.delete(filePath);

        results.push({
          filePath,
          success: true,
          restoredFrom: backup.backupPath
        });

        successCount++;
      } catch (error) {
        results.push({
          filePath,
          success: false,
          error: error.message
        });
      }
    }

    return {
      success: successCount === paths.length,
      message: `Rolled back ${successCount} of ${paths.length} file(s)`,
      filesRolledBack: successCount,
      results
    };
  }

  /**
   * Get list of files with available backups
   * @returns {Array<Object>} List of backed up files
   */
  getBackups() {
    return Array.from(this.backups.entries()).map(([filePath, backup]) => ({
      filePath,
      backupPath: backup.backupPath,
      timestamp: backup.timestamp,
      date: new Date(backup.timestamp)
    }));
  }

  /**
   * Clear all backup information (does not delete backup files)
   * @returns {void}
   */
  clearBackups() {
    this.backups.clear();
  }

  /**
   * Clean up old backups, keeping only the specified number
   * @param {number} keepCount - Number of backups to keep
   * @returns {Promise<Object>} Cleanup result
   */
  async cleanupOldBackups(keepCount = 5) {
    const backups = this.getBackups();
    
    if (backups.length <= keepCount) {
      return {
        success: true,
        message: 'No cleanup needed',
        removedCount: 0
      };
    }

    // Sort by timestamp (oldest first)
    backups.sort((a, b) => a.timestamp - b.timestamp);
    
    const toRemove = backups.slice(0, backups.length - keepCount);
    let removedCount = 0;

    for (const backup of toRemove) {
      try {
        await fs.unlink(backup.backupPath);
        this.backups.delete(backup.filePath);
        removedCount++;
      } catch (error) {
        // Continue with other backups even if one fails
      }
    }

    return {
      success: true,
      message: `Cleaned up ${removedCount} old backup(s)`,
      removedCount
    };
  }
}

export default FileModifier;
