// codestral/src/templates/django-app.js
/**
 * Django App Template
 * Creates a Django web application with REST API, authentication, and admin interface
 */

function generate(projectName, options = {}) {
  const files = [];
  const projectSlug = projectName.toLowerCase().replace(/\s+/g, '_');

  // Requirements.txt
  files.push({
    path: 'requirements.txt',
    content: `Django==4.2.7
djangorestframework==3.14.0
django-cors-headers==4.3.1
djangorestframework-simplejwt==5.3.0
python-decouple==3.8
Pillow==10.1.0
psycopg2-binary==2.9.9
gunicorn==21.2.0
whitenoise==6.6.0
pytest-django==4.7.0
pytest==7.4.3`
  });

  // Django settings
  files.push({
    path: `${projectSlug}/settings.py`,
    content: `"""
Django settings for ${projectName} project.
"""

from pathlib import Path
from decouple import config
from datetime import timedelta

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = config('SECRET_KEY', default='django-insecure-change-me-in-production')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = config('DEBUG', default=True, cast=bool)

ALLOWED_HOSTS = config('ALLOWED_HOSTS', default='localhost,127.0.0.1', cast=lambda v: [s.strip() for s in v.split(',')])

# Application definition
DJANGO_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
]

THIRD_PARTY_APPS = [
    'rest_framework',
    'rest_framework_simplejwt',
    'corsheaders',
]

LOCAL_APPS = [
    'accounts',
    'api',
]

INSTALLED_APPS = DJANGO_APPS + THIRD_PARTY_APPS + LOCAL_APPS

MIDDLEWARE = [
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.security.SecurityMiddleware',
    'whitenoise.middleware.WhiteNoiseMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = '${projectSlug}.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [BASE_DIR / 'templates'],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = '${projectSlug}.wsgi.application'

# Database
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': BASE_DIR / 'db.sqlite3',
    }
}

# Password validation
AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# Internationalization
LANGUAGE_CODE = 'en-us'
TIME_ZONE = 'UTC'
USE_I18N = True
USE_TZ = True

# Static files (CSS, JavaScript, Images)
STATIC_URL = '/static/'
STATIC_ROOT = BASE_DIR / 'staticfiles'
STATICFILES_DIRS = [
    BASE_DIR / 'static',
]

# Media files
MEDIA_URL = '/media/'
MEDIA_ROOT = BASE_DIR / 'media'

# Default primary key field type
DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# Custom user model
AUTH_USER_MODEL = 'accounts.User'

# Django REST Framework
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework_simplejwt.authentication.JWTAuthentication',
    ),
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.IsAuthenticated',
    ],
    'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
    'PAGE_SIZE': 20,
    'DEFAULT_RENDERER_CLASSES': [
        'rest_framework.renderers.JSONRenderer',
    ],
}

# JWT Settings
SIMPLE_JWT = {
    'ACCESS_TOKEN_LIFETIME': timedelta(minutes=60),
    'REFRESH_TOKEN_LIFETIME': timedelta(days=7),
    'ROTATE_REFRESH_TOKENS': True,
    'BLACKLIST_AFTER_ROTATION': True,
}

# CORS settings
CORS_ALLOWED_ORIGINS = [
    "http://localhost:3000",
    "http://127.0.0.1:3000",
    "http://localhost:8080",
    "http://127.0.0.1:8080",
]

CORS_ALLOW_CREDENTIALS = True

# Security settings for production
if not DEBUG:
    SECURE_BROWSER_XSS_FILTER = True
    SECURE_CONTENT_TYPE_NOSNIFF = True
    SECURE_HSTS_INCLUDE_SUBDOMAINS = True
    SECURE_HSTS_SECONDS = 31536000
    SECURE_REDIRECT_EXEMPT = []
    SECURE_SSL_REDIRECT = True
    SESSION_COOKIE_SECURE = True
    CSRF_COOKIE_SECURE = True`
  });

  // Django URLs
  files.push({
    path: `${projectSlug}/urls.py`,
    content: `"""
URL configuration for ${projectName} project.
"""
from django.contrib import admin
from django.urls import path, include
from django.conf import settings
from django.conf.urls.static import static
from django.http import JsonResponse

def api_root(request):
    return JsonResponse({
        'message': 'Welcome to ${projectName} API',
        'version': '1.0.0',
        'admin': '/admin/',
        'api': '/api/',
        'docs': '/api/docs/'
    })

urlpatterns = [
    path('admin/', admin.site.urls),
    path('api/', include('api.urls')),
    path('', api_root, name='api_root'),
]

# Serve media files in development
if settings.DEBUG:
    urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
    urlpatterns += static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)`
  });

  // Django WSGI
  files.push({
    path: `${projectSlug}/wsgi.py`,
    content: `"""
WSGI config for ${projectName} project.
"""

import os
from django.core.wsgi import get_wsgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', '${projectSlug}.settings')

application = get_wsgi_application()`
  });

  // Django ASGI
  files.push({
    path: `${projectSlug}/asgi.py`,
    content: `"""
ASGI config for ${projectName} project.
"""

import os
from django.core.asgi import get_asgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', '${projectSlug}.settings')

application = get_asgi_application()`
  });

  // Django init
  files.push({
    path: `${projectSlug}/__init__.py`,
    content: ``
  });

  // Custom User Model
  files.push({
    path: 'accounts/models.py',
    content: `from django.contrib.auth.models import AbstractUser
from django.db import models

class User(AbstractUser):
    email = models.EmailField(unique=True)
    first_name = models.CharField(max_length=30)
    last_name = models.CharField(max_length=30)
    is_active = models.BooleanField(default=True)
    date_joined = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    USERNAME_FIELD = 'email'
    REQUIRED_FIELDS = ['username', 'first_name', 'last_name']

    class Meta:
        db_table = 'auth_user'
        verbose_name = 'User'
        verbose_name_plural = 'Users'

    def __str__(self):
        return self.email

    @property
    def full_name(self):
        return f"{self.first_name} {self.last_name}".strip()

    def get_short_name(self):
        return self.first_name`
  });

  // Accounts admin
  files.push({
    path: 'accounts/admin.py',
    content: `from django.contrib import admin
from django.contrib.auth.admin import UserAdmin as BaseUserAdmin
from .models import User

@admin.register(User)
class UserAdmin(BaseUserAdmin):
    list_display = ('email', 'username', 'first_name', 'last_name', 'is_staff', 'is_active', 'date_joined')
    list_filter = ('is_staff', 'is_superuser', 'is_active', 'date_joined')
    search_fields = ('email', 'username', 'first_name', 'last_name')
    ordering = ('-date_joined',)
    
    fieldsets = (
        (None, {'fields': ('email', 'username', 'password')}),
        ('Personal info', {'fields': ('first_name', 'last_name')}),
        ('Permissions', {'fields': ('is_active', 'is_staff', 'is_superuser', 'groups', 'user_permissions')}),
        ('Important dates', {'fields': ('last_login', 'date_joined')}),
    )
    
    add_fieldsets = (
        (None, {
            'classes': ('wide',),
            'fields': ('email', 'username', 'first_name', 'last_name', 'password1', 'password2'),
        }),
    )`
  });

  // Accounts apps
  files.push({
    path: 'accounts/apps.py',
    content: `from django.apps import AppConfig

class AccountsConfig(AppConfig):
    default_auto_field = 'django.db.models.BigAutoField'
    name = 'accounts'`
  });

  // Accounts init
  files.push({
    path: 'accounts/__init__.py',
    content: ``
  });

  // API URLs
  files.push({
    path: 'api/urls.py',
    content: `from django.urls import path, include
from rest_framework.routers import DefaultRouter
from rest_framework_simplejwt.views import TokenRefreshView
from . import views

router = DefaultRouter()
router.register(r'users', views.UserViewSet)

urlpatterns = [
    # Authentication
    path('auth/register/', views.RegisterView.as_view(), name='register'),
    path('auth/login/', views.LoginView.as_view(), name='login'),
    path('auth/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
    path('auth/me/', views.CurrentUserView.as_view(), name='current_user'),
    
    # API endpoints
    path('', include(router.urls)),
    
    # General endpoints
    path('status/', views.StatusView.as_view(), name='status'),
    path('docs/', views.DocsView.as_view(), name='docs'),
]`
  });

  // API Views
  files.push({
    path: 'api/views.py',
    content: `from rest_framework import generics, viewsets, status, permissions
from rest_framework.decorators import action
from rest_framework.response import Response
from rest_framework.views import APIView
from rest_framework_simplejwt.tokens import RefreshToken
from django.contrib.auth import authenticate
from accounts.models import User
from .serializers import UserSerializer, UserCreateSerializer, LoginSerializer
from .permissions import IsOwnerOrAdmin

class RegisterView(generics.CreateAPIView):
    """Register a new user"""
    queryset = User.objects.all()
    serializer_class = UserCreateSerializer
    permission_classes = [permissions.AllowAny]

    def create(self, request, *args, **kwargs):
        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        user = serializer.save()
        
        # Generate tokens
        refresh = RefreshToken.for_user(user)
        
        return Response({
            'success': True,
            'message': 'User registered successfully',
            'data': {
                'user': UserSerializer(user).data,
                'access_token': str(refresh.access_token),
                'refresh_token': str(refresh),
            }
        }, status=status.HTTP_201_CREATED)

class LoginView(APIView):
    """Login user"""
    permission_classes = [permissions.AllowAny]

    def post(self, request):
        serializer = LoginSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        
        email = serializer.validated_data['email']
        password = serializer.validated_data['password']
        
        user = authenticate(request, username=email, password=password)
        
        if user is None:
            return Response({
                'success': False,
                'message': 'Invalid credentials'
            }, status=status.HTTP_401_UNAUTHORIZED)
        
        if not user.is_active:
            return Response({
                'success': False,
                'message': 'Account is deactivated'
            }, status=status.HTTP_401_UNAUTHORIZED)
        
        # Generate tokens
        refresh = RefreshToken.for_user(user)
        
        return Response({
            'success': True,
            'message': 'Login successful',
            'data': {
                'user': UserSerializer(user).data,
                'access_token': str(refresh.access_token),
                'refresh_token': str(refresh),
            }
        })

class CurrentUserView(generics.RetrieveUpdateAPIView):
    """Get or update current user"""
    serializer_class = UserSerializer
    permission_classes = [permissions.IsAuthenticated]

    def get_object(self):
        return self.request.user

    def retrieve(self, request, *args, **kwargs):
        instance = self.get_object()
        serializer = self.get_serializer(instance)
        return Response({
            'success': True,
            'data': {'user': serializer.data}
        })

    def update(self, request, *args, **kwargs):
        partial = kwargs.pop('partial', False)
        instance = self.get_object()
        serializer = self.get_serializer(instance, data=request.data, partial=partial)
        serializer.is_valid(raise_exception=True)
        self.perform_update(serializer)
        
        return Response({
            'success': True,
            'message': 'User updated successfully',
            'data': {'user': serializer.data}
        })

class UserViewSet(viewsets.ModelViewSet):
    """User management viewset"""
    queryset = User.objects.filter(is_active=True)
    serializer_class = UserSerializer
    permission_classes = [permissions.IsAuthenticated, IsOwnerOrAdmin]

    def list(self, request, *args, **kwargs):
        # Only admins can list all users
        if not request.user.is_staff:
            return Response({
                'success': False,
                'message': 'Admin privileges required'
            }, status=status.HTTP_403_FORBIDDEN)
        
        queryset = self.filter_queryset(self.get_queryset())
        page = self.paginate_queryset(queryset)
        
        if page is not None:
            serializer = self.get_serializer(page, many=True)
            return self.get_paginated_response({
                'success': True,
                'data': {'users': serializer.data}
            })

        serializer = self.get_serializer(queryset, many=True)
        return Response({
            'success': True,
            'data': {'users': serializer.data}
        })

    def retrieve(self, request, *args, **kwargs):
        instance = self.get_object()
        serializer = self.get_serializer(instance)
        return Response({
            'success': True,
            'data': {'user': serializer.data}
        })

    def update(self, request, *args, **kwargs):
        partial = kwargs.pop('partial', False)
        instance = self.get_object()
        serializer = self.get_serializer(instance, data=request.data, partial=partial)
        serializer.is_valid(raise_exception=True)
        self.perform_update(serializer)
        
        return Response({
            'success': True,
            'message': 'User updated successfully',
            'data': {'user': serializer.data}
        })

    def destroy(self, request, *args, **kwargs):
        # Only admins can delete users
        if not request.user.is_staff:
            return Response({
                'success': False,
                'message': 'Admin privileges required'
            }, status=status.HTTP_403_FORBIDDEN)
        
        instance = self.get_object()
        # Soft delete
        instance.is_active = False
        instance.save()
        
        return Response({
            'success': True,
            'message': 'User deleted successfully'
        }, status=status.HTTP_204_NO_CONTENT)

class StatusView(APIView):
    """API status endpoint"""
    permission_classes = [permissions.AllowAny]

    def get(self, request):
        return Response({
            'success': True,
            'message': '${projectName} API is running',
            'version': '1.0.0'
        })

class DocsView(APIView):
    """API documentation"""
    permission_classes = [permissions.AllowAny]

    def get(self, request):
        return Response({
            'success': True,
            'message': '${projectName} API Documentation',
            'endpoints': {
                'auth': {
                    'POST /api/auth/register/': 'Register a new user',
                    'POST /api/auth/login/': 'Login user',
                    'POST /api/auth/refresh/': 'Refresh access token',
                    'GET /api/auth/me/': 'Get current user (requires auth)',
                    'PUT /api/auth/me/': 'Update current user (requires auth)'
                },
                'users': {
                    'GET /api/users/': 'Get all users (admin only)',
                    'GET /api/users/{id}/': 'Get user by ID (requires auth)',
                    'PUT /api/users/{id}/': 'Update user (requires auth)',
                    'DELETE /api/users/{id}/': 'Delete user (admin only)'
                },
                'general': {
                    'GET /api/status/': 'Get API status',
                    'GET /api/docs/': 'API documentation'
                }
            },
            'admin': '/admin/',
            'authentication': 'JWT Bearer Token'
        })`
  });

  // API Serializers
  files.push({
    path: 'api/serializers.py',
    content: `from rest_framework import serializers
from django.contrib.auth.password_validation import validate_password
from accounts.models import User

class UserSerializer(serializers.ModelSerializer):
    full_name = serializers.ReadOnlyField()

    class Meta:
        model = User
        fields = ['id', 'username', 'email', 'first_name', 'last_name', 'full_name', 
                 'is_active', 'is_staff', 'date_joined', 'updated_at']
        read_only_fields = ['id', 'is_staff', 'date_joined', 'updated_at']

class UserCreateSerializer(serializers.ModelSerializer):
    password = serializers.CharField(write_only=True, validators=[validate_password])
    password_confirm = serializers.CharField(write_only=True)

    class Meta:
        model = User
        fields = ['username', 'email', 'first_name', 'last_name', 'password', 'password_confirm']

    def validate(self, attrs):
        if attrs['password'] != attrs['password_confirm']:
            raise serializers.ValidationError("Passwords don't match")
        return attrs

    def create(self, validated_data):
        validated_data.pop('password_confirm')
        password = validated_data.pop('password')
        user = User.objects.create_user(**validated_data)
        user.set_password(password)
        user.save()
        return user

class LoginSerializer(serializers.Serializer):
    email = serializers.EmailField()
    password = serializers.CharField()`
  });

  // API Permissions
  files.push({
    path: 'api/permissions.py',
    content: `from rest_framework import permissions

class IsOwnerOrAdmin(permissions.BasePermission):
    """
    Custom permission to only allow owners of an object or admins to edit it.
    """

    def has_object_permission(self, request, view, obj):
        # Read permissions are allowed to any authenticated user
        if request.method in permissions.SAFE_METHODS:
            return True

        # Write permissions are only allowed to the owner of the object or admin
        return obj == request.user or request.user.is_staff`
  });

  // API Apps
  files.push({
    path: 'api/apps.py',
    content: `from django.apps import AppConfig

class ApiConfig(AppConfig):
    default_auto_field = 'django.db.models.BigAutoField'
    name = 'api'`
  });

  // API init
  files.push({
    path: 'api/__init__.py',
    content: ``
  });

  // Manage.py
  files.push({
    path: 'manage.py',
    content: `#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys

if __name__ == '__main__':
    """Run administrative tasks."""
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', '${projectSlug}.settings')
    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)`
  });

  // Environment file
  files.push({
    path: '.env.example',
    content: `# Django Configuration
SECRET_KEY=django-insecure-change-me-in-production
DEBUG=True
ALLOWED_HOSTS=localhost,127.0.0.1

# Database (PostgreSQL for production)
# DATABASE_URL=postgres://username:password@localhost:5432/dbname

# CORS
CORS_ALLOWED_ORIGINS=http://localhost:3000,http://localhost:8080`
  });

  // Test file
  files.push({
    path: 'tests/test_auth.py',
    content: `import pytest
from django.test import TestCase
from django.urls import reverse
from rest_framework.test import APIClient
from rest_framework import status
from accounts.models import User

class AuthTestCase(TestCase):
    def setUp(self):
        self.client = APIClient()
        self.register_url = reverse('register')
        self.login_url = reverse('login')
        self.me_url = reverse('current_user')

    def test_register_user(self):
        """Test user registration"""
        data = {
            'username': 'testuser',
            'email': 'test@example.com',
            'first_name': 'Test',
            'last_name': 'User',
            'password': 'testpass123',
            'password_confirm': 'testpass123'
        }
        
        response = self.client.post(self.register_url, data)
        
        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
        self.assertTrue(response.data['success'])
        self.assertIn('access_token', response.data['data'])
        
        # Check user was created
        self.assertTrue(User.objects.filter(email='test@example.com').exists())

    def test_register_password_mismatch(self):
        """Test registration with password mismatch"""
        data = {
            'username': 'testuser',
            'email': 'test@example.com',
            'first_name': 'Test',
            'last_name': 'User',
            'password': 'testpass123',
            'password_confirm': 'differentpass'
        }
        
        response = self.client.post(self.register_url, data)
        
        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)

    def test_login_valid_credentials(self):
        """Test login with valid credentials"""
        # Create user
        user = User.objects.create_user(
            username='testuser',
            email='test@example.com',
            first_name='Test',
            last_name='User',
            password='testpass123'
        )
        
        data = {
            'email': 'test@example.com',
            'password': 'testpass123'
        }
        
        response = self.client.post(self.login_url, data)
        
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertTrue(response.data['success'])
        self.assertIn('access_token', response.data['data'])

    def test_login_invalid_credentials(self):
        """Test login with invalid credentials"""
        data = {
            'email': 'nonexistent@example.com',
            'password': 'wrongpassword'
        }
        
        response = self.client.post(self.login_url, data)
        
        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
        self.assertFalse(response.data['success'])

    def test_get_current_user(self):
        """Test getting current user info"""
        # Create and login user
        user = User.objects.create_user(
            username='testuser',
            email='test@example.com',
            first_name='Test',
            last_name='User',
            password='testpass123'
        )
        
        login_data = {
            'email': 'test@example.com',
            'password': 'testpass123'
        }
        
        login_response = self.client.post(self.login_url, login_data)
        token = login_response.data['data']['access_token']
        
        # Get user info
        self.client.credentials(HTTP_AUTHORIZATION=f'Bearer {token}')
        response = self.client.get(self.me_url)
        
        self.assertEqual(response.status_code, status.HTTP_200_OK)
        self.assertTrue(response.data['success'])
        self.assertEqual(response.data['data']['user']['email'], 'test@example.com')

    def test_get_current_user_no_token(self):
        """Test getting current user without token"""
        response = self.client.get(self.me_url)
        
        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)`
  });

  // Pytest config
  files.push({
    path: 'pytest.ini',
    content: `[tool:pytest]
DJANGO_SETTINGS_MODULE = ${projectSlug}.settings
python_files = tests.py test_*.py *_tests.py
addopts = -v --tb=short --strict-markers
markers =
    slow: marks tests as slow (deselect with '-m "not slow"')`
  });

  // .gitignore
  files.push({
    path: '.gitignore',
    content: `# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class

# C extensions
*.so

# Distribution / packaging
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST

# PyInstaller
*.manifest
*.spec

# Installer logs
pip-log.txt
pip-delete-this-directory.txt

# Unit test / coverage reports
htmlcov/
.tox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
.hypothesis/
.pytest_cache/

# Translations
*.mo
*.pot

# Django stuff:
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal

# Flask stuff:
instance/
.webassets-cache

# Scrapy stuff:
.scrapy

# Sphinx documentation
docs/_build/

# PyBuilder
target/

# Jupyter Notebook
.ipynb_checkpoints

# pyenv
.python-version

# celery beat schedule file
celerybeat-schedule

# SageMath parsed files
*.sage.py

# Environments
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/

# Spyder project settings
.spyderproject
.spyproject

# Rope project settings
.ropeproject

# mkdocs documentation
/site

# mypy
.mypy_cache/
.dmypy.json
dmypy.json

# Media files
media/

# Static files
staticfiles/

# IDE
.vscode/
.idea/
*.swp
*.swo`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A Django web application with REST API, JWT authentication, and admin interface.

## Features

- 🐍 Django 4.2 with Django REST Framework
- 🔐 JWT authentication with refresh tokens
- 👤 Custom user model and management
- 🛡️ Security middleware (CORS, CSRF protection)
- 📊 Admin interface for user management
- ✅ Input validation and serialization
- 🧪 Testing with pytest-django
- 📚 API documentation endpoints
- 🚀 Production-ready configuration

## Getting Started

### Prerequisites

- Python 3.8 or higher
- pip (Python package installer)

### Installation

1. Clone the repository and navigate to the project directory

2. Create a virtual environment:
   \`\`\`bash
   python -m venv venv
   source venv/bin/activate  # On Windows: venv\\Scripts\\activate
   \`\`\`

3. Install dependencies:
   \`\`\`bash
   pip install -r requirements.txt
   \`\`\`

4. Set up environment variables:
   \`\`\`bash
   cp .env.example .env
   \`\`\`
   
   Update the \`.env\` file with your configuration.

5. Run database migrations:
   \`\`\`bash
   python manage.py makemigrations
   python manage.py migrate
   \`\`\`

6. Create a superuser (optional):
   \`\`\`bash
   python manage.py createsuperuser
   \`\`\`

7. Run the development server:
   \`\`\`bash
   python manage.py runserver
   \`\`\`

8. The application will be available at:
   - API: \`http://localhost:8000/\`
   - Admin: \`http://localhost:8000/admin/\`
   - API Documentation: \`http://localhost:8000/api/docs/\`

## API Endpoints

### Authentication
- \`POST /api/auth/register/\` - Register a new user
- \`POST /api/auth/login/\` - Login user
- \`POST /api/auth/refresh/\` - Refresh access token
- \`GET /api/auth/me/\` - Get current user (requires auth)
- \`PUT /api/auth/me/\` - Update current user (requires auth)

### Users
- \`GET /api/users/\` - Get all users (admin only)
- \`GET /api/users/{id}/\` - Get user by ID (requires auth)
- \`PUT /api/users/{id}/\` - Update user (requires auth)
- \`DELETE /api/users/{id}/\` - Delete user (admin only)

### General
- \`GET /\` - Welcome message
- \`GET /api/status/\` - API status
- \`GET /api/docs/\` - API documentation

## Authentication

This API uses JWT (JSON Web Tokens) for authentication. Include the token in the Authorization header:

\`\`\`
Authorization: Bearer <your-jwt-token>
\`\`\`

### Token Refresh

Access tokens expire after 1 hour. Use the refresh token to get a new access token:

\`\`\`bash
curl -X POST http://localhost:8000/api/auth/refresh/ \\
  -H "Content-Type: application/json" \\
  -d '{"refresh": "your-refresh-token"}'
\`\`\`

## Project Structure

\`\`\`
├── ${projectSlug}/              # Django project settings
│   ├── __init__.py
│   ├── settings.py         # Django settings
│   ├── urls.py            # Main URL configuration
│   ├── wsgi.py            # WSGI configuration
│   └── asgi.py            # ASGI configuration
├── accounts/              # Custom user app
│   ├── models.py          # User model
│   ├── admin.py           # Admin configuration
│   └── apps.py            # App configuration
├── api/                   # REST API app
│   ├── views.py           # API views
│   ├── serializers.py     # DRF serializers
│   ├── permissions.py     # Custom permissions
│   ├── urls.py            # API URL patterns
│   └── apps.py            # App configuration
├── tests/                 # Test files
│   └── test_auth.py       # Authentication tests
├── manage.py              # Django management script
├── requirements.txt       # Python dependencies
└── .env.example          # Environment variables template
\`\`\`

## Database

This project uses SQLite by default for development. For production, configure PostgreSQL:

1. Install PostgreSQL and create a database
2. Update your \`.env\` file:
   \`\`\`
   DATABASE_URL=postgres://username:password@localhost:5432/dbname
   \`\`\`
3. Install psycopg2: \`pip install psycopg2-binary\`
4. Run migrations: \`python manage.py migrate\`

## Admin Interface

Django provides a powerful admin interface at \`/admin/\`. Create a superuser to access it:

\`\`\`bash
python manage.py createsuperuser
\`\`\`

The admin interface allows you to:
- Manage users and permissions
- View and edit data
- Monitor application activity

## Testing

Run the test suite:

\`\`\`bash
pytest
\`\`\`

Run tests with coverage:

\`\`\`bash
pytest --cov=. --cov-report=html
\`\`\`

Run specific test files:

\`\`\`bash
pytest tests/test_auth.py
\`\`\`

## Environment Variables

Create a \`.env\` file in the root directory:

\`\`\`env
SECRET_KEY=django-insecure-change-me-in-production
DEBUG=True
ALLOWED_HOSTS=localhost,127.0.0.1
CORS_ALLOWED_ORIGINS=http://localhost:3000,http://localhost:8080
\`\`\`

## Deployment

### Using Gunicorn

For production, use Gunicorn as the WSGI server:

\`\`\`bash
gunicorn ${projectSlug}.wsgi:application --bind 0.0.0.0:8000
\`\`\`

### Using Docker

Create a \`Dockerfile\`:

\`\`\`dockerfile
FROM python:3.9-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install -r requirements.txt

COPY . .

RUN python manage.py collectstatic --noinput

EXPOSE 8000

CMD ["gunicorn", "${projectSlug}.wsgi:application", "--bind", "0.0.0.0:8000"]
\`\`\`

### Heroku Deployment

1. Create a \`Procfile\`:
   \`\`\`
   web: gunicorn ${projectSlug}.wsgi:application --port $PORT
   release: python manage.py migrate
   \`\`\`

2. Create \`runtime.txt\`:
   \`\`\`
   python-3.9.18
   \`\`\`

3. Update settings for production:
   \`\`\`python
   import dj_database_url
   DATABASES['default'] = dj_database_url.parse(os.environ.get('DATABASE_URL'))
   \`\`\`

4. Set environment variables in Heroku dashboard
5. Deploy with Git:
   \`\`\`bash
   git push heroku main
   \`\`\`

## Security Features

- **JWT Authentication**: Secure token-based authentication with refresh tokens
- **Password Validation**: Django's built-in password validators
- **CORS Protection**: Configurable cross-origin resource sharing
- **CSRF Protection**: Built-in CSRF protection for forms
- **SQL Injection Protection**: Django ORM prevents SQL injection
- **XSS Protection**: Template system escapes output by default
- **Security Headers**: Configurable security headers for production

## API Examples

### Register User
\`\`\`bash
curl -X POST "http://localhost:8000/api/auth/register/" \\
  -H "Content-Type: application/json" \\
  -d '{
    "username": "johndoe",
    "email": "john@example.com",
    "first_name": "John",
    "last_name": "Doe",
    "password": "securepass123",
    "password_confirm": "securepass123"
  }'
\`\`\`

### Login
\`\`\`bash
curl -X POST "http://localhost:8000/api/auth/login/" \\
  -H "Content-Type: application/json" \\
  -d '{
    "email": "john@example.com",
    "password": "securepass123"
  }'
\`\`\`

### Access Protected Route
\`\`\`bash
curl -X GET "http://localhost:8000/api/auth/me/" \\
  -H "Authorization: Bearer YOUR_JWT_TOKEN"
\`\`\`

## Development

### Django Management Commands

Common Django commands:

\`\`\`bash
# Create new app
python manage.py startapp myapp

# Make migrations
python manage.py makemigrations

# Apply migrations
python manage.py migrate

# Create superuser
python manage.py createsuperuser

# Collect static files
python manage.py collectstatic

# Run development server
python manage.py runserver

# Django shell
python manage.py shell
\`\`\`

### Adding New API Endpoints

1. Define models in \`models.py\`
2. Create serializers in \`serializers.py\`
3. Implement views in \`views.py\`
4. Add URL patterns in \`urls.py\`
5. Write tests for new functionality

### Custom Permissions

The project includes custom permissions in \`api/permissions.py\`. You can create additional permissions as needed:

\`\`\`python
from rest_framework import permissions

class IsOwnerOrReadOnly(permissions.BasePermission):
    def has_object_permission(self, request, view, obj):
        if request.method in permissions.SAFE_METHODS:
            return True
        return obj.owner == request.user
\`\`\`

## Contributing

1. Fork the repository
2. Create a feature branch
3. Make your changes
4. Add tests for new functionality
5. Ensure all tests pass
6. Submit a pull request

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'Django App',
  description: 'A Django web application with REST API, authentication, and admin interface',
  generate
};
