// codestral/src/templates/fastapi-api.js
/**
 * FastAPI Template
 * Creates a modern FastAPI application with async support, automatic docs, and JWT authentication
 */

function generate(projectName, options = {}) {
  const files = [];

  // Requirements.txt
  files.push({
    path: 'requirements.txt',
    content: `fastapi==0.104.1
uvicorn[standard]==0.24.0
python-jose[cryptography]==3.3.0
passlib[bcrypt]==1.7.4
python-multipart==0.0.6
sqlalchemy==2.0.23
alembic==1.12.1
psycopg2-binary==2.9.9
python-dotenv==1.0.0
pydantic==2.5.0
pydantic-settings==2.1.0
pytest==7.4.3
pytest-asyncio==0.21.1
httpx==0.25.2`
  });

  // Main app file
  files.push({
    path: 'main.py',
    content: `from fastapi import FastAPI, HTTPException, Depends, status
from fastapi.middleware.cors import CORSMiddleware
from fastapi.security import HTTPBearer
from contextlib import asynccontextmanager
import uvicorn

from database import engine, Base
from routers import auth, users, api
from config import settings

# Create tables
@asynccontextmanager
async def lifespan(app: FastAPI):
    # Startup
    Base.metadata.create_all(bind=engine)
    yield
    # Shutdown
    pass

app = FastAPI(
    title="${projectName} API",
    description="A modern FastAPI application with authentication and user management",
    version="1.0.0",
    lifespan=lifespan
)

# CORS middleware
app.add_middleware(
    CORSMiddleware,
    allow_origins=settings.CORS_ORIGINS,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Include routers
app.include_router(auth.router, prefix="/api/auth", tags=["Authentication"])
app.include_router(users.router, prefix="/api/users", tags=["Users"])
app.include_router(api.router, prefix="/api", tags=["General"])

@app.get("/")
async def root():
    """Welcome endpoint"""
    return {
        "message": "Welcome to ${projectName} API",
        "version": "1.0.0",
        "docs": "/docs",
        "redoc": "/redoc",
        "health": "/health"
    }

@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "OK",
        "message": "${projectName} API is running",
        "version": "1.0.0"
    }

if __name__ == "__main__":
    uvicorn.run(
        "main:app",
        host="0.0.0.0",
        port=8000,
        reload=True
    )`
  });

  // Configuration
  files.push({
    path: 'config.py',
    content: `from pydantic_settings import BaseSettings
from typing import List

class Settings(BaseSettings):
    # Database
    DATABASE_URL: str = "sqlite:///./app.db"
    
    # JWT
    SECRET_KEY: str = "your-secret-key-change-in-production"
    ALGORITHM: str = "HS256"
    ACCESS_TOKEN_EXPIRE_MINUTES: int = 30
    
    # CORS
    CORS_ORIGINS: List[str] = ["http://localhost:3000", "http://localhost:8080"]
    
    # App
    APP_NAME: str = "${projectName} API"
    DEBUG: bool = True

    class Config:
        env_file = ".env"

settings = Settings()`
  });

  // Database setup
  files.push({
    path: 'database.py',
    content: `from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from config import settings

engine = create_engine(
    settings.DATABASE_URL,
    connect_args={"check_same_thread": False} if "sqlite" in settings.DATABASE_URL else {}
)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()

def get_db():
    """Dependency to get database session"""
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()`
  });

  // Models
  files.push({
    path: 'models.py',
    content: `from sqlalchemy import Column, Integer, String, Boolean, DateTime
from sqlalchemy.sql import func
from database import Base
from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

class User(Base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    email = Column(String(255), unique=True, index=True, nullable=False)
    hashed_password = Column(String(255), nullable=False)
    role = Column(String(20), default="user")
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    def verify_password(self, password: str) -> bool:
        """Verify password against hash"""
        return pwd_context.verify(password, self.hashed_password)

    @staticmethod
    def get_password_hash(password: str) -> str:
        """Generate password hash"""
        return pwd_context.hash(password)

    def set_password(self, password: str):
        """Set password hash"""
        self.hashed_password = self.get_password_hash(password)`
  });

  // Schemas
  files.push({
    path: 'schemas.py',
    content: `from pydantic import BaseModel, EmailStr, Field
from typing import Optional
from datetime import datetime

# User schemas
class UserBase(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    email: EmailStr

class UserCreate(UserBase):
    password: str = Field(..., min_length=6)

class UserUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    email: Optional[EmailStr] = None

class UserResponse(UserBase):
    id: int
    role: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# Auth schemas
class UserLogin(BaseModel):
    email: EmailStr
    password: str

class Token(BaseModel):
    access_token: str
    token_type: str

class TokenData(BaseModel):
    email: Optional[str] = None

# Response schemas
class APIResponse(BaseModel):
    success: bool
    message: str
    data: Optional[dict] = None

class PaginatedResponse(BaseModel):
    success: bool
    data: dict
    pagination: dict`
  });

  // Auth utilities
  files.push({
    path: 'auth.py',
    content: `from datetime import datetime, timedelta
from typing import Optional
from jose import JWTError, jwt
from fastapi import Depends, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from sqlalchemy.orm import Session
from database import get_db
from models import User
from schemas import TokenData
from config import settings

security = HTTPBearer()

def create_access_token(data: dict, expires_delta: Optional[timedelta] = None):
    """Create JWT access token"""
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt

def verify_token(token: str, credentials_exception):
    """Verify JWT token"""
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
        email: str = payload.get("sub")
        if email is None:
            raise credentials_exception
        token_data = TokenData(email=email)
    except JWTError:
        raise credentials_exception
    return token_data

def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_db)
):
    """Get current authenticated user"""
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    
    token_data = verify_token(credentials.credentials, credentials_exception)
    user = db.query(User).filter(User.email == token_data.email).first()
    if user is None:
        raise credentials_exception
    return user

def get_current_active_user(current_user: User = Depends(get_current_user)):
    """Get current active user"""
    if not current_user.is_active:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user

def require_admin(current_user: User = Depends(get_current_active_user)):
    """Require admin role"""
    if current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Admin privileges required"
        )
    return current_user`
  });

  // Auth router
  files.push({
    path: 'routers/auth.py',
    content: `from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from datetime import timedelta
from database import get_db
from models import User
from schemas import UserCreate, UserLogin, UserResponse, Token, APIResponse
from auth import create_access_token, get_current_active_user
from config import settings

router = APIRouter()

@router.post("/register", response_model=APIResponse, status_code=status.HTTP_201_CREATED)
async def register(user_data: UserCreate, db: Session = Depends(get_db)):
    """Register a new user"""
    # Check if user already exists
    db_user = db.query(User).filter(User.email == user_data.email).first()
    if db_user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User already exists with this email"
        )
    
    # Create new user
    db_user = User(
        name=user_data.name,
        email=user_data.email
    )
    db_user.set_password(user_data.password)
    
    db.add(db_user)
    db.commit()
    db.refresh(db_user)
    
    # Create access token
    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": db_user.email}, expires_delta=access_token_expires
    )
    
    return APIResponse(
        success=True,
        message="User registered successfully",
        data={
            "user": UserResponse.from_orm(db_user).dict(),
            "access_token": access_token,
            "token_type": "bearer"
        }
    )

@router.post("/login", response_model=APIResponse)
async def login(user_credentials: UserLogin, db: Session = Depends(get_db)):
    """Login user"""
    # Find user
    user = db.query(User).filter(User.email == user_credentials.email).first()
    
    if not user or not user.verify_password(user_credentials.password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid credentials",
            headers={"WWW-Authenticate": "Bearer"},
        )
    
    if not user.is_active:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Account is deactivated"
        )
    
    # Create access token
    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.email}, expires_delta=access_token_expires
    )
    
    return APIResponse(
        success=True,
        message="Login successful",
        data={
            "user": UserResponse.from_orm(user).dict(),
            "access_token": access_token,
            "token_type": "bearer"
        }
    )

@router.get("/me", response_model=APIResponse)
async def get_current_user_info(current_user: User = Depends(get_current_active_user)):
    """Get current user information"""
    return APIResponse(
        success=True,
        message="User information retrieved successfully",
        data={"user": UserResponse.from_orm(current_user).dict()}
    )`
  });

  // Users router
  files.push({
    path: 'routers/users.py',
    content: `from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List
from database import get_db
from models import User
from schemas import UserResponse, UserUpdate, APIResponse, PaginatedResponse
from auth import get_current_active_user, require_admin

router = APIRouter()

@router.get("", response_model=PaginatedResponse)
async def get_users(
    page: int = Query(1, ge=1),
    per_page: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin)
):
    """Get all users (admin only)"""
    offset = (page - 1) * per_page
    
    users = db.query(User).filter(User.is_active == True).offset(offset).limit(per_page).all()
    total = db.query(User).filter(User.is_active == True).count()
    
    return PaginatedResponse(
        success=True,
        data={
            "users": [UserResponse.from_orm(user).dict() for user in users]
        },
        pagination={
            "page": page,
            "per_page": per_page,
            "total": total,
            "pages": (total + per_page - 1) // per_page
        }
    )

@router.get("/{user_id}", response_model=APIResponse)
async def get_user(
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get user by ID"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Users can only view their own profile unless they're admin
    if current_user.id != user_id and current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )
    
    return APIResponse(
        success=True,
        message="User retrieved successfully",
        data={"user": UserResponse.from_orm(user).dict()}
    )

@router.put("/{user_id}", response_model=APIResponse)
async def update_user(
    user_id: int,
    user_update: UserUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Update user"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Users can only update their own profile unless they're admin
    if current_user.id != user_id and current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied"
        )
    
    # Check if email is already taken by another user
    if user_update.email and user_update.email != user.email:
        existing_user = db.query(User).filter(User.email == user_update.email).first()
        if existing_user:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already in use"
            )
    
    # Update user
    update_data = user_update.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(user, field, value)
    
    db.commit()
    db.refresh(user)
    
    return APIResponse(
        success=True,
        message="User updated successfully",
        data={"user": UserResponse.from_orm(user).dict()}
    )

@router.delete("/{user_id}", response_model=APIResponse)
async def delete_user(
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin)
):
    """Delete user (soft delete, admin only)"""
    user = db.query(User).filter(User.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Soft delete
    user.is_active = False
    db.commit()
    
    return APIResponse(
        success=True,
        message="User deleted successfully",
        data=None
    )`
  });

  // API router
  files.push({
    path: 'routers/api.py',
    content: `from fastapi import APIRouter, Depends
from schemas import APIResponse
from auth import get_current_active_user
from models import User

router = APIRouter()

@router.get("/status", response_model=APIResponse)
async def get_status():
    """Get API status"""
    return APIResponse(
        success=True,
        message="${projectName} API is running",
        data={"version": "1.0.0"}
    )

@router.get("/protected", response_model=APIResponse)
async def protected_route(current_user: User = Depends(get_current_active_user)):
    """Protected route example"""
    return APIResponse(
        success=True,
        message="This is a protected route",
        data={"user_id": current_user.id, "email": current_user.email}
    )

@router.get("/docs-info", response_model=APIResponse)
async def get_docs_info():
    """API documentation information"""
    return APIResponse(
        success=True,
        message="${projectName} API Documentation",
        data={
            "endpoints": {
                "auth": {
                    "POST /api/auth/register": "Register a new user",
                    "POST /api/auth/login": "Login user",
                    "GET /api/auth/me": "Get current user (requires auth)"
                },
                "users": {
                    "GET /api/users": "Get all users (admin only)",
                    "GET /api/users/{id}": "Get user by ID (requires auth)",
                    "PUT /api/users/{id}": "Update user (requires auth)",
                    "DELETE /api/users/{id}": "Delete user (admin only)"
                },
                "general": {
                    "GET /api/status": "Get API status",
                    "GET /api/protected": "Protected route example (requires auth)",
                    "GET /health": "Health check endpoint"
                }
            },
            "documentation": {
                "swagger": "/docs",
                "redoc": "/redoc",
                "openapi": "/openapi.json"
            }
        }
    )`
  });

  // Routers init
  files.push({
    path: 'routers/__init__.py',
    content: ``
  });

  // Environment file
  files.push({
    path: '.env.example',
    content: `# Database
DATABASE_URL=sqlite:///./app.db

# JWT
SECRET_KEY=your-super-secret-key-here
ALGORITHM=HS256
ACCESS_TOKEN_EXPIRE_MINUTES=30

# CORS
CORS_ORIGINS=["http://localhost:3000", "http://localhost:8080"]

# App
APP_NAME=${projectName} API
DEBUG=true`
  });

  // Test file
  files.push({
    path: 'tests/test_auth.py',
    content: `import pytest
from fastapi.testclient import TestClient
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from database import Base, get_db
from main import app
from models import User

# Test database
SQLALCHEMY_DATABASE_URL = "sqlite:///./test.db"
engine = create_engine(SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False})
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def override_get_db():
    try:
        db = TestingSessionLocal()
        yield db
    finally:
        db.close()

app.dependency_overrides[get_db] = override_get_db

@pytest.fixture
def client():
    Base.metadata.create_all(bind=engine)
    with TestClient(app) as c:
        yield c
    Base.metadata.drop_all(bind=engine)

@pytest.fixture
def sample_user(client):
    response = client.post(
        "/api/auth/register",
        json={
            "name": "Test User",
            "email": "test@example.com",
            "password": "password123"
        }
    )
    return response.json()

def test_register_user(client):
    """Test user registration"""
    response = client.post(
        "/api/auth/register",
        json={
            "name": "New User",
            "email": "new@example.com",
            "password": "password123"
        }
    )
    
    assert response.status_code == 201
    data = response.json()
    assert data["success"] is True
    assert "access_token" in data["data"]

def test_register_duplicate_email(client, sample_user):
    """Test registration with duplicate email"""
    response = client.post(
        "/api/auth/register",
        json={
            "name": "Another User",
            "email": "test@example.com",
            "password": "password123"
        }
    )
    
    assert response.status_code == 400
    data = response.json()
    assert "already exists" in data["detail"]

def test_login_valid_credentials(client, sample_user):
    """Test login with valid credentials"""
    response = client.post(
        "/api/auth/login",
        json={
            "email": "test@example.com",
            "password": "password123"
        }
    )
    
    assert response.status_code == 200
    data = response.json()
    assert data["success"] is True
    assert "access_token" in data["data"]

def test_login_invalid_credentials(client):
    """Test login with invalid credentials"""
    response = client.post(
        "/api/auth/login",
        json={
            "email": "nonexistent@example.com",
            "password": "wrongpassword"
        }
    )
    
    assert response.status_code == 401

def test_get_current_user(client, sample_user):
    """Test getting current user info"""
    token = sample_user["data"]["access_token"]
    
    response = client.get(
        "/api/auth/me",
        headers={"Authorization": f"Bearer {token}"}
    )
    
    assert response.status_code == 200
    data = response.json()
    assert data["success"] is True
    assert data["data"]["user"]["email"] == "test@example.com"

def test_protected_route(client, sample_user):
    """Test protected route access"""
    token = sample_user["data"]["access_token"]
    
    response = client.get(
        "/api/protected",
        headers={"Authorization": f"Bearer {token}"}
    )
    
    assert response.status_code == 200
    data = response.json()
    assert data["success"] is True

def test_protected_route_no_token(client):
    """Test protected route without token"""
    response = client.get("/api/protected")
    
    assert response.status_code == 403`
  });

  // Pytest config
  files.push({
    path: 'pytest.ini',
    content: `[tool:pytest]
testpaths = tests
python_files = test_*.py
python_classes = Test*
python_functions = test_*
addopts = -v --tb=short
asyncio_mode = auto`
  });

  // .gitignore
  files.push({
    path: '.gitignore',
    content: `# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class

# C extensions
*.so

# Distribution / packaging
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST

# PyInstaller
*.manifest
*.spec

# Installer logs
pip-log.txt
pip-delete-this-directory.txt

# Unit test / coverage reports
htmlcov/
.tox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
.hypothesis/
.pytest_cache/

# Translations
*.mo
*.pot

# Django stuff:
*.log
local_settings.py
db.sqlite3

# Flask stuff:
instance/
.webassets-cache

# Scrapy stuff:
.scrapy

# Sphinx documentation
docs/_build/

# PyBuilder
target/

# Jupyter Notebook
.ipynb_checkpoints

# pyenv
.python-version

# celery beat schedule file
celerybeat-schedule

# SageMath parsed files
*.sage.py

# Environments
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/

# Spyder project settings
.spyderproject
.spyproject

# Rope project settings
.ropeproject

# mkdocs documentation
/site

# mypy
.mypy_cache/
.dmypy.json
dmypy.json

# Database
*.db
*.sqlite3
test.db

# IDE
.vscode/
.idea/
*.swp
*.swo`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A modern FastAPI application with automatic API documentation, async support, and JWT authentication.

## Features

- ⚡ FastAPI with async/await support
- 🔐 JWT authentication and authorization
- 👤 User management with roles
- 📊 SQLAlchemy ORM with Pydantic models
- 🛡️ Security middleware (CORS, automatic validation)
- 📚 Automatic interactive API documentation (Swagger UI & ReDoc)
- ✅ Type hints and automatic validation
- 🧪 Testing with pytest and httpx
- 🚀 High performance with async support

## Getting Started

### Prerequisites

- Python 3.8 or higher
- pip (Python package installer)

### Installation

1. Clone the repository and navigate to the project directory

2. Create a virtual environment:
   \`\`\`bash
   python -m venv venv
   source venv/bin/activate  # On Windows: venv\\Scripts\\activate
   \`\`\`

3. Install dependencies:
   \`\`\`bash
   pip install -r requirements.txt
   \`\`\`

4. Set up environment variables:
   \`\`\`bash
   cp .env.example .env
   \`\`\`
   
   Update the \`.env\` file with your configuration.

5. Run the development server:
   \`\`\`bash
   uvicorn main:app --reload
   \`\`\`

6. The API will be available at:
   - API: \`http://localhost:8000\`
   - Interactive docs (Swagger UI): \`http://localhost:8000/docs\`
   - Alternative docs (ReDoc): \`http://localhost:8000/redoc\`

## API Documentation

FastAPI automatically generates interactive API documentation:

- **Swagger UI**: [http://localhost:8000/docs](http://localhost:8000/docs)
- **ReDoc**: [http://localhost:8000/redoc](http://localhost:8000/redoc)
- **OpenAPI JSON**: [http://localhost:8000/openapi.json](http://localhost:8000/openapi.json)

## API Endpoints

### Authentication
- \`POST /api/auth/register\` - Register a new user
- \`POST /api/auth/login\` - Login user
- \`GET /api/auth/me\` - Get current user (requires auth)

### Users
- \`GET /api/users\` - Get all users (admin only)
- \`GET /api/users/{id}\` - Get user by ID (requires auth)
- \`PUT /api/users/{id}\` - Update user (requires auth)
- \`DELETE /api/users/{id}\` - Delete user (admin only)

### General
- \`GET /\` - Welcome message
- \`GET /health\` - Health check
- \`GET /api/status\` - API status
- \`GET /api/docs-info\` - API documentation info
- \`GET /api/protected\` - Protected route example

## Authentication

This API uses JWT (JSON Web Tokens) for authentication. Include the token in the Authorization header:

\`\`\`
Authorization: Bearer <your-jwt-token>
\`\`\`

You can also use the "Authorize" button in the Swagger UI to authenticate.

## Project Structure

\`\`\`
├── main.py                 # Main FastAPI application
├── config.py              # Configuration settings
├── database.py            # Database setup and connection
├── models.py              # SQLAlchemy models
├── schemas.py             # Pydantic schemas
├── auth.py                # Authentication utilities
├── routers/
│   ├── __init__.py
│   ├── auth.py           # Authentication routes
│   ├── users.py          # User management routes
│   └── api.py            # General API routes
├── tests/
│   └── test_auth.py      # Authentication tests
├── requirements.txt      # Python dependencies
└── .env.example         # Environment variables template
\`\`\`

## Database

This project uses SQLAlchemy with SQLite by default. To use PostgreSQL:

1. Install PostgreSQL and create a database
2. Update \`DATABASE_URL\` in your \`.env\` file:
   \`\`\`
   DATABASE_URL=postgresql://username:password@localhost/dbname
   \`\`\`

For production, consider using Alembic for database migrations:

\`\`\`bash
# Initialize Alembic
alembic init alembic

# Create a migration
alembic revision --autogenerate -m "Initial migration"

# Apply migrations
alembic upgrade head
\`\`\`

## Testing

Run the test suite:

\`\`\`bash
pytest
\`\`\`

Run tests with coverage:

\`\`\`bash
pytest --cov=. --cov-report=html
\`\`\`

Run tests in watch mode:

\`\`\`bash
pytest-watch
\`\`\`

## Environment Variables

Create a \`.env\` file in the root directory:

\`\`\`env
DATABASE_URL=sqlite:///./app.db
SECRET_KEY=your-super-secret-key-here
ALGORITHM=HS256
ACCESS_TOKEN_EXPIRE_MINUTES=30
CORS_ORIGINS=["http://localhost:3000", "http://localhost:8080"]
APP_NAME=${projectName} API
DEBUG=true
\`\`\`

## Deployment

### Using Uvicorn

For production, use Uvicorn with multiple workers:

\`\`\`bash
uvicorn main:app --host 0.0.0.0 --port 8000 --workers 4
\`\`\`

### Using Docker

Create a \`Dockerfile\`:

\`\`\`dockerfile
FROM python:3.9-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install -r requirements.txt

COPY . .

EXPOSE 8000

CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000"]
\`\`\`

Build and run:

\`\`\`bash
docker build -t ${projectName.toLowerCase().replace(/\s+/g, '-')}-api .
docker run -p 8000:8000 ${projectName.toLowerCase().replace(/\s+/g, '-')}-api
\`\`\`

### Heroku Deployment

1. Create a \`Procfile\`:
   \`\`\`
   web: uvicorn main:app --host 0.0.0.0 --port $PORT
   \`\`\`

2. Create \`runtime.txt\`:
   \`\`\`
   python-3.9.18
   \`\`\`

3. Set environment variables in Heroku dashboard
4. Deploy with Git:
   \`\`\`bash
   git push heroku main
   \`\`\`

## Performance Features

FastAPI provides excellent performance out of the box:

- **Async Support**: Full async/await support for I/O operations
- **Automatic Validation**: Pydantic models provide fast validation
- **Type Hints**: Full type hint support with IDE integration
- **Dependency Injection**: Efficient dependency injection system
- **Background Tasks**: Built-in support for background tasks

## Security Features

- **JWT Authentication**: Secure token-based authentication
- **Password Hashing**: Bcrypt for secure password storage
- **CORS Protection**: Configurable cross-origin resource sharing
- **Automatic Validation**: Pydantic models prevent invalid data
- **SQL Injection Protection**: SQLAlchemy ORM prevents SQL injection
- **Type Safety**: Type hints catch errors at development time

## API Examples

### Register User
\`\`\`bash
curl -X POST "http://localhost:8000/api/auth/register" \\
  -H "Content-Type: application/json" \\
  -d '{"name": "John Doe", "email": "john@example.com", "password": "password123"}'
\`\`\`

### Login
\`\`\`bash
curl -X POST "http://localhost:8000/api/auth/login" \\
  -H "Content-Type: application/json" \\
  -d '{"email": "john@example.com", "password": "password123"}'
\`\`\`

### Access Protected Route
\`\`\`bash
curl -X GET "http://localhost:8000/api/auth/me" \\
  -H "Authorization: Bearer YOUR_JWT_TOKEN"
\`\`\`

## Development

### Code Quality

This project follows Python best practices:

- Type hints for better IDE support and error catching
- Pydantic models for data validation
- Async/await for better performance
- Dependency injection for testability

### Adding New Endpoints

1. Define Pydantic schemas in \`schemas.py\`
2. Create route handlers in appropriate router files
3. Add authentication/authorization as needed
4. Write tests for new functionality

### Database Models

When adding new models:

1. Define the SQLAlchemy model in \`models.py\`
2. Create corresponding Pydantic schemas
3. Add any necessary relationships
4. Create and run migrations if using Alembic

## Contributing

1. Fork the repository
2. Create a feature branch
3. Make your changes
4. Add tests for new functionality
5. Ensure all tests pass
6. Submit a pull request

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'FastAPI',
  description: 'A modern FastAPI application with async support, automatic docs, and JWT authentication',
  generate
};
