// codestral/src/templates/flask-api.js
/**
 * Flask API Template
 * Creates a RESTful API server with Flask, SQLAlchemy, and JWT authentication
 */

function generate(projectName, options = {}) {
  const files = [];

  // Requirements.txt
  files.push({
    path: 'requirements.txt',
    content: `Flask==2.3.3
Flask-SQLAlchemy==3.0.5
Flask-Migrate==4.0.5
Flask-JWT-Extended==4.5.2
Flask-CORS==4.0.0
Flask-Limiter==3.5.0
python-dotenv==1.0.0
bcrypt==4.0.1
marshmallow==3.20.1
gunicorn==21.2.0
pytest==7.4.2
pytest-flask==1.2.0`
  });

  // Main app file
  files.push({
    path: 'app.py',
    content: `from flask import Flask
from flask_cors import CORS
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address
from config import Config
from extensions import db, migrate, jwt
from routes import auth_bp, users_bp, api_bp
import os

def create_app(config_class=Config):
    app = Flask(__name__)
    app.config.from_object(config_class)
    
    # Initialize extensions
    db.init_app(app)
    migrate.init_app(app, db)
    jwt.init_app(app)
    
    # CORS configuration
    CORS(app, origins=app.config['CORS_ORIGINS'])
    
    # Rate limiting
    limiter = Limiter(
        app,
        key_func=get_remote_address,
        default_limits=["100 per hour"]
    )
    
    # Register blueprints
    app.register_blueprint(auth_bp, url_prefix='/api/auth')
    app.register_blueprint(users_bp, url_prefix='/api/users')
    app.register_blueprint(api_bp, url_prefix='/api')
    
    # Health check endpoint
    @app.route('/health')
    def health_check():
        return {
            'status': 'OK',
            'message': '${projectName} API is running',
            'version': '1.0.0'
        }
    
    # Welcome endpoint
    @app.route('/')
    def welcome():
        return {
            'message': 'Welcome to ${projectName} API',
            'version': '1.0.0',
            'documentation': '/api/docs',
            'health': '/health'
        }
    
    # Error handlers
    @app.errorhandler(404)
    def not_found(error):
        return {'error': 'Resource not found'}, 404
    
    @app.errorhandler(500)
    def internal_error(error):
        return {'error': 'Internal server error'}, 500
    
    return app

if __name__ == '__main__':
    app = create_app()
    app.run(debug=True)`
  });

  // Configuration
  files.push({
    path: 'config.py',
    content: `import os
from datetime import timedelta

class Config:
    # Basic Flask config
    SECRET_KEY = os.environ.get('SECRET_KEY') or 'dev-secret-key-change-in-production'
    
    # Database
    SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL') or 'sqlite:///app.db'
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    
    # JWT Configuration
    JWT_SECRET_KEY = os.environ.get('JWT_SECRET_KEY') or 'jwt-secret-change-in-production'
    JWT_ACCESS_TOKEN_EXPIRES = timedelta(hours=1)
    JWT_REFRESH_TOKEN_EXPIRES = timedelta(days=30)
    
    # CORS
    CORS_ORIGINS = os.environ.get('CORS_ORIGINS', 'http://localhost:3000').split(',')
    
    # Rate limiting
    RATELIMIT_STORAGE_URL = os.environ.get('REDIS_URL') or 'memory://'

class DevelopmentConfig(Config):
    DEBUG = True

class ProductionConfig(Config):
    DEBUG = False

class TestingConfig(Config):
    TESTING = True
    SQLALCHEMY_DATABASE_URI = 'sqlite:///:memory:'

config = {
    'development': DevelopmentConfig,
    'production': ProductionConfig,
    'testing': TestingConfig,
    'default': DevelopmentConfig
}`
  });

  // Extensions
  files.push({
    path: 'extensions.py',
    content: `from flask_sqlalchemy import SQLAlchemy
from flask_migrate import Migrate
from flask_jwt_extended import JWTManager

db = SQLAlchemy()
migrate = Migrate()
jwt = JWTManager()`
  });

  // Models
  files.push({
    path: 'models.py',
    content: `from extensions import db
from datetime import datetime
import bcrypt

class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)
    email = db.Column(db.String(120), unique=True, nullable=False, index=True)
    password_hash = db.Column(db.String(128), nullable=False)
    role = db.Column(db.String(20), default='user')
    is_active = db.Column(db.Boolean, default=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    last_login = db.Column(db.DateTime)

    def set_password(self, password):
        """Hash and set password"""
        self.password_hash = bcrypt.hashpw(
            password.encode('utf-8'), 
            bcrypt.gensalt()
        ).decode('utf-8')

    def check_password(self, password):
        """Check if provided password matches hash"""
        return bcrypt.checkpw(
            password.encode('utf-8'), 
            self.password_hash.encode('utf-8')
        )

    def to_dict(self):
        """Convert user to dictionary (excluding password)"""
        return {
            'id': self.id,
            'name': self.name,
            'email': self.email,
            'role': self.role,
            'is_active': self.is_active,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'last_login': self.last_login.isoformat() if self.last_login else None
        }

    def __repr__(self):
        return f'<User {self.email}>'`
  });

  // Schemas for validation
  files.push({
    path: 'schemas.py',
    content: `from marshmallow import Schema, fields, validate, ValidationError

class UserRegistrationSchema(Schema):
    name = fields.Str(required=True, validate=validate.Length(min=2, max=100))
    email = fields.Email(required=True)
    password = fields.Str(required=True, validate=validate.Length(min=6))

class UserLoginSchema(Schema):
    email = fields.Email(required=True)
    password = fields.Str(required=True)

class UserUpdateSchema(Schema):
    name = fields.Str(validate=validate.Length(min=2, max=100))
    email = fields.Email()

class UserResponseSchema(Schema):
    id = fields.Int()
    name = fields.Str()
    email = fields.Email()
    role = fields.Str()
    is_active = fields.Bool()
    created_at = fields.DateTime()
    updated_at = fields.DateTime()
    last_login = fields.DateTime()`
  });

  // Auth routes
  files.push({
    path: 'routes/auth.py',
    content: `from flask import Blueprint, request, jsonify
from flask_jwt_extended import create_access_token, jwt_required, get_jwt_identity
from marshmallow import ValidationError
from datetime import datetime
from extensions import db
from models import User
from schemas import UserRegistrationSchema, UserLoginSchema, UserResponseSchema

auth_bp = Blueprint('auth', __name__)

@auth_bp.route('/register', methods=['POST'])
def register():
    """Register a new user"""
    try:
        # Validate input
        schema = UserRegistrationSchema()
        data = schema.load(request.json)
        
        # Check if user already exists
        if User.query.filter_by(email=data['email']).first():
            return jsonify({
                'success': False,
                'message': 'User already exists with this email'
            }), 400
        
        # Create new user
        user = User(
            name=data['name'],
            email=data['email']
        )
        user.set_password(data['password'])
        
        db.session.add(user)
        db.session.commit()
        
        # Create access token
        access_token = create_access_token(identity=user.id)
        
        # Prepare response
        user_schema = UserResponseSchema()
        user_data = user_schema.dump(user.to_dict())
        
        return jsonify({
            'success': True,
            'message': 'User registered successfully',
            'data': {
                'user': user_data,
                'access_token': access_token
            }
        }), 201
        
    except ValidationError as err:
        return jsonify({
            'success': False,
            'message': 'Validation failed',
            'errors': err.messages
        }), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({
            'success': False,
            'message': 'Registration failed'
        }), 500

@auth_bp.route('/login', methods=['POST'])
def login():
    """Login user"""
    try:
        # Validate input
        schema = UserLoginSchema()
        data = schema.load(request.json)
        
        # Find user
        user = User.query.filter_by(email=data['email']).first()
        
        if not user or not user.check_password(data['password']):
            return jsonify({
                'success': False,
                'message': 'Invalid credentials'
            }), 401
        
        if not user.is_active:
            return jsonify({
                'success': False,
                'message': 'Account is deactivated'
            }), 401
        
        # Update last login
        user.last_login = datetime.utcnow()
        db.session.commit()
        
        # Create access token
        access_token = create_access_token(identity=user.id)
        
        # Prepare response
        user_schema = UserResponseSchema()
        user_data = user_schema.dump(user.to_dict())
        
        return jsonify({
            'success': True,
            'message': 'Login successful',
            'data': {
                'user': user_data,
                'access_token': access_token
            }
        })
        
    except ValidationError as err:
        return jsonify({
            'success': False,
            'message': 'Validation failed',
            'errors': err.messages
        }), 400
    except Exception as e:
        return jsonify({
            'success': False,
            'message': 'Login failed'
        }), 500

@auth_bp.route('/me', methods=['GET'])
@jwt_required()
def get_current_user():
    """Get current user information"""
    try:
        user_id = get_jwt_identity()
        user = User.query.get(user_id)
        
        if not user:
            return jsonify({
                'success': False,
                'message': 'User not found'
            }), 404
        
        user_schema = UserResponseSchema()
        user_data = user_schema.dump(user.to_dict())
        
        return jsonify({
            'success': True,
            'data': {'user': user_data}
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'message': 'Failed to get user information'
        }), 500`
  });

  // User routes
  files.push({
    path: 'routes/users.py',
    content: `from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from marshmallow import ValidationError
from extensions import db
from models import User
from schemas import UserUpdateSchema, UserResponseSchema
from utils.auth import admin_required

users_bp = Blueprint('users', __name__)

@users_bp.route('', methods=['GET'])
@jwt_required()
@admin_required
def get_users():
    """Get all users (admin only)"""
    try:
        page = request.args.get('page', 1, type=int)
        per_page = min(request.args.get('per_page', 10, type=int), 100)
        
        users = User.query.filter_by(is_active=True).paginate(
            page=page, per_page=per_page, error_out=False
        )
        
        user_schema = UserResponseSchema(many=True)
        users_data = user_schema.dump([user.to_dict() for user in users.items])
        
        return jsonify({
            'success': True,
            'data': {
                'users': users_data,
                'pagination': {
                    'page': page,
                    'per_page': per_page,
                    'total': users.total,
                    'pages': users.pages
                }
            }
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'message': 'Failed to get users'
        }), 500

@users_bp.route('/<int:user_id>', methods=['GET'])
@jwt_required()
def get_user(user_id):
    """Get user by ID"""
    try:
        current_user_id = get_jwt_identity()
        current_user = User.query.get(current_user_id)
        
        user = User.query.get(user_id)
        if not user:
            return jsonify({
                'success': False,
                'message': 'User not found'
            }), 404
        
        # Users can only view their own profile unless they're admin
        if current_user_id != user_id and current_user.role != 'admin':
            return jsonify({
                'success': False,
                'message': 'Access denied'
            }), 403
        
        user_schema = UserResponseSchema()
        user_data = user_schema.dump(user.to_dict())
        
        return jsonify({
            'success': True,
            'data': {'user': user_data}
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'message': 'Failed to get user'
        }), 500

@users_bp.route('/<int:user_id>', methods=['PUT'])
@jwt_required()
def update_user(user_id):
    """Update user"""
    try:
        current_user_id = get_jwt_identity()
        current_user = User.query.get(current_user_id)
        
        user = User.query.get(user_id)
        if not user:
            return jsonify({
                'success': False,
                'message': 'User not found'
            }), 404
        
        # Users can only update their own profile unless they're admin
        if current_user_id != user_id and current_user.role != 'admin':
            return jsonify({
                'success': False,
                'message': 'Access denied'
            }), 403
        
        # Validate input
        schema = UserUpdateSchema()
        data = schema.load(request.json)
        
        # Check if email is already taken by another user
        if 'email' in data and data['email'] != user.email:
            existing_user = User.query.filter_by(email=data['email']).first()
            if existing_user:
                return jsonify({
                    'success': False,
                    'message': 'Email already in use'
                }), 400
        
        # Update user
        for key, value in data.items():
            setattr(user, key, value)
        
        db.session.commit()
        
        user_schema = UserResponseSchema()
        user_data = user_schema.dump(user.to_dict())
        
        return jsonify({
            'success': True,
            'message': 'User updated successfully',
            'data': {'user': user_data}
        })
        
    except ValidationError as err:
        return jsonify({
            'success': False,
            'message': 'Validation failed',
            'errors': err.messages
        }), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({
            'success': False,
            'message': 'Failed to update user'
        }), 500

@users_bp.route('/<int:user_id>', methods=['DELETE'])
@jwt_required()
@admin_required
def delete_user(user_id):
    """Delete user (soft delete, admin only)"""
    try:
        user = User.query.get(user_id)
        if not user:
            return jsonify({
                'success': False,
                'message': 'User not found'
            }), 404
        
        # Soft delete
        user.is_active = False
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': 'User deleted successfully'
        })
        
    except Exception as e:
        db.session.rollback()
        return jsonify({
            'success': False,
            'message': 'Failed to delete user'
        }), 500`
  });

  // API routes
  files.push({
    path: 'routes/api.py',
    content: `from flask import Blueprint, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity

api_bp = Blueprint('api', __name__)

@api_bp.route('/status', methods=['GET'])
def status():
    """Get API status"""
    return jsonify({
        'success': True,
        'message': '${projectName} API is running',
        'version': '1.0.0'
    })

@api_bp.route('/protected', methods=['GET'])
@jwt_required()
def protected():
    """Protected route example"""
    user_id = get_jwt_identity()
    return jsonify({
        'success': True,
        'message': 'This is a protected route',
        'user_id': user_id
    })

@api_bp.route('/docs', methods=['GET'])
def docs():
    """API documentation"""
    return jsonify({
        'success': True,
        'message': '${projectName} API Documentation',
        'endpoints': {
            'auth': {
                'POST /api/auth/register': 'Register a new user',
                'POST /api/auth/login': 'Login user',
                'GET /api/auth/me': 'Get current user (requires auth)'
            },
            'users': {
                'GET /api/users': 'Get all users (admin only)',
                'GET /api/users/<id>': 'Get user by ID (requires auth)',
                'PUT /api/users/<id>': 'Update user (requires auth)',
                'DELETE /api/users/<id>': 'Delete user (admin only)'
            },
            'general': {
                'GET /api/status': 'Get API status',
                'GET /api/protected': 'Protected route example (requires auth)',
                'GET /health': 'Health check endpoint'
            }
        }
    })`
  });

  // Routes init
  files.push({
    path: 'routes/__init__.py',
    content: `from .auth import auth_bp
from .users import users_bp
from .api import api_bp

__all__ = ['auth_bp', 'users_bp', 'api_bp']`
  });

  // Auth utilities
  files.push({
    path: 'utils/auth.py',
    content: `from functools import wraps
from flask import jsonify
from flask_jwt_extended import get_jwt_identity
from models import User

def admin_required(f):
    """Decorator to require admin role"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        user_id = get_jwt_identity()
        user = User.query.get(user_id)
        
        if not user or user.role != 'admin':
            return jsonify({
                'success': False,
                'message': 'Admin privileges required'
            }), 403
            
        return f(*args, **kwargs)
    return decorated_function`
  });

  // Utils init
  files.push({
    path: 'utils/__init__.py',
    content: ``
  });

  // Environment file
  files.push({
    path: '.env.example',
    content: `# Flask Configuration
FLASK_APP=app.py
FLASK_ENV=development
SECRET_KEY=your-secret-key-here

# Database
DATABASE_URL=sqlite:///app.db

# JWT
JWT_SECRET_KEY=your-jwt-secret-here

# CORS
CORS_ORIGINS=http://localhost:3000

# Redis (for rate limiting)
REDIS_URL=redis://localhost:6379/0`
  });

  // Test file
  files.push({
    path: 'tests/test_auth.py',
    content: `import pytest
import json
from app import create_app
from extensions import db
from models import User
from config import TestingConfig

@pytest.fixture
def app():
    app = create_app(TestingConfig)
    with app.app_context():
        db.create_all()
        yield app
        db.drop_all()

@pytest.fixture
def client(app):
    return app.test_client()

@pytest.fixture
def sample_user(app):
    with app.app_context():
        user = User(name='Test User', email='test@example.com')
        user.set_password('password123')
        db.session.add(user)
        db.session.commit()
        return user

def test_register_user(client):
    """Test user registration"""
    response = client.post('/api/auth/register', 
        json={
            'name': 'New User',
            'email': 'new@example.com',
            'password': 'password123'
        }
    )
    
    assert response.status_code == 201
    data = json.loads(response.data)
    assert data['success'] is True
    assert 'access_token' in data['data']

def test_register_duplicate_email(client, sample_user):
    """Test registration with duplicate email"""
    response = client.post('/api/auth/register',
        json={
            'name': 'Another User',
            'email': 'test@example.com',
            'password': 'password123'
        }
    )
    
    assert response.status_code == 400
    data = json.loads(response.data)
    assert data['success'] is False

def test_login_valid_credentials(client, sample_user):
    """Test login with valid credentials"""
    response = client.post('/api/auth/login',
        json={
            'email': 'test@example.com',
            'password': 'password123'
        }
    )
    
    assert response.status_code == 200
    data = json.loads(response.data)
    assert data['success'] is True
    assert 'access_token' in data['data']

def test_login_invalid_credentials(client):
    """Test login with invalid credentials"""
    response = client.post('/api/auth/login',
        json={
            'email': 'nonexistent@example.com',
            'password': 'wrongpassword'
        }
    )
    
    assert response.status_code == 401
    data = json.loads(response.data)
    assert data['success'] is False

def test_get_current_user(client, sample_user):
    """Test getting current user info"""
    # First login to get token
    login_response = client.post('/api/auth/login',
        json={
            'email': 'test@example.com',
            'password': 'password123'
        }
    )
    
    token = json.loads(login_response.data)['data']['access_token']
    
    # Then get user info
    response = client.get('/api/auth/me',
        headers={'Authorization': f'Bearer {token}'}
    )
    
    assert response.status_code == 200
    data = json.loads(response.data)
    assert data['success'] is True
    assert data['data']['user']['email'] == 'test@example.com'`
  });

  // Pytest config
  files.push({
    path: 'pytest.ini',
    content: `[tool:pytest]
testpaths = tests
python_files = test_*.py
python_classes = Test*
python_functions = test_*
addopts = -v --tb=short`
  });

  // .gitignore
  files.push({
    path: '.gitignore',
    content: `# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class

# C extensions
*.so

# Distribution / packaging
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST

# PyInstaller
*.manifest
*.spec

# Installer logs
pip-log.txt
pip-delete-this-directory.txt

# Unit test / coverage reports
htmlcov/
.tox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
.hypothesis/
.pytest_cache/

# Translations
*.mo
*.pot

# Django stuff:
*.log
local_settings.py
db.sqlite3

# Flask stuff:
instance/
.webassets-cache

# Scrapy stuff:
.scrapy

# Sphinx documentation
docs/_build/

# PyBuilder
target/

# Jupyter Notebook
.ipynb_checkpoints

# pyenv
.python-version

# celery beat schedule file
celerybeat-schedule

# SageMath parsed files
*.sage.py

# Environments
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/

# Spyder project settings
.spyderproject
.spyproject

# Rope project settings
.ropeproject

# mkdocs documentation
/site

# mypy
.mypy_cache/
.dmypy.json
dmypy.json

# Database
*.db
*.sqlite3

# IDE
.vscode/
.idea/
*.swp
*.swo`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A RESTful API built with Flask, SQLAlchemy, and JWT authentication.

## Features

- 🐍 Flask web framework
- 🔐 JWT authentication and authorization
- 👤 User management with roles
- 🛡️ Security middleware (CORS, Rate limiting)
- 📊 SQLAlchemy ORM with migrations
- ✅ Input validation with Marshmallow
- 🧪 Testing with pytest
- 📝 Comprehensive API documentation

## Getting Started

### Prerequisites

- Python 3.8 or higher
- pip (Python package installer)

### Installation

1. Clone the repository and navigate to the project directory

2. Create a virtual environment:
   \`\`\`bash
   python -m venv venv
   source venv/bin/activate  # On Windows: venv\\Scripts\\activate
   \`\`\`

3. Install dependencies:
   \`\`\`bash
   pip install -r requirements.txt
   \`\`\`

4. Set up environment variables:
   \`\`\`bash
   cp .env.example .env
   \`\`\`
   
   Update the \`.env\` file with your configuration.

5. Initialize the database:
   \`\`\`bash
   flask db init
   flask db migrate -m "Initial migration"
   flask db upgrade
   \`\`\`

6. Run the development server:
   \`\`\`bash
   python app.py
   \`\`\`

7. The API will be available at \`http://localhost:5000\`

## API Endpoints

### Authentication
- \`POST /api/auth/register\` - Register a new user
- \`POST /api/auth/login\` - Login user
- \`GET /api/auth/me\` - Get current user (requires auth)

### Users
- \`GET /api/users\` - Get all users (admin only)
- \`GET /api/users/<id>\` - Get user by ID (requires auth)
- \`PUT /api/users/<id>\` - Update user (requires auth)
- \`DELETE /api/users/<id>\` - Delete user (admin only)

### General
- \`GET /\` - Welcome message
- \`GET /health\` - Health check
- \`GET /api/status\` - API status
- \`GET /api/docs\` - API documentation
- \`GET /api/protected\` - Protected route example

## Authentication

This API uses JWT (JSON Web Tokens) for authentication. Include the token in the Authorization header:

\`\`\`
Authorization: Bearer <your-jwt-token>
\`\`\`

## Project Structure

\`\`\`
├── app.py                  # Main application file
├── config.py              # Configuration settings
├── extensions.py          # Flask extensions
├── models.py              # Database models
├── schemas.py             # Marshmallow schemas
├── routes/
│   ├── __init__.py
│   ├── auth.py           # Authentication routes
│   ├── users.py          # User management routes
│   └── api.py            # General API routes
├── utils/
│   ├── __init__.py
│   └── auth.py           # Authentication utilities
├── tests/
│   └── test_auth.py      # Authentication tests
├── migrations/           # Database migrations
├── requirements.txt      # Python dependencies
└── .env.example         # Environment variables template
\`\`\`

## Database Migrations

This project uses Flask-Migrate for database migrations:

\`\`\`bash
# Create a new migration
flask db migrate -m "Description of changes"

# Apply migrations
flask db upgrade

# Downgrade to previous migration
flask db downgrade
\`\`\`

## Testing

Run the test suite:

\`\`\`bash
pytest
\`\`\`

Run tests with coverage:

\`\`\`bash
pytest --cov=. --cov-report=html
\`\`\`

## Environment Variables

Create a \`.env\` file in the root directory:

\`\`\`env
FLASK_APP=app.py
FLASK_ENV=development
SECRET_KEY=your-secret-key-here
DATABASE_URL=sqlite:///app.db
JWT_SECRET_KEY=your-jwt-secret-here
CORS_ORIGINS=http://localhost:3000
REDIS_URL=redis://localhost:6379/0
\`\`\`

## Deployment

### Using Gunicorn

1. Install Gunicorn (included in requirements.txt)
2. Run with Gunicorn:
   \`\`\`bash
   gunicorn -w 4 -b 0.0.0.0:5000 app:app
   \`\`\`

### Using Docker

Create a \`Dockerfile\`:

\`\`\`dockerfile
FROM python:3.9-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install -r requirements.txt

COPY . .

EXPOSE 5000

CMD ["gunicorn", "-w", "4", "-b", "0.0.0.0:5000", "app:app"]
\`\`\`

### Heroku Deployment

1. Create a \`Procfile\`:
   \`\`\`
   web: gunicorn app:app
   \`\`\`

2. Set environment variables in Heroku dashboard
3. Deploy with Git:
   \`\`\`bash
   git push heroku main
   \`\`\`

## Security Features

- **JWT Authentication**: Secure token-based authentication
- **Password Hashing**: Bcrypt for secure password storage
- **CORS Protection**: Configurable cross-origin resource sharing
- **Rate Limiting**: Prevents abuse with configurable limits
- **Input Validation**: Marshmallow schemas for request validation
- **SQL Injection Protection**: SQLAlchemy ORM prevents SQL injection

## API Documentation

Visit \`/api/docs\` for interactive API documentation, or use tools like Postman with the following example requests:

### Register User
\`\`\`bash
curl -X POST http://localhost:5000/api/auth/register \\
  -H "Content-Type: application/json" \\
  -d '{"name": "John Doe", "email": "john@example.com", "password": "password123"}'
\`\`\`

### Login
\`\`\`bash
curl -X POST http://localhost:5000/api/auth/login \\
  -H "Content-Type: application/json" \\
  -d '{"email": "john@example.com", "password": "password123"}'
\`\`\`

### Access Protected Route
\`\`\`bash
curl -X GET http://localhost:5000/api/auth/me \\
  -H "Authorization: Bearer YOUR_JWT_TOKEN"
\`\`\`

## Contributing

1. Fork the repository
2. Create a feature branch
3. Make your changes
4. Add tests for new functionality
5. Ensure all tests pass
6. Submit a pull request

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'Flask API',
  description: 'A RESTful API server with Flask, SQLAlchemy, and JWT authentication',
  generate
};
