// codestral/src/templates/full-stack.js
/**
 * Full-Stack Template (React + Express)
 * Creates a complete full-stack application with React frontend and Express backend
 */

import reactTemplate from './react-app.js';
import expressTemplate from './express-api.js';

function generate(projectName, options = {}) {
  const files = [];

  // Generate React frontend files in 'client' directory
  const reactFiles = reactTemplate.generate(projectName, options);
  reactFiles.forEach(file => {
    files.push({
      path: `client/${file.path}`,
      content: file.content
    });
  });

  // Generate Express backend files in 'server' directory
  const expressFiles = expressTemplate.generate(`${projectName} API`, options);
  expressFiles.forEach(file => {
    files.push({
      path: `server/${file.path}`,
      content: file.content
    });
  });

  // Root package.json for managing both client and server
  files.push({
    path: 'package.json',
    content: `{
  "name": "${projectName.toLowerCase().replace(/\s+/g, '-')}",
  "version": "1.0.0",
  "description": "${projectName} - Full-stack React and Express application",
  "main": "server/server.js",
  "scripts": {
    "dev": "concurrently \\"npm run server\\" \\"npm run client\\"",
    "server": "cd server && npm run dev",
    "client": "cd client && npm start",
    "build": "cd client && npm run build",
    "install-server": "cd server && npm install",
    "install-client": "cd client && npm install",
    "install-all": "npm run install-server && npm run install-client",
    "start": "cd server && npm start",
    "test": "cd server && npm test && cd ../client && npm test",
    "heroku-postbuild": "npm run install-all && npm run build"
  },
  "keywords": ["react", "express", "fullstack", "nodejs"],
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "concurrently": "^8.2.0"
  }
}`
  });

  // Update client package.json to add proxy for development
  const clientPackageIndex = files.findIndex(f => f.path === 'client/package.json');
  if (clientPackageIndex !== -1) {
    const clientPackage = JSON.parse(files[clientPackageIndex].content);
    clientPackage.proxy = "http://localhost:3001";
    files[clientPackageIndex].content = JSON.stringify(clientPackage, null, 2);
  }

  // Update server package.json to use port 3001
  const serverPackageIndex = files.findIndex(f => f.path === 'server/package.json');
  if (serverPackageIndex !== -1) {
    const serverPackage = JSON.parse(files[serverPackageIndex].content);
    serverPackage.scripts.start = "PORT=3001 node server.js";
    serverPackage.scripts.dev = "PORT=3001 nodemon server.js";
    files[serverPackageIndex].content = JSON.stringify(serverPackage, null, 2);
  }

  // Update server .env.example to use port 3001
  const serverEnvIndex = files.findIndex(f => f.path === 'server/.env.example');
  if (serverEnvIndex !== -1) {
    files[serverEnvIndex].content = files[serverEnvIndex].content.replace('PORT=3000', 'PORT=3001');
    files[serverEnvIndex].content = files[serverEnvIndex].content.replace(
      'FRONTEND_URL=http://localhost:3000',
      'FRONTEND_URL=http://localhost:3000'
    );
  }

  // Enhanced React App.js with API integration
  const appJsIndex = files.findIndex(f => f.path === 'client/src/App.js');
  if (appJsIndex !== -1) {
    files[appJsIndex].content = `import React, { useState, useEffect } from 'react';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import Header from './components/Header';
import Footer from './components/Footer';
import Home from './pages/Home';
import About from './pages/About';
import Contact from './pages/Contact';
import Login from './pages/Login';
import Register from './pages/Register';
import Dashboard from './pages/Dashboard';
import { AuthProvider } from './context/AuthContext';
import ProtectedRoute from './components/ProtectedRoute';
import './App.css';

function App() {
  return (
    <AuthProvider>
      <Router>
        <div className="App">
          <Header />
          <main className="main-content">
            <Routes>
              <Route path="/" element={<Home />} />
              <Route path="/about" element={<About />} />
              <Route path="/contact" element={<Contact />} />
              <Route path="/login" element={<Login />} />
              <Route path="/register" element={<Register />} />
              <Route 
                path="/dashboard" 
                element={
                  <ProtectedRoute>
                    <Dashboard />
                  </ProtectedRoute>
                } 
              />
            </Routes>
          </main>
          <Footer />
        </div>
      </Router>
    </AuthProvider>
  );
}

export default App;`;
  }

  // Add Auth Context
  files.push({
    path: 'client/src/context/AuthContext.js',
    content: `import React, { createContext, useContext, useState, useEffect } from 'react';

const AuthContext = createContext();

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);
  const [token, setToken] = useState(localStorage.getItem('token'));

  useEffect(() => {
    if (token) {
      fetchUser();
    } else {
      setLoading(false);
    }
  }, [token]);

  const fetchUser = async () => {
    try {
      const response = await fetch('/api/auth/me', {
        headers: {
          'Authorization': \`Bearer \${token}\`
        }
      });

      if (response.ok) {
        const data = await response.json();
        setUser(data.data.user);
      } else {
        logout();
      }
    } catch (error) {
      console.error('Error fetching user:', error);
      logout();
    } finally {
      setLoading(false);
    }
  };

  const login = async (email, password) => {
    try {
      const response = await fetch('/api/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({ email, password })
      });

      const data = await response.json();

      if (response.ok) {
        setToken(data.data.token);
        setUser(data.data.user);
        localStorage.setItem('token', data.data.token);
        return { success: true };
      } else {
        return { success: false, message: data.message };
      }
    } catch (error) {
      return { success: false, message: 'Network error' };
    }
  };

  const register = async (name, email, password) => {
    try {
      const response = await fetch('/api/auth/register', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({ name, email, password })
      });

      const data = await response.json();

      if (response.ok) {
        setToken(data.data.token);
        setUser(data.data.user);
        localStorage.setItem('token', data.data.token);
        return { success: true };
      } else {
        return { success: false, message: data.message };
      }
    } catch (error) {
      return { success: false, message: 'Network error' };
    }
  };

  const logout = () => {
    setUser(null);
    setToken(null);
    localStorage.removeItem('token');
  };

  const value = {
    user,
    token,
    loading,
    login,
    register,
    logout
  };

  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
};`
  });

  // Add Protected Route component
  files.push({
    path: 'client/src/components/ProtectedRoute.js',
    content: `import React from 'react';
import { Navigate } from 'react-router-dom';
import { useAuth } from '../context/AuthContext';

const ProtectedRoute = ({ children }) => {
  const { user, loading } = useAuth();

  if (loading) {
    return (
      <div className="loading-container">
        <div className="loading-spinner">Loading...</div>
      </div>
    );
  }

  return user ? children : <Navigate to="/login" />;
};

export default ProtectedRoute;`
  });

  // Add Login page
  files.push({
    path: 'client/src/pages/Login.js',
    content: `import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useAuth } from '../context/AuthContext';
import './Auth.css';

const Login = () => {
  const [formData, setFormData] = useState({
    email: '',
    password: ''
  });
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  
  const { login } = useAuth();
  const navigate = useNavigate();

  const handleChange = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value
    });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setError('');
    setLoading(true);

    const result = await login(formData.email, formData.password);
    
    if (result.success) {
      navigate('/dashboard');
    } else {
      setError(result.message);
    }
    
    setLoading(false);
  };

  return (
    <div className="auth-container">
      <div className="auth-card">
        <h1>Login</h1>
        <p>Welcome back! Please sign in to your account.</p>
        
        {error && <div className="error-message">{error}</div>}
        
        <form onSubmit={handleSubmit} className="auth-form">
          <div className="form-group">
            <label htmlFor="email">Email</label>
            <input
              type="email"
              id="email"
              name="email"
              value={formData.email}
              onChange={handleChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="password">Password</label>
            <input
              type="password"
              id="password"
              name="password"
              value={formData.password}
              onChange={handleChange}
              required
            />
          </div>

          <button type="submit" className="btn btn-primary" disabled={loading}>
            {loading ? 'Signing in...' : 'Sign In'}
          </button>
        </form>

        <p className="auth-link">
          Don't have an account? <Link to="/register">Sign up</Link>
        </p>
      </div>
    </div>
  );
};

export default Login;`
  });

  // Add Register page
  files.push({
    path: 'client/src/pages/Register.js',
    content: `import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useAuth } from '../context/AuthContext';
import './Auth.css';

const Register = () => {
  const [formData, setFormData] = useState({
    name: '',
    email: '',
    password: '',
    confirmPassword: ''
  });
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  
  const { register } = useAuth();
  const navigate = useNavigate();

  const handleChange = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value
    });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    setError('');

    if (formData.password !== formData.confirmPassword) {
      setError('Passwords do not match');
      return;
    }

    if (formData.password.length < 6) {
      setError('Password must be at least 6 characters');
      return;
    }

    setLoading(true);

    const result = await register(formData.name, formData.email, formData.password);
    
    if (result.success) {
      navigate('/dashboard');
    } else {
      setError(result.message);
    }
    
    setLoading(false);
  };

  return (
    <div className="auth-container">
      <div className="auth-card">
        <h1>Sign Up</h1>
        <p>Create your account to get started.</p>
        
        {error && <div className="error-message">{error}</div>}
        
        <form onSubmit={handleSubmit} className="auth-form">
          <div className="form-group">
            <label htmlFor="name">Full Name</label>
            <input
              type="text"
              id="name"
              name="name"
              value={formData.name}
              onChange={handleChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="email">Email</label>
            <input
              type="email"
              id="email"
              name="email"
              value={formData.email}
              onChange={handleChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="password">Password</label>
            <input
              type="password"
              id="password"
              name="password"
              value={formData.password}
              onChange={handleChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="confirmPassword">Confirm Password</label>
            <input
              type="password"
              id="confirmPassword"
              name="confirmPassword"
              value={formData.confirmPassword}
              onChange={handleChange}
              required
            />
          </div>

          <button type="submit" className="btn btn-primary" disabled={loading}>
            {loading ? 'Creating account...' : 'Sign Up'}
          </button>
        </form>

        <p className="auth-link">
          Already have an account? <Link to="/login">Sign in</Link>
        </p>
      </div>
    </div>
  );
};

export default Register;`
  });

  // Add Dashboard page
  files.push({
    path: 'client/src/pages/Dashboard.js',
    content: `import React, { useState, useEffect } from 'react';
import { useAuth } from '../context/AuthContext';
import './Dashboard.css';

const Dashboard = () => {
  const { user, token, logout } = useAuth();
  const [apiStatus, setApiStatus] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetchApiStatus();
  }, []);

  const fetchApiStatus = async () => {
    try {
      const response = await fetch('/api/status');
      const data = await response.json();
      setApiStatus(data);
    } catch (error) {
      console.error('Error fetching API status:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleLogout = () => {
    logout();
  };

  if (loading) {
    return (
      <div className="dashboard">
        <div className="container">
          <div className="loading">Loading dashboard...</div>
        </div>
      </div>
    );
  }

  return (
    <div className="dashboard">
      <div className="container">
        <div className="dashboard-header">
          <h1>Welcome, {user?.name}!</h1>
          <button onClick={handleLogout} className="btn btn-secondary">
            Logout
          </button>
        </div>

        <div className="dashboard-grid">
          <div className="dashboard-card">
            <h3>👤 Profile Information</h3>
            <div className="profile-info">
              <p><strong>Name:</strong> {user?.name}</p>
              <p><strong>Email:</strong> {user?.email}</p>
              <p><strong>Role:</strong> {user?.role}</p>
              <p><strong>Member since:</strong> {new Date(user?.createdAt).toLocaleDateString()}</p>
            </div>
          </div>

          <div className="dashboard-card">
            <h3>🚀 API Status</h3>
            {apiStatus ? (
              <div className="api-status">
                <p><strong>Status:</strong> <span className="status-ok">✅ Online</span></p>
                <p><strong>Version:</strong> {apiStatus.version}</p>
                <p><strong>Message:</strong> {apiStatus.message}</p>
              </div>
            ) : (
              <p>Unable to fetch API status</p>
            )}
          </div>

          <div className="dashboard-card">
            <h3>📊 Quick Stats</h3>
            <div className="stats">
              <div className="stat-item">
                <span className="stat-number">1</span>
                <span className="stat-label">Active Sessions</span>
              </div>
              <div className="stat-item">
                <span className="stat-number">{user?.lastLogin ? '✅' : '❌'}</span>
                <span className="stat-label">Last Login</span>
              </div>
            </div>
          </div>

          <div className="dashboard-card">
            <h3>🔧 Quick Actions</h3>
            <div className="actions">
              <button className="btn btn-primary">Update Profile</button>
              <button className="btn btn-secondary">View Settings</button>
              <button className="btn btn-secondary">Download Data</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Dashboard;`
  });

  // Add Auth CSS
  files.push({
    path: 'client/src/pages/Auth.css',
    content: `.auth-container {
  min-height: 80vh;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem 0;
}

.auth-card {
  background: white;
  padding: 3rem;
  border-radius: 10px;
  box-shadow: 0 10px 30px rgba(0, 0, 0, 0.1);
  width: 100%;
  max-width: 400px;
}

.auth-card h1 {
  text-align: center;
  margin-bottom: 0.5rem;
  color: #333;
}

.auth-card p {
  text-align: center;
  color: #666;
  margin-bottom: 2rem;
}

.auth-form {
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}

.error-message {
  background-color: #fee;
  color: #c33;
  padding: 0.75rem;
  border-radius: 5px;
  border: 1px solid #fcc;
  margin-bottom: 1rem;
  text-align: center;
}

.auth-link {
  text-align: center;
  margin-top: 2rem;
  color: #666;
}

.auth-link a {
  color: #007bff;
  text-decoration: none;
}

.auth-link a:hover {
  text-decoration: underline;
}

.loading-container {
  min-height: 80vh;
  display: flex;
  align-items: center;
  justify-content: center;
}

.loading-spinner {
  font-size: 1.2rem;
  color: #666;
}`
  });

  // Add Dashboard CSS
  files.push({
    path: 'client/src/pages/Dashboard.css',
    content: `.dashboard {
  padding: 2rem 0;
  min-height: 80vh;
}

.dashboard-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 3rem;
  padding-bottom: 1rem;
  border-bottom: 1px solid #eee;
}

.dashboard-header h1 {
  color: #333;
  margin: 0;
}

.dashboard-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  gap: 2rem;
}

.dashboard-card {
  background: white;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
}

.dashboard-card h3 {
  margin-bottom: 1.5rem;
  color: #333;
  font-size: 1.2rem;
}

.profile-info p {
  margin-bottom: 0.5rem;
  color: #555;
}

.profile-info strong {
  color: #333;
}

.api-status p {
  margin-bottom: 0.5rem;
  color: #555;
}

.status-ok {
  color: #28a745;
}

.stats {
  display: flex;
  justify-content: space-around;
}

.stat-item {
  text-align: center;
}

.stat-number {
  display: block;
  font-size: 2rem;
  font-weight: bold;
  color: #007bff;
  margin-bottom: 0.5rem;
}

.stat-label {
  font-size: 0.9rem;
  color: #666;
}

.actions {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.actions .btn {
  width: 100%;
}

.loading {
  text-align: center;
  padding: 3rem;
  color: #666;
  font-size: 1.2rem;
}

@media (max-width: 768px) {
  .dashboard-header {
    flex-direction: column;
    gap: 1rem;
    text-align: center;
  }
  
  .dashboard-grid {
    grid-template-columns: 1fr;
  }
  
  .stats {
    flex-direction: column;
    gap: 1rem;
  }
}`
  });

  // Update Header component to include auth links
  const headerIndex = files.findIndex(f => f.path === 'client/src/components/Header.js');
  if (headerIndex !== -1) {
    files[headerIndex].content = `import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { useAuth } from '../context/AuthContext';
import './Header.css';

const Header = () => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);
  const { user, logout } = useAuth();

  const toggleMenu = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  const handleLogout = () => {
    logout();
    setIsMenuOpen(false);
  };

  return (
    <header className="header">
      <nav className="navbar">
        <div className="container">
          <Link to="/" className="nav-brand">
            ${projectName}
          </Link>
          
          <button 
            className={\`nav-toggle \${isMenuOpen ? 'active' : ''}\`}
            onClick={toggleMenu}
            aria-label="Toggle navigation"
          >
            <span></span>
            <span></span>
            <span></span>
          </button>

          <ul className={\`nav-menu \${isMenuOpen ? 'active' : ''}\`}>
            <li className="nav-item">
              <Link to="/" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                Home
              </Link>
            </li>
            <li className="nav-item">
              <Link to="/about" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                About
              </Link>
            </li>
            <li className="nav-item">
              <Link to="/contact" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                Contact
              </Link>
            </li>
            {user ? (
              <>
                <li className="nav-item">
                  <Link to="/dashboard" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                    Dashboard
                  </Link>
                </li>
                <li className="nav-item">
                  <button className="nav-link nav-button" onClick={handleLogout}>
                    Logout
                  </button>
                </li>
              </>
            ) : (
              <>
                <li className="nav-item">
                  <Link to="/login" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                    Login
                  </Link>
                </li>
                <li className="nav-item">
                  <Link to="/register" className="nav-link nav-link-primary" onClick={() => setIsMenuOpen(false)}>
                    Sign Up
                  </Link>
                </li>
              </>
            )}
          </ul>
        </div>
      </nav>
    </header>
  );
};

export default Header;`;
  }

  // Update Header CSS to include auth button styles
  const headerCssIndex = files.findIndex(f => f.path === 'client/src/components/Header.css');
  if (headerCssIndex !== -1) {
    files[headerCssIndex].content += `

.nav-button {
  background: none;
  border: none;
  cursor: pointer;
  font-weight: 500;
  font-size: inherit;
  font-family: inherit;
}

.nav-link-primary {
  background-color: #007bff;
  color: white !important;
  padding: 0.5rem 1rem;
  border-radius: 5px;
  transition: background-color 0.3s ease;
}

.nav-link-primary:hover {
  background-color: #0056b3;
  color: white !important;
}

@media (max-width: 768px) {
  .nav-link-primary {
    background-color: transparent;
    color: #333 !important;
    padding: 0;
    border-radius: 0;
  }
  
  .nav-link-primary:hover {
    background-color: transparent;
    color: #007bff !important;
  }
}`;
  }

  // Root README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A full-stack web application built with React frontend and Express.js backend.

## Features

### Frontend (React)
- ⚛️ React 18 with Hooks
- 🚀 React Router for navigation
- 🔐 JWT authentication with context
- 📱 Fully responsive design
- 🎨 Modern CSS with animations

### Backend (Express.js)
- 🚀 Express.js REST API
- 🔐 JWT authentication and authorization
- 👤 User management with roles
- 🛡️ Security middleware (Helmet, CORS, Rate limiting)
- 📊 MongoDB with Mongoose ODM
- ✅ Input validation

## Getting Started

### Prerequisites

- Node.js (version 14 or higher)
- MongoDB (local or cloud instance)
- npm or yarn

### Installation

1. Clone the repository:
   \`\`\`bash
   git clone <repository-url>
   cd ${projectName.toLowerCase().replace(/\s+/g, '-')}
   \`\`\`

2. Install dependencies for both client and server:
   \`\`\`bash
   npm run install-all
   \`\`\`

3. Set up environment variables:
   \`\`\`bash
   cd server
   cp .env.example .env
   \`\`\`
   
   Update the \`server/.env\` file with your configuration.

4. Start the development servers:
   \`\`\`bash
   npm run dev
   \`\`\`

   This will start both the React development server (port 3000) and the Express API server (port 3001).

5. Open [http://localhost:3000](http://localhost:3000) to view the application.

## Project Structure

\`\`\`
${projectName.toLowerCase().replace(/\s+/g, '-')}/
├── client/                 # React frontend
│   ├── public/
│   ├── src/
│   │   ├── components/     # Reusable components
│   │   ├── pages/          # Page components
│   │   ├── context/        # React context (Auth)
│   │   └── ...
│   └── package.json
├── server/                 # Express backend
│   ├── src/
│   │   ├── config/         # Database configuration
│   │   ├── middleware/     # Express middleware
│   │   ├── models/         # Mongoose models
│   │   ├── routes/         # API routes
│   │   └── ...
│   ├── tests/              # API tests
│   └── package.json
├── package.json            # Root package.json
└── README.md
\`\`\`

## Available Scripts

### Root Level
- \`npm run dev\` - Start both client and server in development mode
- \`npm run install-all\` - Install dependencies for both client and server
- \`npm run build\` - Build the React app for production
- \`npm start\` - Start the production server
- \`npm test\` - Run tests for both client and server

### Client (React)
- \`npm start\` - Start development server
- \`npm run build\` - Build for production
- \`npm test\` - Run tests

### Server (Express)
- \`npm run dev\` - Start with nodemon
- \`npm start\` - Start production server
- \`npm test\` - Run API tests

## API Endpoints

### Authentication
- \`POST /api/auth/register\` - Register a new user
- \`POST /api/auth/login\` - Login user
- \`GET /api/auth/me\` - Get current user

### Users
- \`GET /api/users\` - Get all users (admin only)
- \`GET /api/users/:id\` - Get user by ID
- \`PUT /api/users/:id\` - Update user
- \`DELETE /api/users/:id\` - Delete user (admin only)

### General
- \`GET /api/status\` - API status
- \`GET /health\` - Health check

## Authentication Flow

1. User registers or logs in through the React frontend
2. Backend validates credentials and returns JWT token
3. Frontend stores token and includes it in subsequent API requests
4. Protected routes require valid JWT token
5. User context provides authentication state throughout the app

## Deployment

### Development
The app is configured to run the React dev server on port 3000 and the API server on port 3001. The React app proxies API requests to the backend.

### Production

#### Option 1: Heroku
1. Set up Heroku app
2. Configure environment variables
3. The \`heroku-postbuild\` script will build the React app
4. Deploy with \`git push heroku main\`

#### Option 2: Traditional Server
1. Build the React app: \`npm run build\`
2. Serve the built files from the Express server
3. Configure Express to serve static files from \`client/build\`

### Environment Variables

Server environment variables (in \`server/.env\`):
\`\`\`
PORT=3001
NODE_ENV=production
MONGODB_URI=your-mongodb-connection-string
JWT_SECRET=your-super-secret-jwt-key
FRONTEND_URL=http://localhost:3000
\`\`\`

## Security Features

- JWT-based authentication
- Password hashing with bcrypt
- CORS protection
- Rate limiting
- Input validation
- Helmet security headers
- Protected routes on both frontend and backend

## Contributing

1. Fork the repository
2. Create a feature branch
3. Make your changes
4. Add tests for new functionality
5. Ensure all tests pass
6. Submit a pull request

## License

This project is licensed under the MIT License.
`
  });

  // .gitignore for root
  files.push({
    path: '.gitignore',
    content: `# Dependencies
node_modules/
client/node_modules/
server/node_modules/

# Production builds
client/build/
server/dist/

# Environment variables
.env
.env.local
.env.development.local
.env.test.local
.env.production.local
server/.env
client/.env

# Logs
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# Runtime data
pids
*.pid
*.seed
*.pid.lock

# Coverage directory used by tools like istanbul
coverage/
.nyc_output

# Dependency directories
jspm_packages/

# Optional npm cache directory
.npm

# Optional REPL history
.node_repl_history

# Output of 'npm pack'
*.tgz

# Yarn Integrity file
.yarn-integrity

# IDE files
.vscode/
.idea/
*.swp
*.swo

# OS generated files
.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
ehthumbs.db
Thumbs.db`
  });

  return files;
}

export default {
  name: 'Full-Stack App',
  description: 'A complete full-stack application with React frontend and Express backend',
  generate
};
