// codestral/src/templates/next-app.js
/**
 * Next.js App Template
 * Creates a modern Next.js application with App Router and TypeScript
 */

function generate(projectName, options = {}) {
  const files = [];

  // Package.json
  files.push({
    path: 'package.json',
    content: `{
  "name": "${projectName.toLowerCase().replace(/\s+/g, '-')}",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "type-check": "tsc --noEmit"
  },
  "dependencies": {
    "next": "14.0.0",
    "react": "^18",
    "react-dom": "^18"
  },
  "devDependencies": {
    "typescript": "^5",
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "eslint": "^8",
    "eslint-config-next": "14.0.0"
  }
}`
  });

  // Next.js config
  files.push({
    path: 'next.config.js',
    content: `/** @type {import('next').NextConfig} */
const nextConfig = {
  experimental: {
    appDir: true,
  },
}

module.exports = nextConfig`
  });

  // TypeScript config
  files.push({
    path: 'tsconfig.json',
    content: `{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "es6"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}`
  });

  // Root layout
  files.push({
    path: 'src/app/layout.tsx',
    content: `import type { Metadata } from 'next'
import { Inter } from 'next/font/google'
import './globals.css'
import Header from '@/components/Header'
import Footer from '@/components/Footer'

const inter = Inter({ subsets: ['latin'] })

export const metadata: Metadata = {
  title: '${projectName}',
  description: 'A modern Next.js application',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <div className="app">
          <Header />
          <main className="main-content">
            {children}
          </main>
          <Footer />
        </div>
      </body>
    </html>
  )
}`
  });

  // Global CSS
  files.push({
    path: 'src/app/globals.css',
    content: `* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
  font-family: inherit;
  line-height: 1.6;
  color: #333;
}

.app {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

.main-content {
  flex: 1;
  padding-top: 80px; /* Account for fixed header */
}

/* Container */
.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 0 20px;
}

/* Buttons */
.btn {
  display: inline-block;
  padding: 12px 24px;
  background-color: #0070f3;
  color: white;
  text-decoration: none;
  border-radius: 5px;
  border: none;
  cursor: pointer;
  font-size: 16px;
  font-weight: 500;
  transition: all 0.2s ease;
}

.btn:hover {
  background-color: #0051cc;
  transform: translateY(-1px);
}

.btn-secondary {
  background-color: #6c757d;
}

.btn-secondary:hover {
  background-color: #545b62;
}

/* Form Styles */
.form-group {
  margin-bottom: 1rem;
}

.form-group label {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: 600;
  color: #333;
}

.form-group input,
.form-group textarea {
  width: 100%;
  padding: 12px;
  border: 2px solid #e1e5e9;
  border-radius: 6px;
  font-size: 16px;
  transition: border-color 0.2s ease;
}

.form-group input:focus,
.form-group textarea:focus {
  outline: none;
  border-color: #0070f3;
}

/* Cards */
.card {
  background: white;
  padding: 2rem;
  border-radius: 12px;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1);
  transition: transform 0.2s ease, box-shadow 0.2s ease;
}

.card:hover {
  transform: translateY(-2px);
  box-shadow: 0 10px 25px -3px rgba(0, 0, 0, 0.1);
}

/* Grid */
.grid {
  display: grid;
  gap: 2rem;
}

.grid-cols-1 {
  grid-template-columns: repeat(1, minmax(0, 1fr));
}

.grid-cols-2 {
  grid-template-columns: repeat(2, minmax(0, 1fr));
}

.grid-cols-3 {
  grid-template-columns: repeat(3, minmax(0, 1fr));
}

@media (max-width: 768px) {
  .grid-cols-2,
  .grid-cols-3 {
    grid-template-columns: repeat(1, minmax(0, 1fr));
  }
  
  .container {
    padding: 0 15px;
  }
  
  .main-content {
    padding-top: 70px;
  }
}

/* Animations */
@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

.animate-fade-in-up {
  animation: fadeInUp 0.6s ease forwards;
}

/* Loading */
.loading {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 2rem;
}

.spinner {
  width: 40px;
  height: 40px;
  border: 4px solid #f3f3f3;
  border-top: 4px solid #0070f3;
  border-radius: 50%;
  animation: spin 1s linear infinite;
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}`
  });

  // Home page
  files.push({
    path: 'src/app/page.tsx',
    content: `import Hero from '@/components/Hero'
import Features from '@/components/Features'

export default function Home() {
  return (
    <>
      <Hero />
      <Features />
    </>
  )
}`
  });

  // About page
  files.push({
    path: 'src/app/about/page.tsx',
    content: `import type { Metadata } from 'next'

export const metadata: Metadata = {
  title: 'About - ${projectName}',
  description: 'Learn more about ${projectName}',
}

export default function About() {
  const stats = [
    { value: '100%', label: 'TypeScript' },
    { value: '14+', label: 'Next.js Version' },
    { value: '∞', label: 'Possibilities' }
  ]

  return (
    <div className="about">
      <div className="container">
        <section className="about-hero">
          <h1>About ${projectName}</h1>
          <p>Learn more about our mission and values</p>
        </section>

        <section className="about-content">
          <div className="about-text">
            <h2>Our Story</h2>
            <p>
              ${projectName} was created to demonstrate modern Next.js development practices
              and provide a solid foundation for building scalable web applications.
            </p>
            <p>
              This template includes the App Router, TypeScript, and responsive design
              to help you get started quickly with your next project.
            </p>
          </div>
          
          <div className="about-stats">
            {stats.map((stat, index) => (
              <div key={index} className="stat">
                <h3>{stat.value}</h3>
                <p>{stat.label}</p>
              </div>
            ))}
          </div>
        </section>
      </div>
    </div>
  )
}

// Styles would typically be in a separate CSS module or styled-components
// For simplicity, including them here as a style tag would be added
`
  });

  // Contact page
  files.push({
    path: 'src/app/contact/page.tsx',
    content: `'use client'

import { useState } from 'react'
import type { Metadata } from 'next'

// Note: metadata export doesn't work in client components
// You would typically handle this differently in a real app

interface FormData {
  name: string
  email: string
  message: string
}

export default function Contact() {
  const [formData, setFormData] = useState<FormData>({
    name: '',
    email: '',
    message: ''
  })
  const [isSubmitting, setIsSubmitting] = useState(false)

  const contactDetails = [
    {
      icon: '📧',
      title: 'Email',
      content: 'hello@${projectName.toLowerCase().replace(/\s+/g, '')}.com'
    },
    {
      icon: '📱',
      title: 'Phone',
      content: '+1 (555) 123-4567'
    },
    {
      icon: '📍',
      title: 'Address',
      content: '123 Main Street\\nCity, State 12345'
    }
  ]

  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value
    })
  }

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault()
    
    if (!formData.name || !formData.email || !formData.message) {
      alert('Please fill in all fields')
      return
    }

    setIsSubmitting(true)

    // Simulate form submission
    setTimeout(() => {
      alert('Thank you for your message! We will get back to you soon.')
      setFormData({ name: '', email: '', message: '' })
      setIsSubmitting(false)
    }, 1000)
  }

  return (
    <div className="contact">
      <div className="container">
        <section className="contact-hero">
          <h1>Contact Us</h1>
          <p>Get in touch with our team</p>
        </section>

        <section className="contact-content">
          <div className="contact-info">
            <h2>Get In Touch</h2>
            <p>
              We'd love to hear from you. Send us a message and we'll respond
              as soon as possible.
            </p>
            
            <div className="contact-details">
              {contactDetails.map((item, index) => (
                <div key={index} className="contact-item">
                  <h3>{item.icon} {item.title}</h3>
                  <p>{item.content}</p>
                </div>
              ))}
            </div>
          </div>

          <form className="contact-form" onSubmit={handleSubmit}>
            <div className="form-group">
              <label htmlFor="name">Name</label>
              <input
                type="text"
                id="name"
                name="name"
                value={formData.name}
                onChange={handleChange}
                required
              />
            </div>

            <div className="form-group">
              <label htmlFor="email">Email</label>
              <input
                type="email"
                id="email"
                name="email"
                value={formData.email}
                onChange={handleChange}
                required
              />
            </div>

            <div className="form-group">
              <label htmlFor="message">Message</label>
              <textarea
                id="message"
                name="message"
                rows={5}
                value={formData.message}
                onChange={handleChange}
                required
              />
            </div>

            <button type="submit" className="btn" disabled={isSubmitting}>
              {isSubmitting ? 'Sending...' : 'Send Message'}
            </button>
          </form>
        </section>
      </div>
    </div>
  )
}`
  });

  // Header component
  files.push({
    path: 'src/components/Header.tsx',
    content: `'use client'

import { useState } from 'react'
import Link from 'next/link'
import { usePathname } from 'next/navigation'

export default function Header() {
  const [isMenuOpen, setIsMenuOpen] = useState(false)
  const pathname = usePathname()

  const toggleMenu = () => {
    setIsMenuOpen(!isMenuOpen)
  }

  const closeMenu = () => {
    setIsMenuOpen(false)
  }

  const isActive = (path: string) => pathname === path

  return (
    <header className="header">
      <nav className="navbar">
        <div className="container">
          <Link href="/" className="nav-brand">
            ${projectName}
          </Link>
          
          <button 
            className={\`nav-toggle \${isMenuOpen ? 'active' : ''}\`}
            onClick={toggleMenu}
            aria-label="Toggle navigation"
          >
            <span></span>
            <span></span>
            <span></span>
          </button>

          <ul className={\`nav-menu \${isMenuOpen ? 'active' : ''}\`}>
            <li className="nav-item">
              <Link 
                href="/" 
                className={\`nav-link \${isActive('/') ? 'active' : ''}\`}
                onClick={closeMenu}
              >
                Home
              </Link>
            </li>
            <li className="nav-item">
              <Link 
                href="/about" 
                className={\`nav-link \${isActive('/about') ? 'active' : ''}\`}
                onClick={closeMenu}
              >
                About
              </Link>
            </li>
            <li className="nav-item">
              <Link 
                href="/contact" 
                className={\`nav-link \${isActive('/contact') ? 'active' : ''}\`}
                onClick={closeMenu}
              >
                Contact
              </Link>
            </li>
          </ul>
        </div>
      </nav>

      <style jsx>{\`
        .header {
          position: fixed;
          top: 0;
          left: 0;
          right: 0;
          z-index: 1000;
          background-color: rgba(255, 255, 255, 0.95);
          backdrop-filter: blur(10px);
          box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
        }

        .navbar {
          padding: 1rem 0;
        }

        .navbar .container {
          display: flex;
          justify-content: space-between;
          align-items: center;
        }

        .nav-brand {
          font-size: 1.5rem;
          font-weight: bold;
          color: #0070f3;
          text-decoration: none;
        }

        .nav-menu {
          display: flex;
          list-style: none;
          gap: 2rem;
          margin: 0;
        }

        .nav-link {
          color: #333;
          text-decoration: none;
          font-weight: 500;
          transition: color 0.3s ease;
          position: relative;
        }

        .nav-link:hover,
        .nav-link.active {
          color: #0070f3;
        }

        .nav-link.active::after {
          content: '';
          position: absolute;
          bottom: -5px;
          left: 0;
          right: 0;
          height: 2px;
          background-color: #0070f3;
        }

        .nav-toggle {
          display: none;
          flex-direction: column;
          background: none;
          border: none;
          cursor: pointer;
          padding: 0;
          width: 30px;
          height: 30px;
          justify-content: space-around;
        }

        .nav-toggle span {
          display: block;
          height: 3px;
          width: 100%;
          background-color: #333;
          border-radius: 3px;
          transition: all 0.3s ease;
        }

        .nav-toggle.active span:nth-child(1) {
          transform: rotate(45deg) translate(5px, 5px);
        }

        .nav-toggle.active span:nth-child(2) {
          opacity: 0;
        }

        .nav-toggle.active span:nth-child(3) {
          transform: rotate(-45deg) translate(7px, -6px);
        }

        @media (max-width: 768px) {
          .nav-toggle {
            display: flex;
          }

          .nav-menu {
            position: fixed;
            left: -100%;
            top: 70px;
            flex-direction: column;
            background-color: rgba(255, 255, 255, 0.95);
            backdrop-filter: blur(10px);
            width: 100%;
            text-align: center;
            transition: 0.3s;
            box-shadow: 0 10px 27px rgba(0, 0, 0, 0.05);
            padding: 2rem 0;
            gap: 1rem;
          }

          .nav-menu.active {
            left: 0;
          }
        }
      \`}</style>
    </header>
  )
}`
  });

  // Footer component
  files.push({
    path: 'src/components/Footer.tsx',
    content: `export default function Footer() {
  return (
    <footer className="footer">
      <div className="container">
        <p>&copy; 2024 ${projectName}. All rights reserved.</p>
      </div>

      <style jsx>{\`
        .footer {
          background-color: #f8f9fa;
          padding: 2rem 0;
          margin-top: auto;
          border-top: 1px solid #e9ecef;
        }

        .footer .container {
          text-align: center;
        }

        .footer p {
          margin: 0;
          color: #6c757d;
        }
      \`}</style>
    </footer>
  )
}`
  });

  // Hero component
  files.push({
    path: 'src/components/Hero.tsx',
    content: `export default function Hero() {
  return (
    <section className="hero">
      <div className="container">
        <div className="hero-content">
          <h1 className="hero-title">Welcome to ${projectName}</h1>
          <p className="hero-description">
            A modern Next.js application built with the App Router and TypeScript
          </p>
          <button className="btn">Get Started</button>
        </div>
      </div>

      <style jsx>{\`
        .hero {
          background: linear-gradient(135deg, #0070f3 0%, #00d4ff 100%);
          color: white;
          padding: 120px 0 80px;
          text-align: center;
        }

        .hero-content {
          max-width: 600px;
          margin: 0 auto;
        }

        .hero-title {
          font-size: 3rem;
          margin-bottom: 1rem;
          animation: fadeInUp 1s ease;
        }

        .hero-description {
          font-size: 1.2rem;
          margin-bottom: 2rem;
          animation: fadeInUp 1s ease 0.2s both;
        }

        .btn {
          animation: fadeInUp 1s ease 0.4s both;
        }

        @keyframes fadeInUp {
          from {
            opacity: 0;
            transform: translateY(30px);
          }
          to {
            opacity: 1;
            transform: translateY(0);
          }
        }

        @media (max-width: 768px) {
          .hero-title {
            font-size: 2rem;
          }
          
          .hero-description {
            font-size: 1rem;
          }
        }
      \`}</style>
    </section>
  )
}`
  });

  // Features component
  files.push({
    path: 'src/components/Features.tsx',
    content: `export default function Features() {
  const features = [
    {
      icon: '⚡',
      title: 'Fast',
      description: 'Built with Next.js 14 and the App Router for optimal performance'
    },
    {
      icon: '📱',
      title: 'Responsive',
      description: 'Works perfectly on all devices and screen sizes'
    },
    {
      icon: '🎨',
      title: 'Modern',
      description: 'Clean, modern design with TypeScript and CSS-in-JS'
    }
  ]

  return (
    <section className="features">
      <div className="container">
        <h2>Features</h2>
        <div className="features-grid">
          {features.map((feature, index) => (
            <div key={index} className="feature-card">
              <h3>{feature.icon} {feature.title}</h3>
              <p>{feature.description}</p>
            </div>
          ))}
        </div>
      </div>

      <style jsx>{\`
        .features {
          padding: 80px 0;
          background-color: #f8f9fa;
        }

        .features h2 {
          text-align: center;
          margin-bottom: 3rem;
          font-size: 2.5rem;
          color: #333;
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
          gap: 2rem;
        }

        .feature-card {
          background: white;
          padding: 2rem;
          border-radius: 12px;
          box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1);
          text-align: center;
          transition: transform 0.3s ease, box-shadow 0.3s ease;
        }

        .feature-card:hover {
          transform: translateY(-5px);
          box-shadow: 0 10px 25px -3px rgba(0, 0, 0, 0.1);
        }

        .feature-card h3 {
          font-size: 1.5rem;
          margin-bottom: 1rem;
          color: #333;
        }

        .feature-card p {
          color: #666;
          line-height: 1.6;
        }

        @media (max-width: 768px) {
          .features-grid {
            grid-template-columns: 1fr;
          }
        }
      \`}</style>
    </section>
  )
}`
  });

  // ESLint config
  files.push({
    path: '.eslintrc.json',
    content: `{
  "extends": "next/core-web-vitals"
}`
  });

  // .gitignore
  files.push({
    path: '.gitignore',
    content: `# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.

# dependencies
/node_modules
/.pnp
.pnp.js

# testing
/coverage

# next.js
/.next/
/out/

# production
/build

# misc
.DS_Store
*.pem

# debug
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# local env files
.env*.local

# vercel
.vercel

# typescript
*.tsbuildinfo
next-env.d.ts`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A modern Next.js application built with the App Router and TypeScript.

## Features

- ⚡ Next.js 14 with App Router
- 🔷 TypeScript for type safety
- 📱 Fully responsive design
- 🎨 CSS-in-JS with styled-jsx
- 🚀 Optimized for performance
- 📦 Built-in SEO optimization

## Getting Started

### Prerequisites

- Node.js 18.17 or later
- npm, yarn, or pnpm

### Installation

1. Install dependencies:
   \`\`\`bash
   npm install
   # or
   yarn install
   # or
   pnpm install
   \`\`\`

2. Run the development server:
   \`\`\`bash
   npm run dev
   # or
   yarn dev
   # or
   pnpm dev
   \`\`\`

3. Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

## Available Scripts

- \`npm run dev\` - Runs the app in development mode
- \`npm run build\` - Builds the app for production
- \`npm run start\` - Runs the built app in production mode
- \`npm run lint\` - Runs ESLint to check for code issues
- \`npm run type-check\` - Runs TypeScript compiler to check for type errors

## Project Structure

\`\`\`
src/
├── app/                    # App Router pages and layouts
│   ├── about/
│   │   └── page.tsx       # About page
│   ├── contact/
│   │   └── page.tsx       # Contact page
│   ├── globals.css        # Global styles
│   ├── layout.tsx         # Root layout
│   └── page.tsx           # Home page
└── components/            # Reusable components
    ├── Header.tsx
    ├── Footer.tsx
    ├── Hero.tsx
    └── Features.tsx
\`\`\`

## App Router

This project uses Next.js 14's App Router, which provides:

- **File-based routing**: Pages are created by adding files to the \`app\` directory
- **Layouts**: Shared UI components that wrap pages
- **Server Components**: Components that render on the server by default
- **Streaming**: Progressive rendering for better user experience

## Styling

This project uses CSS-in-JS with styled-jsx for component-level styling:

- Scoped styles that don't leak to other components
- Dynamic styling based on props and state
- No build step required for CSS
- Full CSS feature support

For global styles, see \`src/app/globals.css\`.

## TypeScript

The project is fully typed with TypeScript:

- Type safety for props, state, and API responses
- Better IDE support with autocomplete and error checking
- Easier refactoring and maintenance
- Run \`npm run type-check\` to check for type errors

## SEO and Metadata

Next.js App Router provides built-in SEO optimization:

- **Metadata API**: Define page titles, descriptions, and Open Graph tags
- **Static and Dynamic Metadata**: Generate metadata at build time or runtime
- **Automatic Sitemap**: Generate sitemaps automatically

Example metadata in a page:

\`\`\`tsx
import type { Metadata } from 'next'

export const metadata: Metadata = {
  title: 'About - ${projectName}',
  description: 'Learn more about ${projectName}',
}
\`\`\`

## Performance

Next.js provides many performance optimizations out of the box:

- **Automatic Code Splitting**: Only load the code needed for each page
- **Image Optimization**: Automatic image optimization with the \`next/image\` component
- **Font Optimization**: Automatic font optimization with \`next/font\`
- **Static Generation**: Pre-render pages at build time when possible

## Deployment

### Vercel (Recommended)

The easiest way to deploy your Next.js app is to use [Vercel](https://vercel.com):

1. Push your code to GitHub
2. Connect your repository to Vercel
3. Deploy automatically on every push

### Other Platforms

You can also deploy to:

- **Netlify**: Build command: \`npm run build\`, Publish directory: \`out\`
- **AWS Amplify**: Connect your GitHub repository
- **Docker**: Use the included Dockerfile for containerized deployment

### Self-Hosting

To deploy on your own server:

1. Build the application: \`npm run build\`
2. Start the production server: \`npm run start\`

## Learn More

To learn more about Next.js and the technologies used:

- [Next.js Documentation](https://nextjs.org/docs) - learn about Next.js features and API
- [App Router Documentation](https://nextjs.org/docs/app) - learn about the App Router
- [TypeScript Documentation](https://www.typescriptlang.org/docs/) - learn about TypeScript
- [React Documentation](https://react.dev/) - learn about React

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'Next.js App',
  description: 'A modern Next.js application with App Router and TypeScript',
  generate
};
