// codestral/src/templates/react-app.js
/**
 * React App Template
 * Creates a modern React application with routing and components
 */

function generate(projectName, options = {}) {
  const files = [];

  // Package.json
  files.push({
    path: 'package.json',
    content: `{
  "name": "${projectName.toLowerCase().replace(/\s+/g, '-')}",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/user-event": "^13.5.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.3.0",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}`
  });

  // Public files
  files.push({
    path: 'public/index.html',
    content: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#000000" />
    <meta
      name="description"
      content="${projectName} - A React application"
    />
    <link rel="apple-touch-icon" href="%PUBLIC_URL%/logo192.png" />
    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
    <title>${projectName}</title>
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id="root"></div>
  </body>
</html>`
  });

  files.push({
    path: 'public/manifest.json',
    content: `{
  "short_name": "${projectName}",
  "name": "${projectName} React App",
  "icons": [
    {
      "src": "favicon.ico",
      "sizes": "64x64 32x32 24x24 16x16",
      "type": "image/x-icon"
    }
  ],
  "start_url": ".",
  "display": "standalone",
  "theme_color": "#000000",
  "background_color": "#ffffff"
}`
  });

  // Main App component
  files.push({
    path: 'src/App.js',
    content: `import React from 'react';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import Header from './components/Header';
import Footer from './components/Footer';
import Home from './pages/Home';
import About from './pages/About';
import Contact from './pages/Contact';
import './App.css';

function App() {
  return (
    <Router>
      <div className="App">
        <Header />
        <main className="main-content">
          <Routes>
            <Route path="/" element={<Home />} />
            <Route path="/about" element={<About />} />
            <Route path="/contact" element={<Contact />} />
          </Routes>
        </main>
        <Footer />
      </div>
    </Router>
  );
}

export default App;`
  });

  // App CSS
  files.push({
    path: 'src/App.css',
    content: `.App {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

.main-content {
  flex: 1;
  padding-top: 80px; /* Account for fixed header */
}

/* Global Styles */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  line-height: 1.6;
  color: #333;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

/* Container */
.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 0 20px;
}

/* Buttons */
.btn {
  display: inline-block;
  padding: 12px 24px;
  background-color: #007bff;
  color: white;
  text-decoration: none;
  border-radius: 5px;
  border: none;
  cursor: pointer;
  font-size: 16px;
  transition: background-color 0.3s ease;
}

.btn:hover {
  background-color: #0056b3;
}

.btn-secondary {
  background-color: #6c757d;
}

.btn-secondary:hover {
  background-color: #545b62;
}

/* Form Styles */
.form-group {
  margin-bottom: 1rem;
}

.form-group label {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: bold;
}

.form-group input,
.form-group textarea {
  width: 100%;
  padding: 10px;
  border: 1px solid #ddd;
  border-radius: 4px;
  font-size: 16px;
}

.form-group input:focus,
.form-group textarea:focus {
  outline: none;
  border-color: #007bff;
  box-shadow: 0 0 0 2px rgba(0, 123, 255, 0.25);
}

/* Responsive */
@media (max-width: 768px) {
  .container {
    padding: 0 15px;
  }
  
  .main-content {
    padding-top: 70px;
  }
}`
  });

  // Index.js
  files.push({
    path: 'src/index.js',
    content: `import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();`
  });

  // Index CSS
  files.push({
    path: 'src/index.css',
    content: `body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}`
  });

  // Header component
  files.push({
    path: 'src/components/Header.js',
    content: `import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import './Header.css';

const Header = () => {
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  const toggleMenu = () => {
    setIsMenuOpen(!isMenuOpen);
  };

  return (
    <header className="header">
      <nav className="navbar">
        <div className="container">
          <Link to="/" className="nav-brand">
            ${projectName}
          </Link>
          
          <button 
            className={\`nav-toggle \${isMenuOpen ? 'active' : ''}\`}
            onClick={toggleMenu}
            aria-label="Toggle navigation"
          >
            <span></span>
            <span></span>
            <span></span>
          </button>

          <ul className={\`nav-menu \${isMenuOpen ? 'active' : ''}\`}>
            <li className="nav-item">
              <Link to="/" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                Home
              </Link>
            </li>
            <li className="nav-item">
              <Link to="/about" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                About
              </Link>
            </li>
            <li className="nav-item">
              <Link to="/contact" className="nav-link" onClick={() => setIsMenuOpen(false)}>
                Contact
              </Link>
            </li>
          </ul>
        </div>
      </nav>
    </header>
  );
};

export default Header;`
  });

  // Header CSS
  files.push({
    path: 'src/components/Header.css',
    content: `.header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1000;
  background-color: #fff;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
}

.navbar {
  padding: 1rem 0;
}

.navbar .container {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.nav-brand {
  font-size: 1.5rem;
  font-weight: bold;
  color: #007bff;
  text-decoration: none;
}

.nav-menu {
  display: flex;
  list-style: none;
  gap: 2rem;
  margin: 0;
}

.nav-link {
  color: #333;
  text-decoration: none;
  font-weight: 500;
  transition: color 0.3s ease;
}

.nav-link:hover {
  color: #007bff;
}

.nav-toggle {
  display: none;
  flex-direction: column;
  background: none;
  border: none;
  cursor: pointer;
  padding: 0;
  width: 30px;
  height: 30px;
  justify-content: space-around;
}

.nav-toggle span {
  display: block;
  height: 3px;
  width: 100%;
  background-color: #333;
  border-radius: 3px;
  transition: all 0.3s ease;
}

.nav-toggle.active span:nth-child(1) {
  transform: rotate(45deg) translate(5px, 5px);
}

.nav-toggle.active span:nth-child(2) {
  opacity: 0;
}

.nav-toggle.active span:nth-child(3) {
  transform: rotate(-45deg) translate(7px, -6px);
}

@media (max-width: 768px) {
  .nav-toggle {
    display: flex;
  }

  .nav-menu {
    position: fixed;
    left: -100%;
    top: 70px;
    flex-direction: column;
    background-color: #fff;
    width: 100%;
    text-align: center;
    transition: 0.3s;
    box-shadow: 0 10px 27px rgba(0, 0, 0, 0.05);
    padding: 2rem 0;
    gap: 1rem;
  }

  .nav-menu.active {
    left: 0;
  }
}`
  });

  // Footer component
  files.push({
    path: 'src/components/Footer.js',
    content: `import React from 'react';
import './Footer.css';

const Footer = () => {
  return (
    <footer className="footer">
      <div className="container">
        <p>&copy; 2024 ${projectName}. All rights reserved.</p>
      </div>
    </footer>
  );
};

export default Footer;`
  });

  // Footer CSS
  files.push({
    path: 'src/components/Footer.css',
    content: `.footer {
  background-color: #f8f9fa;
  padding: 2rem 0;
  margin-top: auto;
  border-top: 1px solid #e9ecef;
}

.footer .container {
  text-align: center;
}

.footer p {
  margin: 0;
  color: #6c757d;
}`
  });

  // Home page
  files.push({
    path: 'src/pages/Home.js',
    content: `import React from 'react';
import './Home.css';

const Home = () => {
  return (
    <div className="home">
      <section className="hero">
        <div className="container">
          <div className="hero-content">
            <h1 className="hero-title">Welcome to ${projectName}</h1>
            <p className="hero-description">
              A modern React application built with the latest technologies
            </p>
            <button className="btn btn-primary">Get Started</button>
          </div>
        </div>
      </section>

      <section className="features">
        <div className="container">
          <h2>Features</h2>
          <div className="features-grid">
            <div className="feature-card">
              <h3>⚡ Fast</h3>
              <p>Built with React 18 for optimal performance</p>
            </div>
            <div className="feature-card">
              <h3>📱 Responsive</h3>
              <p>Works perfectly on all devices and screen sizes</p>
            </div>
            <div className="feature-card">
              <h3>🎨 Modern</h3>
              <p>Clean, modern design with smooth animations</p>
            </div>
          </div>
        </div>
      </section>
    </div>
  );
};

export default Home;`
  });

  // Home CSS
  files.push({
    path: 'src/pages/Home.css',
    content: `.hero {
  background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
  color: white;
  padding: 100px 0;
  text-align: center;
}

.hero-content {
  max-width: 600px;
  margin: 0 auto;
}

.hero-title {
  font-size: 3rem;
  margin-bottom: 1rem;
  animation: fadeInUp 1s ease;
}

.hero-description {
  font-size: 1.2rem;
  margin-bottom: 2rem;
  animation: fadeInUp 1s ease 0.2s both;
}

.features {
  padding: 80px 0;
  background-color: #f8f9fa;
}

.features h2 {
  text-align: center;
  margin-bottom: 3rem;
  font-size: 2.5rem;
  color: #333;
}

.features-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  gap: 2rem;
}

.feature-card {
  background: white;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
  text-align: center;
  transition: transform 0.3s ease;
}

.feature-card:hover {
  transform: translateY(-5px);
}

.feature-card h3 {
  font-size: 1.5rem;
  margin-bottom: 1rem;
  color: #333;
}

@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

@media (max-width: 768px) {
  .hero-title {
    font-size: 2rem;
  }
  
  .hero-description {
    font-size: 1rem;
  }
  
  .features-grid {
    grid-template-columns: 1fr;
  }
}`
  });

  // About page
  files.push({
    path: 'src/pages/About.js',
    content: `import React from 'react';
import './About.css';

const About = () => {
  return (
    <div className="about">
      <div className="container">
        <section className="about-hero">
          <h1>About ${projectName}</h1>
          <p>Learn more about our mission and values</p>
        </section>

        <section className="about-content">
          <div className="about-text">
            <h2>Our Story</h2>
            <p>
              ${projectName} was created to demonstrate modern React development practices
              and provide a solid foundation for building scalable web applications.
            </p>
            <p>
              This template includes routing, responsive design, and component-based
              architecture to help you get started quickly with your next project.
            </p>
          </div>
          
          <div className="about-stats">
            <div className="stat">
              <h3>100%</h3>
              <p>React Hooks</p>
            </div>
            <div className="stat">
              <h3>0</h3>
              <p>Class Components</p>
            </div>
            <div className="stat">
              <h3>∞</h3>
              <p>Possibilities</p>
            </div>
          </div>
        </section>
      </div>
    </div>
  );
};

export default About;`
  });

  // About CSS
  files.push({
    path: 'src/pages/About.css',
    content: `.about {
  padding: 2rem 0;
}

.about-hero {
  text-align: center;
  padding: 3rem 0;
  background-color: #f8f9fa;
  margin-bottom: 3rem;
}

.about-hero h1 {
  font-size: 2.5rem;
  margin-bottom: 1rem;
  color: #333;
}

.about-hero p {
  font-size: 1.2rem;
  color: #6c757d;
}

.about-content {
  display: grid;
  grid-template-columns: 2fr 1fr;
  gap: 3rem;
  align-items: start;
}

.about-text h2 {
  margin-bottom: 1.5rem;
  color: #333;
}

.about-text p {
  margin-bottom: 1rem;
  line-height: 1.8;
  color: #555;
}

.about-stats {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.stat {
  text-align: center;
  padding: 1.5rem;
  background: white;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
}

.stat h3 {
  font-size: 2rem;
  color: #007bff;
  margin-bottom: 0.5rem;
}

.stat p {
  color: #6c757d;
  margin: 0;
}

@media (max-width: 768px) {
  .about-content {
    grid-template-columns: 1fr;
    gap: 2rem;
  }
  
  .about-stats {
    flex-direction: row;
    justify-content: space-around;
  }
  
  .stat {
    flex: 1;
    margin: 0 0.5rem;
  }
}`
  });

  // Contact page
  files.push({
    path: 'src/pages/Contact.js',
    content: `import React, { useState } from 'react';
import './Contact.css';

const Contact = () => {
  const [formData, setFormData] = useState({
    name: '',
    email: '',
    message: ''
  });

  const handleChange = (e) => {
    setFormData({
      ...formData,
      [e.target.name]: e.target.value
    });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    
    // Basic validation
    if (!formData.name || !formData.email || !formData.message) {
      alert('Please fill in all fields');
      return;
    }

    // Simulate form submission
    alert('Thank you for your message! We will get back to you soon.');
    setFormData({ name: '', email: '', message: '' });
  };

  return (
    <div className="contact">
      <div className="container">
        <section className="contact-hero">
          <h1>Contact Us</h1>
          <p>Get in touch with our team</p>
        </section>

        <section className="contact-content">
          <div className="contact-info">
            <h2>Get In Touch</h2>
            <p>
              We'd love to hear from you. Send us a message and we'll respond
              as soon as possible.
            </p>
            
            <div className="contact-details">
              <div className="contact-item">
                <h3>📧 Email</h3>
                <p>hello@${projectName.toLowerCase().replace(/\s+/g, '')}.com</p>
              </div>
              <div className="contact-item">
                <h3>📱 Phone</h3>
                <p>+1 (555) 123-4567</p>
              </div>
              <div className="contact-item">
                <h3>📍 Address</h3>
                <p>123 Main Street<br />City, State 12345</p>
              </div>
            </div>
          </div>

          <form className="contact-form" onSubmit={handleSubmit}>
            <div className="form-group">
              <label htmlFor="name">Name</label>
              <input
                type="text"
                id="name"
                name="name"
                value={formData.name}
                onChange={handleChange}
                required
              />
            </div>

            <div className="form-group">
              <label htmlFor="email">Email</label>
              <input
                type="email"
                id="email"
                name="email"
                value={formData.email}
                onChange={handleChange}
                required
              />
            </div>

            <div className="form-group">
              <label htmlFor="message">Message</label>
              <textarea
                id="message"
                name="message"
                rows="5"
                value={formData.message}
                onChange={handleChange}
                required
              ></textarea>
            </div>

            <button type="submit" className="btn">
              Send Message
            </button>
          </form>
        </section>
      </div>
    </div>
  );
};

export default Contact;`
  });

  // Contact CSS
  files.push({
    path: 'src/pages/Contact.css',
    content: `.contact {
  padding: 2rem 0;
}

.contact-hero {
  text-align: center;
  padding: 3rem 0;
  background-color: #f8f9fa;
  margin-bottom: 3rem;
}

.contact-hero h1 {
  font-size: 2.5rem;
  margin-bottom: 1rem;
  color: #333;
}

.contact-hero p {
  font-size: 1.2rem;
  color: #6c757d;
}

.contact-content {
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 3rem;
}

.contact-info h2 {
  margin-bottom: 1rem;
  color: #333;
}

.contact-info p {
  margin-bottom: 2rem;
  line-height: 1.8;
  color: #555;
}

.contact-details {
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}

.contact-item h3 {
  margin-bottom: 0.5rem;
  color: #333;
  font-size: 1.1rem;
}

.contact-item p {
  margin: 0;
  color: #6c757d;
}

.contact-form {
  background: white;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
}

@media (max-width: 768px) {
  .contact-content {
    grid-template-columns: 1fr;
    gap: 2rem;
  }
  
  .contact-form {
    padding: 1.5rem;
  }
}`
  });

  // reportWebVitals.js
  files.push({
    path: 'src/reportWebVitals.js',
    content: `const reportWebVitals = onPerfEntry => {
  if (onPerfEntry && onPerfEntry instanceof Function) {
    import('web-vitals').then(({ getCLS, getFID, getFCP, getLCP, getTTFB }) => {
      getCLS(onPerfEntry);
      getFID(onPerfEntry);
      getFCP(onPerfEntry);
      getLCP(onPerfEntry);
      getTTFB(onPerfEntry);
    });
  }
};

export default reportWebVitals;`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A modern React application built with Create React App.

## Features

- ⚛️ React 18 with Hooks
- 🚀 React Router for navigation
- 📱 Fully responsive design
- 🎨 Modern CSS with animations
- 🧪 Testing setup with Jest and React Testing Library
- 📦 Optimized build process

## Getting Started

### Prerequisites

- Node.js (version 14 or higher)
- npm or yarn

### Installation

1. Install dependencies:
   \`\`\`bash
   npm install
   \`\`\`

2. Start the development server:
   \`\`\`bash
   npm start
   \`\`\`

3. Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

## Available Scripts

- \`npm start\` - Runs the app in development mode
- \`npm test\` - Launches the test runner
- \`npm run build\` - Builds the app for production
- \`npm run eject\` - Ejects from Create React App (one-way operation)

## Project Structure

\`\`\`
src/
├── components/          # Reusable components
│   ├── Header.js
│   ├── Header.css
│   ├── Footer.js
│   └── Footer.css
├── pages/              # Page components
│   ├── Home.js
│   ├── Home.css
│   ├── About.js
│   ├── About.css
│   ├── Contact.js
│   └── Contact.css
├── App.js              # Main App component
├── App.css             # Global styles
├── index.js            # Entry point
└── index.css           # Base styles
\`\`\`

## Customization

### Adding New Pages

1. Create a new component in \`src/pages/\`
2. Add the route in \`src/App.js\`
3. Update the navigation in \`src/components/Header.js\`

### Styling

- Global styles are in \`src/App.css\` and \`src/index.css\`
- Component-specific styles are co-located with components
- The project uses CSS modules pattern for better organization

### Adding Dependencies

\`\`\`bash
npm install package-name
\`\`\`

## Deployment

### Build for Production

\`\`\`bash
npm run build
\`\`\`

This creates a \`build\` folder with optimized production files.

### Deploy to Netlify

1. Build the project: \`npm run build\`
2. Drag and drop the \`build\` folder to Netlify
3. Or connect your Git repository for automatic deployments

### Deploy to Vercel

\`\`\`bash
npm install -g vercel
vercel
\`\`\`

## Learn More

- [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started)
- [React documentation](https://reactjs.org/)
- [React Router documentation](https://reactrouter.com/)

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'React App',
  description: 'A modern React application with routing and components',
  generate
};
