// codestral/src/templates/vue-app.js
/**
 * Vue App Template
 * Creates a modern Vue.js application with Vue Router and Composition API
 */

function generate(projectName, options = {}) {
  const files = [];

  // Package.json
  files.push({
    path: 'package.json',
    content: `{
  "name": "${projectName.toLowerCase().replace(/\s+/g, '-')}",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "serve": "vue-cli-service serve",
    "build": "vue-cli-service build",
    "test:unit": "vue-cli-service test:unit",
    "lint": "vue-cli-service lint"
  },
  "dependencies": {
    "vue": "^3.3.4",
    "vue-router": "^4.2.4",
    "axios": "^1.4.0"
  },
  "devDependencies": {
    "@vue/cli-plugin-router": "~5.0.0",
    "@vue/cli-plugin-unit-jest": "~5.0.0",
    "@vue/cli-service": "~5.0.0",
    "@vue/test-utils": "^2.4.0",
    "jest": "^27.0.5"
  }
}`
  });

  // Main HTML file
  files.push({
    path: 'public/index.html',
    content: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="icon" href="<%= BASE_URL %>favicon.ico">
    <title>${projectName}</title>
  </head>
  <body>
    <noscript>
      <strong>We're sorry but ${projectName} doesn't work properly without JavaScript enabled. Please enable it to continue.</strong>
    </noscript>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>`
  });

  // Main App.vue
  files.push({
    path: 'src/App.vue',
    content: `<template>
  <div id="app">
    <NavBar />
    <main class="main-content">
      <router-view />
    </main>
    <AppFooter />
  </div>
</template>

<script>
import NavBar from './components/NavBar.vue'
import AppFooter from './components/AppFooter.vue'

export default {
  name: 'App',
  components: {
    NavBar,
    AppFooter
  }
}
</script>

<style>
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  line-height: 1.6;
  color: #333;
}

#app {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
}

.main-content {
  flex: 1;
  padding-top: 80px; /* Account for fixed header */
}

/* Container */
.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 0 20px;
}

/* Buttons */
.btn {
  display: inline-block;
  padding: 12px 24px;
  background-color: #42b883;
  color: white;
  text-decoration: none;
  border-radius: 5px;
  border: none;
  cursor: pointer;
  font-size: 16px;
  transition: background-color 0.3s ease;
}

.btn:hover {
  background-color: #369870;
}

.btn-secondary {
  background-color: #6c757d;
}

.btn-secondary:hover {
  background-color: #545b62;
}

/* Form Styles */
.form-group {
  margin-bottom: 1rem;
}

.form-group label {
  display: block;
  margin-bottom: 0.5rem;
  font-weight: bold;
}

.form-group input,
.form-group textarea {
  width: 100%;
  padding: 10px;
  border: 1px solid #ddd;
  border-radius: 4px;
  font-size: 16px;
}

.form-group input:focus,
.form-group textarea:focus {
  outline: none;
  border-color: #42b883;
  box-shadow: 0 0 0 2px rgba(66, 184, 131, 0.25);
}

/* Animations */
.fade-enter-active, .fade-leave-active {
  transition: opacity 0.3s;
}
.fade-enter-from, .fade-leave-to {
  opacity: 0;
}

/* Responsive */
@media (max-width: 768px) {
  .container {
    padding: 0 15px;
  }
  
  .main-content {
    padding-top: 70px;
  }
}
</style>`
  });

  // Main.js
  files.push({
    path: 'src/main.js',
    content: `import { createApp } from 'vue'
import App from './App.vue'
import router from './router'

createApp(App).use(router).mount('#app')`
  });

  // Router
  files.push({
    path: 'src/router/index.js',
    content: `import { createRouter, createWebHistory } from 'vue-router'
import Home from '../views/Home.vue'
import About from '../views/About.vue'
import Contact from '../views/Contact.vue'

const routes = [
  {
    path: '/',
    name: 'Home',
    component: Home
  },
  {
    path: '/about',
    name: 'About',
    component: About
  },
  {
    path: '/contact',
    name: 'Contact',
    component: Contact
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
})

export default router`
  });

  // NavBar component
  files.push({
    path: 'src/components/NavBar.vue',
    content: `<template>
  <header class="header">
    <nav class="navbar">
      <div class="container">
        <router-link to="/" class="nav-brand">
          ${projectName}
        </router-link>
        
        <button 
          class="nav-toggle"
          :class="{ active: isMenuOpen }"
          @click="toggleMenu"
          aria-label="Toggle navigation"
        >
          <span></span>
          <span></span>
          <span></span>
        </button>

        <ul class="nav-menu" :class="{ active: isMenuOpen }">
          <li class="nav-item">
            <router-link to="/" class="nav-link" @click="closeMenu">
              Home
            </router-link>
          </li>
          <li class="nav-item">
            <router-link to="/about" class="nav-link" @click="closeMenu">
              About
            </router-link>
          </li>
          <li class="nav-item">
            <router-link to="/contact" class="nav-link" @click="closeMenu">
              Contact
            </router-link>
          </li>
        </ul>
      </div>
    </nav>
  </header>
</template>

<script>
import { ref } from 'vue'

export default {
  name: 'NavBar',
  setup() {
    const isMenuOpen = ref(false)

    const toggleMenu = () => {
      isMenuOpen.value = !isMenuOpen.value
    }

    const closeMenu = () => {
      isMenuOpen.value = false
    }

    return {
      isMenuOpen,
      toggleMenu,
      closeMenu
    }
  }
}
</script>

<style scoped>
.header {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1000;
  background-color: #fff;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
}

.navbar {
  padding: 1rem 0;
}

.navbar .container {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.nav-brand {
  font-size: 1.5rem;
  font-weight: bold;
  color: #42b883;
  text-decoration: none;
}

.nav-menu {
  display: flex;
  list-style: none;
  gap: 2rem;
  margin: 0;
}

.nav-link {
  color: #333;
  text-decoration: none;
  font-weight: 500;
  transition: color 0.3s ease;
}

.nav-link:hover,
.nav-link.router-link-active {
  color: #42b883;
}

.nav-toggle {
  display: none;
  flex-direction: column;
  background: none;
  border: none;
  cursor: pointer;
  padding: 0;
  width: 30px;
  height: 30px;
  justify-content: space-around;
}

.nav-toggle span {
  display: block;
  height: 3px;
  width: 100%;
  background-color: #333;
  border-radius: 3px;
  transition: all 0.3s ease;
}

.nav-toggle.active span:nth-child(1) {
  transform: rotate(45deg) translate(5px, 5px);
}

.nav-toggle.active span:nth-child(2) {
  opacity: 0;
}

.nav-toggle.active span:nth-child(3) {
  transform: rotate(-45deg) translate(7px, -6px);
}

@media (max-width: 768px) {
  .nav-toggle {
    display: flex;
  }

  .nav-menu {
    position: fixed;
    left: -100%;
    top: 70px;
    flex-direction: column;
    background-color: #fff;
    width: 100%;
    text-align: center;
    transition: 0.3s;
    box-shadow: 0 10px 27px rgba(0, 0, 0, 0.05);
    padding: 2rem 0;
    gap: 1rem;
  }

  .nav-menu.active {
    left: 0;
  }
}
</style>`
  });

  // Footer component
  files.push({
    path: 'src/components/AppFooter.vue',
    content: `<template>
  <footer class="footer">
    <div class="container">
      <p>&copy; 2024 ${projectName}. All rights reserved.</p>
    </div>
  </footer>
</template>

<script>
export default {
  name: 'AppFooter'
}
</script>

<style scoped>
.footer {
  background-color: #f8f9fa;
  padding: 2rem 0;
  margin-top: auto;
  border-top: 1px solid #e9ecef;
}

.footer .container {
  text-align: center;
}

.footer p {
  margin: 0;
  color: #6c757d;
}
</style>`
  });

  // Home view
  files.push({
    path: 'src/views/Home.vue',
    content: `<template>
  <div class="home">
    <section class="hero">
      <div class="container">
        <div class="hero-content">
          <h1 class="hero-title">Welcome to ${projectName}</h1>
          <p class="hero-description">
            A modern Vue.js application built with the Composition API
          </p>
          <button class="btn" @click="scrollToFeatures">Get Started</button>
        </div>
      </div>
    </section>

    <section class="features" ref="featuresSection">
      <div class="container">
        <h2>Features</h2>
        <div class="features-grid">
          <div class="feature-card" v-for="feature in features" :key="feature.id">
            <h3>{{ feature.icon }} {{ feature.title }}</h3>
            <p>{{ feature.description }}</p>
          </div>
        </div>
      </div>
    </section>
  </div>
</template>

<script>
import { ref, reactive } from 'vue'

export default {
  name: 'Home',
  setup() {
    const featuresSection = ref(null)
    
    const features = reactive([
      {
        id: 1,
        icon: '⚡',
        title: 'Fast',
        description: 'Built with Vue 3 and the Composition API for optimal performance'
      },
      {
        id: 2,
        icon: '📱',
        title: 'Responsive',
        description: 'Works perfectly on all devices and screen sizes'
      },
      {
        id: 3,
        icon: '🎨',
        title: 'Modern',
        description: 'Clean, modern design with smooth animations'
      }
    ])

    const scrollToFeatures = () => {
      featuresSection.value?.scrollIntoView({ behavior: 'smooth' })
    }

    return {
      features,
      featuresSection,
      scrollToFeatures
    }
  }
}
</script>

<style scoped>
.hero {
  background: linear-gradient(135deg, #42b883 0%, #35495e 100%);
  color: white;
  padding: 100px 0;
  text-align: center;
}

.hero-content {
  max-width: 600px;
  margin: 0 auto;
}

.hero-title {
  font-size: 3rem;
  margin-bottom: 1rem;
  animation: fadeInUp 1s ease;
}

.hero-description {
  font-size: 1.2rem;
  margin-bottom: 2rem;
  animation: fadeInUp 1s ease 0.2s both;
}

.features {
  padding: 80px 0;
  background-color: #f8f9fa;
}

.features h2 {
  text-align: center;
  margin-bottom: 3rem;
  font-size: 2.5rem;
  color: #333;
}

.features-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
  gap: 2rem;
}

.feature-card {
  background: white;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
  text-align: center;
  transition: transform 0.3s ease;
}

.feature-card:hover {
  transform: translateY(-5px);
}

.feature-card h3 {
  font-size: 1.5rem;
  margin-bottom: 1rem;
  color: #333;
}

@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translateY(30px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}

@media (max-width: 768px) {
  .hero-title {
    font-size: 2rem;
  }
  
  .hero-description {
    font-size: 1rem;
  }
  
  .features-grid {
    grid-template-columns: 1fr;
  }
}
</style>`
  });

  // About view
  files.push({
    path: 'src/views/About.vue',
    content: `<template>
  <div class="about">
    <div class="container">
      <section class="about-hero">
        <h1>About ${projectName}</h1>
        <p>Learn more about our mission and values</p>
      </section>

      <section class="about-content">
        <div class="about-text">
          <h2>Our Story</h2>
          <p>
            ${projectName} was created to demonstrate modern Vue.js development practices
            and provide a solid foundation for building scalable web applications.
          </p>
          <p>
            This template includes Vue Router, the Composition API, and responsive design
            to help you get started quickly with your next project.
          </p>
        </div>
        
        <div class="about-stats">
          <div class="stat" v-for="stat in stats" :key="stat.id">
            <h3>{{ stat.value }}</h3>
            <p>{{ stat.label }}</p>
          </div>
        </div>
      </section>
    </div>
  </div>
</template>

<script>
import { reactive } from 'vue'

export default {
  name: 'About',
  setup() {
    const stats = reactive([
      { id: 1, value: '100%', label: 'Composition API' },
      { id: 2, value: '0', label: 'Options API' },
      { id: 3, value: '∞', label: 'Possibilities' }
    ])

    return {
      stats
    }
  }
}
</script>

<style scoped>
.about {
  padding: 2rem 0;
}

.about-hero {
  text-align: center;
  padding: 3rem 0;
  background-color: #f8f9fa;
  margin-bottom: 3rem;
}

.about-hero h1 {
  font-size: 2.5rem;
  margin-bottom: 1rem;
  color: #333;
}

.about-hero p {
  font-size: 1.2rem;
  color: #6c757d;
}

.about-content {
  display: grid;
  grid-template-columns: 2fr 1fr;
  gap: 3rem;
  align-items: start;
}

.about-text h2 {
  margin-bottom: 1.5rem;
  color: #333;
}

.about-text p {
  margin-bottom: 1rem;
  line-height: 1.8;
  color: #555;
}

.about-stats {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.stat {
  text-align: center;
  padding: 1.5rem;
  background: white;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
}

.stat h3 {
  font-size: 2rem;
  color: #42b883;
  margin-bottom: 0.5rem;
}

.stat p {
  color: #6c757d;
  margin: 0;
}

@media (max-width: 768px) {
  .about-content {
    grid-template-columns: 1fr;
    gap: 2rem;
  }
  
  .about-stats {
    flex-direction: row;
    justify-content: space-around;
  }
  
  .stat {
    flex: 1;
    margin: 0 0.5rem;
  }
}
</style>`
  });

  // Contact view
  files.push({
    path: 'src/views/Contact.vue',
    content: `<template>
  <div class="contact">
    <div class="container">
      <section class="contact-hero">
        <h1>Contact Us</h1>
        <p>Get in touch with our team</p>
      </section>

      <section class="contact-content">
        <div class="contact-info">
          <h2>Get In Touch</h2>
          <p>
            We'd love to hear from you. Send us a message and we'll respond
            as soon as possible.
          </p>
          
          <div class="contact-details">
            <div class="contact-item" v-for="item in contactDetails" :key="item.id">
              <h3>{{ item.icon }} {{ item.title }}</h3>
              <p v-html="item.content"></p>
            </div>
          </div>
        </div>

        <form class="contact-form" @submit.prevent="handleSubmit">
          <div class="form-group">
            <label for="name">Name</label>
            <input
              type="text"
              id="name"
              v-model="form.name"
              required
            />
          </div>

          <div class="form-group">
            <label for="email">Email</label>
            <input
              type="email"
              id="email"
              v-model="form.email"
              required
            />
          </div>

          <div class="form-group">
            <label for="message">Message</label>
            <textarea
              id="message"
              rows="5"
              v-model="form.message"
              required
            ></textarea>
          </div>

          <button type="submit" class="btn" :disabled="isSubmitting">
            {{ isSubmitting ? 'Sending...' : 'Send Message' }}
          </button>
        </form>
      </section>
    </div>
  </div>
</template>

<script>
import { reactive, ref } from 'vue'

export default {
  name: 'Contact',
  setup() {
    const form = reactive({
      name: '',
      email: '',
      message: ''
    })

    const isSubmitting = ref(false)

    const contactDetails = reactive([
      {
        id: 1,
        icon: '📧',
        title: 'Email',
        content: 'hello@${projectName.toLowerCase().replace(/\s+/g, '')}.com'
      },
      {
        id: 2,
        icon: '📱',
        title: 'Phone',
        content: '+1 (555) 123-4567'
      },
      {
        id: 3,
        icon: '📍',
        title: 'Address',
        content: '123 Main Street<br />City, State 12345'
      }
    ])

    const handleSubmit = async () => {
      if (!form.name || !form.email || !form.message) {
        alert('Please fill in all fields')
        return
      }

      isSubmitting.value = true

      // Simulate form submission
      setTimeout(() => {
        alert('Thank you for your message! We will get back to you soon.')
        
        // Reset form
        form.name = ''
        form.email = ''
        form.message = ''
        
        isSubmitting.value = false
      }, 1000)
    }

    return {
      form,
      isSubmitting,
      contactDetails,
      handleSubmit
    }
  }
}
</script>

<style scoped>
.contact {
  padding: 2rem 0;
}

.contact-hero {
  text-align: center;
  padding: 3rem 0;
  background-color: #f8f9fa;
  margin-bottom: 3rem;
}

.contact-hero h1 {
  font-size: 2.5rem;
  margin-bottom: 1rem;
  color: #333;
}

.contact-hero p {
  font-size: 1.2rem;
  color: #6c757d;
}

.contact-content {
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 3rem;
}

.contact-info h2 {
  margin-bottom: 1rem;
  color: #333;
}

.contact-info p {
  margin-bottom: 2rem;
  line-height: 1.8;
  color: #555;
}

.contact-details {
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}

.contact-item h3 {
  margin-bottom: 0.5rem;
  color: #333;
  font-size: 1.1rem;
}

.contact-item p {
  margin: 0;
  color: #6c757d;
}

.contact-form {
  background: white;
  padding: 2rem;
  border-radius: 10px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
}

@media (max-width: 768px) {
  .contact-content {
    grid-template-columns: 1fr;
    gap: 2rem;
  }
  
  .contact-form {
    padding: 1.5rem;
  }
}
</style>`
  });

  // Vue config
  files.push({
    path: 'vue.config.js',
    content: `const { defineConfig } = require('@vue/cli-service')
module.exports = defineConfig({
  transpileDependencies: true,
  devServer: {
    port: 8080,
    open: true
  }
})`
  });

  // Test file
  files.push({
    path: 'tests/unit/Home.spec.js',
    content: `import { mount } from '@vue/test-utils'
import Home from '@/views/Home.vue'

describe('Home.vue', () => {
  it('renders welcome message', () => {
    const wrapper = mount(Home)
    expect(wrapper.text()).toMatch('Welcome to ${projectName}')
  })

  it('has three feature cards', () => {
    const wrapper = mount(Home)
    const featureCards = wrapper.findAll('.feature-card')
    expect(featureCards).toHaveLength(3)
  })

  it('scrolls to features when button is clicked', async () => {
    const wrapper = mount(Home)
    const scrollSpy = jest.fn()
    
    // Mock scrollIntoView
    Element.prototype.scrollIntoView = scrollSpy
    
    const button = wrapper.find('.btn')
    await button.trigger('click')
    
    expect(scrollSpy).toHaveBeenCalledWith({ behavior: 'smooth' })
  })
})`
  });

  // Jest config
  files.push({
    path: 'jest.config.js',
    content: `module.exports = {
  preset: '@vue/cli-plugin-unit-jest',
  testEnvironment: 'jsdom',
  transform: {
    '^.+\\.vue$': '@vue/vue3-jest'
  }
}`
  });

  // .gitignore
  files.push({
    path: '.gitignore',
    content: `.DS_Store
node_modules
/dist

# local env files
.env.local
.env.*.local

# Log files
npm-debug.log*
yarn-debug.log*
yarn-error.log*
pnpm-debug.log*

# Editor directories and files
.idea
.vscode
*.suo
*.ntvs*
*.njsproj
*.sln
*.sw?`
  });

  // README
  files.push({
    path: 'README.md',
    content: `# ${projectName}

A modern Vue.js application built with Vue 3 and the Composition API.

## Features

- ⚡ Vue 3 with Composition API
- 🚀 Vue Router for navigation
- 📱 Fully responsive design
- 🎨 Modern CSS with animations
- 🧪 Testing setup with Jest and Vue Test Utils
- 📦 Vue CLI for development and build

## Getting Started

### Prerequisites

- Node.js (version 14 or higher)
- npm or yarn

### Installation

1. Install dependencies:
   \`\`\`bash
   npm install
   \`\`\`

2. Start the development server:
   \`\`\`bash
   npm run serve
   \`\`\`

3. Open [http://localhost:8080](http://localhost:8080) to view it in the browser.

## Available Scripts

- \`npm run serve\` - Runs the app in development mode
- \`npm run build\` - Builds the app for production
- \`npm run test:unit\` - Runs unit tests
- \`npm run lint\` - Lints and fixes files

## Project Structure

\`\`\`
src/
├── components/          # Reusable components
│   ├── NavBar.vue
│   └── AppFooter.vue
├── views/              # Page components
│   ├── Home.vue
│   ├── About.vue
│   └── Contact.vue
├── router/             # Vue Router configuration
│   └── index.js
├── App.vue             # Main App component
└── main.js             # Entry point
\`\`\`

## Customization

### Adding New Pages

1. Create a new component in \`src/views/\`
2. Add the route in \`src/router/index.js\`
3. Update the navigation in \`src/components/NavBar.vue\`

### Styling

- Global styles are in \`src/App.vue\`
- Component-specific styles use scoped CSS
- The project uses CSS Grid and Flexbox for layouts

### State Management

For larger applications, consider adding Vuex or Pinia:

\`\`\`bash
npm install vuex@next
# or
npm install pinia
\`\`\`

## Testing

Run the test suite:

\`\`\`bash
npm run test:unit
\`\`\`

Tests are located in the \`tests/unit/\` directory and use Jest with Vue Test Utils.

## Building for Production

\`\`\`bash
npm run build
\`\`\`

This creates a \`dist\` folder with optimized production files.

## Deployment

### Netlify

1. Build the project: \`npm run build\`
2. Drag and drop the \`dist\` folder to Netlify
3. Or connect your Git repository for automatic deployments

### Vercel

\`\`\`bash
npm install -g vercel
vercel
\`\`\`

### GitHub Pages

1. Install gh-pages: \`npm install --save-dev gh-pages\`
2. Add deploy script to package.json:
   \`\`\`json
   "scripts": {
     "deploy": "npm run build && gh-pages -d dist"
   }
   \`\`\`
3. Run: \`npm run deploy\`

## Browser Support

This project works in all modern browsers:
- Chrome 60+
- Firefox 60+
- Safari 12+
- Edge 79+

## Learn More

- [Vue.js Documentation](https://vuejs.org/)
- [Vue Router Documentation](https://router.vuejs.org/)
- [Vue CLI Documentation](https://cli.vuejs.org/)
- [Composition API Guide](https://vuejs.org/guide/extras/composition-api-faq.html)

## License

This project is licensed under the MIT License.
`
  });

  return files;
}

export default {
  name: 'Vue App',
  description: 'A modern Vue.js application with Vue Router and Composition API',
  generate
};
