// codestral/src/executor/CommandExecutor.js
import { exec } from 'child_process';
import { promisify } from 'util';
import logger from '../utils/logger.js';

const execAsync = promisify(exec);

/**
 * CommandExecutor handles shell command execution with safety checks
 */
class CommandExecutor {
  /**
   * @param {Object} options - Executor options
   * @param {number} options.timeout - Command timeout in milliseconds (default: 30000)
   * @param {string} options.cwd - Working directory (default: process.cwd())
   */
  constructor(options = {}) {
    this.timeout = options.timeout || 30000; // 30 seconds default
    this.cwd = options.cwd || process.cwd();
    
    // List of dangerous commands that should never be auto-approved
    this.dangerousCommands = [
      'rm -rf',
      'sudo',
      'chmod',
      'chown',
      'mkfs',
      'dd',
      'format',
      ':(){:|:&};:', // fork bomb
      '> /dev/sda',
      'mv /* ',
      'wget',
      'curl'
    ];
  }

  /**
   * Execute a shell command
   * @param {string} command - Command to execute
   * @param {Object} options - Execution options
   * @param {string} options.cwd - Working directory (overrides default)
   * @param {number} options.timeout - Timeout in milliseconds (overrides default)
   * @returns {Promise<Object>} Execution result
   */
  async execute(command, options = {}) {
    if (!command || typeof command !== 'string') {
      throw new Error('Command must be a non-empty string');
    }

    // Check for dangerous commands
    if (this.isDangerous(command)) {
      throw new Error(
        `Dangerous command detected: "${command}". This command requires explicit user approval.`
      );
    }

    const cwd = options.cwd || this.cwd;
    const timeout = options.timeout || this.timeout;

    logger.info(`Executing command: ${command}`, { cwd, timeout });

    try {
      const { stdout, stderr } = await execAsync(command, {
        cwd,
        timeout,
        maxBuffer: 1024 * 1024 * 10, // 10MB buffer
        shell: true
      });

      logger.info(`Command executed successfully: ${command}`);

      return {
        success: true,
        command,
        stdout: stdout.trim(),
        stderr: stderr.trim(),
        exitCode: 0
      };
    } catch (error) {
      logger.error(`Command execution failed: ${command}`, { error: error.message });

      // Handle timeout
      if (error.killed && error.signal === 'SIGTERM') {
        throw new Error(`Command timed out after ${timeout}ms: ${command}`);
      }

      // Handle shell not found
      if (error.code === 'ENOENT') {
        throw new Error(`Shell not found or command not executable: ${command}`);
      }

      return {
        success: false,
        command,
        stdout: error.stdout ? error.stdout.trim() : '',
        stderr: error.stderr ? error.stderr.trim() : error.message,
        exitCode: typeof error.code === 'number' ? error.code : 1,
        error: error.message
      };
    }
  }

  /**
   * Check if a command is dangerous
   * @param {string} command - Command to check
   * @returns {boolean} True if command is dangerous
   */
  isDangerous(command) {
    const lowerCommand = command.toLowerCase().trim();
    
    return this.dangerousCommands.some(dangerous => 
      lowerCommand.includes(dangerous.toLowerCase())
    );
  }

  /**
   * Get list of dangerous command patterns
   * @returns {Array<string>} Array of dangerous command patterns
   */
  getDangerousCommands() {
    return [...this.dangerousCommands];
  }

  /**
   * Add a dangerous command pattern
   * @param {string} pattern - Command pattern to add
   */
  addDangerousCommand(pattern) {
    if (!this.dangerousCommands.includes(pattern)) {
      this.dangerousCommands.push(pattern);
    }
  }
}

export default CommandExecutor;
