// codestral/src/indexer/CodebaseIndexer.js
import path from 'path';
import fg from 'fast-glob';
import chokidar from 'chokidar';

/**
 * CodebaseIndexer scans and indexes the project directory structure
 * with performance optimizations including lazy loading and caching
 */
class CodebaseIndexer {
  /**
   * @param {string} rootPath - Root directory path to index
   * @param {FileSystemManager} fileSystemManager - File system manager instance
   * @param {Object} options - Configuration options
   */
  constructor(rootPath = process.cwd(), fileSystemManager, options = {}) {
    this.rootPath = path.resolve(rootPath);
    this.fileSystemManager = fileSystemManager;
    this.index = new Map(); // Map<filePath, FileMetadata>
    this.languages = new Set();
    this.frameworks = new Set();
    this.dependencies = new Map();
    this.lastIndexed = null;
    this.watcher = null;
    this.isWatching = false;

    // Performance optimization options
    this.options = {
      maxFileSize: options.maxFileSize || 10 * 1024 * 1024, // 10MB default
      lazyLoadThreshold: options.lazyLoadThreshold || 1000, // Files count threshold for lazy loading
      cacheSize: options.cacheSize || 100, // Number of files to keep in memory cache
      batchSize: options.batchSize || 50, // Files to process in each batch
      enableProfiling: options.enableProfiling || false,
      ...options
    };

    // Caching for frequently accessed files
    this.fileContentCache = new Map(); // LRU cache for file contents
    this.cacheAccessOrder = []; // Track access order for LRU eviction
    
    // Lazy loading state
    this.isLazyMode = false;
    this.lazyLoadedPaths = new Set();
    
    // Performance metrics
    this.metrics = {
      indexingTime: 0,
      filesProcessed: 0,
      cacheHits: 0,
      cacheMisses: 0,
      memoryPeakUsage: 0
    };
  }

  /**
   * Build the complete index by scanning the project directory
   * @param {Object} options - Build options
   * @returns {Promise<Object>} Index summary
   */
  async buildIndex(options = {}) {
    const startTime = Date.now();
    
    try {
      if (this.options.enableProfiling) {
        console.log('🔍 Starting codebase indexing...');
      }

      // Load .gitignore patterns
      await this.fileSystemManager.loadGitignore();

      // Use fast-glob to discover all files efficiently
      const files = await fg('**/*', {
        cwd: this.rootPath,
        dot: false, // Don't include hidden files by default
        ignore: this._getIgnorePatterns(),
        onlyFiles: true,
        absolute: false,
        stats: true
      });

      // Determine if we should use lazy loading
      this.isLazyMode = files.length > this.options.lazyLoadThreshold;
      
      if (this.options.enableProfiling) {
        console.log(`📁 Found ${files.length} files, lazy mode: ${this.isLazyMode}`);
      }

      // Clear existing index and caches
      this.index.clear();
      this.languages.clear();
      this.frameworks.clear();
      this._clearCache();

      let result;
      if (this.isLazyMode) {
        result = await this._buildIndexLazy(files);
      } else {
        result = await this._buildIndexEager(files);
      }

      // Detect frameworks after indexing
      const detectedFrameworks = await this.detectFrameworks();
      detectedFrameworks.forEach(framework => this.frameworks.add(framework));

      return result;
    } catch (error) {
      throw new Error(`Failed to build index: ${error.message}`);
    } finally {
      this.metrics.indexingTime = Date.now() - startTime;
      if (this.options.enableProfiling) {
        this._logPerformanceMetrics();
      }
    }
  }

  /**
   * Build index eagerly for smaller codebases
   * @param {Array} files - Array of file entries
   * @returns {Promise<Object>} Index summary
   * @private
   */
  async _buildIndexEager(files) {
    // Process files in batches to avoid memory spikes
    for (let i = 0; i < files.length; i += this.options.batchSize) {
      const batch = files.slice(i, i + this.options.batchSize);
      await this._processBatch(batch);
      
      // Allow event loop to breathe
      if (i % (this.options.batchSize * 4) === 0) {
        await new Promise(resolve => setImmediate(resolve));
      }
    }

    this.lastIndexed = new Date();
    this.metrics.filesProcessed = this.index.size;

    return {
      fileCount: this.index.size,
      languages: Array.from(this.languages),
      lastIndexed: this.lastIndexed,
      lazyMode: false,
      metrics: this.metrics
    };
  }

  /**
   * Build index lazily for large codebases
   * @param {Array} files - Array of file entries
   * @returns {Promise<Object>} Index summary
   * @private
   */
  async _buildIndexLazy(files) {
    // Only process metadata for all files, defer content analysis
    for (let i = 0; i < files.length; i += this.options.batchSize) {
      const batch = files.slice(i, i + this.options.batchSize);
      await this._processMetadataBatch(batch);
      
      // Allow event loop to breathe
      if (i % (this.options.batchSize * 4) === 0) {
        await new Promise(resolve => setImmediate(resolve));
      }
    }

    this.lastIndexed = new Date();
    this.metrics.filesProcessed = this.index.size;

    return {
      fileCount: this.index.size,
      languages: Array.from(this.languages),
      lastIndexed: this.lastIndexed,
      lazyMode: true,
      metrics: this.metrics
    };
  }

  /**
   * Process a batch of files with full analysis
   * @param {Array} batch - Batch of file entries
   * @private
   */
  async _processBatch(batch) {
    for (const entry of batch) {
      const filePath = entry.path || entry;
      const stats = entry.stats;

      try {
        // Skip files that are too large
        if (stats && stats.size > this.options.maxFileSize) {
          if (this.options.enableProfiling) {
            console.warn(`⚠️  Skipping large file: ${filePath} (${(stats.size / 1024 / 1024).toFixed(2)}MB)`);
          }
          continue;
        }

        // Skip binary files
        const isBinary = await this.fileSystemManager.isBinary(filePath);
        if (isBinary) {
          continue;
        }

        // Detect language from extension
        const language = this._detectLanguage(filePath);
        if (language) {
          this.languages.add(language);
        }

        // Create file metadata
        const metadata = {
          path: filePath,
          size: stats ? stats.size : 0,
          language: language || 'unknown',
          imports: [],
          exports: [],
          lastModified: stats ? stats.mtime : new Date(),
          analyzed: true
        };

        this.index.set(filePath, metadata);
      } catch (error) {
        // Skip files that can't be processed
        if (this.options.enableProfiling) {
          console.warn(`Warning: Could not process file ${filePath}: ${error.message}`);
        }
      }
    }
  }

  /**
   * Process a batch of files with metadata only (lazy mode)
   * @param {Array} batch - Batch of file entries
   * @private
   */
  async _processMetadataBatch(batch) {
    for (const entry of batch) {
      const filePath = entry.path || entry;
      const stats = entry.stats;

      try {
        // Skip files that are too large
        if (stats && stats.size > this.options.maxFileSize) {
          continue;
        }

        // Quick binary check without reading file
        if (this._isLikelyBinary(filePath)) {
          continue;
        }

        // Detect language from extension
        const language = this._detectLanguage(filePath);
        if (language) {
          this.languages.add(language);
        }

        // Create minimal metadata for lazy loading
        const metadata = {
          path: filePath,
          size: stats ? stats.size : 0,
          language: language || 'unknown',
          imports: [],
          exports: [],
          lastModified: stats ? stats.mtime : new Date(),
          analyzed: false // Mark as not fully analyzed
        };

        this.index.set(filePath, metadata);
      } catch (error) {
        // Skip files that can't be processed
        if (this.options.enableProfiling) {
          console.warn(`Warning: Could not process metadata for ${filePath}: ${error.message}`);
        }
      }
    }
  }

  /**
   * Quick binary file detection based on extension
   * @param {string} filePath - Path to file
   * @returns {boolean} True if likely binary
   * @private
   */
  _isLikelyBinary(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const binaryExtensions = new Set([
      '.exe', '.dll', '.so', '.dylib', '.bin', '.dat',
      '.jpg', '.jpeg', '.png', '.gif', '.bmp', '.ico', '.svg',
      '.mp3', '.mp4', '.avi', '.mov', '.wav', '.flac',
      '.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx',
      '.zip', '.tar', '.gz', '.rar', '.7z',
      '.ttf', '.otf', '.woff', '.woff2', '.eot'
    ]);
    return binaryExtensions.has(ext);
  }

  /**
   * Get ignore patterns for fast-glob
   * @returns {string[]} Array of ignore patterns
   * @private
   */
  _getIgnorePatterns() {
    return [
      '**/node_modules/**',
      '**/.git/**',
      '**/dist/**',
      '**/build/**',
      '**/.next/**',
      '**/.nuxt/**',
      '**/coverage/**',
      '**/.cache/**',
      '**/tmp/**',
      '**/temp/**',
      '**/*.log',
      '**/.DS_Store',
      '**/Thumbs.db'
    ];
  }

  /**
   * Detect programming language from file extension
   * @param {string} filePath - Path to file
   * @returns {string|null} Language name or null
   * @private
   */
  _detectLanguage(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    
    const languageMap = {
      '.js': 'JavaScript',
      '.jsx': 'JavaScript',
      '.ts': 'TypeScript',
      '.tsx': 'TypeScript',
      '.py': 'Python',
      '.java': 'Java',
      '.c': 'C',
      '.cpp': 'C++',
      '.cc': 'C++',
      '.cxx': 'C++',
      '.h': 'C/C++',
      '.hpp': 'C++',
      '.cs': 'C#',
      '.go': 'Go',
      '.rs': 'Rust',
      '.rb': 'Ruby',
      '.php': 'PHP',
      '.swift': 'Swift',
      '.kt': 'Kotlin',
      '.scala': 'Scala',
      '.r': 'R',
      '.m': 'Objective-C',
      '.sh': 'Shell',
      '.bash': 'Bash',
      '.zsh': 'Zsh',
      '.fish': 'Fish',
      '.ps1': 'PowerShell',
      '.html': 'HTML',
      '.htm': 'HTML',
      '.css': 'CSS',
      '.scss': 'SCSS',
      '.sass': 'Sass',
      '.less': 'Less',
      '.json': 'JSON',
      '.xml': 'XML',
      '.yaml': 'YAML',
      '.yml': 'YAML',
      '.toml': 'TOML',
      '.md': 'Markdown',
      '.sql': 'SQL',
      '.graphql': 'GraphQL',
      '.gql': 'GraphQL',
      '.vue': 'Vue',
      '.svelte': 'Svelte'
    };

    return languageMap[ext] || null;
  }

  /**
   * Get the complete index
   * @returns {Map<string, Object>} File index
   */
  getIndex() {
    return this.index;
  }

  /**
   * Get project structure summary
   * @returns {Object} Project structure information
   */
  getProjectStructure() {
    return {
      totalFiles: this.index.size,
      languages: Array.from(this.languages),
      frameworks: this.frameworks,
      dependencies: Object.fromEntries(this.dependencies),
      lastIndexed: this.lastIndexed
    };
  }

  /**
   * Get file metadata by path with lazy loading support
   * @param {string} filePath - Path to file
   * @param {boolean} forceAnalyze - Force full analysis if not done
   * @returns {Promise<Object|null>} File metadata or null if not found
   */
  async getFileMetadata(filePath, forceAnalyze = false) {
    const metadata = this.index.get(filePath);
    
    if (!metadata) {
      return null;
    }

    // If in lazy mode and file hasn't been fully analyzed, analyze it now
    if (this.isLazyMode && !metadata.analyzed && forceAnalyze) {
      await this._analyzeFile(filePath, metadata);
    }

    return metadata;
  }

  /**
   * Get file content with caching
   * @param {string} filePath - Path to file
   * @returns {Promise<string|null>} File content or null if not found
   */
  async getFileContent(filePath) {
    // Check cache first
    if (this.fileContentCache.has(filePath)) {
      this._updateCacheAccess(filePath);
      this.metrics.cacheHits++;
      return this.fileContentCache.get(filePath);
    }

    this.metrics.cacheMisses++;

    try {
      const content = await this.fileSystemManager.readFile(filePath);
      this._addToCache(filePath, content);
      return content;
    } catch (error) {
      return null;
    }
  }

  /**
   * Analyze a file fully (used in lazy mode)
   * @param {string} filePath - Path to file
   * @param {Object} metadata - Existing metadata
   * @private
   */
  async _analyzeFile(filePath, metadata) {
    try {
      // Skip if already analyzed
      if (metadata.analyzed) {
        return;
      }

      // Perform full binary check
      const isBinary = await this.fileSystemManager.isBinary(filePath);
      if (isBinary) {
        // Remove from index if it's binary
        this.index.delete(filePath);
        return;
      }

      // TODO: Add import/export analysis here if needed
      // This could include parsing the file to find imports/exports

      // Mark as analyzed
      metadata.analyzed = true;
      this.lazyLoadedPaths.add(filePath);

      if (this.options.enableProfiling) {
        console.log(`🔍 Analyzed file: ${filePath}`);
      }
    } catch (error) {
      if (this.options.enableProfiling) {
        console.warn(`Warning: Could not analyze file ${filePath}: ${error.message}`);
      }
    }
  }

  /**
   * Add content to cache with LRU eviction
   * @param {string} filePath - Path to file
   * @param {string} content - File content
   * @private
   */
  _addToCache(filePath, content) {
    // Remove if already exists to update position
    if (this.fileContentCache.has(filePath)) {
      this.fileContentCache.delete(filePath);
      const index = this.cacheAccessOrder.indexOf(filePath);
      if (index > -1) {
        this.cacheAccessOrder.splice(index, 1);
      }
    }

    // Add to cache
    this.fileContentCache.set(filePath, content);
    this.cacheAccessOrder.push(filePath);

    // Evict oldest if cache is full
    while (this.fileContentCache.size > this.options.cacheSize) {
      const oldest = this.cacheAccessOrder.shift();
      this.fileContentCache.delete(oldest);
    }
  }

  /**
   * Update cache access order for LRU
   * @param {string} filePath - Path to file
   * @private
   */
  _updateCacheAccess(filePath) {
    const index = this.cacheAccessOrder.indexOf(filePath);
    if (index > -1) {
      this.cacheAccessOrder.splice(index, 1);
      this.cacheAccessOrder.push(filePath);
    }
  }

  /**
   * Clear file content cache
   * @private
   */
  _clearCache() {
    this.fileContentCache.clear();
    this.cacheAccessOrder.length = 0;
    this.metrics.cacheHits = 0;
    this.metrics.cacheMisses = 0;
  }

  /**
   * Get all files of a specific language
   * @param {string} language - Language name
   * @returns {Array<Object>} Array of file metadata
   */
  getFilesByLanguage(language) {
    const files = [];
    for (const [filePath, metadata] of this.index.entries()) {
      if (metadata.language === language) {
        files.push(metadata);
      }
    }
    return files;
  }

  /**
   * Detect languages used in the project
   * @returns {Array<string>} Array of detected languages
   */
  detectLanguages() {
    return Array.from(this.languages);
  }

  /**
   * Detect frameworks used in the project
   * @returns {Promise<Array<string>>} Array of detected frameworks
   */
  async detectFrameworks() {
    const frameworks = new Set();

    try {
      // Check for package.json (Node.js projects)
      if (this.index.has('package.json')) {
        const packageJson = await this._readPackageJson();
        if (packageJson) {
          const allDeps = {
            ...packageJson.dependencies,
            ...packageJson.devDependencies
          };

          // Detect JavaScript/TypeScript frameworks
          if (allDeps['react'] || allDeps['react-dom']) {
            frameworks.add('React');
          }
          if (allDeps['vue']) {
            frameworks.add('Vue');
          }
          if (allDeps['@angular/core']) {
            frameworks.add('Angular');
          }
          if (allDeps['svelte']) {
            frameworks.add('Svelte');
          }
          if (allDeps['next']) {
            frameworks.add('Next.js');
          }
          if (allDeps['nuxt']) {
            frameworks.add('Nuxt.js');
          }
          if (allDeps['express']) {
            frameworks.add('Express');
          }
          if (allDeps['fastify']) {
            frameworks.add('Fastify');
          }
          if (allDeps['nest']) {
            frameworks.add('NestJS');
          }
          if (allDeps['gatsby']) {
            frameworks.add('Gatsby');
          }
          if (allDeps['electron']) {
            frameworks.add('Electron');
          }
          if (allDeps['jest']) {
            frameworks.add('Jest');
          }
          if (allDeps['vitest']) {
            frameworks.add('Vitest');
          }
          if (allDeps['webpack']) {
            frameworks.add('Webpack');
          }
          if (allDeps['vite']) {
            frameworks.add('Vite');
          }
        }
      }

      // Check for requirements.txt (Python projects)
      if (this.index.has('requirements.txt')) {
        const requirements = await this._readRequirementsTxt();
        if (requirements.includes('django')) {
          frameworks.add('Django');
        }
        if (requirements.includes('flask')) {
          frameworks.add('Flask');
        }
        if (requirements.includes('fastapi')) {
          frameworks.add('FastAPI');
        }
        if (requirements.includes('pytest')) {
          frameworks.add('Pytest');
        }
      }

      // Check for Gemfile (Ruby projects)
      if (this.index.has('Gemfile')) {
        const gemfile = await this._readGemfile();
        if (gemfile.includes('rails')) {
          frameworks.add('Ruby on Rails');
        }
        if (gemfile.includes('sinatra')) {
          frameworks.add('Sinatra');
        }
      }

      // Check for pom.xml or build.gradle (Java projects)
      if (this.index.has('pom.xml')) {
        frameworks.add('Maven');
        const pom = await this._readFile('pom.xml');
        if (pom.includes('spring-boot')) {
          frameworks.add('Spring Boot');
        }
      }
      if (this.index.has('build.gradle') || this.index.has('build.gradle.kts')) {
        frameworks.add('Gradle');
      }

      // Check for Cargo.toml (Rust projects)
      if (this.index.has('Cargo.toml')) {
        frameworks.add('Cargo');
      }

      // Check for go.mod (Go projects)
      if (this.index.has('go.mod')) {
        const goMod = await this._readFile('go.mod');
        if (goMod.includes('gin-gonic/gin')) {
          frameworks.add('Gin');
        }
        if (goMod.includes('gorilla/mux')) {
          frameworks.add('Gorilla');
        }
      }

      // Check for composer.json (PHP projects)
      if (this.index.has('composer.json')) {
        const composer = await this._readComposerJson();
        if (composer && composer.require) {
          if (composer.require['laravel/framework']) {
            frameworks.add('Laravel');
          }
          if (composer.require['symfony/symfony']) {
            frameworks.add('Symfony');
          }
        }
      }

    } catch (error) {
      console.warn(`Warning: Error detecting frameworks: ${error.message}`);
    }

    this.frameworks = frameworks;
    return Array.from(frameworks);
  }

  /**
   * Find and parse dependencies from package.json
   * @returns {Promise<Map<string, string>>} Map of dependency name to version
   */
  async findDependencies() {
    this.dependencies.clear();

    try {
      // Node.js dependencies
      if (this.index.has('package.json')) {
        const packageJson = await this._readPackageJson();
        if (packageJson) {
          const allDeps = {
            ...packageJson.dependencies,
            ...packageJson.devDependencies
          };
          
          for (const [name, version] of Object.entries(allDeps)) {
            this.dependencies.set(name, version);
          }
        }
      }

      // Python dependencies
      if (this.index.has('requirements.txt')) {
        const requirements = await this._readRequirementsTxt();
        const lines = requirements.split('\n');
        for (const line of lines) {
          const trimmed = line.trim();
          if (trimmed && !trimmed.startsWith('#')) {
            const match = trimmed.match(/^([a-zA-Z0-9-_]+)([>=<~!]+.*)?$/);
            if (match) {
              this.dependencies.set(match[1], match[2] || '*');
            }
          }
        }
      }

    } catch (error) {
      console.warn(`Warning: Error finding dependencies: ${error.message}`);
    }

    return this.dependencies;
  }

  /**
   * Read and parse package.json
   * @returns {Promise<Object|null>} Parsed package.json or null
   * @private
   */
  async _readPackageJson() {
    try {
      const content = await this.fileSystemManager.readFile('package.json');
      return JSON.parse(content);
    } catch (error) {
      return null;
    }
  }

  /**
   * Read and parse composer.json
   * @returns {Promise<Object|null>} Parsed composer.json or null
   * @private
   */
  async _readComposerJson() {
    try {
      const content = await this.fileSystemManager.readFile('composer.json');
      return JSON.parse(content);
    } catch (error) {
      return null;
    }
  }

  /**
   * Read requirements.txt
   * @returns {Promise<string>} Content of requirements.txt
   * @private
   */
  async _readRequirementsTxt() {
    try {
      return await this.fileSystemManager.readFile('requirements.txt');
    } catch (error) {
      return '';
    }
  }

  /**
   * Read Gemfile
   * @returns {Promise<string>} Content of Gemfile
   * @private
   */
  async _readGemfile() {
    try {
      return await this.fileSystemManager.readFile('Gemfile');
    } catch (error) {
      return '';
    }
  }

  /**
   * Read a file from the project
   * @param {string} filePath - Path to file
   * @returns {Promise<string>} File content
   * @private
   */
  async _readFile(filePath) {
    try {
      return await this.fileSystemManager.readFile(filePath);
    } catch (error) {
      return '';
    }
  }

  /**
   * Update index for specific files (incremental update)
   * @param {Array<string>} changedFiles - Array of file paths that changed
   * @returns {Promise<Object>} Update summary
   */
  async updateIndex(changedFiles) {
    let added = 0;
    let updated = 0;
    let removed = 0;

    for (const filePath of changedFiles) {
      try {
        // Check if file exists
        const exists = await this.fileSystemManager.exists(filePath);

        if (!exists) {
          // File was deleted
          if (this.index.has(filePath)) {
            this.index.delete(filePath);
            removed++;
          }
          continue;
        }

        // Check if file should be ignored
        const stats = await this.fileSystemManager.getStats(filePath);
        if (stats.isDirectory) {
          continue;
        }

        // Skip binary files
        const isBinary = await this.fileSystemManager.isBinary(filePath);
        if (isBinary) {
          if (this.index.has(filePath)) {
            this.index.delete(filePath);
            removed++;
          }
          continue;
        }

        // Detect language
        const language = this._detectLanguage(filePath);
        if (language) {
          this.languages.add(language);
        }

        // Create or update metadata
        const metadata = {
          path: filePath,
          size: stats.size,
          language: language || 'unknown',
          imports: [],
          exports: [],
          lastModified: stats.modified
        };

        if (this.index.has(filePath)) {
          this.index.set(filePath, metadata);
          updated++;
        } else {
          this.index.set(filePath, metadata);
          added++;
        }
      } catch (error) {
        console.warn(`Warning: Could not update index for ${filePath}: ${error.message}`);
      }
    }

    return { added, updated, removed };
  }

  /**
   * Start watching for file changes and update index incrementally
   * @param {Object} options - Watch options
   * @returns {void}
   */
  startWatching(options = {}) {
    if (this.isWatching) {
      console.warn('Already watching for changes');
      return;
    }

    const watchOptions = {
      cwd: this.rootPath,
      ignored: this._getIgnorePatterns(),
      ignoreInitial: true,
      persistent: true,
      awaitWriteFinish: {
        stabilityThreshold: 300,
        pollInterval: 100
      },
      ...options
    };

    this.watcher = chokidar.watch('**/*', watchOptions);

    this.watcher
      .on('add', async (filePath) => {
        await this.updateIndex([filePath]);
      })
      .on('change', async (filePath) => {
        await this.updateIndex([filePath]);
      })
      .on('unlink', async (filePath) => {
        await this.updateIndex([filePath]);
      })
      .on('error', (error) => {
        console.error(`Watcher error: ${error.message}`);
      });

    this.isWatching = true;
  }

  /**
   * Stop watching for file changes
   * @returns {Promise<void>}
   */
  async stopWatching() {
    if (this.watcher) {
      await this.watcher.close();
      this.watcher = null;
      this.isWatching = false;
    }
  }

  /**
   * Check if currently watching for changes
   * @returns {boolean}
   */
  isCurrentlyWatching() {
    return this.isWatching;
  }

  /**
   * Get memory usage statistics with detailed breakdown
   * @returns {Object} Memory usage information
   */
  getMemoryUsage() {
    const indexSize = this.index.size;
    const languagesSize = this.languages.size;
    const dependenciesSize = this.dependencies.size;
    const cacheSize = this.fileContentCache.size;
    
    // Estimate memory usage (more accurate approximation)
    let estimatedBytes = 0;
    
    // Index entries: path + metadata
    for (const [path, metadata] of this.index.entries()) {
      estimatedBytes += path.length * 2; // UTF-16 encoding
      estimatedBytes += 200; // Base metadata object
      estimatedBytes += (metadata.imports?.length || 0) * 30; // Import paths
      estimatedBytes += (metadata.exports?.length || 0) * 20; // Export names
    }
    
    // Languages set: string storage
    for (const lang of this.languages) {
      estimatedBytes += lang.length * 2;
    }
    
    // Dependencies map
    for (const [name, version] of this.dependencies.entries()) {
      estimatedBytes += (name.length + version.length) * 2;
    }
    
    // Frameworks array
    for (const framework of this.frameworks) {
      estimatedBytes += framework.length * 2;
    }
    
    // File content cache
    let cacheBytes = 0;
    for (const content of this.fileContentCache.values()) {
      cacheBytes += content.length * 2; // UTF-16 encoding
    }
    estimatedBytes += cacheBytes;

    // Update peak usage
    if (estimatedBytes > this.metrics.memoryPeakUsage) {
      this.metrics.memoryPeakUsage = estimatedBytes;
    }

    return {
      indexedFiles: indexSize,
      languages: languagesSize,
      dependencies: dependenciesSize,
      frameworks: this.frameworks.length,
      cachedFiles: cacheSize,
      lazyMode: this.isLazyMode,
      analyzedFiles: this.lazyLoadedPaths.size,
      estimatedMemoryBytes: estimatedBytes,
      estimatedMemoryMB: (estimatedBytes / (1024 * 1024)).toFixed(2),
      cacheMemoryMB: (cacheBytes / (1024 * 1024)).toFixed(2),
      peakMemoryMB: (this.metrics.memoryPeakUsage / (1024 * 1024)).toFixed(2),
      cacheHitRate: this.metrics.cacheHits + this.metrics.cacheMisses > 0 
        ? ((this.metrics.cacheHits / (this.metrics.cacheHits + this.metrics.cacheMisses)) * 100).toFixed(1) + '%'
        : '0%'
    };
  }

  /**
   * Get performance metrics
   * @returns {Object} Performance metrics
   */
  getPerformanceMetrics() {
    return {
      ...this.metrics,
      indexingTimeMs: this.metrics.indexingTime,
      indexingTimeSeconds: (this.metrics.indexingTime / 1000).toFixed(2),
      filesPerSecond: this.metrics.indexingTime > 0 
        ? Math.round(this.metrics.filesProcessed / (this.metrics.indexingTime / 1000))
        : 0,
      cacheHitRate: this.metrics.cacheHits + this.metrics.cacheMisses > 0 
        ? ((this.metrics.cacheHits / (this.metrics.cacheHits + this.metrics.cacheMisses)) * 100).toFixed(1) + '%'
        : '0%'
    };
  }

  /**
   * Log performance metrics to console
   * @private
   */
  _logPerformanceMetrics() {
    const metrics = this.getPerformanceMetrics();
    const memory = this.getMemoryUsage();
    
    console.log('\n📊 Indexing Performance Metrics:');
    console.log(`   ⏱️  Time: ${metrics.indexingTimeSeconds}s`);
    console.log(`   📁 Files: ${metrics.filesProcessed} (${metrics.filesPerSecond}/s)`);
    console.log(`   🧠 Memory: ${memory.estimatedMemoryMB}MB (peak: ${memory.peakMemoryMB}MB)`);
    console.log(`   💾 Cache: ${memory.cachedFiles} files, ${memory.cacheMemoryMB}MB, ${memory.cacheHitRate} hit rate`);
    console.log(`   🚀 Mode: ${this.isLazyMode ? 'Lazy' : 'Eager'}`);
    if (this.isLazyMode) {
      console.log(`   🔍 Analyzed: ${memory.analyzedFiles}/${memory.indexedFiles} files`);
    }
  }

  /**
   * Optimize memory usage by clearing caches and unused data
   * @param {Object} options - Optimization options
   * @returns {Object} Optimization results
   */
  optimizeMemory(options = {}) {
    const beforeMemory = this.getMemoryUsage();
    let optimizations = [];

    // Clear file content cache if requested or if it's large
    if (options.clearCache || this.fileContentCache.size > this.options.cacheSize * 2) {
      const cacheSize = this.fileContentCache.size;
      this._clearCache();
      optimizations.push(`Cleared ${cacheSize} cached files`);
    }

    // In lazy mode, clear analysis data for files not recently accessed
    if (this.isLazyMode && options.clearUnusedAnalysis) {
      let cleared = 0;
      for (const [filePath, metadata] of this.index.entries()) {
        if (metadata.analyzed && !this.cacheAccessOrder.includes(filePath)) {
          metadata.analyzed = false;
          metadata.imports = [];
          metadata.exports = [];
          this.lazyLoadedPaths.delete(filePath);
          cleared++;
        }
      }
      if (cleared > 0) {
        optimizations.push(`Reset analysis for ${cleared} unused files`);
      }
    }

    // Compact arrays and sets
    if (options.compact) {
      // Force garbage collection of unused language entries
      const usedLanguages = new Set();
      for (const metadata of this.index.values()) {
        if (metadata.language && metadata.language !== 'unknown') {
          usedLanguages.add(metadata.language);
        }
      }
      const removedLanguages = this.languages.size - usedLanguages.size;
      this.languages = usedLanguages;
      
      if (removedLanguages > 0) {
        optimizations.push(`Removed ${removedLanguages} unused language entries`);
      }
    }

    const afterMemory = this.getMemoryUsage();
    const savedMB = ((beforeMemory.estimatedMemoryBytes - afterMemory.estimatedMemoryBytes) / (1024 * 1024)).toFixed(2);

    return {
      optimizations,
      memoryBefore: beforeMemory.estimatedMemoryMB + 'MB',
      memoryAfter: afterMemory.estimatedMemoryMB + 'MB',
      memorySaved: savedMB + 'MB'
    };
  }

  /**
   * Clear the index to free memory
   * @returns {void}
   */
  clearIndex() {
    this.index.clear();
    this.languages.clear();
    this.frameworks = new Set();
    this.dependencies.clear();
    this.lastIndexed = null;
    this._clearCache();
    this.lazyLoadedPaths.clear();
    
    // Reset metrics
    this.metrics = {
      indexingTime: 0,
      filesProcessed: 0,
      cacheHits: 0,
      cacheMisses: 0,
      memoryPeakUsage: 0
    };
  }

  /**
   * Profile indexing performance on a directory
   * @param {string} targetPath - Path to profile (defaults to root)
   * @param {Object} options - Profiling options
   * @returns {Promise<Object>} Profiling results
   */
  async profileIndexing(targetPath = null, options = {}) {
    const originalProfiling = this.options.enableProfiling;
    this.options.enableProfiling = true;

    const startMemory = process.memoryUsage();
    const startTime = Date.now();

    try {
      if (targetPath) {
        // Profile specific directory
        const files = await fg('**/*', {
          cwd: path.resolve(this.rootPath, targetPath),
          dot: false,
          ignore: this._getIgnorePatterns(),
          onlyFiles: true,
          absolute: false,
          stats: true
        });

        console.log(`🎯 Profiling directory: ${targetPath}`);
        console.log(`📁 Files found: ${files.length}`);

        if (files.length > this.options.lazyLoadThreshold) {
          await this._buildIndexLazy(files);
        } else {
          await this._buildIndexEager(files);
        }
      } else {
        // Profile full indexing
        await this.buildIndex();
      }

      const endTime = Date.now();
      const endMemory = process.memoryUsage();

      const results = {
        duration: endTime - startTime,
        filesProcessed: this.metrics.filesProcessed,
        filesPerSecond: Math.round(this.metrics.filesProcessed / ((endTime - startTime) / 1000)),
        memoryUsage: {
          heapUsedMB: ((endMemory.heapUsed - startMemory.heapUsed) / 1024 / 1024).toFixed(2),
          heapTotalMB: ((endMemory.heapTotal - startMemory.heapTotal) / 1024 / 1024).toFixed(2),
          rssMB: ((endMemory.rss - startMemory.rss) / 1024 / 1024).toFixed(2)
        },
        indexMemory: this.getMemoryUsage(),
        performance: this.getPerformanceMetrics(),
        recommendations: this._getPerformanceRecommendations()
      };

      console.log('\n🏁 Profiling Results:');
      console.log(`   ⏱️  Duration: ${results.duration}ms`);
      console.log(`   📊 Throughput: ${results.filesPerSecond} files/second`);
      console.log(`   🧠 Heap Used: ${results.memoryUsage.heapUsedMB}MB`);
      console.log(`   💾 RSS: ${results.memoryUsage.rssMB}MB`);

      return results;
    } finally {
      this.options.enableProfiling = originalProfiling;
    }
  }

  /**
   * Get performance recommendations based on current metrics
   * @returns {Array<string>} Array of recommendations
   * @private
   */
  _getPerformanceRecommendations() {
    const recommendations = [];
    const memory = this.getMemoryUsage();
    const metrics = this.getPerformanceMetrics();

    // Memory recommendations
    if (parseFloat(memory.estimatedMemoryMB) > 500) {
      recommendations.push('Consider enabling lazy loading for large codebases (set lazyLoadThreshold lower)');
    }

    if (parseFloat(memory.cacheMemoryMB) > 100) {
      recommendations.push('File cache is large - consider reducing cacheSize option');
    }

    // Performance recommendations
    if (metrics.filesPerSecond < 100) {
      recommendations.push('Indexing is slow - consider increasing batchSize or enabling lazy mode');
    }

    if (parseFloat(memory.cacheHitRate) < 50 && this.metrics.cacheHits + this.metrics.cacheMisses > 10) {
      recommendations.push('Low cache hit rate - consider increasing cache size or reviewing access patterns');
    }

    // Lazy loading recommendations
    if (this.isLazyMode && this.lazyLoadedPaths.size / this.index.size > 0.8) {
      recommendations.push('Most files have been analyzed - consider switching to eager mode');
    }

    if (!this.isLazyMode && this.index.size > 5000) {
      recommendations.push('Large codebase detected - consider enabling lazy loading');
    }

    return recommendations;
  }

  /**
   * Benchmark different indexing strategies
   * @param {Object} options - Benchmark options
   * @returns {Promise<Object>} Benchmark results
   */
  async benchmarkStrategies(options = {}) {
    const strategies = [
      { name: 'Eager (small batch)', lazyLoadThreshold: Infinity, batchSize: 10 },
      { name: 'Eager (medium batch)', lazyLoadThreshold: Infinity, batchSize: 50 },
      { name: 'Eager (large batch)', lazyLoadThreshold: Infinity, batchSize: 200 },
      { name: 'Lazy (small batch)', lazyLoadThreshold: 100, batchSize: 10 },
      { name: 'Lazy (medium batch)', lazyLoadThreshold: 100, batchSize: 50 },
      { name: 'Lazy (large batch)', lazyLoadThreshold: 100, batchSize: 200 }
    ];

    const results = [];

    console.log('🏃 Benchmarking indexing strategies...\n');

    for (const strategy of strategies) {
      // Save original options
      const originalOptions = { ...this.options };
      
      // Apply strategy options
      Object.assign(this.options, strategy);
      
      try {
        const startTime = Date.now();
        const startMemory = process.memoryUsage();
        
        await this.buildIndex();
        
        const endTime = Date.now();
        const endMemory = process.memoryUsage();
        
        const result = {
          strategy: strategy.name,
          duration: endTime - startTime,
          filesProcessed: this.metrics.filesProcessed,
          filesPerSecond: Math.round(this.metrics.filesProcessed / ((endTime - startTime) / 1000)),
          memoryUsedMB: ((endMemory.heapUsed - startMemory.heapUsed) / 1024 / 1024).toFixed(2),
          indexMemoryMB: this.getMemoryUsage().estimatedMemoryMB
        };
        
        results.push(result);
        
        console.log(`✅ ${strategy.name}: ${result.duration}ms, ${result.filesPerSecond} files/s, ${result.memoryUsedMB}MB heap`);
        
        // Clear for next test
        this.clearIndex();
        
      } catch (error) {
        console.log(`❌ ${strategy.name}: Failed - ${error.message}`);
      } finally {
        // Restore original options
        Object.assign(this.options, originalOptions);
      }
    }

    // Find best strategy
    const fastest = results.reduce((best, current) => 
      current.filesPerSecond > best.filesPerSecond ? current : best
    );
    
    const mostMemoryEfficient = results.reduce((best, current) => 
      parseFloat(current.memoryUsedMB) < parseFloat(best.memoryUsedMB) ? current : best
    );

    console.log(`\n🏆 Fastest: ${fastest.strategy} (${fastest.filesPerSecond} files/s)`);
    console.log(`🧠 Most Memory Efficient: ${mostMemoryEfficient.strategy} (${mostMemoryEfficient.memoryUsedMB}MB)`);

    return {
      results,
      fastest,
      mostMemoryEfficient,
      recommendations: this._getBenchmarkRecommendations(results)
    };
  }

  /**
   * Get recommendations based on benchmark results
   * @param {Array} results - Benchmark results
   * @returns {Array<string>} Recommendations
   * @private
   */
  _getBenchmarkRecommendations(results) {
    const recommendations = [];
    
    if (results.length === 0) return recommendations;

    const avgDuration = results.reduce((sum, r) => sum + r.duration, 0) / results.length;
    const avgMemory = results.reduce((sum, r) => sum + parseFloat(r.memoryUsedMB), 0) / results.length;

    // Performance recommendations
    const fastStrategies = results.filter(r => r.duration < avgDuration * 0.8);
    if (fastStrategies.length > 0) {
      recommendations.push(`Fast strategies: ${fastStrategies.map(s => s.strategy).join(', ')}`);
    }

    // Memory recommendations
    const memoryEfficientStrategies = results.filter(r => parseFloat(r.memoryUsedMB) < avgMemory * 0.8);
    if (memoryEfficientStrategies.length > 0) {
      recommendations.push(`Memory efficient strategies: ${memoryEfficientStrategies.map(s => s.strategy).join(', ')}`);
    }

    return recommendations;
  }

  /**
   * Get statistics about the indexed codebase
   * @returns {Object} Statistics object
   */
  getStats() {
    const stats = {
      totalFiles: this.index.size,
      languages: Array.from(this.languages),
      frameworks: Array.from(this.frameworks),
      dependencies: Array.from(this.dependencies),
      filesByLanguage: {},
      totalLines: 0,
      totalSize: 0
    };

    // Calculate file counts by language
    for (const [filePath, metadata] of this.index.entries()) {
      const language = metadata.language || 'unknown';
      stats.filesByLanguage[language] = (stats.filesByLanguage[language] || 0) + 1;
      stats.totalLines += metadata.lines || 0;
      stats.totalSize += metadata.size || 0;
    }

    return stats;
  }
}

export default CodebaseIndexer;
