// codestral/src/ai/MistralClient.js
import { Mistral } from '@mistralai/mistralai';

/**
 * MistralClient - Interface for Mistral AI API
 * Handles authentication, message sending, streaming, and tool calling
 */
class MistralClient {
  /**
   * @param {string} apiKey - Mistral AI API key
   * @param {Object} options - Configuration options
   * @param {string} options.model - Model to use (default: mistral-large-latest)
   * @param {number} options.maxTokens - Maximum tokens in response
   * @param {number} options.temperature - Temperature for response generation
   */
  constructor(apiKey, options = {}) {
    if (!apiKey) {
      throw new Error('API key is required for MistralClient');
    }

    this.apiKey = apiKey;
    this.model = options.model || 'mistral-large-latest';
    this.maxTokens = options.maxTokens || 4096;
    this.temperature = options.temperature || 0.7;
    this.tools = [];

    try {
      this.client = new Mistral({ apiKey: this.apiKey });
    } catch (error) {
      throw new Error(`Failed to initialize Mistral client: ${error.message}`);
    }
  }

  /**
   * Send a message to Mistral AI and get a complete response
   * @param {Array} messages - Array of message objects with role and content
   * @param {Object} options - Optional parameters to override defaults
   * @returns {Promise<Object>} Response from Mistral AI
   */
  async sendMessage(messages, options = {}) {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('Messages must be a non-empty array');
    }

    // Validate message format
    for (const msg of messages) {
      if (!msg.role || !msg.content) {
        throw new Error('Each message must have role and content properties');
      }
    }

    const requestOptions = {
      model: options.model || this.model,
      messages: messages,
      maxTokens: options.maxTokens || this.maxTokens,
      temperature: options.temperature !== undefined ? options.temperature : this.temperature,
    };

    try {
      const response = await this.client.chat.complete(requestOptions);
      
      if (!response || !response.choices || response.choices.length === 0) {
        throw new Error('Invalid response from Mistral AI');
      }

      return {
        content: response.choices[0].message.content,
        role: response.choices[0].message.role,
        finishReason: response.choices[0].finishReason,
        usage: response.usage,
        model: response.model,
      };
    } catch (error) {
      // Handle authentication errors
      if (error.message?.includes('401') || error.message?.includes('Unauthorized')) {
        throw new Error('Authentication failed: Invalid API key');
      }
      
      // Handle other API errors
      if (error.message?.includes('API')) {
        throw new Error(`Mistral AI API error: ${error.message}`);
      }

      throw error;
    }
  }

  /**
   * Stream a response from Mistral AI with real-time token delivery
   * @param {Array} messages - Array of message objects
   * @param {Function} onChunk - Callback function called for each token chunk
   * @param {Object} options - Optional parameters
   * @returns {Promise<Object>} Complete response after streaming finishes
   */
  async streamResponse(messages, onChunk, options = {}) {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('Messages must be a non-empty array');
    }

    if (typeof onChunk !== 'function') {
      throw new Error('onChunk must be a function');
    }

    const requestOptions = {
      model: options.model || this.model,
      messages: messages,
      maxTokens: options.maxTokens || this.maxTokens,
      temperature: options.temperature !== undefined ? options.temperature : this.temperature,
    };

    try {
      const stream = await this.client.chat.stream(requestOptions);
      
      let fullContent = '';
      let fullResponse = null;

      for await (const chunk of stream) {
        if (chunk.data.choices && chunk.data.choices.length > 0) {
          const delta = chunk.data.choices[0].delta;
          
          if (delta.content) {
            fullContent += delta.content;
            onChunk(delta.content, chunk.data);
          }

          // Store the full response data
          if (chunk.data.choices[0].finishReason) {
            fullResponse = chunk.data;
          }
        }
      }

      return {
        content: fullContent,
        role: 'assistant',
        finishReason: fullResponse?.choices[0]?.finishReason || 'stop',
        usage: fullResponse?.usage,
        model: fullResponse?.model || this.model,
      };
    } catch (error) {
      if (error.message?.includes('401') || error.message?.includes('Unauthorized')) {
        throw new Error('Authentication failed: Invalid API key');
      }
      throw error;
    }
  }

  /**
   * Retry a function with exponential backoff
   * @param {Function} fn - Async function to retry
   * @param {number} maxRetries - Maximum number of retry attempts
   * @returns {Promise<any>} Result of the function
   */
  async retryWithBackoff(fn, maxRetries = 3) {
    let lastError;
    
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;
        
        // Don't retry on authentication errors
        if (error.message?.includes('Authentication failed')) {
          throw error;
        }

        // Check if we should retry
        if (attempt < maxRetries) {
          const shouldRetry = this.shouldRetry(error);
          
          if (shouldRetry) {
            const delay = this.calculateBackoff(attempt);
            await this.sleep(delay);
            continue;
          }
        }
        
        throw error;
      }
    }
    
    throw lastError;
  }

  /**
   * Determine if an error should trigger a retry
   * @param {Error} error - The error to check
   * @returns {boolean} Whether to retry
   */
  shouldRetry(error) {
    const errorMessage = error.message?.toLowerCase() || '';
    
    // Retry on rate limit errors
    if (errorMessage.includes('rate limit') || errorMessage.includes('429')) {
      return true;
    }
    
    // Retry on network errors
    if (errorMessage.includes('network') || errorMessage.includes('timeout') || 
        errorMessage.includes('econnreset') || errorMessage.includes('econnrefused')) {
      return true;
    }
    
    // Retry on 5xx server errors
    if (errorMessage.includes('500') || errorMessage.includes('502') || 
        errorMessage.includes('503') || errorMessage.includes('504')) {
      return true;
    }
    
    return false;
  }

  /**
   * Calculate backoff delay with jitter
   * @param {number} attempt - Current attempt number (0-indexed)
   * @returns {number} Delay in milliseconds
   */
  calculateBackoff(attempt) {
    // Exponential backoff: 1s, 2s, 4s
    const baseDelay = Math.pow(2, attempt) * 1000;
    
    // Add jitter (random value between 0 and 20% of base delay)
    const jitter = Math.random() * baseDelay * 0.2;
    
    return baseDelay + jitter;
  }

  /**
   * Handle rate limit errors
   * @param {Error} error - Rate limit error
   * @returns {Promise<void>}
   */
  async handleRateLimit(error) {
    // Extract retry-after header if available
    const retryAfter = error.response?.headers?.['retry-after'];
    
    if (retryAfter) {
      const delay = parseInt(retryAfter) * 1000;
      await this.sleep(delay);
    } else {
      // Default to 60 seconds if no retry-after header
      await this.sleep(60000);
    }
  }

  /**
   * Sleep for a specified duration
   * @param {number} ms - Milliseconds to sleep
   * @returns {Promise<void>}
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Register tools for function calling
   * @param {Array} toolDefinitions - Array of tool definition objects
   */
  registerTools(toolDefinitions) {
    if (!Array.isArray(toolDefinitions)) {
      throw new Error('Tool definitions must be an array');
    }

    // Validate tool definitions
    for (const tool of toolDefinitions) {
      if (!tool.type || !tool.function) {
        throw new Error('Each tool must have type and function properties');
      }
      if (!tool.function.name || !tool.function.description) {
        throw new Error('Each tool function must have name and description');
      }
    }

    this.tools = toolDefinitions;
  }

  /**
   * Send a message with tool definitions for function calling
   * @param {Array} messages - Array of message objects
   * @param {Array} tools - Tool definitions (optional, uses registered tools if not provided)
   * @param {Object} options - Optional parameters
   * @returns {Promise<Object>} Response from Mistral AI with potential tool calls
   */
  async sendWithTools(messages, tools = null, options = {}) {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('Messages must be a non-empty array');
    }

    const toolsToUse = tools || this.tools;
    
    if (!toolsToUse || toolsToUse.length === 0) {
      throw new Error('No tools available. Register tools first or provide them as parameter');
    }

    const requestOptions = {
      model: options.model || this.model,
      messages: messages,
      tools: toolsToUse,
      toolChoice: options.toolChoice || 'auto',
      maxTokens: options.maxTokens || this.maxTokens,
      temperature: options.temperature !== undefined ? options.temperature : this.temperature,
    };

    try {
      const response = await this.client.chat.complete(requestOptions);
      
      if (!response || !response.choices || response.choices.length === 0) {
        throw new Error('Invalid response from Mistral AI');
      }

      const message = response.choices[0].message;

      return {
        content: message.content || null,
        role: message.role,
        toolCalls: message.toolCalls || null,
        finishReason: response.choices[0].finishReason,
        usage: response.usage,
        model: response.model,
      };
    } catch (error) {
      if (error.message?.includes('401') || error.message?.includes('Unauthorized')) {
        throw new Error('Authentication failed: Invalid API key');
      }
      throw error;
    }
  }

  /**
   * Stream a response with tool support
   * @param {Array} messages - Array of message objects
   * @param {Function} onChunk - Callback for content chunks
   * @param {Function} onToolCall - Callback for tool calls
   * @param {Array} tools - Tool definitions
   * @param {Object} options - Optional parameters
   * @returns {Promise<Object>} Complete response
   */
  async streamWithTools(messages, onChunk, onToolCall, tools = null, options = {}) {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('Messages must be a non-empty array');
    }

    if (typeof onChunk !== 'function') {
      throw new Error('onChunk must be a function');
    }

    const toolsToUse = tools || this.tools;
    
    if (!toolsToUse || toolsToUse.length === 0) {
      throw new Error('No tools available');
    }

    const requestOptions = {
      model: options.model || this.model,
      messages: messages,
      tools: toolsToUse,
      toolChoice: options.toolChoice || 'auto',
      maxTokens: options.maxTokens || this.maxTokens,
      temperature: options.temperature !== undefined ? options.temperature : this.temperature,
    };

    try {
      const stream = await this.client.chat.stream(requestOptions);
      
      let fullContent = '';
      let toolCalls = [];
      let fullResponse = null;

      for await (const chunk of stream) {
        if (chunk.data.choices && chunk.data.choices.length > 0) {
          const delta = chunk.data.choices[0].delta;
          
          // Handle content chunks
          if (delta.content) {
            fullContent += delta.content;
            onChunk(delta.content, chunk.data);
          }

          // Handle tool calls
          if (delta.toolCalls) {
            toolCalls = delta.toolCalls;
            if (onToolCall) {
              onToolCall(delta.toolCalls, chunk.data);
            }
          }

          // Store final response
          if (chunk.data.choices[0].finishReason) {
            fullResponse = chunk.data;
          }
        }
      }

      return {
        content: fullContent || null,
        role: 'assistant',
        toolCalls: toolCalls.length > 0 ? toolCalls : null,
        finishReason: fullResponse?.choices[0]?.finishReason || 'stop',
        usage: fullResponse?.usage,
        model: fullResponse?.model || this.model,
      };
    } catch (error) {
      if (error.message?.includes('401') || error.message?.includes('Unauthorized')) {
        throw new Error('Authentication failed: Invalid API key');
      }
      throw error;
    }
  }
}

export default MistralClient;
