// codestral/src/parser/InputParser.js
/**
 * InputParser - Parses user input to extract mentions, commands, and sanitize content
 */
export default class InputParser {
    constructor() {
        // Regex pattern to match @-mentions (e.g., @filename.js, @src/file.js, @"file with spaces.js")
        // Matches: @"quoted", @'quoted', or @unquoted (alphanumeric, -, _, /, .)
        this.mentionPattern = /@(?:"([^"]+)"|'([^']+)'|([\w\-./]+))/g;

        // Regex pattern to match special commands (e.g., /help, /clear, /search)
        this.commandPattern = /^\/(\w+)(?:\s+(.*))?$/;

        // List of supported commands
        this.supportedCommands = [
            'help',
            'clear',
            'search',
            'files',
            'config',
            'setup',
            'set',
            'mode',
            'tools'
        ];
    }

    /**
     * Parse user input to extract mentions, commands, and sanitized content
     * @param {string} input - Raw user input
     * @returns {Object} Parsed input object
     */
    parse(input) {
        if (!input || typeof input !== 'string') {
            return {
                original: input || '',
                sanitized: '',
                mentions: [],
                command: null,
                isCommand: false
            };
        }

        const mentions = this.extractMentions(input);
        const command = this.extractCommands(input);
        const sanitized = this.sanitizeInput(input);

        return {
            original: input,
            sanitized,
            mentions,
            command,
            isCommand: command !== null
        };
    }

    /**
     * Extract @-mentions from input
     * @param {string} input - User input
     * @returns {Array<string>} Array of mentioned file paths
     */
    extractMentions(input) {
        if (!input || typeof input !== 'string') {
            return [];
        }

        const mentions = [];
        let match;

        // Reset regex state
        this.mentionPattern.lastIndex = 0;

        while ((match = this.mentionPattern.exec(input)) !== null) {
            // Extract the captured group (quoted or unquoted)
            const mention = match[1] || match[2] || match[3];
            if (mention) {
                mentions.push(mention);
            }
        }

        return mentions;
    }

    /**
     * Extract special commands from input
     * @param {string} input - User input
     * @returns {Object|null} Command object or null if no command found
     */
    extractCommands(input) {
        if (!input || typeof input !== 'string') {
            return null;
        }

        const trimmedInput = input.trim();
        const match = this.commandPattern.exec(trimmedInput);

        if (!match) {
            return null;
        }

        const commandName = match[1].toLowerCase();
        const commandArgs = match[2] ? match[2].trim() : '';

        // Check if command is supported
        if (!this.supportedCommands.includes(commandName)) {
            return {
                name: commandName,
                args: commandArgs,
                isSupported: false
            };
        }

        return {
            name: commandName,
            args: commandArgs,
            isSupported: true
        };
    }

    /**
     * Sanitize user input to prevent injection attacks and normalize whitespace
     * @param {string} input - User input
     * @returns {string} Sanitized input
     */
    sanitizeInput(input) {
        if (!input || typeof input !== 'string') {
            return '';
        }

        // Trim leading/trailing whitespace
        let sanitized = input.trim();

        // Normalize multiple spaces to single space (but preserve newlines)
        sanitized = sanitized.replace(/[^\S\n]+/g, ' ');

        // Remove null bytes
        sanitized = sanitized.replace(/\0/g, '');

        // Remove control characters except newlines, tabs, and carriage returns
        sanitized = sanitized.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');

        return sanitized;
    }

    /**
     * Check if input contains any mentions
     * @param {string} input - User input
     * @returns {boolean} True if input contains mentions
     */
    hasMentions(input) {
        return this.extractMentions(input).length > 0;
    }

    /**
     * Check if input is a command
     * @param {string} input - User input
     * @returns {boolean} True if input is a command
     */
    isCommand(input) {
        return this.extractCommands(input) !== null;
    }

    /**
     * Remove mentions from input text
     * @param {string} input - User input
     * @returns {string} Input with mentions removed
     */
    removeMentions(input) {
        if (!input || typeof input !== 'string') {
            return '';
        }

        // Remove mentions and normalize multiple spaces
        return input.replace(this.mentionPattern, '').replace(/\s+/g, ' ').trim();
    }
}
