// codestral/src/tools/handlers/approveProject.js
/**
 * Approve project tool handler
 * Handles project approval workflow and executes project creation
 */

import { createCreateMultipleFilesHandler } from './createMultipleFiles.js';

/**
 * Create approve_project tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createApproveProjectHandler(fileSystemManager) {
  const createMultipleFilesHandler = createCreateMultipleFilesHandler(fileSystemManager);
  
  return async (params) => {
    const { 
      approved, 
      files, 
      projectName, 
      template,
      createInSubdirectory = true,
      directoryName = null
    } = params;
    
    if (!approved) {
      return {
        success: false,
        message: 'Project creation cancelled by user',
        action: 'cancelled'
      };
    }
    
    if (!files || !Array.isArray(files)) {
      throw new Error('Files array is required for project approval');
    }
    
    if (files.length === 0) {
      throw new Error('No files to create');
    }
    
    try {
      let finalFiles = files;
      let projectDirectory = '';
      
      // Create in subdirectory if requested
      if (createInSubdirectory) {
        projectDirectory = directoryName || projectName.toLowerCase().replace(/[^a-z0-9-]/g, '-');
        finalFiles = files.map(file => ({
          ...file,
          path: `${projectDirectory}/${file.path}`
        }));
      }
      
      console.log(`Creating project: ${projectName}`);
      console.log(`Template: ${template}`);
      console.log(`Files: ${finalFiles.length}`);
      if (projectDirectory) {
        console.log(`Directory: ${projectDirectory}`);
      }
      
      // Create the project files
      const result = await createMultipleFilesHandler({
        files: finalFiles,
        rollbackOnError: true,
        createDirectories: true
      });
      
      if (result.success) {
        const setupInstructions = generatePostCreationInstructions(
          template, 
          projectName, 
          projectDirectory
        );
        
        return {
          success: true,
          message: `✅ Project "${projectName}" created successfully!`,
          projectName,
          template,
          projectDirectory: projectDirectory || '.',
          filesCreated: result.successCount,
          totalSize: result.totalSize,
          setupInstructions,
          nextSteps: [
            projectDirectory ? `Navigate to project: cd ${projectDirectory}` : 'You are in the project directory',
            ...setupInstructions,
            'Start developing! 🚀'
          ],
          action: 'created'
        };
      } else {
        return {
          success: false,
          message: `❌ Project creation failed: ${result.message}`,
          errors: result.errors,
          action: 'failed'
        };
      }
      
    } catch (error) {
      return {
        success: false,
        message: `❌ Project creation failed: ${error.message}`,
        error: error.message,
        action: 'failed'
      };
    }
  };
}

/**
 * Generate post-creation setup instructions
 * @param {string} template - Template name
 * @param {string} projectName - Project name
 * @param {string} projectDirectory - Project directory
 * @returns {Array} Setup instructions
 */
function generatePostCreationInstructions(template, projectName, projectDirectory) {
  const instructions = [];
  
  // Navigation instruction
  if (projectDirectory) {
    instructions.push(`cd ${projectDirectory}`);
  }
  
  // Template-specific setup
  switch (template) {
    case 'static-website':
      instructions.push('Open index.html in your web browser');
      break;
      
    case 'react-app':
      instructions.push('npm install');
      instructions.push('npm run dev');
      instructions.push('Open http://localhost:3000');
      break;
      
    case 'vue-app':
      instructions.push('npm install');
      instructions.push('npm run serve');
      instructions.push('Open http://localhost:8080');
      break;
      
    case 'next-app':
      instructions.push('npm install');
      instructions.push('npm run dev');
      instructions.push('Open http://localhost:3000');
      break;
      
    case 'express-api':
      instructions.push('npm install');
      instructions.push('cp .env.example .env');
      instructions.push('npm run dev');
      instructions.push('API available at http://localhost:3000');
      break;
      
    case 'full-stack':
      instructions.push('npm run install-all');
      instructions.push('cd server && cp .env.example .env && cd ..');
      instructions.push('npm run dev');
      instructions.push('Frontend: http://localhost:3000, Backend: http://localhost:3001');
      break;
      
    case 'flask-api':
      instructions.push('python -m venv venv');
      instructions.push('source venv/bin/activate  # or venv\\\\Scripts\\\\activate on Windows');
      instructions.push('pip install -r requirements.txt');
      instructions.push('cp .env.example .env');
      instructions.push('python app.py');
      instructions.push('API available at http://localhost:5000');
      break;
      
    case 'fastapi-api':
      instructions.push('python -m venv venv');
      instructions.push('source venv/bin/activate  # or venv\\\\Scripts\\\\activate on Windows');
      instructions.push('pip install -r requirements.txt');
      instructions.push('cp .env.example .env');
      instructions.push('uvicorn main:app --reload');
      instructions.push('API available at http://localhost:8000');
      instructions.push('Docs at http://localhost:8000/docs');
      break;
      
    case 'django-app':
      instructions.push('python -m venv venv');
      instructions.push('source venv/bin/activate  # or venv\\\\Scripts\\\\activate on Windows');
      instructions.push('pip install -r requirements.txt');
      instructions.push('cp .env.example .env');
      instructions.push('python manage.py migrate');
      instructions.push('python manage.py createsuperuser');
      instructions.push('python manage.py runserver');
      instructions.push('App available at http://localhost:8000');
      break;
      
    default:
      instructions.push('Follow the README.md for setup instructions');
  }
  
  return instructions;
}

/**
 * Tool schema for approve_project
 */
export const approveProjectSchema = {
  description: 'Approve and execute project creation after reviewing the project plan',
  parameters: {
    type: 'object',
    properties: {
      approved: {
        type: 'boolean',
        description: 'Whether the user approves the project creation (true to create, false to cancel)'
      },
      files: {
        type: 'array',
        description: 'Array of files from the project plan to create',
        items: {
          type: 'object',
          properties: {
            path: { type: 'string' },
            content: { type: 'string' }
          },
          required: ['path', 'content']
        }
      },
      projectName: {
        type: 'string',
        description: 'Name of the project being created'
      },
      template: {
        type: 'string',
        description: 'Template used for the project'
      },
      createInSubdirectory: {
        type: 'boolean',
        description: 'Whether to create the project in a subdirectory (default: true)',
        default: true
      },
      directoryName: {
        type: 'string',
        description: 'Custom directory name (defaults to project name in kebab-case)'
      }
    },
    required: ['approved']
  }
};
