// codestral/src/tools/handlers/createDirectory.js
import fs from 'fs/promises';

/**
 * Create directory tool handler
 * Creates a new directory (and parent directories if needed)
 */

/**
 * Create create_directory tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createCreateDirectoryHandler(fileSystemManager) {
  return async (params) => {
    const { path } = params;

    if (!path) {
      throw new Error('Path parameter is required');
    }

    try {
      // Validate path
      const validatedPath = fileSystemManager.validatePath(path);

      // Create directory with recursive option
      await fs.mkdir(validatedPath, { recursive: true });

      // Check if directory was created
      const exists = await fileSystemManager.exists(path);

      if (!exists) {
        throw new Error('Directory creation failed');
      }

      return {
        path,
        success: true,
        message: `Directory created successfully: ${path}`
      };
    } catch (error) {
      throw new Error(`Failed to create directory: ${error.message}`);
    }
  };
}

/**
 * Tool schema for create_directory
 */
export const createDirectorySchema = {
  description: 'Create a new directory (creates parent directories if needed)',
  parameters: {
    type: 'object',
    properties: {
      path: {
        type: 'string',
        description: 'Directory path to create (relative to project root)'
      }
    },
    required: ['path']
  }
};
