// codestral/src/tools/handlers/createMultipleFiles.js
/**
 * Create multiple files tool handler
 * Creates multiple files in a single operation for project scaffolding
 * Enhanced with progress indicators, validation, and transaction-like behavior
 */

import fs from 'fs/promises';
import path from 'path';

/**
 * Validate file paths and content
 * @param {Array} files - Array of file objects
 * @returns {Object} Validation result
 */
function validateFiles(files) {
  const errors = [];
  const validFiles = [];

  for (let i = 0; i < files.length; i++) {
    const file = files[i];
    const fileIndex = `File ${i + 1}`;

    // Check required properties
    if (!file.path) {
      errors.push({
        index: i,
        file: fileIndex,
        error: 'Missing required property: path'
      });
      continue;
    }

    if (file.content === undefined || file.content === null) {
      errors.push({
        index: i,
        file: file.path,
        error: 'Missing required property: content'
      });
      continue;
    }

    // Validate path format
    if (typeof file.path !== 'string' || file.path.trim() === '') {
      errors.push({
        index: i,
        file: file.path,
        error: 'Path must be a non-empty string'
      });
      continue;
    }

    // Check for path traversal attempts
    const normalizedPath = path.normalize(file.path);
    if (normalizedPath.includes('..') || path.isAbsolute(normalizedPath)) {
      errors.push({
        index: i,
        file: file.path,
        error: 'Invalid path: absolute paths and path traversal not allowed'
      });
      continue;
    }

    // Validate content type
    if (typeof file.content !== 'string') {
      errors.push({
        index: i,
        file: file.path,
        error: 'Content must be a string'
      });
      continue;
    }

    // Check file size (limit to 10MB per file)
    const contentSize = Buffer.byteLength(file.content, 'utf8');
    if (contentSize > 10 * 1024 * 1024) {
      errors.push({
        index: i,
        file: file.path,
        error: `File too large: ${Math.round(contentSize / 1024 / 1024)}MB (max 10MB)`
      });
      continue;
    }

    validFiles.push({
      ...file,
      index: i,
      size: contentSize
    });
  }

  return { validFiles, errors };
}

/**
 * Create progress reporter function
 * @param {number} total - Total number of files
 * @param {Function} onProgress - Progress callback
 * @returns {Function} Progress reporter
 */
function createProgressReporter(total, onProgress) {
  let completed = 0;
  
  return (filePath, success = true) => {
    completed++;
    const percentage = Math.round((completed / total) * 100);
    
    if (onProgress) {
      onProgress({
        completed,
        total,
        percentage,
        currentFile: filePath,
        success
      });
    }
    
    // Log progress for large operations
    if (total > 10 && (completed % Math.ceil(total / 10) === 0 || completed === total)) {
      console.log(`Progress: ${completed}/${total} files (${percentage}%)`);
    }
  };
}

/**
 * Create create_multiple_files tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createCreateMultipleFilesHandler(fileSystemManager) {
  return async (params) => {
    const { files, rollbackOnError = true, createDirectories = true } = params;

    // Input validation
    if (!files || !Array.isArray(files)) {
      throw new Error('Files parameter must be an array');
    }

    if (files.length === 0) {
      throw new Error('Files array cannot be empty');
    }

    // Validate file limit for batch operations
    if (files.length > 1000) {
      throw new Error(`Too many files: ${files.length} (maximum 1000 files per batch)`);
    }

    console.log(`Starting batch file creation: ${files.length} files`);

    // Validate all files first
    const { validFiles, errors: validationErrors } = validateFiles(files);
    
    // If rollback is enabled and there are validation errors, fail fast
    if (rollbackOnError && validationErrors.length > 0) {
      return {
        totalFiles: files.length,
        successCount: 0,
        errorCount: validationErrors.length,
        success: false,
        createdFiles: [],
        errors: validationErrors,
        message: `Validation failed for ${validationErrors.length} file(s)`
      };
    }

    const results = [];
    const errors = [...validationErrors]; // Start with validation errors
    const createdFiles = [];
    const createdDirectories = new Set();

    // Progress tracking
    const reportProgress = createProgressReporter(validFiles.length, null);

    try {
      // Calculate total size
      const totalSize = validFiles.reduce((sum, file) => sum + file.size, 0);
      const totalSizeMB = Math.round(totalSize / 1024 / 1024 * 100) / 100;
      
      console.log(`Total size: ${totalSizeMB}MB across ${validFiles.length} files`);

      // Process each file
      for (const file of validFiles) {
        try {
          // Create directory if needed
          if (createDirectories) {
            const dir = path.dirname(file.path);
            if (dir !== '.' && !createdDirectories.has(dir)) {
              try {
                await fileSystemManager.createDirectory(dir);
                createdDirectories.add(dir);
              } catch (dirError) {
                // Directory might already exist, continue
                if (!dirError.message.includes('already exists')) {
                  throw dirError;
                }
              }
            }
          }

          // Write file
          await fileSystemManager.writeFile(file.path, file.content);
          
          createdFiles.push(file.path);
          results.push({
            path: file.path,
            success: true,
            size: file.size
          });

          reportProgress(file.path, true);

        } catch (error) {
          const fileError = {
            index: file.index,
            file: file.path,
            error: error.message
          };
          
          errors.push(fileError);
          reportProgress(file.path, false);

          // If rollback is enabled and we have an error, stop and rollback
          if (rollbackOnError) {
            console.log(`Error creating ${file.path}, rolling back...`);
            await rollbackCreatedFiles(createdFiles, fileSystemManager);
            
            throw new Error(`File creation failed at ${file.path}: ${error.message}. All changes rolled back.`);
          }
        }
      }

      const successCount = results.length;
      const errorCount = errors.length;
      const isSuccess = errorCount === 0;

      console.log(`Batch creation completed: ${successCount} success, ${errorCount} errors`);

      return {
        totalFiles: files.length,
        successCount,
        errorCount,
        success: isSuccess,
        createdFiles: results.map(r => r.path),
        createdDirectories: Array.from(createdDirectories),
        totalSize: totalSizeMB,
        errors: errors.length > 0 ? errors : undefined,
        message: isSuccess 
          ? `Successfully created ${successCount} file(s) (${totalSizeMB}MB)`
          : `Created ${successCount} file(s) with ${errorCount} error(s)`
      };

    } catch (error) {
      // If we get here, it's either a rollback scenario or unexpected error
      if (error.message.includes('rolled back')) {
        throw error; // Re-throw rollback errors as-is
      }
      
      throw new Error(`Batch file creation failed: ${error.message}`);
    }
  };
}

/**
 * Rollback created files in case of failure
 * @param {Array} createdFiles - Array of file paths that were created
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 */
async function rollbackCreatedFiles(createdFiles, fileSystemManager) {
  console.log(`Rolling back ${createdFiles.length} created files...`);
  
  for (const filePath of createdFiles.reverse()) {
    try {
      await fs.unlink(filePath);
      console.log(`Rolled back: ${filePath}`);
    } catch (error) {
      console.warn(`Failed to rollback ${filePath}: ${error.message}`);
    }
  }
}

/**
 * Tool schema for create_multiple_files
 */
export const createMultipleFilesSchema = {
  description: 'Create multiple files in a single operation for project scaffolding with enhanced validation and progress tracking',
  parameters: {
    type: 'object',
    properties: {
      files: {
        type: 'array',
        description: 'Array of file objects to create (max 1000 files)',
        maxItems: 1000,
        items: {
          type: 'object',
          properties: {
            path: {
              type: 'string',
              description: 'File path (relative to project root, no absolute paths or path traversal)'
            },
            content: {
              type: 'string',
              description: 'File content (max 10MB per file)'
            }
          },
          required: ['path', 'content']
        }
      },
      rollbackOnError: {
        type: 'boolean',
        description: 'Whether to rollback all created files if any file creation fails (default: true)',
        default: true
      },
      createDirectories: {
        type: 'boolean',
        description: 'Whether to automatically create parent directories (default: true)',
        default: true
      }
    },
    required: ['files']
  }
};
