// codestral/src/tools/handlers/executeCommand.js
/**
 * Execute command tool handler
 * Executes a shell command with safety checks
 */

/**
 * Create execute_command tool handler
 * @param {CommandExecutor} commandExecutor - Command executor instance
 * @returns {Function} Tool handler function
 */
export function createExecuteCommandHandler(commandExecutor) {
  return async (params) => {
    const { command, cwd } = params;

    if (!command) {
      throw new Error('Command parameter is required');
    }

    try {
      // Execute command
      const result = await commandExecutor.execute(command, { cwd });

      return {
        command,
        cwd: cwd || commandExecutor.cwd,
        success: result.success,
        stdout: result.stdout,
        stderr: result.stderr,
        exitCode: result.exitCode
      };
    } catch (error) {
      throw new Error(`Failed to execute command: ${error.message}`);
    }
  };
}

/**
 * Tool schema for execute_command
 */
export const executeCommandSchema = {
  description: 'Execute a shell command (requires user approval for dangerous operations)',
  parameters: {
    type: 'object',
    properties: {
      command: {
        type: 'string',
        description: 'Shell command to execute'
      },
      cwd: {
        type: 'string',
        description: 'Working directory for command execution (optional)'
      }
    },
    required: ['command']
  }
};
