// codestral/src/tools/handlers/getProjectTemplate.js
/**
 * Get project template tool handler
 * Returns predefined project templates for common project types
 * Enhanced to use the comprehensive template system
 */

import { getTemplateNames, generateTemplate } from '../../templates/index.js';

/**
 * Create get_project_template tool handler
 * @returns {Function} Tool handler function
 */
export function createGetProjectTemplateHandler() {
  return async (params) => {
    const { template, projectName = 'My Project', options = {} } = params;

    if (!template) {
      // Return list of available templates
      const templateNames = getTemplateNames();
      const availableTemplates = [];

      // Import and get template info for each template
      for (const templateName of templateNames) {
        try {
          const templateModule = await import(`../templates/${templateName}.js`);
          availableTemplates.push({
            id: templateName,
            name: templateModule.default.name,
            description: templateModule.default.description
          });
        } catch (error) {
          console.warn(`Failed to load template ${templateName}:`, error.message);
        }
      }

      return {
        success: true,
        availableTemplates,
        totalTemplates: availableTemplates.length,
        message: 'Available project templates. Specify a template parameter to generate files.',
        usage: 'Use template parameter with one of the available template IDs'
      };
    }

    try {
      // Generate template files
      const files = generateTemplate(template, projectName, options);
      
      // Calculate total size
      const totalSize = files.reduce((sum, file) => {
        return sum + Buffer.byteLength(file.content, 'utf8');
      }, 0);
      
      const totalSizeMB = Math.round(totalSize / 1024 / 1024 * 100) / 100;

      return {
        success: true,
        template,
        projectName,
        fileCount: files.length,
        totalSize: totalSizeMB,
        files,
        message: `Generated ${files.length} files for ${template} template (${totalSizeMB}MB total)`,
        nextSteps: [
          'Review the generated files',
          'Use create_multiple_files tool to create the project',
          'Follow the README.md instructions for setup'
        ]
      };

    } catch (error) {
      const templateNames = getTemplateNames();
      
      return {
        success: false,
        error: error.message,
        availableTemplates: templateNames,
        message: `Template "${template}" not found or failed to generate. Available templates: ${templateNames.join(', ')}`
      };
    }
  };
}

/**
 * Tool schema for get_project_template
 */
export const getProjectTemplateSchema = {
  description: 'Get a comprehensive predefined project template (static-website, react-app, express-api, full-stack, vue-app, next-app, flask-api, fastapi-api, django-app)',
  parameters: {
    type: 'object',
    properties: {
      template: {
        type: 'string',
        description: 'Template name (omit to see all available templates)',
        enum: [
          'static-website',
          'react-app', 
          'express-api',
          'full-stack',
          'vue-app',
          'next-app',
          'flask-api',
          'fastapi-api',
          'django-app'
        ]
      },
      projectName: {
        type: 'string',
        description: 'Project name to use in template files (defaults to "My Project")',
        default: 'My Project'
      },
      options: {
        type: 'object',
        description: 'Additional template-specific options',
        properties: {},
        additionalProperties: true
      }
    },
    required: []
  }
};
