// codestral/src/tools/handlers/listDirectory.js
/**
 * List directory tool handler
 * Lists files and directories in a given path
 */

/**
 * Create list_directory tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createListDirectoryHandler(fileSystemManager) {
  return async (params) => {
    const { path = '.', recursive = false } = params;

    try {
      // List directory contents
      const entries = await fileSystemManager.listDirectory(path, recursive, true);

      // Format entries for better readability
      const files = entries.filter(e => e.isFile).map(e => e.path);
      const directories = entries.filter(e => e.isDirectory).map(e => e.path);

      return {
        path,
        recursive,
        totalEntries: entries.length,
        files,
        directories,
        fileCount: files.length,
        directoryCount: directories.length
      };
    } catch (error) {
      throw new Error(`Failed to list directory: ${error.message}`);
    }
  };
}

/**
 * Tool schema for list_directory
 */
export const listDirectorySchema = {
  description: 'List files and directories in a given path',
  parameters: {
    type: 'object',
    properties: {
      path: {
        type: 'string',
        description: 'Directory path to list (relative to project root, defaults to ".")'
      },
      recursive: {
        type: 'boolean',
        description: 'Whether to list recursively (defaults to false)'
      }
    },
    required: []
  }
};
