// codestral/src/tools/handlers/planProject.js
/**
 * Plan project tool handler
 * Generates project plans and file tree previews before creation
 */

import { generateTemplate } from '../../templates/index.js';

/**
 * Generate a visual file tree representation
 * @param {Array} files - Array of file objects with path and content
 * @returns {string} ASCII file tree
 */
function generateFileTree(files) {
  // Sort files by path for better tree structure
  const sortedFiles = [...files].sort((a, b) => a.path.localeCompare(b.path));
  
  // Build tree structure
  const tree = {};
  
  sortedFiles.forEach(file => {
    const parts = file.path.split('/');
    let current = tree;
    
    parts.forEach((part, index) => {
      if (!current[part]) {
        current[part] = index === parts.length - 1 ? { 
          _isFile: true, 
          _size: Buffer.byteLength(file.content, 'utf8'),
          _content: file.content
        } : {};
      }
      current = current[part];
    });
  });
  
  // Convert tree to ASCII representation
  function renderTree(node, prefix = '', isLast = true) {
    let result = '';
    const entries = Object.entries(node).filter(([key]) => !key.startsWith('_'));
    
    entries.forEach(([name, child], index) => {
      const isLastEntry = index === entries.length - 1;
      const connector = isLastEntry ? '└── ' : '├── ';
      const nextPrefix = prefix + (isLastEntry ? '    ' : '│   ');
      
      if (child._isFile) {
        const sizeKB = Math.round(child._size / 1024 * 100) / 100;
        const sizeStr = child._size < 1024 ? `${child._size}B` : `${sizeKB}KB`;
        result += `${prefix}${connector}${name} (${sizeStr})\\n`;
      } else {
        result += `${prefix}${connector}${name}/\\n`;
        result += renderTree(child, nextPrefix, isLastEntry);
      }
    });
    
    return result;
  }
  
  return renderTree(tree);
}

/**
 * Calculate project statistics
 * @param {Array} files - Array of file objects
 * @returns {Object} Project statistics
 */
function calculateProjectStats(files) {
  const stats = {
    totalFiles: files.length,
    totalSize: 0,
    fileTypes: {},
    directories: new Set(),
    largestFile: null,
    averageFileSize: 0
  };
  
  files.forEach(file => {
    const size = Buffer.byteLength(file.content, 'utf8');
    stats.totalSize += size;
    
    // Track file types
    const ext = file.path.split('.').pop().toLowerCase();
    stats.fileTypes[ext] = (stats.fileTypes[ext] || 0) + 1;
    
    // Track directories
    const dir = file.path.split('/').slice(0, -1).join('/');
    if (dir) {
      stats.directories.add(dir);
    }
    
    // Track largest file
    if (!stats.largestFile || size > stats.largestFile.size) {
      stats.largestFile = { path: file.path, size };
    }
  });
  
  stats.averageFileSize = Math.round(stats.totalSize / stats.totalFiles);
  stats.totalSizeMB = Math.round(stats.totalSize / 1024 / 1024 * 100) / 100;
  stats.directoryCount = stats.directories.size;
  
  return stats;
}

/**
 * Generate setup instructions based on project type
 * @param {string} template - Template name
 * @param {string} projectName - Project name
 * @param {Array} files - Generated files
 * @returns {Array} Setup instructions
 */
function generateSetupInstructions(template, projectName, files) {
  const instructions = [];
  
  // Check for package.json
  const hasPackageJson = files.some(f => f.path === 'package.json');
  const hasRequirementsTxt = files.some(f => f.path === 'requirements.txt');
  const hasManagePy = files.some(f => f.path === 'manage.py');
  
  if (hasPackageJson) {
    instructions.push('📦 Install dependencies: `npm install`');
    
    // Check for specific scripts
    const packageJson = files.find(f => f.path === 'package.json');
    if (packageJson) {
      try {
        const pkg = JSON.parse(packageJson.content);
        if (pkg.scripts) {
          if (pkg.scripts.dev) {
            instructions.push('🚀 Start development: `npm run dev`');
          } else if (pkg.scripts.start) {
            instructions.push('🚀 Start application: `npm start`');
          }
          if (pkg.scripts.build) {
            instructions.push('🏗️  Build for production: `npm run build`');
          }
        }
      } catch (e) {
        // Ignore JSON parse errors
      }
    }
  }
  
  if (hasRequirementsTxt) {
    instructions.push('🐍 Create virtual environment: `python -m venv venv`');
    instructions.push('🔧 Activate virtual environment: `source venv/bin/activate` (Linux/Mac) or `venv\\\\Scripts\\\\activate` (Windows)');
    instructions.push('📦 Install dependencies: `pip install -r requirements.txt`');
  }
  
  if (hasManagePy) {
    instructions.push('🗄️  Run migrations: `python manage.py migrate`');
    instructions.push('👤 Create superuser: `python manage.py createsuperuser`');
    instructions.push('🚀 Start development server: `python manage.py runserver`');
  }
  
  // Template-specific instructions
  switch (template) {
    case 'static-website':
      instructions.push('🌐 Open `index.html` in your web browser');
      break;
    case 'react-app':
    case 'vue-app':
      instructions.push('🌐 Open http://localhost:3000 in your browser');
      break;
    case 'next-app':
      instructions.push('🌐 Open http://localhost:3000 in your browser');
      break;
    case 'express-api':
      instructions.push('🌐 API will be available at http://localhost:3000');
      instructions.push('📚 Check API documentation at /api/docs');
      break;
    case 'flask-api':
      instructions.push('🌐 API will be available at http://localhost:5000');
      break;
    case 'fastapi-api':
      instructions.push('🌐 API will be available at http://localhost:8000');
      instructions.push('📚 Interactive docs at http://localhost:8000/docs');
      break;
    case 'django-app':
      instructions.push('🌐 Application will be available at http://localhost:8000');
      instructions.push('👑 Admin interface at http://localhost:8000/admin/');
      break;
    case 'full-stack':
      instructions.push('🌐 Frontend: http://localhost:3000');
      instructions.push('🔧 Backend: http://localhost:3001');
      instructions.push('📚 Use `npm run dev` to start both servers');
      break;
  }
  
  // Environment setup
  const hasEnvExample = files.some(f => f.path === '.env.example');
  if (hasEnvExample) {
    instructions.push('⚙️  Copy environment file: `cp .env.example .env`');
    instructions.push('🔧 Update .env with your configuration');
  }
  
  return instructions;
}

/**
 * Create plan_project tool handler
 * @returns {Function} Tool handler function
 */
export function createPlanProjectHandler() {
  return async (params) => {
    const { template, projectName = 'My Project', options = {} } = params;
    
    if (!template) {
      throw new Error('Template parameter is required for project planning');
    }
    
    try {
      // Generate template files
      const files = generateTemplate(template, projectName, options);
      
      // Calculate statistics
      const stats = calculateProjectStats(files);
      
      // Generate file tree
      const fileTree = generateFileTree(files);
      
      // Generate setup instructions
      const setupInstructions = generateSetupInstructions(template, projectName, files);
      
      // Create project plan
      const plan = {
        projectName,
        template,
        createdAt: new Date().toISOString(),
        
        // Overview
        overview: {
          description: `${projectName} - A ${template.replace('-', ' ')} project`,
          totalFiles: stats.totalFiles,
          totalSize: `${stats.totalSizeMB}MB`,
          directories: stats.directoryCount,
          estimatedSetupTime: getEstimatedSetupTime(template, stats)
        },
        
        // File structure
        structure: {
          fileTree,
          fileTypes: stats.fileTypes,
          largestFile: stats.largestFile
        },
        
        // Setup process
        setup: {
          instructions: setupInstructions,
          requirements: getProjectRequirements(template),
          nextSteps: [
            'Review the project structure above',
            'Confirm project creation',
            'Follow setup instructions',
            'Start development'
          ]
        },
        
        // Technical details
        technical: {
          template,
          fileCount: stats.totalFiles,
          totalSizeBytes: stats.totalSize,
          averageFileSize: `${Math.round(stats.averageFileSize / 1024 * 100) / 100}KB`,
          directories: Array.from(stats.directories).sort()
        }
      };
      
      return {
        success: true,
        plan,
        files, // Include files for potential creation
        message: `Project plan generated for ${projectName}. Review the structure and confirm to proceed with creation.`,
        warnings: generateWarnings(stats, template),
        readyToCreate: true
      };
      
    } catch (error) {
      return {
        success: false,
        error: error.message,
        message: `Failed to generate project plan: ${error.message}`
      };
    }
  };
}

/**
 * Get estimated setup time based on template and stats
 * @param {string} template - Template name
 * @param {Object} stats - Project statistics
 * @returns {string} Estimated setup time
 */
function getEstimatedSetupTime(template, stats) {
  let baseTime = 2; // minutes
  
  // Add time based on complexity
  if (stats.totalFiles > 50) baseTime += 3;
  if (stats.totalFiles > 100) baseTime += 5;
  
  // Template-specific adjustments
  switch (template) {
    case 'static-website':
      return '1-2 minutes';
    case 'react-app':
    case 'vue-app':
      return '3-5 minutes';
    case 'next-app':
      return '4-6 minutes';
    case 'express-api':
    case 'flask-api':
      return '2-4 minutes';
    case 'fastapi-api':
      return '3-5 minutes';
    case 'django-app':
      return '5-8 minutes';
    case 'full-stack':
      return '8-12 minutes';
    default:
      return `${baseTime}-${baseTime + 2} minutes`;
  }
}

/**
 * Get project requirements based on template
 * @param {string} template - Template name
 * @returns {Array} Requirements list
 */
function getProjectRequirements(template) {
  const requirements = [];
  
  switch (template) {
    case 'static-website':
      requirements.push('Web browser');
      break;
    case 'react-app':
    case 'vue-app':
    case 'next-app':
    case 'express-api':
    case 'full-stack':
      requirements.push('Node.js 16+', 'npm or yarn');
      break;
    case 'flask-api':
    case 'fastapi-api':
    case 'django-app':
      requirements.push('Python 3.8+', 'pip');
      break;
  }
  
  return requirements;
}

/**
 * Generate warnings based on project statistics
 * @param {Object} stats - Project statistics
 * @param {string} template - Template name
 * @returns {Array} Warning messages
 */
function generateWarnings(stats, template) {
  const warnings = [];
  
  if (stats.totalSizeMB > 50) {
    warnings.push(`Large project size: ${stats.totalSizeMB}MB - may take longer to create`);
  }
  
  if (stats.totalFiles > 200) {
    warnings.push(`Many files: ${stats.totalFiles} files - creation may take some time`);
  }
  
  if (stats.largestFile && stats.largestFile.size > 1024 * 1024) {
    const sizeMB = Math.round(stats.largestFile.size / 1024 / 1024 * 100) / 100;
    warnings.push(`Large file detected: ${stats.largestFile.path} (${sizeMB}MB)`);
  }
  
  return warnings;
}

/**
 * Tool schema for plan_project
 */
export const planProjectSchema = {
  description: 'Generate a comprehensive project plan with file tree preview and setup instructions before creating files',
  parameters: {
    type: 'object',
    properties: {
      template: {
        type: 'string',
        description: 'Template name to plan',
        enum: [
          'static-website',
          'react-app',
          'express-api',
          'full-stack',
          'vue-app',
          'next-app',
          'flask-api',
          'fastapi-api',
          'django-app'
        ]
      },
      projectName: {
        type: 'string',
        description: 'Project name to use in the plan',
        default: 'My Project'
      },
      options: {
        type: 'object',
        description: 'Additional template-specific options',
        properties: {},
        additionalProperties: true
      }
    },
    required: ['template']
  }
};
