// codestral/src/tools/handlers/readFile.js
/**
 * Read file tool handler
 * Reads the contents of a file from the file system
 */

/**
 * Create read_file tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createReadFileHandler(fileSystemManager) {
  return async (params) => {
    const { path } = params;

    if (!path) {
      throw new Error('Path parameter is required');
    }

    try {
      // Check if file is binary
      const isBinary = await fileSystemManager.isBinary(path);
      
      if (isBinary) {
        return {
          path,
          content: '[Binary file - content not displayed]',
          isBinary: true
        };
      }

      // Read file content
      const content = await fileSystemManager.readFile(path);
      
      // Get file stats
      const stats = await fileSystemManager.getStats(path);

      return {
        path,
        content,
        size: stats.size,
        isBinary: false,
        modified: stats.modified
      };
    } catch (error) {
      throw new Error(`Failed to read file: ${error.message}`);
    }
  };
}

/**
 * Tool schema for read_file
 */
export const readFileSchema = {
  description: 'Read the contents of a file',
  parameters: {
    type: 'object',
    properties: {
      path: {
        type: 'string',
        description: 'Path to the file to read (relative to project root)'
      }
    },
    required: ['path']
  }
};
