// codestral/src/tools/handlers/replaceInFile.js
/**
 * Replace in file tool handler
 * Replaces text in a file using search and replace
 */

/**
 * Create replace_in_file tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createReplaceInFileHandler(fileSystemManager) {
  return async (params) => {
    const { path, search, replace } = params;

    if (!path) {
      throw new Error('Path parameter is required');
    }

    if (search === undefined || search === null) {
      throw new Error('Search parameter is required');
    }

    if (replace === undefined || replace === null) {
      throw new Error('Replace parameter is required');
    }

    try {
      // Read current file content
      const content = await fileSystemManager.readFile(path);

      // Perform replacement
      const newContent = content.replace(new RegExp(search, 'g'), replace);

      // Count replacements
      const matches = content.match(new RegExp(search, 'g'));
      const replacementCount = matches ? matches.length : 0;

      if (replacementCount === 0) {
        return {
          path,
          success: false,
          replacementCount: 0,
          message: `No matches found for "${search}" in ${path}`
        };
      }

      // Write updated content
      await fileSystemManager.writeFile(path, newContent);

      return {
        path,
        success: true,
        replacementCount,
        message: `Replaced ${replacementCount} occurrence(s) of "${search}" with "${replace}" in ${path}`
      };
    } catch (error) {
      throw new Error(`Failed to replace in file: ${error.message}`);
    }
  };
}

/**
 * Tool schema for replace_in_file
 */
export const replaceInFileSchema = {
  description: 'Replace text in a file using search and replace',
  parameters: {
    type: 'object',
    properties: {
      path: {
        type: 'string',
        description: 'Path to the file (relative to project root)'
      },
      search: {
        type: 'string',
        description: 'Text to search for (will be used as regex pattern)'
      },
      replace: {
        type: 'string',
        description: 'Replacement text'
      }
    },
    required: ['path', 'search', 'replace']
  }
};
