// codestral/src/tools/handlers/searchCode.js
/**
 * Search code tool handler
 * Searches for text or patterns in the codebase
 */

/**
 * Create search_code tool handler
 * @param {SearchEngine} searchEngine - Search engine instance
 * @returns {Function} Tool handler function
 */
export function createSearchCodeHandler(searchEngine) {
  return async (params) => {
    const { 
      query, 
      regex = false, 
      fileType = null,
      caseSensitive = false,
      contextLines = 2
    } = params;

    if (!query) {
      throw new Error('Query parameter is required');
    }

    try {
      let results;

      // Perform search based on type
      if (regex) {
        results = await searchEngine.searchRegex(query, {
          flags: caseSensitive ? 'g' : 'gi',
          contextLines
        });
      } else {
        results = await searchEngine.search(query, {
          caseSensitive,
          contextLines
        });
      }

      // Filter by file type if specified
      if (fileType) {
        results = searchEngine.filterByFileType(results, [fileType]);
      }

      // Rank results by relevance
      results = searchEngine.rankResults(results, query);

      // Limit results to top 50 for performance
      const limitedResults = results.slice(0, 50);

      return {
        query,
        regex,
        fileType,
        totalMatches: results.length,
        displayedMatches: limitedResults.length,
        results: limitedResults.map(r => ({
          file: r.filePath,
          line: r.lineNumber,
          content: r.line,
          score: r.score,
          language: r.language
        }))
      };
    } catch (error) {
      throw new Error(`Failed to search code: ${error.message}`);
    }
  };
}

/**
 * Tool schema for search_code
 */
export const searchCodeSchema = {
  description: 'Search for text or patterns in the codebase',
  parameters: {
    type: 'object',
    properties: {
      query: {
        type: 'string',
        description: 'Search query or regex pattern'
      },
      regex: {
        type: 'boolean',
        description: 'Whether to use regex pattern matching (defaults to false)'
      },
      fileType: {
        type: 'string',
        description: 'Filter by file extension (e.g., ".js", ".py")'
      },
      caseSensitive: {
        type: 'boolean',
        description: 'Whether search should be case sensitive (defaults to false)'
      },
      contextLines: {
        type: 'number',
        description: 'Number of context lines to include around matches (defaults to 2)'
      }
    },
    required: ['query']
  }
};
