// codestral/src/tools/handlers/writeFile.js
/**
 * Write file tool handler
 * Writes content to a file, creating it if it doesn't exist
 */

/**
 * Create write_file tool handler
 * @param {FileSystemManager} fileSystemManager - File system manager instance
 * @returns {Function} Tool handler function
 */
export function createWriteFileHandler(fileSystemManager) {
  return async (params) => {
    const { path, content } = params;

    if (!path) {
      throw new Error('Path parameter is required');
    }

    if (content === undefined || content === null) {
      throw new Error('Content parameter is required');
    }

    try {
      // Write file content
      await fileSystemManager.writeFile(path, content);
      
      // Get file stats to confirm
      const stats = await fileSystemManager.getStats(path);

      return {
        path,
        size: stats.size,
        success: true,
        message: `File written successfully: ${path}`
      };
    } catch (error) {
      throw new Error(`Failed to write file: ${error.message}`);
    }
  };
}

/**
 * Tool schema for write_file
 */
export const writeFileSchema = {
  description: 'Write content to a file (creates file if it doesn\'t exist)',
  parameters: {
    type: 'object',
    properties: {
      path: {
        type: 'string',
        description: 'Path to the file to write (relative to project root)'
      },
      content: {
        type: 'string',
        description: 'Content to write to the file'
      }
    },
    required: ['path', 'content']
  }
};
