// codestral/src/context/ContextManager.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * ContextManager - Manages conversation context and history
 * 
 * Responsibilities:
 * - Maintain conversation history
 * - Track files in context
 * - Implement context window management
 * - Estimate token usage
 * - Persist and restore sessions
 */

class ContextManager {
  /**
   * Create a new ContextManager
   * @param {number} maxTokens - Maximum tokens allowed in context window
   */
  constructor(maxTokens = 4096) {
    this.maxTokens = maxTokens;
    this.messages = [];
    this.fileContexts = new Map(); // Map<filePath, FileContext>
    this.sessionId = null;
  }

  /**
   * Add a message to the conversation history
   * @param {string} role - Message role ('user', 'assistant', 'system')
   * @param {string} content - Message content
   * @param {Object} metadata - Optional metadata (files, tokens, etc.)
   * @returns {Object} The added message
   */
  addMessage(role, content, metadata = {}) {
    if (!['user', 'assistant', 'system'].includes(role)) {
      throw new Error(`Invalid role: ${role}. Must be 'user', 'assistant', or 'system'`);
    }

    if (typeof content !== 'string') {
      throw new Error('Message content must be a string');
    }

    const message = {
      role,
      content,
      timestamp: new Date(),
      metadata: {
        files: metadata.files || [],
        tokens: metadata.tokens || this.estimateTokens(content),
        ...metadata
      }
    };

    this.messages.push(message);
    return message;
  }

  /**
   * Get all messages in the conversation history
   * @param {Object} options - Optional filtering options
   * @returns {Array} Array of messages
   */
  getMessages(options = {}) {
    let messages = [...this.messages];

    // Filter by role if specified
    if (options.role) {
      messages = messages.filter(msg => msg.role === options.role);
    }

    // Limit number of messages if specified
    if (options.limit) {
      messages = messages.slice(-options.limit);
    }

    // Return only specific fields if specified
    if (options.fields) {
      messages = messages.map(msg => {
        const filtered = {};
        options.fields.forEach(field => {
          if (msg[field] !== undefined) {
            filtered[field] = msg[field];
          }
        });
        return filtered;
      });
    }

    return messages;
  }

  /**
   * Clear all messages and reset context
   */
  clear() {
    this.messages = [];
    this.fileContexts.clear();
    this.sessionId = null;
  }

  /**
   * Get the total number of messages
   * @returns {number} Message count
   */
  getMessageCount() {
    return this.messages.length;
  }

  /**
   * Get the most recent message
   * @returns {Object|null} The last message or null if no messages
   */
  getLastMessage() {
    return this.messages.length > 0 ? this.messages[this.messages.length - 1] : null;
  }

  /**
   * Remove the last message from history
   * @returns {Object|null} The removed message or null if no messages
   */
  removeLastMessage() {
    return this.messages.pop() || null;
  }

  /**
   * Estimate the number of tokens in a text string
   * @param {string} text - Text to estimate tokens for
   * @returns {number} Estimated token count
   */
  estimateTokens(text) {
    // Rough estimation: ~4 characters per token on average
    // This is a simple heuristic and not exact
    if (!text || typeof text !== 'string') {
      return 0;
    }
    return Math.ceil(text.length / 4);
  }

  /**
   * Calculate total tokens in all messages
   * @returns {number} Total token count
   */
  getTotalTokens() {
    return this.messages.reduce((total, msg) => {
      return total + (msg.metadata.tokens || this.estimateTokens(msg.content));
    }, 0);
  }

  /**
   * Truncate messages to fit within token limit using smart strategy
   * Keeps system messages and recent messages, removes older messages
   * @param {Object} options - Truncation options
   * @returns {Array} Removed messages
   */
  truncate(options = {}) {
    const {
      keepSystemMessages = true,
      keepRecentCount = 10,
      targetTokens = this.maxTokens * 0.8 // Keep 80% of max to leave room
    } = options;

    const totalTokens = this.getTotalTokens();
    
    // No truncation needed
    if (totalTokens <= targetTokens) {
      return [];
    }

    const removedMessages = [];
    
    // Separate messages by type
    const systemMessages = [];
    const conversationMessages = [];
    
    this.messages.forEach(msg => {
      if (msg.role === 'system' && keepSystemMessages) {
        systemMessages.push(msg);
      } else {
        conversationMessages.push(msg);
      }
    });

    // Calculate tokens for system messages
    const systemTokens = systemMessages.reduce((total, msg) => {
      return total + (msg.metadata.tokens || this.estimateTokens(msg.content));
    }, 0);

    // Calculate how many tokens we have for conversation messages
    const availableTokens = targetTokens - systemTokens;

    // Keep recent messages up to available tokens
    const recentMessages = conversationMessages.slice(-keepRecentCount);
    let currentTokens = 0;
    const keptMessages = [];

    // Add messages from most recent backwards until we hit token limit
    for (let i = recentMessages.length - 1; i >= 0; i--) {
      const msg = recentMessages[i];
      const msgTokens = msg.metadata.tokens || this.estimateTokens(msg.content);
      
      if (currentTokens + msgTokens <= availableTokens) {
        keptMessages.unshift(msg);
        currentTokens += msgTokens;
      } else {
        removedMessages.unshift(msg);
      }
    }

    // Add any older messages that were removed
    const olderMessages = conversationMessages.slice(0, -keepRecentCount);
    removedMessages.unshift(...olderMessages);

    // Update messages array with system messages + kept conversation messages
    this.messages = [...systemMessages, ...keptMessages];

    return removedMessages;
  }

  /**
   * Check if context is near token limit
   * @param {number} threshold - Threshold percentage (0-1)
   * @returns {boolean} True if near limit
   */
  isNearTokenLimit(threshold = 0.9) {
    const totalTokens = this.getTotalTokens();
    return totalTokens >= (this.maxTokens * threshold);
  }

  /**
   * Get context statistics
   * @returns {Object} Statistics about the context
   */
  getStats() {
    const totalTokens = this.getTotalTokens();
    const messageCount = this.messages.length;
    const fileCount = this.fileContexts.size;
    
    const roleCount = this.messages.reduce((acc, msg) => {
      acc[msg.role] = (acc[msg.role] || 0) + 1;
      return acc;
    }, {});

    return {
      messageCount,
      totalTokens,
      maxTokens: this.maxTokens,
      tokenUsagePercent: (totalTokens / this.maxTokens) * 100,
      fileCount,
      roleCount,
      isNearLimit: this.isNearTokenLimit()
    };
  }

  /**
   * Add file context to track files included in conversation
   * @param {string} filePath - Path to the file
   * @param {string} content - File content
   * @param {Object} metadata - Optional metadata (language, size, etc.)
   */
  addFileContext(filePath, content, metadata = {}) {
    if (!filePath || typeof filePath !== 'string') {
      throw new Error('File path must be a non-empty string');
    }

    if (typeof content !== 'string') {
      throw new Error('File content must be a string');
    }

    const fileContext = {
      path: filePath,
      content,
      language: metadata.language || this._detectLanguage(filePath),
      size: content.length,
      lastModified: metadata.lastModified || new Date(),
      addedAt: new Date()
    };

    this.fileContexts.set(filePath, fileContext);
    return fileContext;
  }

  /**
   * Get file context by path
   * @param {string} filePath - Path to the file
   * @returns {Object|null} File context or null if not found
   */
  getFileContext(filePath) {
    return this.fileContexts.get(filePath) || null;
  }

  /**
   * Get all file contexts
   * @returns {Array} Array of file contexts
   */
  getAllFileContexts() {
    return Array.from(this.fileContexts.values());
  }

  /**
   * Remove file context
   * @param {string} filePath - Path to the file
   * @returns {boolean} True if removed, false if not found
   */
  removeFileContext(filePath) {
    return this.fileContexts.delete(filePath);
  }

  /**
   * Detect programming language from file extension
   * @private
   * @param {string} filePath - File path
   * @returns {string} Detected language
   */
  _detectLanguage(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const languageMap = {
      '.js': 'javascript',
      '.jsx': 'javascript',
      '.ts': 'typescript',
      '.tsx': 'typescript',
      '.py': 'python',
      '.java': 'java',
      '.cpp': 'cpp',
      '.c': 'c',
      '.cs': 'csharp',
      '.go': 'go',
      '.rs': 'rust',
      '.rb': 'ruby',
      '.php': 'php',
      '.html': 'html',
      '.css': 'css',
      '.json': 'json',
      '.md': 'markdown',
      '.yaml': 'yaml',
      '.yml': 'yaml',
      '.xml': 'xml',
      '.sh': 'bash',
      '.sql': 'sql'
    };
    return languageMap[ext] || 'text';
  }

  /**
   * Save session to disk
   * @param {string} sessionId - Session identifier
   * @param {string} sessionsDir - Directory to save sessions (optional)
   * @returns {Promise<string>} Path to saved session file
   */
  async save(sessionId, sessionsDir = null) {
    if (!sessionId || typeof sessionId !== 'string') {
      throw new Error('Session ID must be a non-empty string');
    }

    // Default sessions directory
    const defaultDir = path.join(process.cwd(), '.codestral', 'sessions');
    const saveDir = sessionsDir || defaultDir;

    // Ensure directory exists
    await fs.mkdir(saveDir, { recursive: true });

    // Prepare session data
    const sessionData = {
      sessionId,
      maxTokens: this.maxTokens,
      messages: this.messages,
      fileContexts: Array.from(this.fileContexts.entries()),
      savedAt: new Date().toISOString(),
      stats: this.getStats()
    };

    // Save to file
    const filePath = path.join(saveDir, `${sessionId}.json`);
    await fs.writeFile(filePath, JSON.stringify(sessionData, null, 2), 'utf-8');

    this.sessionId = sessionId;
    return filePath;
  }

  /**
   * Load session from disk
   * @param {string} sessionId - Session identifier
   * @param {string} sessionsDir - Directory to load sessions from (optional)
   * @returns {Promise<boolean>} True if loaded successfully
   */
  async load(sessionId, sessionsDir = null) {
    if (!sessionId || typeof sessionId !== 'string') {
      throw new Error('Session ID must be a non-empty string');
    }

    // Default sessions directory
    const defaultDir = path.join(process.cwd(), '.codestral', 'sessions');
    const loadDir = sessionsDir || defaultDir;

    const filePath = path.join(loadDir, `${sessionId}.json`);

    try {
      // Read session file
      const data = await fs.readFile(filePath, 'utf-8');
      const sessionData = JSON.parse(data);

      // Restore session data
      this.sessionId = sessionData.sessionId;
      this.maxTokens = sessionData.maxTokens;
      
      // Restore messages with Date objects
      this.messages = sessionData.messages.map(msg => ({
        ...msg,
        timestamp: new Date(msg.timestamp)
      }));

      // Restore file contexts
      this.fileContexts.clear();
      sessionData.fileContexts.forEach(([filePath, context]) => {
        this.fileContexts.set(filePath, {
          ...context,
          lastModified: new Date(context.lastModified),
          addedAt: new Date(context.addedAt)
        });
      });

      return true;
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`Session not found: ${sessionId}`);
      }
      throw error;
    }
  }

  /**
   * List all saved sessions
   * @param {string} sessionsDir - Directory to list sessions from (optional)
   * @returns {Promise<Array>} Array of session metadata
   */
  async listSessions(sessionsDir = null) {
    const defaultDir = path.join(process.cwd(), '.codestral', 'sessions');
    const loadDir = sessionsDir || defaultDir;

    try {
      const files = await fs.readdir(loadDir);
      const sessions = [];

      for (const file of files) {
        if (file.endsWith('.json')) {
          const filePath = path.join(loadDir, file);
          const data = await fs.readFile(filePath, 'utf-8');
          const sessionData = JSON.parse(data);
          
          sessions.push({
            sessionId: sessionData.sessionId,
            savedAt: sessionData.savedAt,
            messageCount: sessionData.messages.length,
            fileCount: sessionData.fileContexts.length,
            stats: sessionData.stats
          });
        }
      }

      return sessions;
    } catch (error) {
      if (error.code === 'ENOENT') {
        return [];
      }
      throw error;
    }
  }

  /**
   * Delete a saved session
   * @param {string} sessionId - Session identifier
   * @param {string} sessionsDir - Directory containing sessions (optional)
   * @returns {Promise<boolean>} True if deleted successfully
   */
  async deleteSession(sessionId, sessionsDir = null) {
    if (!sessionId || typeof sessionId !== 'string') {
      throw new Error('Session ID must be a non-empty string');
    }

    const defaultDir = path.join(process.cwd(), '.codestral', 'sessions');
    const loadDir = sessionsDir || defaultDir;
    const filePath = path.join(loadDir, `${sessionId}.json`);

    try {
      await fs.unlink(filePath);
      return true;
    } catch (error) {
      if (error.code === 'ENOENT') {
        return false;
      }
      throw error;
    }
  }
}

export default ContextManager;
