// codestral/src/config/ApiKeyManager.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import chalk from 'chalk';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * ApiKeyManager - Manages hardcoded API key in the codebase
 * Creates and updates the HardcodedApiKey.js file dynamically
 */
class ApiKeyManager {
  constructor() {
    this.apiKeyFilePath = path.join(__dirname, 'HardcodedApiKey.js');
  }

  /**
   * Create the HardcodedApiKey.js file with the provided API key
   * @param {string} apiKey - The Mistral AI API key to hardcode
   * @returns {Promise<void>}
   */
  async createApiKeyFile(apiKey) {
    const fileContent = `/**
 * Hardcoded API Key Module
 * This module stores the Mistral AI API key directly in the codebase
 * 
 * SECURITY WARNING: Hardcoding API keys is NOT recommended for production use
 * This is only suitable for:
 * - Personal projects
 * - Development/testing environments
 * - Single-user applications
 * 
 * For production, use environment variables or secure key management systems
 */

/**
 * HARDCODED MISTRAL AI API KEY
 * Get your API key from https://console.mistral.ai/
 */
const HARDCODED_API_KEY = '${apiKey}';

/**
 * Get the hardcoded API key
 * @returns {string} The hardcoded Mistral AI API key
 */
export function getHardcodedApiKey() {
  return HARDCODED_API_KEY;
}

/**
 * Check if a valid API key is hardcoded
 * @returns {boolean} True if API key is set and not the default placeholder
 */
export function hasValidHardcodedApiKey() {
  return HARDCODED_API_KEY && 
         HARDCODED_API_KEY !== 'your_mistral_api_key_here' &&
         HARDCODED_API_KEY.length > 10;
}

/**
 * Get API key (returns hardcoded value)
 * @returns {string} The API key to use
 */
export function getApiKey() {
  if (hasValidHardcodedApiKey()) {
    return HARDCODED_API_KEY;
  }
  
  return '';
}

export default {
  getHardcodedApiKey,
  hasValidHardcodedApiKey,
  getApiKey
};
`;

    try {
      // Ensure the config directory exists
      const configDir = path.dirname(this.apiKeyFilePath);
      await fs.mkdir(configDir, { recursive: true });

      // Write the file
      await fs.writeFile(this.apiKeyFilePath, fileContent, 'utf8');
      
      console.log(chalk.green(`✓ API key hardcoded successfully!`));
      console.log(chalk.gray(`  File: ${this.apiKeyFilePath}`));
      
      return true;
    } catch (error) {
      console.error(chalk.red(`✗ Failed to create API key file: ${error.message}`));
      throw error;
    }
  }

  /**
   * Update the API key in the existing file
   * @param {string} newApiKey - The new API key
   * @returns {Promise<void>}
   */
  async updateApiKey(newApiKey) {
    try {
      // Check if file exists
      try {
        await fs.access(this.apiKeyFilePath);
      } catch {
        // File doesn't exist, create it
        return await this.createApiKeyFile(newApiKey);
      }

      // Read existing file
      let content = await fs.readFile(this.apiKeyFilePath, 'utf8');

      // Replace the API key value
      const regex = /const HARDCODED_API_KEY = ['"].*?['"];/;
      content = content.replace(regex, `const HARDCODED_API_KEY = '${newApiKey}';`);

      // Write back
      await fs.writeFile(this.apiKeyFilePath, content, 'utf8');

      console.log(chalk.green(`✓ API key updated successfully!`));
      
      return true;
    } catch (error) {
      console.error(chalk.red(`✗ Failed to update API key: ${error.message}`));
      throw error;
    }
  }

  /**
   * Read the current API key from the file
   * @returns {Promise<string|null>} The current API key or null if not found
   */
  async readApiKey() {
    try {
      const content = await fs.readFile(this.apiKeyFilePath, 'utf8');
      const match = content.match(/const HARDCODED_API_KEY = ['"](.*)['"];/);
      
      if (match && match[1] && match[1] !== 'your_mistral_api_key_here') {
        return match[1];
      }
      
      return null;
    } catch (error) {
      return null;
    }
  }

  /**
   * Check if API key file exists and has a valid key
   * @returns {Promise<boolean>}
   */
  async hasValidApiKey() {
    const apiKey = await this.readApiKey();
    return apiKey !== null && apiKey.length > 10;
  }

  /**
   * Initialize the API key file if it doesn't exist
   * @returns {Promise<void>}
   */
  async initialize() {
    try {
      await fs.access(this.apiKeyFilePath);
      // File exists, check if it has a valid key
      const hasValid = await this.hasValidApiKey();
      if (!hasValid) {
        console.log(chalk.yellow('⚠ API key file exists but no valid key is set'));
      }
    } catch {
      // File doesn't exist, create with placeholder
      await this.createApiKeyFile('your_mistral_api_key_here');
      console.log(chalk.yellow('⚠ Created API key file with placeholder'));
      console.log(chalk.cyan('  Run setup to configure your API key'));
    }
  }
}

export default ApiKeyManager;
