// codestral/src/config/ConfigManager.js
import { cosmiconfig } from 'cosmiconfig';
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
import prompts from 'prompts';
import chalk from 'chalk';
import { getApiKey } from './HardcodedApiKey.js';
import ApiKeyManager from './ApiKeyManager.js';

/**
 * ConfigManager handles loading, saving, and validating application configuration
 */
class ConfigManager {
  constructor() {
    this.configName = 'codestral';
    this.explorer = cosmiconfig(this.configName);
    this.config = null;
    this.configPath = null;
    this.apiKeyManager = new ApiKeyManager();
  }

  /**
   * Get default configuration
   * @returns {Object} Default configuration object
   */
  getDefaultConfig() {
    return {
      mistral: {
        apiKey: getApiKey(), // Uses hardcoded key from HardcodedApiKey.js
        model: 'mistral-large-latest',
        maxTokens: 4096,
        temperature: 0.7
      },
      context: {
        maxMessages: 50,
        maxTokens: 8000
      },
      files: {
        maxFileSize: 1048576, // 1MB
        ignorePatterns: [
          'node_modules/**',
          '.git/**',
          'dist/**',
          'build/**',
          '*.log'
        ]
      },
      ui: {
        theme: 'default',
        syntaxHighlight: true
      },
      logging: {
        level: 'info',
        file: 'logs/codestral.log'
      },
      agent: {
        autonomyMode: 'safe',
        autoApproveTools: [],
        dangerousCommands: [
          'rm -rf',
          'sudo',
          'chmod',
          'chown',
          'mkfs',
          'dd',
          'format'
        ],
        maxToolCalls: 20
      }
    };
  }

  /**
   * Load configuration from file or environment
   * @returns {Promise<Object>} Loaded configuration
   */
  async load() {
    try {
      // Try to load from config file
      const result = await this.explorer.search();
      
      if (result && !result.isEmpty) {
        // Get defaults (which may include env vars)
        const defaults = this.getDefaultConfig();
        
        // Merge with file config, giving priority to file config over defaults
        // Only use default values for fields that are missing in the file config
        this.config = {
          mistral: {
            apiKey: result.config.mistral?.apiKey ?? defaults.mistral.apiKey,
            model: result.config.mistral?.model ?? defaults.mistral.model,
            maxTokens: result.config.mistral?.maxTokens ?? defaults.mistral.maxTokens,
            temperature: result.config.mistral?.temperature ?? defaults.mistral.temperature
          },
          context: {
            maxMessages: result.config.context?.maxMessages ?? defaults.context.maxMessages,
            maxTokens: result.config.context?.maxTokens ?? defaults.context.maxTokens
          },
          files: {
            maxFileSize: result.config.files?.maxFileSize ?? defaults.files.maxFileSize,
            ignorePatterns: result.config.files?.ignorePatterns ?? defaults.files.ignorePatterns
          },
          ui: {
            theme: result.config.ui?.theme ?? defaults.ui.theme,
            syntaxHighlight: result.config.ui?.syntaxHighlight ?? defaults.ui.syntaxHighlight
          },
          logging: {
            level: result.config.logging?.level ?? defaults.logging.level,
            file: result.config.logging?.file ?? defaults.logging.file
          },
          agent: {
            autonomyMode: result.config.agent?.autonomyMode ?? defaults.agent.autonomyMode,
            autoApproveTools: result.config.agent?.autoApproveTools ?? defaults.agent.autoApproveTools,
            dangerousCommands: result.config.agent?.dangerousCommands ?? defaults.agent.dangerousCommands,
            maxToolCalls: result.config.agent?.maxToolCalls ?? defaults.agent.maxToolCalls
          }
        };
        this.configPath = result.filepath;
      } else {
        // No config file found, use defaults
        this.config = this.getDefaultConfig();
        this.configPath = this._getDefaultConfigPath();
      }

      // Validate the loaded configuration
      this.validate(this.config);

      return this.config;
    } catch (error) {
      throw new Error(`Failed to load configuration: ${error.message}`);
    }
  }

  /**
   * Save configuration to file
   * @param {Object} config - Configuration object to save
   * @returns {Promise<void>}
   */
  async save(config = null) {
    try {
      const configToSave = config || this.config;
      
      if (!configToSave) {
        throw new Error('No configuration to save');
      }

      // Validate before saving
      this.validate(configToSave);

      // Determine save path
      const savePath = this.configPath || this._getDefaultConfigPath();
      
      // Ensure directory exists
      const dir = path.dirname(savePath);
      await fs.mkdir(dir, { recursive: true });

      // Write configuration as JSON
      const configJson = JSON.stringify(configToSave, null, 2);
      await fs.writeFile(savePath, configJson, 'utf8');

      this.config = configToSave;
      this.configPath = savePath;
    } catch (error) {
      throw new Error(`Failed to save configuration: ${error.message}`);
    }
  }

  /**
   * Validate configuration object
   * @param {Object} config - Configuration to validate
   * @throws {Error} If configuration is invalid
   */
  validate(config) {
    if (!config || typeof config !== 'object') {
      throw new Error('Configuration must be an object');
    }

    // Validate mistral configuration
    if (!config.mistral || typeof config.mistral !== 'object') {
      throw new Error('Configuration must include mistral settings');
    }

    // API key can be empty string (will be prompted on first run)
    if (config.mistral.apiKey !== undefined && 
        config.mistral.apiKey !== null && 
        typeof config.mistral.apiKey !== 'string') {
      throw new Error('mistral.apiKey must be a string');
    }

    if (!config.mistral.model || typeof config.mistral.model !== 'string') {
      throw new Error('mistral.model must be a non-empty string');
    }

    if (typeof config.mistral.maxTokens !== 'number' || config.mistral.maxTokens <= 0) {
      throw new Error('mistral.maxTokens must be a positive number');
    }

    if (typeof config.mistral.temperature !== 'number' || 
        config.mistral.temperature < 0 || 
        config.mistral.temperature > 2) {
      throw new Error('mistral.temperature must be a number between 0 and 2');
    }

    // Validate context configuration
    if (!config.context || typeof config.context !== 'object') {
      throw new Error('Configuration must include context settings');
    }

    if (typeof config.context.maxMessages !== 'number' || config.context.maxMessages <= 0) {
      throw new Error('context.maxMessages must be a positive number');
    }

    if (typeof config.context.maxTokens !== 'number' || config.context.maxTokens <= 0) {
      throw new Error('context.maxTokens must be a positive number');
    }

    // Validate files configuration
    if (!config.files || typeof config.files !== 'object') {
      throw new Error('Configuration must include files settings');
    }

    if (typeof config.files.maxFileSize !== 'number' || config.files.maxFileSize <= 0) {
      throw new Error('files.maxFileSize must be a positive number');
    }

    if (!Array.isArray(config.files.ignorePatterns)) {
      throw new Error('files.ignorePatterns must be an array');
    }

    // Validate agent configuration
    if (!config.agent || typeof config.agent !== 'object') {
      throw new Error('Configuration must include agent settings');
    }

    const validModes = ['safe', 'autonomous', 'full'];
    if (!validModes.includes(config.agent.autonomyMode)) {
      throw new Error(`agent.autonomyMode must be one of: ${validModes.join(', ')}`);
    }

    if (!Array.isArray(config.agent.autoApproveTools)) {
      throw new Error('agent.autoApproveTools must be an array');
    }

    if (!Array.isArray(config.agent.dangerousCommands)) {
      throw new Error('agent.dangerousCommands must be an array');
    }

    if (typeof config.agent.maxToolCalls !== 'number' || config.agent.maxToolCalls <= 0) {
      throw new Error('agent.maxToolCalls must be a positive number');
    }

    // Validate UI configuration
    if (!config.ui || typeof config.ui !== 'object') {
      throw new Error('Configuration must include ui settings');
    }

    if (typeof config.ui.syntaxHighlight !== 'boolean') {
      throw new Error('ui.syntaxHighlight must be a boolean');
    }

    // Validate logging configuration
    if (!config.logging || typeof config.logging !== 'object') {
      throw new Error('Configuration must include logging settings');
    }

    const validLogLevels = ['error', 'warn', 'info', 'debug'];
    if (!validLogLevels.includes(config.logging.level)) {
      throw new Error(`logging.level must be one of: ${validLogLevels.join(', ')}`);
    }

    return true;
  }

  /**
   * Get the default configuration file path
   * @returns {string} Default config path
   * @private
   */
  _getDefaultConfigPath() {
    const homeDir = os.homedir();
    return path.join(homeDir, `.${this.configName}rc.json`);
  }

  /**
   * Prompt user for API key
   * @returns {Promise<string>} API key entered by user
   */
  async promptForApiKey() {
    const response = await prompts({
      type: 'password',
      name: 'apiKey',
      message: 'Enter your Mistral AI API key:',
      validate: value => value.length > 0 ? true : 'API key cannot be empty'
    });

    if (!response.apiKey) {
      throw new Error('API key is required');
    }

    return response.apiKey;
  }

  /**
   * Run interactive setup wizard
   * @returns {Promise<Object>} Configuration created by wizard
   */
  async runSetupWizard() {
    console.log(chalk.bold.cyan('\n🤖 AI Terminal Assistant Setup Wizard\n'));
    console.log('Let\'s configure your AI assistant. Press Ctrl+C to cancel.\n');

    const responses = await prompts([
      {
        type: 'password',
        name: 'apiKey',
        message: 'Mistral AI API key:',
        validate: value => value.length > 0 ? true : 'API key is required'
      },
      {
        type: 'select',
        name: 'model',
        message: 'Select Mistral AI model:',
        choices: [
          { title: 'mistral-large-latest (Recommended)', value: 'mistral-large-latest' },
          { title: 'mistral-medium-latest', value: 'mistral-medium-latest' },
          { title: 'mistral-small-latest', value: 'mistral-small-latest' }
        ],
        initial: 0
      },
      {
        type: 'number',
        name: 'maxTokens',
        message: 'Maximum tokens per response:',
        initial: 4096,
        min: 100,
        max: 32000
      },
      {
        type: 'number',
        name: 'temperature',
        message: 'Temperature (0.0 - 2.0, higher = more creative):',
        initial: 0.7,
        min: 0,
        max: 2,
        float: true
      },
      {
        type: 'select',
        name: 'autonomyMode',
        message: 'Agent autonomy mode:',
        choices: [
          { 
            title: 'Safe (Requires approval for file writes and commands)', 
            value: 'safe',
            description: 'Most secure, requires confirmation for modifications'
          },
          { 
            title: 'Autonomous (Auto-approve reads and writes)', 
            value: 'autonomous',
            description: 'Moderate security, auto-approves most operations'
          },
          { 
            title: 'Full (Auto-approve all except dangerous commands)', 
            value: 'full',
            description: 'Least secure, maximum automation'
          }
        ],
        initial: 0
      },
      {
        type: 'select',
        name: 'logLevel',
        message: 'Logging level:',
        choices: [
          { title: 'Error (Only errors)', value: 'error' },
          { title: 'Warn (Warnings and errors)', value: 'warn' },
          { title: 'Info (General information)', value: 'info' },
          { title: 'Debug (Verbose logging)', value: 'debug' }
        ],
        initial: 2
      },
      {
        type: 'confirm',
        name: 'syntaxHighlight',
        message: 'Enable syntax highlighting?',
        initial: true
      }
    ]);

    // Check if user cancelled
    if (!responses.apiKey) {
      throw new Error('Setup cancelled by user');
    }

    // Build configuration from responses
    const config = this.getDefaultConfig();
    
    // Hardcode the API key into the codebase
    await this.apiKeyManager.createApiKeyFile(responses.apiKey);
    
    config.mistral.apiKey = responses.apiKey;
    config.mistral.model = responses.model;
    config.mistral.maxTokens = responses.maxTokens;
    config.mistral.temperature = responses.temperature;
    config.agent.autonomyMode = responses.autonomyMode;
    config.logging.level = responses.logLevel;
    config.ui.syntaxHighlight = responses.syntaxHighlight;

    // Save configuration
    await this.save(config);

    console.log(chalk.green('\n✓ Configuration saved successfully!'));
    console.log(chalk.green('✓ API key hardcoded into src/config/HardcodedApiKey.js'));
    console.log(chalk.gray(`Config file: ${this.configPath}\n`));

    return config;
  }

  /**
   * Update a specific configuration value
   * @param {string} key - Configuration key (dot notation supported, e.g., 'mistral.apiKey')
   * @param {*} value - New value
   * @returns {Promise<void>}
   */
  async updateConfig(key, value) {
    if (!this.config) {
      await this.load();
    }

    // Parse dot notation key
    const keys = key.split('.');
    let current = this.config;
    
    // Navigate to the parent object
    for (let i = 0; i < keys.length - 1; i++) {
      if (!current[keys[i]]) {
        throw new Error(`Invalid configuration key: ${key}`);
      }
      current = current[keys[i]];
    }

    // Get the final key
    const finalKey = keys[keys.length - 1];
    
    if (!(finalKey in current)) {
      throw new Error(`Invalid configuration key: ${key}`);
    }

    // Type conversion based on existing value type
    const existingValue = current[finalKey];
    let convertedValue = value;

    if (typeof existingValue === 'number') {
      convertedValue = parseFloat(value);
      if (isNaN(convertedValue)) {
        throw new Error(`Value for ${key} must be a number`);
      }
    } else if (typeof existingValue === 'boolean') {
      if (typeof value === 'string') {
        convertedValue = value.toLowerCase() === 'true';
      } else {
        convertedValue = Boolean(value);
      }
    }

    // Update the value
    current[finalKey] = convertedValue;

    // Validate and save
    this.validate(this.config);
    await this.save();

    console.log(chalk.green(`✓ Updated ${key} = ${convertedValue}`));
  }

  /**
   * Display current configuration
   * @returns {void}
   */
  displayConfig() {
    if (!this.config) {
      console.log(chalk.yellow('No configuration loaded. Run setup first.'));
      return;
    }

    console.log(chalk.bold.cyan('\n📋 Current Configuration\n'));
    console.log(chalk.gray(`Config file: ${this.configPath || 'Not saved yet'}\n`));

    // Display Mistral settings
    console.log(chalk.bold('Mistral AI:'));
    console.log(`  API Key: ${this.config.mistral.apiKey ? chalk.green('✓ Set') : chalk.red('✗ Not set')}`);
    console.log(`  Model: ${chalk.cyan(this.config.mistral.model)}`);
    console.log(`  Max Tokens: ${chalk.cyan(this.config.mistral.maxTokens)}`);
    console.log(`  Temperature: ${chalk.cyan(this.config.mistral.temperature)}`);

    // Display Context settings
    console.log(chalk.bold('\nContext:'));
    console.log(`  Max Messages: ${chalk.cyan(this.config.context.maxMessages)}`);
    console.log(`  Max Tokens: ${chalk.cyan(this.config.context.maxTokens)}`);

    // Display Files settings
    console.log(chalk.bold('\nFiles:'));
    console.log(`  Max File Size: ${chalk.cyan(this.config.files.maxFileSize)} bytes`);
    console.log(`  Ignore Patterns: ${chalk.cyan(this.config.files.ignorePatterns.length)} patterns`);

    // Display Agent settings
    console.log(chalk.bold('\nAgent:'));
    console.log(`  Autonomy Mode: ${chalk.cyan(this.config.agent.autonomyMode)}`);
    console.log(`  Max Tool Calls: ${chalk.cyan(this.config.agent.maxToolCalls)}`);
    console.log(`  Auto-approve Tools: ${chalk.cyan(this.config.agent.autoApproveTools.length)} tools`);

    // Display UI settings
    console.log(chalk.bold('\nUI:'));
    console.log(`  Theme: ${chalk.cyan(this.config.ui.theme)}`);
    console.log(`  Syntax Highlight: ${this.config.ui.syntaxHighlight ? chalk.green('✓ Enabled') : chalk.red('✗ Disabled')}`);

    // Display Logging settings
    console.log(chalk.bold('\nLogging:'));
    console.log(`  Level: ${chalk.cyan(this.config.logging.level)}`);
    console.log(`  File: ${chalk.cyan(this.config.logging.file)}`);

    console.log('');
  }

  /**
   * Get a configuration value
   * @param {string} key - Configuration key (dot notation supported)
   * @param {*} defaultValue - Default value if key not found
   * @returns {*} Configuration value
   */
  get(key, defaultValue = null) {
    if (!this.config) {
      return defaultValue;
    }

    const keys = key.split('.');
    let current = this.config;

    for (const k of keys) {
      if (current && typeof current === 'object' && k in current) {
        current = current[k];
      } else {
        return defaultValue;
      }
    }

    return current;
  }

  /**
   * Set a configuration value (in memory only, call save() to persist)
   * @param {string} key - Configuration key (dot notation supported)
   * @param {*} value - Value to set
   */
  set(key, value) {
    if (!this.config) {
      this.config = this.getDefaultConfig();
    }

    const keys = key.split('.');
    let current = this.config;

    for (let i = 0; i < keys.length - 1; i++) {
      if (!current[keys[i]]) {
        current[keys[i]] = {};
      }
      current = current[keys[i]];
    }

    current[keys[keys.length - 1]] = value;
  }
}

export default ConfigManager;
