// codestral/tests/MultiStepAgentReasoning.integration.test.js
/**
 * Integration tests for multi-step agent reasoning
 * Tests complex scenarios where the agent needs to use multiple tools
 * and reason through multi-step processes
 */

import { jest } from '@jest/globals';
import AgentOrchestrator from '../src/agent/AgentOrchestrator.js';
import ToolRegistry from '../src/tools/ToolRegistry.js';
import ApprovalManager from '../src/agent/ApprovalManager.js';
import ContextManager from '../src/context/ContextManager.js';
import MistralClient from '../src/ai/MistralClient.js';
import FileSystemManager from '../src/filesystem/FileSystemManager.js';
import SearchEngine from '../src/search/SearchEngine.js';
import CodebaseIndexer from '../src/indexer/CodebaseIndexer.js';
import TerminalInterface from '../src/cli/TerminalInterface.js';
import fs from 'fs/promises';
import path from 'path';
import os from 'os';

// Import tool handlers
import {
  createReadFileHandler, readFileSchema,
  createWriteFileHandler, writeFileSchema,
  createListDirectoryHandler, listDirectorySchema,
  createSearchCodeHandler, searchCodeSchema,
  createReplaceInFileHandler, replaceInFileSchema,
  createCreateDirectoryHandler, createDirectorySchema
} from '../src/tools/handlers/index.js';

describe('Multi-Step Agent Reasoning Integration', () => {
  let tempDir;
  let agentOrchestrator;
  let toolRegistry;
  let approvalManager;
  let contextManager;
  let mistralClient;
  let fileSystemManager;
  let searchEngine;
  let codebaseIndexer;
  let terminalInterface;

  beforeEach(async () => {
    // Create temporary directory for test files
    tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'codestral-agent-test-'));
    
    // Mock configuration
    const mockConfig = {
      mistral: {
        apiKey: 'test-api-key',
        model: 'mistral-large-latest',
        maxTokens: 4000,
        temperature: 0.7
      },
      context: {
        maxTokens: 8000
      },
      agent: {
        autonomyMode: 'autonomous',
        autoApproveTools: ['read_file', 'list_directory', 'search_code'],
        dangerousCommands: ['rm -rf', 'sudo'],
        maxToolCalls: 20
      },
      ui: {
        syntaxHighlight: true,
        theme: 'default'
      }
    };

    // Initialize components
    fileSystemManager = new FileSystemManager(tempDir, []);
    codebaseIndexer = new CodebaseIndexer(tempDir, fileSystemManager);
    searchEngine = new SearchEngine(codebaseIndexer);
    terminalInterface = new TerminalInterface(mockConfig.ui);
    
    // Mock Mistral client
    mistralClient = new MistralClient(mockConfig.mistral.apiKey, mockConfig.mistral);
    mistralClient.registerTools = jest.fn();

    // Initialize tool registry with all tools
    toolRegistry = new ToolRegistry(fileSystemManager, searchEngine);
    toolRegistry.registerTool('read_file', readFileSchema, createReadFileHandler(fileSystemManager));
    toolRegistry.registerTool('write_file', writeFileSchema, createWriteFileHandler(fileSystemManager));
    toolRegistry.registerTool('list_directory', listDirectorySchema, createListDirectoryHandler(fileSystemManager));
    toolRegistry.registerTool('search_code', searchCodeSchema, createSearchCodeHandler(searchEngine));
    toolRegistry.registerTool('replace_in_file', replaceInFileSchema, createReplaceInFileHandler(fileSystemManager));
    toolRegistry.registerTool('create_directory', createDirectorySchema, createCreateDirectoryHandler(fileSystemManager));

    // Initialize approval manager (auto-approve for testing)
    approvalManager = new ApprovalManager(mockConfig, terminalInterface);
    approvalManager.requestApproval = jest.fn().mockResolvedValue(true);

    // Initialize context manager
    contextManager = new ContextManager(mockConfig.context.maxTokens);

    // Initialize agent orchestrator
    agentOrchestrator = new AgentOrchestrator(
      mistralClient,
      toolRegistry,
      approvalManager,
      {
        maxToolCalls: mockConfig.agent.maxToolCalls,
        maxIterations: 10,
        streamResponses: false,
        displayToolCalls: false
      }
    );

    // Create test project structure
    await createTestProject();
    await codebaseIndexer.buildIndex();
  });

  afterEach(async () => {
    // Clean up temporary directory
    try {
      await fs.rm(tempDir, { recursive: true, force: true });
    } catch (error) {
      // Ignore cleanup errors
    }
  });

  async function createTestProject() {
    // Create a realistic project structure
    await fs.mkdir(path.join(tempDir, 'src'), { recursive: true });
    await fs.mkdir(path.join(tempDir, 'tests'), { recursive: true });
    await fs.mkdir(path.join(tempDir, 'docs'), { recursive: true });

    // Main application file with bugs
    await fs.writeFile(path.join(tempDir, 'src', 'app.js'), `
const express = require('express');
const app = express();

// TODO: Add input validation
app.post('/users', (req, res) => {
  const { name, email } = req.body;
  // BUG: No validation of input
  const user = { id: Date.now(), name, email };
  res.json(user);
});

// TODO: Add error handling
app.get('/users/:id', (req, res) => {
  const id = req.params.id;
  // BUG: No error handling for invalid ID
  const user = findUserById(id);
  res.json(user);
});

function findUserById(id) {
  // TODO: Implement database lookup
  return { id, name: 'Test User', email: 'test@example.com' };
}

module.exports = app;
`);

    // Utility file with incomplete functions
    await fs.writeFile(path.join(tempDir, 'src', 'utils.js'), `
// Utility functions
function validateEmail(email) {
  // TODO: Implement proper email validation
  return email.includes('@');
}

function sanitizeInput(input) {
  // TODO: Add proper sanitization
  return input;
}

function hashPassword(password) {
  // TODO: Implement secure password hashing
  return password;
}

module.exports = {
  validateEmail,
  sanitizeInput,
  hashPassword
};
`);

    // Test file with missing tests
    await fs.writeFile(path.join(tempDir, 'tests', 'app.test.js'), `
const request = require('supertest');
const app = require('../src/app');

describe('App', () => {
  // TODO: Add tests for POST /users
  
  // TODO: Add tests for GET /users/:id
  
  // TODO: Add tests for error cases
});
`);

    // Package.json
    await fs.writeFile(path.join(tempDir, 'package.json'), `
{
  "name": "test-app",
  "version": "1.0.0",
  "description": "Test application",
  "main": "src/app.js",
  "scripts": {
    "start": "node src/app.js",
    "test": "jest"
  },
  "dependencies": {
    "express": "^4.18.0"
  },
  "devDependencies": {
    "jest": "^29.0.0",
    "supertest": "^6.0.0"
  }
}
`);

    // README with incomplete documentation
    await fs.writeFile(path.join(tempDir, 'README.md'), `
# Test Application

A simple Express.js application.

## Installation

\`\`\`bash
npm install
\`\`\`

## Usage

TODO: Add usage instructions

## API

TODO: Document API endpoints

## Testing

TODO: Add testing instructions
`);
  }

  describe('Code Analysis and Improvement Workflow', () => {
    it('should analyze codebase and suggest improvements', async () => {
      // Mock AI responses for multi-step analysis
      const responses = [
        // Step 1: List project structure
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'list_directory',
              arguments: JSON.stringify({ path: '.', recursive: true })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Search for TODO comments
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'search_code',
              arguments: JSON.stringify({ query: 'TODO' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Read main app file
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Read utils file
        {
          content: null,
          toolCalls: [{
            id: 'call_4',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/utils.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 5: Final analysis
        {
          content: 'I\'ve analyzed your codebase and found several areas for improvement:\\n\\n1. Input validation is missing in the POST /users endpoint\\n2. Error handling is incomplete\\n3. Password hashing is not implemented\\n4. Tests are missing\\n\\nWould you like me to implement these improvements?',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3])
        .mockResolvedValueOnce(responses[4]);

      const result = await agentOrchestrator.processMessage(
        'Analyze my codebase and suggest improvements',
        contextManager
      );

      expect(result.content).toContain('analyzed your codebase');
      expect(result.content).toContain('Input validation');
      expect(result.content).toContain('Error handling');
      expect(result.iterations).toBe(5);
      expect(result.toolCallsUsed).toBe(4);

      // Verify all tools were called
      expect(mistralClient.sendWithTools).toHaveBeenCalledTimes(5);
    });

    it('should implement security improvements across multiple files', async () => {
      // Mock AI responses for security implementation
      const responses = [
        // Step 1: Search for security issues
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'search_code',
              arguments: JSON.stringify({ query: 'req.body' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Read app.js to understand current implementation
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Update utils.js with proper validation
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'replace_in_file',
              arguments: JSON.stringify({
                path: 'src/utils.js',
                search: 'function validateEmail(email) {\\n  // TODO: Implement proper email validation\\n  return email.includes(\'@\');\\n}',
                replace: 'function validateEmail(email) {\\n  const emailRegex = /^[^\\\\s@]+@[^\\\\s@]+\\\\.[^\\\\s@]+$/;\\n  return emailRegex.test(email);\\n}'
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Update sanitization function
        {
          content: null,
          toolCalls: [{
            id: 'call_4',
            function: {
              name: 'replace_in_file',
              arguments: JSON.stringify({
                path: 'src/utils.js',
                search: 'function sanitizeInput(input) {\\n  // TODO: Add proper sanitization\\n  return input;\\n}',
                replace: 'function sanitizeInput(input) {\\n  return input.trim().replace(/[<>\"\'&]/g, \'\');\\n}'
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 5: Update app.js with validation
        {
          content: null,
          toolCalls: [{
            id: 'call_5',
            function: {
              name: 'replace_in_file',
              arguments: JSON.stringify({
                path: 'src/app.js',
                search: 'app.post(\'/users\', (req, res) => {\\n  const { name, email } = req.body;\\n  // BUG: No validation of input\\n  const user = { id: Date.now(), name, email };\\n  res.json(user);\\n});',
                replace: 'app.post(\'/users\', (req, res) => {\\n  const { name, email } = req.body;\\n  \\n  if (!name || !email) {\\n    return res.status(400).json({ error: \'Name and email are required\' });\\n  }\\n  \\n  if (!validateEmail(email)) {\\n    return res.status(400).json({ error: \'Invalid email format\' });\\n  }\\n  \\n  const user = { \\n    id: Date.now(), \\n    name: sanitizeInput(name), \\n    email: sanitizeInput(email) \\n  };\\n  res.json(user);\\n});'
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 6: Final confirmation
        {
          content: 'I\'ve successfully implemented security improvements:\\n\\n1. ✅ Added proper email validation using regex\\n2. ✅ Implemented input sanitization to prevent XSS\\n3. ✅ Added input validation in the POST /users endpoint\\n4. ✅ Added proper error responses\\n\\nYour application is now more secure!',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3])
        .mockResolvedValueOnce(responses[4])
        .mockResolvedValueOnce(responses[5]);

      const result = await agentOrchestrator.processMessage(
        'Implement security improvements for input validation and sanitization',
        contextManager
      );

      expect(result.content).toContain('security improvements');
      expect(result.content).toContain('email validation');
      expect(result.content).toContain('input sanitization');
      expect(result.iterations).toBe(6);
      expect(result.toolCallsUsed).toBe(5);

      // Verify the agent successfully completed the security improvements
      // The agent should have made multiple tool calls to implement security features
      expect(result.toolCallsUsed).toBeGreaterThan(3); // Should have made several replace_in_file calls
      expect(result.content).toContain('security'); // Response should mention security improvements
      expect(result.finishReason).not.toBe('error'); // Should complete successfully
    });

    it('should create comprehensive test suite', async () => {
      // Mock AI responses for test creation
      const responses = [
        // Step 1: Read existing test file
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'tests/app.test.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Read app.js to understand endpoints
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Replace test file with comprehensive tests
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'tests/app.test.js',
                content: `const request = require('supertest');\\nconst app = require('../src/app');\\n\\ndescribe('App', () => {\\n  describe('POST /users', () => {\\n    it('should create user with valid data', async () => {\\n      const userData = { name: 'John Doe', email: 'john@example.com' };\\n      const response = await request(app)\\n        .post('/users')\\n        .send(userData)\\n        .expect(200);\\n      \\n      expect(response.body).toHaveProperty('id');\\n      expect(response.body.name).toBe(userData.name);\\n      expect(response.body.email).toBe(userData.email);\\n    });\\n    \\n    it('should return 400 for missing name', async () => {\\n      const userData = { email: 'john@example.com' };\\n      await request(app)\\n        .post('/users')\\n        .send(userData)\\n        .expect(400);\\n    });\\n    \\n    it('should return 400 for invalid email', async () => {\\n      const userData = { name: 'John', email: 'invalid-email' };\\n      await request(app)\\n        .post('/users')\\n        .send(userData)\\n        .expect(400);\\n    });\\n  });\\n  \\n  describe('GET /users/:id', () => {\\n    it('should return user by id', async () => {\\n      const response = await request(app)\\n        .get('/users/123')\\n        .expect(200);\\n      \\n      expect(response.body).toHaveProperty('id');\\n      expect(response.body).toHaveProperty('name');\\n      expect(response.body).toHaveProperty('email');\\n    });\\n  });\\n});`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Create utils test file
        {
          content: null,
          toolCalls: [{
            id: 'call_4',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'tests/utils.test.js',
                content: `const { validateEmail, sanitizeInput } = require('../src/utils');\\n\\ndescribe('Utils', () => {\\n  describe('validateEmail', () => {\\n    it('should validate correct email', () => {\\n      expect(validateEmail('test@example.com')).toBe(true);\\n    });\\n    \\n    it('should reject invalid email', () => {\\n      expect(validateEmail('invalid-email')).toBe(false);\\n      expect(validateEmail('test@')).toBe(false);\\n      expect(validateEmail('@example.com')).toBe(false);\\n    });\\n  });\\n  \\n  describe('sanitizeInput', () => {\\n    it('should remove dangerous characters', () => {\\n      expect(sanitizeInput('<script>alert(1)</script>')).toBe('scriptalert(1)/script');\\n      expect(sanitizeInput('test\"quote')).toBe('testquote');\\n    });\\n    \\n    it('should trim whitespace', () => {\\n      expect(sanitizeInput('  test  ')).toBe('test');\\n    });\\n  });\\n});`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 5: Final summary
        {
          content: 'I\'ve created a comprehensive test suite:\\n\\n1. ✅ Added tests for POST /users endpoint (success and error cases)\\n2. ✅ Added tests for GET /users/:id endpoint\\n3. ✅ Created utils.test.js with validation and sanitization tests\\n4. ✅ Tests cover both happy path and error scenarios\\n\\nYou can now run `npm test` to execute the test suite!',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3])
        .mockResolvedValueOnce(responses[4]);

      const result = await agentOrchestrator.processMessage(
        'Create comprehensive tests for all the application functionality',
        contextManager
      );

      expect(result.content).toContain('comprehensive test suite');
      expect(result.content).toContain('POST /users');
      expect(result.content).toContain('utils.test.js');
      expect(result.iterations).toBe(5);
      expect(result.toolCallsUsed).toBe(4);

      // Verify test files were created
      const appTests = await fs.readFile(path.join(tempDir, 'tests', 'app.test.js'), 'utf8');
      expect(appTests).toContain('should create user with valid data');
      expect(appTests).toContain('should return 400 for missing name');

      const utilsTests = await fs.readFile(path.join(tempDir, 'tests', 'utils.test.js'), 'utf8');
      expect(utilsTests).toContain('validateEmail');
      expect(utilsTests).toContain('sanitizeInput');
    });
  });

  describe('Documentation Generation Workflow', () => {
    it('should analyze code and generate comprehensive documentation', async () => {
      const responses = [
        // Step 1: List all files to understand structure
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'list_directory',
              arguments: JSON.stringify({ path: '.', recursive: true })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Read package.json for project info
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'package.json' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Read main app file
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Update README with comprehensive documentation
        {
          content: null,
          toolCalls: [{
            id: 'call_4',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'README.md',
                content: `# Test Application\\n\\nA simple Express.js REST API application with user management functionality.\\n\\n## Features\\n\\n- User creation with validation\\n- User retrieval by ID\\n- Input sanitization and security\\n- Comprehensive test suite\\n\\n## Installation\\n\\n\`\`\`bash\\nnpm install\\n\`\`\`\\n\\n## Usage\\n\\n### Starting the Server\\n\\n\`\`\`bash\\nnpm start\\n\`\`\`\\n\\nThe server will start on port 3000.\\n\\n### Running Tests\\n\\n\`\`\`bash\\nnpm test\\n\`\`\`\\n\\n## API Endpoints\\n\\n### POST /users\\n\\nCreate a new user.\\n\\n**Request Body:**\\n\`\`\`json\\n{\\n  \"name\": \"John Doe\",\\n  \"email\": \"john@example.com\"\\n}\\n\`\`\`\\n\\n**Response:**\\n\`\`\`json\\n{\\n  \"id\": 1234567890,\\n  \"name\": \"John Doe\",\\n  \"email\": \"john@example.com\"\\n}\\n\`\`\`\\n\\n### GET /users/:id\\n\\nRetrieve a user by ID.\\n\\n**Response:**\\n\`\`\`json\\n{\\n  \"id\": \"123\",\\n  \"name\": \"Test User\",\\n  \"email\": \"test@example.com\"\\n}\\n\`\`\`\\n\\n## Project Structure\\n\\n\`\`\`\\nsrc/\\n  app.js          # Main application file\\n  utils.js        # Utility functions\\ntests/\\n  app.test.js     # Application tests\\n  utils.test.js   # Utility tests\\npackage.json      # Project configuration\\nREADME.md         # This file\\n\`\`\`\\n\\n## Security Features\\n\\n- Email validation using regex\\n- Input sanitization to prevent XSS\\n- Request validation with error handling\\n\\n## Contributing\\n\\n1. Fork the repository\\n2. Create a feature branch\\n3. Make your changes\\n4. Add tests\\n5. Submit a pull request`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 5: Create API documentation
        {
          content: null,
          toolCalls: [{
            id: 'call_5',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'docs/API.md',
                content: `# API Documentation\\n\\n## Overview\\n\\nThis document describes the REST API endpoints available in the Test Application.\\n\\n## Base URL\\n\\n\`http://localhost:3000\`\\n\\n## Endpoints\\n\\n### Users\\n\\n#### Create User\\n\\n\`POST /users\`\\n\\nCreates a new user with the provided information.\\n\\n**Parameters:**\\n\\n| Name  | Type   | Required | Description           |\\n|-------|--------|----------|-----------------------|\\n| name  | string | Yes      | User's full name      |\\n| email | string | Yes      | User's email address  |\\n\\n**Example Request:**\\n\\n\`\`\`bash\\ncurl -X POST http://localhost:3000/users \\\\\\n  -H \"Content-Type: application/json\" \\\\\\n  -d '{\\n    \"name\": \"John Doe\",\\n    \"email\": \"john@example.com\"\\n  }'\\n\`\`\`\\n\\n**Example Response:**\\n\\n\`\`\`json\\n{\\n  \"id\": 1640995200000,\\n  \"name\": \"John Doe\",\\n  \"email\": \"john@example.com\"\\n}\\n\`\`\`\\n\\n**Error Responses:**\\n\\n- \`400 Bad Request\`: Missing required fields or invalid email format\\n\\n#### Get User\\n\\n\`GET /users/:id\`\\n\\nRetrieves a user by their ID.\\n\\n**Parameters:**\\n\\n| Name | Type   | Required | Description |\\n|------|--------|----------|-------------|\\n| id   | string | Yes      | User ID     |\\n\\n**Example Request:**\\n\\n\`\`\`bash\\ncurl http://localhost:3000/users/123\\n\`\`\`\\n\\n**Example Response:**\\n\\n\`\`\`json\\n{\\n  \"id\": \"123\",\\n  \"name\": \"Test User\",\\n  \"email\": \"test@example.com\"\\n}\\n\`\`\`\\n\\n## Error Handling\\n\\nAll endpoints return appropriate HTTP status codes:\\n\\n- \`200 OK\`: Successful request\\n- \`400 Bad Request\`: Invalid input data\\n- \`404 Not Found\`: Resource not found\\n- \`500 Internal Server Error\`: Server error`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 6: Final summary
        {
          content: 'I\'ve generated comprehensive documentation for your project:\\n\\n1. ✅ Updated README.md with complete project overview\\n2. ✅ Added installation and usage instructions\\n3. ✅ Documented all API endpoints with examples\\n4. ✅ Created detailed API documentation in docs/API.md\\n5. ✅ Included project structure and security features\\n\\nYour project now has professional-grade documentation!',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3])
        .mockResolvedValueOnce(responses[4])
        .mockResolvedValueOnce(responses[5]);

      const result = await agentOrchestrator.processMessage(
        'Generate comprehensive documentation for this project including API docs',
        contextManager
      );

      expect(result.content).toContain('comprehensive documentation');
      expect(result.content).toContain('README.md');
      expect(result.content).toContain('API documentation');
      expect(result.iterations).toBe(6);
      expect(result.toolCallsUsed).toBe(5);

      // Verify documentation was created
      const readme = await fs.readFile(path.join(tempDir, 'README.md'), 'utf8');
      expect(readme).toContain('## Features');
      expect(readme).toContain('## API Endpoints');
      expect(readme).toContain('POST /users');

      const apiDocs = await fs.readFile(path.join(tempDir, 'docs', 'API.md'), 'utf8');
      expect(apiDocs).toContain('# API Documentation');
      expect(apiDocs).toContain('Create User');
      expect(apiDocs).toContain('Example Request');
    });
  });

  describe('Complex Refactoring Workflow', () => {
    it('should refactor code structure and improve organization', async () => {
      const responses = [
        // Step 1: Analyze current structure
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'list_directory',
              arguments: JSON.stringify({ path: 'src', recursive: true })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Read app.js to understand current structure
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Create controllers directory
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'create_directory',
              arguments: JSON.stringify({ path: 'src/controllers' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Create user controller
        {
          content: null,
          toolCalls: [{
            id: 'call_4',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'src/controllers/userController.js',
                content: `const { validateEmail, sanitizeInput } = require('../utils');\\n\\nclass UserController {\\n  static createUser(req, res) {\\n    const { name, email } = req.body;\\n    \\n    if (!name || !email) {\\n      return res.status(400).json({ error: 'Name and email are required' });\\n    }\\n    \\n    if (!validateEmail(email)) {\\n      return res.status(400).json({ error: 'Invalid email format' });\\n    }\\n    \\n    const user = { \\n      id: Date.now(), \\n      name: sanitizeInput(name), \\n      email: sanitizeInput(email) \\n    };\\n    \\n    res.json(user);\\n  }\\n  \\n  static getUserById(req, res) {\\n    const id = req.params.id;\\n    \\n    try {\\n      const user = UserController.findUserById(id);\\n      res.json(user);\\n    } catch (error) {\\n      res.status(404).json({ error: 'User not found' });\\n    }\\n  }\\n  \\n  static findUserById(id) {\\n    // TODO: Implement database lookup\\n    return { id, name: 'Test User', email: 'test@example.com' };\\n  }\\n}\\n\\nmodule.exports = UserController;`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 5: Create routes directory
        {
          content: null,
          toolCalls: [{
            id: 'call_5',
            function: {
              name: 'create_directory',
              arguments: JSON.stringify({ path: 'src/routes' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 6: Create user routes
        {
          content: null,
          toolCalls: [{
            id: 'call_6',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'src/routes/userRoutes.js',
                content: `const express = require('express');\\nconst UserController = require('../controllers/userController');\\n\\nconst router = express.Router();\\n\\nrouter.post('/users', UserController.createUser);\\nrouter.get('/users/:id', UserController.getUserById);\\n\\nmodule.exports = router;`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 7: Update main app.js
        {
          content: null,
          toolCalls: [{
            id: 'call_7',
            function: {
              name: 'write_file',
              arguments: JSON.stringify({
                path: 'src/app.js',
                content: `const express = require('express');\\nconst userRoutes = require('./routes/userRoutes');\\n\\nconst app = express();\\n\\n// Middleware\\napp.use(express.json());\\n\\n// Routes\\napp.use('/', userRoutes);\\n\\n// Health check endpoint\\napp.get('/health', (req, res) => {\\n  res.json({ status: 'OK', timestamp: new Date().toISOString() });\\n});\\n\\nmodule.exports = app;`
              })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 8: Final summary
        {
          content: 'I\'ve successfully refactored your application with improved organization:\\n\\n1. ✅ Created MVC structure with controllers and routes\\n2. ✅ Separated user logic into UserController class\\n3. ✅ Created dedicated route files for better organization\\n4. ✅ Updated main app.js to use the new structure\\n5. ✅ Added proper middleware and health check endpoint\\n\\nYour code is now more maintainable and follows best practices!',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3])
        .mockResolvedValueOnce(responses[4])
        .mockResolvedValueOnce(responses[5])
        .mockResolvedValueOnce(responses[6])
        .mockResolvedValueOnce(responses[7]);

      const result = await agentOrchestrator.processMessage(
        'Refactor the application to follow MVC pattern with proper separation of concerns',
        contextManager
      );

      expect(result.content).toContain('refactored your application');
      expect(result.content).toContain('MVC structure');
      expect(result.content).toContain('UserController');
      expect(result.iterations).toBe(8);
      expect(result.toolCallsUsed).toBe(7);

      // Verify new structure was created
      const controllerExists = await fs.access(path.join(tempDir, 'src', 'controllers', 'userController.js')).then(() => true).catch(() => false);
      expect(controllerExists).toBe(true);

      const routesExists = await fs.access(path.join(tempDir, 'src', 'routes', 'userRoutes.js')).then(() => true).catch(() => false);
      expect(routesExists).toBe(true);

      const newApp = await fs.readFile(path.join(tempDir, 'src', 'app.js'), 'utf8');
      expect(newApp).toContain('userRoutes');
      expect(newApp).toContain('/health');
    });
  });

  describe('Error Handling and Recovery', () => {
    it('should handle tool failures and adapt strategy', async () => {
      const responses = [
        // Step 1: Try to read non-existent file
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'nonexistent.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 2: Adapt and list directory instead
        {
          content: null,
          toolCalls: [{
            id: 'call_2',
            function: {
              name: 'list_directory',
              arguments: JSON.stringify({ path: '.' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 3: Read existing file
        {
          content: null,
          toolCalls: [{
            id: 'call_3',
            function: {
              name: 'read_file',
              arguments: JSON.stringify({ path: 'src/app.js' })
            }
          }],
          finishReason: 'tool_calls'
        },
        // Step 4: Provide helpful response
        {
          content: 'I couldn\'t find the file you mentioned, but I was able to analyze your existing codebase. Here\'s what I found in your project structure...',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1])
        .mockResolvedValueOnce(responses[2])
        .mockResolvedValueOnce(responses[3]);

      const result = await agentOrchestrator.processMessage(
        'Analyze the nonexistent.js file',
        contextManager
      );

      expect(result.content).toContain('couldn\'t find the file');
      expect(result.content).toContain('existing codebase');
      expect(result.iterations).toBe(4);
      expect(result.toolCallsUsed).toBe(3);

      // Verify that the agent adapted its strategy
      const stats = agentOrchestrator.getToolCallStats();
      expect(stats.failedCalls).toBe(1);
      expect(stats.successfulCalls).toBe(2);
    });

    it('should handle context overflow gracefully', async () => {
      // Fill context with many messages
      for (let i = 0; i < 100; i++) {
        contextManager.addMessage('user', `Message ${i}: ${'x'.repeat(100)}`);
        contextManager.addMessage('assistant', `Response ${i}: ${'y'.repeat(100)}`);
      }

      const responses = [
        {
          content: null,
          toolCalls: [{
            id: 'call_1',
            function: {
              name: 'list_directory',
              arguments: JSON.stringify({ path: '.' })
            }
          }],
          finishReason: 'tool_calls'
        },
        {
          content: 'I can see your project structure. The context was getting full, but I managed to process your request.',
          toolCalls: null,
          finishReason: 'stop'
        }
      ];

      mistralClient.sendWithTools = jest.fn()
        .mockResolvedValueOnce(responses[0])
        .mockResolvedValueOnce(responses[1]);

      const result = await agentOrchestrator.processMessage(
        'What files do I have?',
        contextManager
      );

      expect(result.content).toContain('project structure');
      expect(result.iterations).toBe(2);
      expect(result.toolCallsUsed).toBe(1);

      // Context should have been managed
      expect(contextManager.getTotalTokens()).toBeLessThanOrEqual(contextManager.maxTokens);
    });
  });
});
