// codestral/tests/SearchFunctionality.integration.test.js
/**
 * Integration tests for search functionality
 * Tests the complete search workflow including indexing, searching, and filtering
 */

import { jest } from '@jest/globals';
import SearchEngine from '../src/search/SearchEngine.js';
import CodebaseIndexer from '../src/indexer/CodebaseIndexer.js';
import FileSystemManager from '../src/filesystem/FileSystemManager.js';
import fs from 'fs/promises';
import path from 'path';
import os from 'os';

describe('Search Functionality Integration', () => {
    let tempDir;
    let fileSystemManager;
    let codebaseIndexer;
    let searchEngine;

    beforeEach(async () => {
        // Create temporary directory for test files
        tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'codestral-search-test-'));

        // Initialize components
        fileSystemManager = new FileSystemManager(tempDir);
        codebaseIndexer = new CodebaseIndexer(tempDir, fileSystemManager);
        searchEngine = new SearchEngine(codebaseIndexer);

        // Create test file structure
        await createTestFiles();

        // Build initial index
        await codebaseIndexer.buildIndex();
    });

    afterEach(async () => {
        // Clean up temporary directory
        try {
            await fs.rm(tempDir, { recursive: true, force: true });
        } catch (error) {
            // Ignore cleanup errors
        }
    });

    async function createTestFiles() {
        // Create directory structure
        await fs.mkdir(path.join(tempDir, 'src'), { recursive: true });
        await fs.mkdir(path.join(tempDir, 'tests'), { recursive: true });
        await fs.mkdir(path.join(tempDir, 'docs'), { recursive: true });

        // Create JavaScript files
        await fs.writeFile(path.join(tempDir, 'src', 'index.js'), `
// Main application entry point
const express = require('express');
const app = express();

// TODO: Add authentication middleware
app.get('/', (req, res) => {
  res.send('Hello World');
});

function startServer() {
  app.listen(3000, () => {
    console.log('Server running on port 3000');
  });
}

module.exports = { startServer };
`);

        await fs.writeFile(path.join(tempDir, 'src', 'utils.js'), `
// Utility functions
function formatDate(date) {
  return date.toISOString().split('T')[0];
}

function validateEmail(email) {
  const regex = /^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$/;
  return regex.test(email);
}

// TODO: Add more validation functions
function sanitizeInput(input) {
  return input.trim().replace(/[<>]/g, '');
}

module.exports = {
  formatDate,
  validateEmail,
  sanitizeInput
};
`);

        await fs.writeFile(path.join(tempDir, 'src', 'database.js'), `
// Database connection and queries
const mysql = require('mysql2');

const connection = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'password',
  database: 'myapp'
});

function getUserById(id) {
  return new Promise((resolve, reject) => {
    connection.query('SELECT * FROM users WHERE id = ?', [id], (err, results) => {
      if (err) reject(err);
      else resolve(results[0]);
    });
  });
}

function createUser(userData) {
  const { name, email, password } = userData;
  return new Promise((resolve, reject) => {
    connection.query(
      'INSERT INTO users (name, email, password) VALUES (?, ?, ?)',
      [name, email, password],
      (err, results) => {
        if (err) reject(err);
        else resolve(results.insertId);
      }
    );
  });
}

module.exports = { getUserById, createUser };
`);

        // Create test files
        await fs.writeFile(path.join(tempDir, 'tests', 'utils.test.js'), `
const { formatDate, validateEmail } = require('../src/utils');

describe('Utils', () => {
  test('formatDate should format date correctly', () => {
    const date = new Date('2023-01-01');
    expect(formatDate(date)).toBe('2023-01-01');
  });

  test('validateEmail should validate email addresses', () => {
    expect(validateEmail('test@example.com')).toBe(true);
    expect(validateEmail('invalid-email')).toBe(false);
  });
});
`);

        // Create Python files
        await fs.writeFile(path.join(tempDir, 'src', 'helper.py'), `
# Python helper functions
import re
from datetime import datetime

def validate_phone(phone):
    """Validate phone number format"""
    pattern = r'^\\+?1?\\d{9,15}$'
    return re.match(pattern, phone) is not None

def get_current_timestamp():
    """Get current timestamp"""
    return datetime.now().isoformat()

# TODO: Add more helper functions
class DataProcessor:
    def __init__(self):
        self.data = []
    
    def process_item(self, item):
        """Process a single data item"""
        return item.strip().lower()
`);

        // Create documentation
        await fs.writeFile(path.join(tempDir, 'README.md'), `
# My Application

This is a sample application for testing search functionality.

## Features

- Express.js server
- Database integration
- Utility functions
- TODO: Add more features

## Installation

\`\`\`bash
npm install
\`\`\`

## Usage

\`\`\`javascript
const { startServer } = require('./src/index');
startServer();
\`\`\`
`);

        await fs.writeFile(path.join(tempDir, 'docs', 'api.md'), `
# API Documentation

## Endpoints

### GET /
Returns a welcome message.

### POST /users
Creates a new user.

TODO: Document more endpoints
`);

        // Create package.json
        await fs.writeFile(path.join(tempDir, 'package.json'), `
{
  "name": "test-app",
  "version": "1.0.0",
  "description": "Test application",
  "main": "src/index.js",
  "scripts": {
    "start": "node src/index.js",
    "test": "jest"
  },
  "dependencies": {
    "express": "^4.18.0",
    "mysql2": "^3.0.0"
  },
  "devDependencies": {
    "jest": "^29.0.0"
  }
}
`);
    }

    describe('Basic Text Search', () => {
        it('should find text matches across multiple files', async () => {
            const results = await searchEngine.search('TODO');

            expect(results.length).toBeGreaterThan(0);
            expect(results.some(r => r.filePath.includes('index.js'))).toBe(true);
            expect(results.some(r => r.filePath.includes('utils.js'))).toBe(true);
            expect(results.some(r => r.filePath.includes('README.md'))).toBe(true);
        });

        it('should return results with correct structure', async () => {
            const results = await searchEngine.search('function');

            expect(results.length).toBeGreaterThan(0);
            const result = results[0];

            expect(result).toHaveProperty('filePath');
            expect(result).toHaveProperty('lineNumber');
            expect(result).toHaveProperty('line');
            expect(result).toHaveProperty('context');
            expect(result.context).toHaveProperty('before');
            expect(result.context).toHaveProperty('after');
        });

        it('should include context lines around matches', async () => {
            const results = await searchEngine.search('startServer');

            const result = results.find(r => r.filePath.includes('index.js'));
            expect(result).toBeDefined();
            expect(result.context.before.length).toBeGreaterThan(0);
            expect(result.context.after.length).toBeGreaterThan(0);
        });

        it('should support case insensitive search', async () => {
            const results = await searchEngine.search('todo', { caseSensitive: false });

            // Should find 'TODO' when searching for 'todo' without case sensitivity
            expect(results.length).toBeGreaterThan(0);
        });

        it('should support whole word matching', async () => {
            const results = await searchEngine.search('TODO', { wholeWord: true });

            // Should find standalone 'TODO' but not 'TODO' within other words
            expect(results.length).toBeGreaterThan(0);
            expect(results.every(r => /\bTODO\b/i.test(r.line))).toBe(true);
        });

        it('should search in multiple file types', async () => {
            const results = await searchEngine.search('function');

            const jsResults = results.filter(r => r.filePath.endsWith('.js'));
            const pyResults = results.filter(r => r.filePath.endsWith('.py'));

            expect(jsResults.length).toBeGreaterThan(0);
            expect(pyResults.length).toBeGreaterThan(0);
        });
    });

    describe('Regex Search', () => {
        it('should find regex pattern matches', async () => {
            const results = await searchEngine.searchRegex('function\\s+\\w+');

            expect(results.length).toBeGreaterThan(0);
            expect(results.some(r => r.line.includes('function startServer'))).toBe(true);
        });

        it('should support regex flags', async () => {
            const results = await searchEngine.searchRegex('FUNCTION', { flags: 'i' });

            // Should find 'function' with case-insensitive flag
            expect(results.length).toBeGreaterThan(0);
        });

        it('should throw error for invalid regex', async () => {
            await expect(searchEngine.searchRegex('[')).rejects.toThrow();
        });

        it('should match complex patterns', async () => {
            const results = await searchEngine.searchRegex('console\\.log\\([\'"].*?[\'"]\\)');

            expect(results.length).toBeGreaterThan(0);
            expect(results.some(r => r.line.includes('console.log'))).toBe(true);
        });

        it('should include context lines for regex matches', async () => {
            const results = await searchEngine.searchRegex('require\\([\'"].*?[\'"]\\)');

            const result = results[0];
            expect(result.context).toBeDefined();
            expect(result.context.before).toBeDefined();
            expect(result.context.after).toBeDefined();
        });
    });

    describe('Filtering and Sorting', () => {
        it('should filter by file type', async () => {
            const allResults = await searchEngine.search('function');
            const jsResults = searchEngine.filterByFileType(allResults, ['.js']);

            expect(jsResults.length).toBeLessThanOrEqual(allResults.length);
            expect(jsResults.every(r => r.filePath.endsWith('.js'))).toBe(true);
        });

        it('should filter by directory', async () => {
            const allResults = await searchEngine.search('function');
            const srcResults = searchEngine.filterByDirectory(allResults, 'src');

            expect(srcResults.length).toBeLessThanOrEqual(allResults.length);
            expect(srcResults.every(r => r.filePath.includes('src/'))).toBe(true);
        });

        it('should filter by multiple criteria', async () => {
            const results = await searchEngine.search('function', {
                fileTypes: ['.js'],
                directories: ['src'],
                excludeDirectories: ['node_modules']
            });

            expect(results.every(r => r.filePath.endsWith('.js'))).toBe(true);
            expect(results.every(r => r.filePath.includes('src/'))).toBe(true);
            expect(results.every(r => !r.filePath.includes('node_modules'))).toBe(true);
        });

        it('should sort results by relevance', async () => {
            const results = await searchEngine.search('function');
            const ranked = searchEngine.rankResults(results, 'function');

            // Results should be sorted by score (descending)
            for (let i = 1; i < ranked.length; i++) {
                expect(ranked[i - 1].score).toBeGreaterThanOrEqual(ranked[i].score);
            }
        });

        it('should boost exact matches', async () => {
            const results = await searchEngine.search('TODO');
            const ranked = searchEngine.rankResults(results, 'TODO');

            // Results with exact matches should have higher scores
            const exactMatch = ranked.find(r => r.line.toLowerCase().includes('todo'));
            expect(exactMatch).toBeDefined();
            expect(exactMatch.score).toBeGreaterThan(1.0);
        });

        it('should boost matches at start of line', async () => {
            const results = await searchEngine.search('function');
            const ranked = searchEngine.rankResults(results, 'function');

            const startMatch = ranked.find(r =>
                r.line.trim().toLowerCase().startsWith('function')
            );

            expect(startMatch).toBeDefined();
            expect(startMatch.score).toBeGreaterThan(1.0);
        });
    });

    describe('Advanced Search Options', () => {
        it('should limit number of results', async () => {
            const results = await searchEngine.search('function', { limit: 5 });

            expect(results.length).toBeLessThanOrEqual(5);
        });

        it('should provide context with configurable size', async () => {
            const results = await searchEngine.search('startServer', {
                contextLines: 3
            });

            const result = results[0];
            expect(result.context.before.length).toBeLessThanOrEqual(3);
            expect(result.context.after.length).toBeLessThanOrEqual(3);
        });

        it('should exclude binary files', async () => {
            // Create a binary file
            const binaryContent = Buffer.from([0x00, 0x01, 0x02, 0xFF]);
            await fs.writeFile(path.join(tempDir, 'binary.dat'), binaryContent);

            // Rebuild index to include binary file
            await codebaseIndexer.buildIndex();

            const results = await searchEngine.search('function');

            // Should not include binary files in results
            expect(results.every(r => !r.filePath.endsWith('.dat'))).toBe(true);
        });

        it('should respect .gitignore patterns', async () => {
            // Create .gitignore
            await fs.writeFile(path.join(tempDir, '.gitignore'), 'node_modules/\\n*.log\\n');

            // Create ignored files
            await fs.mkdir(path.join(tempDir, 'node_modules'));
            await fs.writeFile(path.join(tempDir, 'node_modules', 'ignored.js'), 'function ignored() {}');
            await fs.writeFile(path.join(tempDir, 'debug.log'), 'function logged() {}');

            // Rebuild index
            await codebaseIndexer.buildIndex();

            const results = await searchEngine.search('function');

            // Should not include ignored files
            expect(results.every(r => !r.filePath.includes('node_modules'))).toBe(true);
            expect(results.every(r => !r.filePath.endsWith('.log'))).toBe(true);
        });
    });

    describe('Performance and Scalability', () => {
        it('should handle large search results efficiently', async () => {
            // Create many files with the search term
            for (let i = 0; i < 50; i++) {
                await fs.writeFile(
                    path.join(tempDir, `file${i}.js`),
                    `// File ${i}\\nfunction test${i}() {\\n  return ${i};\\n}`
                );
            }

            // Rebuild index
            await codebaseIndexer.buildIndex();

            const startTime = Date.now();
            const results = await searchEngine.search('function');
            const endTime = Date.now();

            expect(results.length).toBeGreaterThan(50);
            expect(endTime - startTime).toBeLessThan(1000); // Should complete within 1 second
        });

        it('should handle concurrent searches', async () => {
            const searches = [
                searchEngine.search('function'),
                searchEngine.search('TODO'),
                searchEngine.search('const'),
                searchEngine.search('module.exports')
            ];

            const results = await Promise.all(searches);

            expect(results).toHaveLength(4);
            expect(results.every(r => Array.isArray(r))).toBe(true);
        });

        it('should cache search results for repeated queries', async () => {
            const query = 'function';

            // First search
            const startTime1 = Date.now();
            const results1 = await searchEngine.search(query);
            const endTime1 = Date.now();

            // Second search (should be faster if cached)
            const startTime2 = Date.now();
            const results2 = await searchEngine.search(query);
            const endTime2 = Date.now();

            expect(results1).toEqual(results2);
            // Second search should be faster (allowing for some variance)
            expect(endTime2 - startTime2).toBeLessThanOrEqual((endTime1 - startTime1) + 50);
        });
    });

    describe('Integration with Codebase Indexer', () => {
        it('should update search results when files change', async () => {
            const initialResults = await searchEngine.search('newFunction');
            expect(initialResults.length).toBe(0);

            // Add new file with the search term
            await fs.writeFile(
                path.join(tempDir, 'new.js'),
                'function newFunction() { return "new"; }'
            );

            // Update index
            await codebaseIndexer.updateIndex(['new.js']);

            const updatedResults = await searchEngine.search('newFunction');
            expect(updatedResults.length).toBeGreaterThan(0);
            expect(updatedResults.some(r => r.filePath.includes('new.js'))).toBe(true);
        });

        it('should handle file deletions', async () => {
            const initialResults = await searchEngine.search('startServer');
            expect(initialResults.length).toBeGreaterThan(0);

            // Delete file containing the search term
            await fs.unlink(path.join(tempDir, 'src', 'index.js'));

            // Update index
            await codebaseIndexer.buildIndex();

            const updatedResults = await searchEngine.search('startServer');
            expect(updatedResults.every(r => !r.filePath.includes('index.js'))).toBe(true);
        });

        it('should detect language and framework information', async () => {
            const stats = codebaseIndexer.getStats();

            expect(stats.languages).toContain('JavaScript');
            expect(stats.languages).toContain('Python');
            expect(stats.frameworks).toContain('Express');
            expect(stats.totalFiles).toBeGreaterThan(5);
        });
    });

    describe('Error Handling', () => {
        it('should handle search in empty codebase', async () => {
            // Create empty search engine
            const emptyDir = await fs.mkdtemp(path.join(os.tmpdir(), 'empty-'));
            const emptyFsManager = new FileSystemManager(emptyDir);
            const emptyIndexer = new CodebaseIndexer(emptyDir, emptyFsManager);
            const emptySearchEngine = new SearchEngine(emptyIndexer);

            await emptyIndexer.buildIndex();

            const results = await emptySearchEngine.search('anything');
            expect(results).toEqual([]);

            // Cleanup
            await fs.rm(emptyDir, { recursive: true, force: true });
        });

        it('should handle invalid search patterns gracefully', async () => {
            const results = await searchEngine.search('');
            expect(results).toEqual([]);
        });

        it('should handle file system errors during search', async () => {
            // Mock file system error
            const originalReadFile = fileSystemManager.readFile;
            fileSystemManager.readFile = jest.fn().mockRejectedValue(new Error('File system error'));

            // Search should still work with available files
            const results = await searchEngine.search('function');

            // Restore original method
            fileSystemManager.readFile = originalReadFile;

            expect(results).toBeDefined();
        });
    });
});
