// codestral/src/search/SearchEngine.js
import path from 'path';

/**
 * SearchEngine provides text search capabilities across the codebase
 */
class SearchEngine {
  /**
   * @param {CodebaseIndexer} indexer - Codebase indexer instance
   */
  constructor(indexer) {
    this.indexer = indexer;
    this.fileSystemManager = indexer.fileSystemManager;
  }

  /**
   * Search for text across all indexed files
   * @param {string} query - Search query
   * @param {Object} options - Search options
   * @param {boolean} options.caseSensitive - Case sensitive search (default: false)
   * @param {boolean} options.wholeWord - Match whole words only (default: false)
   * @param {number} options.contextLines - Number of context lines to include (default: 2)
   * @param {number} options.limit - Maximum number of results to return
   * @returns {Promise<Array<SearchResult>>} Array of search results
   */
  async search(query, options = {}) {
    const {
      caseSensitive = false,
      wholeWord = false,
      contextLines = 2,
      limit,
      fileTypes,
      directories,
      excludeDirectories
    } = options;

    if (!query || query.trim().length === 0) {
      return [];
    }

    const results = [];
    const index = this.indexer.getIndex();

    // Prepare search pattern
    let searchPattern;
    if (wholeWord) {
      const escapedQuery = this._escapeRegex(query);
      searchPattern = new RegExp(`\\b${escapedQuery}\\b`, caseSensitive ? 'g' : 'gi');
    } else {
      searchPattern = caseSensitive ? query : query.toLowerCase();
    }

    // Search through all indexed files
    for (const [filePath, metadata] of index.entries()) {
      try {
        // Apply file type filter
        if (fileTypes && fileTypes.length > 0) {
          const ext = path.extname(filePath).toLowerCase();
          const normalizedTypes = fileTypes.map(t => t.startsWith('.') ? t.toLowerCase() : `.${t.toLowerCase()}`);
          if (!normalizedTypes.includes(ext)) {
            continue;
          }
        }

        // Apply directory filter
        if (directories && directories.length > 0) {
          const normalizedPath = filePath.replace(/\\/g, '/');
          const matchesDir = directories.some(dir => {
            const normalizedDir = dir.replace(/\\/g, '/');
            const dirPrefix = normalizedDir.endsWith('/') ? normalizedDir : `${normalizedDir}/`;
            return normalizedPath.startsWith(dirPrefix) || normalizedPath === normalizedDir;
          });
          if (!matchesDir) {
            continue;
          }
        }

        // Apply exclude directory filter
        if (excludeDirectories && excludeDirectories.length > 0) {
          const normalizedPath = filePath.replace(/\\/g, '/');
          const matchesExcludeDir = excludeDirectories.some(dir => {
            const normalizedDir = dir.replace(/\\/g, '/');
            const dirPrefix = normalizedDir.endsWith('/') ? normalizedDir : `${normalizedDir}/`;
            return normalizedPath.startsWith(dirPrefix) || normalizedPath === normalizedDir;
          });
          if (matchesExcludeDir) {
            continue;
          }
        }

        // Skip binary files
        if (await this.fileSystemManager.isBinary(filePath)) {
          continue;
        }

        // Read file content
        const content = await this.fileSystemManager.readFile(filePath);
        const lines = content.split('\n');

        // Search each line
        for (let i = 0; i < lines.length; i++) {
          const line = lines[i];
          const lineToSearch = caseSensitive ? line : line.toLowerCase();
          
          let hasMatch = false;
          if (wholeWord) {
            hasMatch = searchPattern.test(line);
          } else {
            hasMatch = lineToSearch.includes(searchPattern);
          }

          if (hasMatch) {
            // Get context lines
            const contextBefore = [];
            const contextAfter = [];

            for (let j = Math.max(0, i - contextLines); j < i; j++) {
              contextBefore.push(lines[j]);
            }

            for (let j = i + 1; j <= Math.min(lines.length - 1, i + contextLines); j++) {
              contextAfter.push(lines[j]);
            }

            results.push({
              filePath,
              lineNumber: i + 1, // 1-indexed
              line: line,
              context: {
                before: contextBefore,
                after: contextAfter
              },
              language: metadata.language,
              score: 1.0 // Base score, can be adjusted by ranking
            });

            // Continue collecting results
          }
        }
      } catch (error) {
        // Skip files that can't be read
        console.warn(`Warning: Could not search file ${filePath}: ${error.message}`);
      }
    }

    // Apply limit if specified
    if (limit && results.length > limit) {
      return results.slice(0, limit);
    }
    
    return results;
  }

  /**
   * Search using regex pattern
   * @param {string} pattern - Regex pattern
   * @param {Object} options - Search options
   * @param {string} options.flags - Regex flags (default: 'g')
   * @param {number} options.contextLines - Number of context lines to include (default: 2)
   * @returns {Promise<Array<SearchResult>>} Array of search results
   */
  async searchRegex(pattern, options = {}) {
    const {
      flags = 'g',
      contextLines = 2
    } = options;

    if (!pattern || pattern.trim().length === 0) {
      throw new Error('Search pattern cannot be empty');
    }

    let regex;
    try {
      regex = new RegExp(pattern, flags);
    } catch (error) {
      throw new Error(`Invalid regex pattern: ${error.message}`);
    }

    const results = [];
    const index = this.indexer.getIndex();

    // Search through all indexed files
    for (const [filePath, metadata] of index.entries()) {
      try {
        // Skip binary files
        if (await this.fileSystemManager.isBinary(filePath)) {
          continue;
        }

        // Read file content
        const content = await this.fileSystemManager.readFile(filePath);
        const lines = content.split('\n');

        // Search each line
        for (let i = 0; i < lines.length; i++) {
          const line = lines[i];
          
          // Reset regex lastIndex for global flag
          regex.lastIndex = 0;
          
          if (regex.test(line)) {
            // Get context lines
            const contextBefore = [];
            const contextAfter = [];

            for (let j = Math.max(0, i - contextLines); j < i; j++) {
              contextBefore.push(lines[j]);
            }

            for (let j = i + 1; j <= Math.min(lines.length - 1, i + contextLines); j++) {
              contextAfter.push(lines[j]);
            }

            results.push({
              filePath,
              lineNumber: i + 1, // 1-indexed
              line: line,
              context: {
                before: contextBefore,
                after: contextAfter
              },
              language: metadata.language,
              score: 1.0 // Base score, can be adjusted by ranking
            });
          }
        }
      } catch (error) {
        // Skip files that can't be read
        console.warn(`Warning: Could not search file ${filePath}: ${error.message}`);
      }
    }

    return results;
  }

  /**
   * Filter search results by file type (extension)
   * @param {Array<SearchResult>} results - Search results to filter
   * @param {Array<string>} extensions - File extensions to include (e.g., ['.js', '.ts'])
   * @returns {Array<SearchResult>} Filtered results
   */
  filterByFileType(results, extensions) {
    if (!extensions || extensions.length === 0) {
      return results;
    }

    // Normalize extensions to include leading dot
    const normalizedExtensions = extensions.map(ext => 
      ext.startsWith('.') ? ext.toLowerCase() : `.${ext.toLowerCase()}`
    );

    return results.filter(result => {
      const ext = path.extname(result.filePath).toLowerCase();
      return normalizedExtensions.includes(ext);
    });
  }

  /**
   * Filter search results by directory path
   * @param {Array<SearchResult>} results - Search results to filter
   * @param {string} dirPath - Directory path to filter by
   * @returns {Array<SearchResult>} Filtered results
   */
  filterByDirectory(results, dirPath) {
    if (!dirPath) {
      return results;
    }

    // Normalize directory path
    const normalizedDir = dirPath.replace(/\\/g, '/');
    const dirPrefix = normalizedDir.endsWith('/') ? normalizedDir : `${normalizedDir}/`;

    return results.filter(result => {
      const normalizedPath = result.filePath.replace(/\\/g, '/');
      return normalizedPath.startsWith(dirPrefix) || normalizedPath === normalizedDir;
    });
  }

  /**
   * Rank search results by relevance
   * @param {Array<SearchResult>} results - Search results to rank
   * @param {string} query - Original search query
   * @returns {Array<SearchResult>} Ranked results (sorted by score descending)
   */
  rankResults(results, query) {
    if (!results || results.length === 0) {
      return results;
    }

    const queryLower = query.toLowerCase();
    const queryWords = queryLower.split(/\s+/).filter(w => w.length > 0);

    // Calculate relevance score for each result
    const scoredResults = results.map(result => {
      let score = result.score || 1.0;
      const lineLower = result.line.toLowerCase();
      const filePathLower = result.filePath.toLowerCase();
      const fileName = path.basename(result.filePath).toLowerCase();

      // Boost score for exact matches
      if (lineLower.includes(queryLower)) {
        score += 2.0;
      }

      // Boost score for matches at start of line
      if (lineLower.trimStart().startsWith(queryLower)) {
        score += 1.5;
      }

      // Boost score for multiple query word matches
      const wordMatches = queryWords.filter(word => lineLower.includes(word)).length;
      score += wordMatches * 0.5;

      // Boost score for matches in file name
      if (fileName.includes(queryLower)) {
        score += 1.0;
      }

      // Boost score for matches in file path
      if (filePathLower.includes(queryLower)) {
        score += 0.5;
      }

      // Penalize very long lines (likely minified or generated code)
      if (result.line.length > 200) {
        score *= 0.5;
      }

      // Boost score for common important file patterns
      if (fileName.includes('readme') || fileName.includes('index')) {
        score += 0.3;
      }

      // Penalize test files slightly (unless searching for test-related terms)
      if ((fileName.includes('test') || fileName.includes('spec')) && 
          !queryLower.includes('test') && !queryLower.includes('spec')) {
        score *= 0.8;
      }

      return {
        ...result,
        score
      };
    });

    // Sort by score descending
    return scoredResults.sort((a, b) => b.score - a.score);
  }

  /**
   * Escape special regex characters in a string
   * @param {string} str - String to escape
   * @returns {string} Escaped string
   * @private
   */
  _escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
}

export default SearchEngine;
