// codestral/src/filesystem/FileSystemManager.js
import fs from 'fs/promises';
import path from 'path';
import fsSync from 'fs';
import ignore from 'ignore';
import * as diff from 'diff';

/**
 * FileSystemManager handles all file system operations with safety checks
 */
class FileSystemManager {
  /**
   * @param {string} rootPath - Root directory path for operations
   * @param {string[]} ignorePatterns - Patterns to ignore
   */
  constructor(rootPath = process.cwd(), ignorePatterns = []) {
    this.rootPath = path.resolve(rootPath);
    this.ignorePatterns = ignorePatterns;
    this.ig = ignore().add(ignorePatterns);
    this.gitignoreLoaded = false;
    this.backupDir = path.join(this.rootPath, '.backups');
    this.backups = new Map(); // Track backups: filePath -> backupPath
  }

  /**
   * Load .gitignore file if it exists
   * @returns {Promise<void>}
   */
  async loadGitignore() {
    if (this.gitignoreLoaded) {
      return;
    }

    try {
      const gitignorePath = path.join(this.rootPath, '.gitignore');
      const content = await fs.readFile(gitignorePath, 'utf8');
      this.ig.add(content);
      this.gitignoreLoaded = true;
    } catch (error) {
      // .gitignore doesn't exist or can't be read, that's okay
      this.gitignoreLoaded = true;
    }
  }

  /**
   * Validate and resolve a file path to prevent directory traversal attacks
   * @param {string} filePath - Path to validate
   * @returns {string} Resolved absolute path
   * @throws {Error} If path is invalid or outside root
   */
  validatePath(filePath) {
    const resolvedPath = path.resolve(this.rootPath, filePath);
    
    // Ensure the resolved path is within the root directory
    if (!resolvedPath.startsWith(this.rootPath)) {
      throw new Error(`Invalid path: ${filePath} is outside the root directory`);
    }
    
    return resolvedPath;
  }

  /**
   * Read file contents
   * @param {string} filePath - Path to file relative to root
   * @param {string} encoding - File encoding (default: 'utf8')
   * @returns {Promise<string>} File contents
   */
  async readFile(filePath, encoding = 'utf8') {
    const validatedPath = this.validatePath(filePath);
    
    try {
      const content = await fs.readFile(validatedPath, encoding);
      return content;
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`File not found: ${filePath}`);
      } else if (error.code === 'EACCES') {
        throw new Error(`Permission denied: ${filePath}`);
      } else if (error.code === 'EISDIR') {
        throw new Error(`Path is a directory, not a file: ${filePath}`);
      }
      throw error;
    }
  }

  /**
   * Write content to a file with safety checks
   * @param {string} filePath - Path to file relative to root
   * @param {string} content - Content to write
   * @param {Object} options - Write options
   * @returns {Promise<void>}
   */
  async writeFile(filePath, content, options = {}) {
    const validatedPath = this.validatePath(filePath);
    
    // Ensure parent directory exists
    const dirPath = path.dirname(validatedPath);
    await fs.mkdir(dirPath, { recursive: true });
    
    try {
      await fs.writeFile(validatedPath, content, { encoding: 'utf8', ...options });
    } catch (error) {
      if (error.code === 'EACCES') {
        throw new Error(`Permission denied: ${filePath}`);
      } else if (error.code === 'ENOSPC') {
        throw new Error(`No space left on device: ${filePath}`);
      }
      throw error;
    }
  }

  /**
   * List directory contents
   * @param {string} dirPath - Directory path relative to root
   * @param {boolean} recursive - Whether to list recursively
   * @param {boolean} respectIgnore - Whether to filter by .gitignore patterns
   * @returns {Promise<Array<{name: string, path: string, isDirectory: boolean}>>}
   */
  async listDirectory(dirPath = '.', recursive = false, respectIgnore = true) {
    const validatedPath = this.validatePath(dirPath);
    
    try {
      const entries = await fs.readdir(validatedPath, { withFileTypes: true });
      const results = [];
      
      for (const entry of entries) {
        const entryPath = path.join(dirPath, entry.name);
        const normalizedPath = entryPath.replace(/\\/g, '/');
        const isDir = entry.isDirectory();
        
        // Skip if should be ignored
        if (respectIgnore && this.shouldIgnore(normalizedPath, isDir)) {
          continue;
        }
        
        const fullPath = path.join(validatedPath, entry.name);
        
        results.push({
          name: entry.name,
          path: entryPath,
          isDirectory: isDir,
          isFile: entry.isFile()
        });
        
        // Recursively list subdirectories if requested
        if (recursive && isDir) {
          const subEntries = await this.listDirectory(entryPath, true, respectIgnore);
          results.push(...subEntries);
        }
      }
      
      return results;
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`Directory not found: ${dirPath}`);
      } else if (error.code === 'EACCES') {
        throw new Error(`Permission denied: ${dirPath}`);
      } else if (error.code === 'ENOTDIR') {
        throw new Error(`Path is not a directory: ${dirPath}`);
      }
      throw error;
    }
  }

  /**
   * Check if a file exists
   * @param {string} filePath - Path to check
   * @returns {Promise<boolean>}
   */
  async exists(filePath) {
    try {
      const validatedPath = this.validatePath(filePath);
      await fs.access(validatedPath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Get file stats
   * @param {string} filePath - Path to file
   * @returns {Promise<Object>} File stats
   */
  async getStats(filePath) {
    const validatedPath = this.validatePath(filePath);
    
    try {
      const stats = await fs.stat(validatedPath);
      return {
        size: stats.size,
        isDirectory: stats.isDirectory(),
        isFile: stats.isFile(),
        modified: stats.mtime,
        created: stats.birthtime
      };
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`File not found: ${filePath}`);
      }
      throw error;
    }
  }

  /**
   * Check if a file is binary
   * @param {string} filePath - Path to file
   * @returns {Promise<boolean>} True if file is binary
   */
  async isBinary(filePath) {
    const validatedPath = this.validatePath(filePath);
    
    try {
      // Read first 8000 bytes to check for binary content
      const buffer = Buffer.alloc(8000);
      const fd = await fs.open(validatedPath, 'r');
      
      try {
        const { bytesRead } = await fd.read(buffer, 0, 8000, 0);
        
        // Check for null bytes which indicate binary content
        for (let i = 0; i < bytesRead; i++) {
          if (buffer[i] === 0) {
            return true;
          }
        }
        
        // Check for high ratio of non-text characters
        let nonTextChars = 0;
        for (let i = 0; i < bytesRead; i++) {
          const byte = buffer[i];
          // Count characters outside printable ASCII range (except common whitespace)
          if (byte < 32 && byte !== 9 && byte !== 10 && byte !== 13) {
            nonTextChars++;
          } else if (byte > 126 && byte < 128) {
            nonTextChars++;
          }
        }
        
        // If more than 30% non-text characters, consider it binary
        return (nonTextChars / bytesRead) > 0.3;
      } finally {
        await fd.close();
      }
    } catch (error) {
      if (error.code === 'ENOENT') {
        throw new Error(`File not found: ${filePath}`);
      } else if (error.code === 'EISDIR') {
        return false; // Directories are not binary files
      }
      throw error;
    }
  }

  /**
   * Check if a path should be ignored based on .gitignore patterns
   * @param {string} filePath - Path to check (relative to root)
   * @param {boolean} isDirectory - Whether the path is a directory
   * @returns {boolean} True if path should be ignored
   */
  shouldIgnore(filePath, isDirectory = false) {
    // Normalize path to use forward slashes and remove leading ./
    let normalizedPath = filePath.replace(/\\/g, '/');
    if (normalizedPath.startsWith('./')) {
      normalizedPath = normalizedPath.substring(2);
    }
    
    // Check both with and without trailing slash for directories
    if (isDirectory) {
      return this.ig.ignores(normalizedPath) || this.ig.ignores(normalizedPath + '/');
    }
    
    return this.ig.ignores(normalizedPath);
  }

  /**
   * Filter a list of files based on .gitignore patterns
   * @param {Array<string>} files - Array of file paths
   * @returns {Array<string>} Filtered array of file paths
   */
  filterIgnored(files) {
    return files.filter(file => !this.shouldIgnore(file));
  }
}

export default FileSystemManager;
