// codestral/src/utils/ErrorHandler.js
/**
 * ErrorHandler - Enhanced centralized error handling with categorization, recovery strategies,
 * and improved user experience with helpful suggestions and clear messaging
 * 
 * Handles different types of errors and provides user-friendly messages
 * Requirements: 9.1, 9.4, 9.5, 9.6, 11.7
 */

export default class ErrorHandler {
  constructor(logger = null, options = {}) {
    this.logger = logger;
    this.options = {
      showSuggestions: options.showSuggestions !== false,
      showTechnicalDetails: options.showTechnicalDetails || false,
      colorOutput: options.colorOutput !== false,
      includeContext: options.includeContext !== false,
      ...options
    };
    
    // Error categories
    this.ERROR_TYPES = {
      API: 'API',
      FILE_SYSTEM: 'FILE_SYSTEM',
      USER_INPUT: 'USER_INPUT',
      SYSTEM: 'SYSTEM',
      NETWORK: 'NETWORK',
      TOOL_EXECUTION: 'TOOL_EXECUTION',
      CONFIGURATION: 'CONFIGURATION',
      UNKNOWN: 'UNKNOWN'
    };

    // Common error patterns and their solutions
    this.errorSolutions = new Map([
      ['ENOENT', {
        title: 'File or Directory Not Found',
        suggestions: [
          'Check if the file path is correct',
          'Verify the file exists in the specified location',
          'Use absolute path if relative path is not working',
          'Check for typos in the filename'
        ]
      }],
      ['EACCES', {
        title: 'Permission Denied',
        suggestions: [
          'Check file permissions with `ls -la`',
          'Run with appropriate permissions',
          'Ensure you have read/write access to the directory',
          'Contact system administrator if needed'
        ]
      }],
      ['ENOSPC', {
        title: 'No Space Left on Device',
        suggestions: [
          'Free up disk space by deleting unnecessary files',
          'Check disk usage with `df -h`',
          'Move files to external storage',
          'Clear temporary files and caches'
        ]
      }],
      ['ETIMEDOUT', {
        title: 'Request Timed Out',
        suggestions: [
          'Check your internet connection',
          'Try again in a few moments',
          'Verify the API endpoint is accessible',
          'Consider increasing timeout settings'
        ]
      }],
      ['ENOTFOUND', {
        title: 'Host Not Found',
        suggestions: [
          'Check your internet connection',
          'Verify the API URL is correct',
          'Check DNS settings',
          'Try using a different network'
        ]
      }],
      ['401', {
        title: 'Authentication Failed',
        suggestions: [
          'Check your API key is correct',
          'Verify the API key has not expired',
          'Run setup wizard: /setup',
          'Update API key: /set apiKey YOUR_KEY'
        ]
      }],
      ['429', {
        title: 'Rate Limit Exceeded',
        suggestions: [
          'Wait a few minutes before trying again',
          'Consider upgrading your API plan',
          'Reduce the frequency of requests',
          'Check your usage limits'
        ]
      }],
      ['500', {
        title: 'Server Error',
        suggestions: [
          'The API service is experiencing issues',
          'Try again in a few minutes',
          'Check the service status page',
          'Contact support if the issue persists'
        ]
      }]
    ]);

    // Context-specific help
    this.contextualHelp = new Map([
      ['file_mention', [
        'Use @filename to mention files',
        'Try autocomplete with Tab key',
        'Use relative paths from project root',
        'Check if file is in .gitignore'
      ]],
      ['tool_execution', [
        'Check tool parameters are valid',
        'Verify you have necessary permissions',
        'Try running in different autonomy mode',
        'Check if file paths exist'
      ]],
      ['configuration', [
        'Run setup wizard: /setup',
        'Check configuration file location',
        'Verify all required fields are set',
        'Use /config to view current settings'
      ]],
      ['search', [
        'Use /search command for codebase search',
        'Try different search terms',
        'Check if files are indexed',
        'Use regex patterns for advanced search'
      ]]
    ]);
  }

  /**
   * Main error handling method with enhanced user experience
   * @param {Error} error - The error object
   * @param {Object} context - Additional context about where the error occurred
   * @returns {Object} - Enhanced error information with suggestions and recovery options
   */
  handle(error, context = {}) {
    // Log error details
    if (this.logger) {
      this.logger.error('Error occurred', {
        error: error.message,
        stack: error.stack,
        context,
        timestamp: new Date().toISOString()
      });
    }

    // Determine error type
    const errorType = this.categorize(error);

    // Get enhanced error information
    const errorInfo = this.getEnhancedErrorInfo(errorType, error, context);

    // Check if error is recoverable
    const recoverable = this.isRecoverable(errorType, error);

    // Check if error is critical
    const critical = this.isCritical(errorType, error);

    // Get suggestions based on error and context
    const suggestions = this.getSuggestions(error, context, errorType);

    // Get contextual help
    const contextualHelp = this.getContextualHelp(context);

    return {
      type: errorType,
      title: errorInfo.title,
      message: errorInfo.message,
      suggestions: this.options.showSuggestions ? suggestions : [],
      contextualHelp: this.options.includeContext ? contextualHelp : [],
      technicalDetails: this.options.showTechnicalDetails ? {
        code: error.code,
        stack: error.stack,
        originalMessage: error.message
      } : null,
      recoverable,
      critical,
      originalError: error,
      recovery: recoverable ? this.getRecoveryStrategy(errorType, error, context) : null,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Get enhanced error information with better titles and messages
   * @param {string} errorType - The error type
   * @param {Error} error - The error object
   * @param {Object} context - Error context
   * @returns {Object} Enhanced error info
   */
  getEnhancedErrorInfo(errorType, error, context) {
    const code = error.code;
    const message = error.message;
    const httpStatus = this.extractHttpStatus(error);

    // Check for known error patterns
    const knownError = this.errorSolutions.get(code) || this.errorSolutions.get(httpStatus);
    if (knownError) {
      return {
        title: knownError.title,
        message: this.getUserMessage(errorType, error)
      };
    }

    // Generate contextual title
    let title = 'Error';
    switch (errorType) {
      case this.ERROR_TYPES.API:
        title = 'AI Service Error';
        break;
      case this.ERROR_TYPES.FILE_SYSTEM:
        title = 'File System Error';
        break;
      case this.ERROR_TYPES.USER_INPUT:
        title = 'Input Validation Error';
        break;
      case this.ERROR_TYPES.NETWORK:
        title = 'Network Connection Error';
        break;
      case this.ERROR_TYPES.TOOL_EXECUTION:
        title = 'Tool Execution Error';
        break;
      case this.ERROR_TYPES.CONFIGURATION:
        title = 'Configuration Error';
        break;
      case this.ERROR_TYPES.SYSTEM:
        title = 'System Error';
        break;
      default:
        title = 'Unexpected Error';
    }

    return {
      title,
      message: this.getUserMessage(errorType, error)
    };
  }

  /**
   * Extract HTTP status code from error
   * @param {Error} error - The error object
   * @returns {string|null} HTTP status code
   */
  extractHttpStatus(error) {
    if (error.status) return error.status.toString();
    if (error.statusCode) return error.statusCode.toString();
    
    const message = error.message || '';
    const statusMatch = message.match(/\b(4\d{2}|5\d{2})\b/);
    return statusMatch ? statusMatch[1] : null;
  }

  /**
   * Categorize error by type with enhanced detection
   * @param {Error} error - The error object
   * @returns {string} - Error type
   */
  categorize(error) {
    const message = error.message?.toLowerCase() || '';
    const code = error.code;
    const name = error.name;

    // Configuration errors
    if (name === 'ConfigurationError' ||
        message.includes('config') ||
        message.includes('api key') ||
        message.includes('setup') ||
        message.includes('missing required')) {
      return this.ERROR_TYPES.CONFIGURATION;
    }

    // Tool execution errors
    if (name === 'ToolExecutionError' ||
        message.includes('tool') ||
        message.includes('command failed') ||
        message.includes('execution') ||
        message.includes('parameter')) {
      return this.ERROR_TYPES.TOOL_EXECUTION;
    }

    // API errors (enhanced detection)
    if (name === 'APIError' || 
        message.includes('api') || 
        message.includes('authentication') ||
        message.includes('rate limit') ||
        message.includes('mistral') ||
        message.includes('unauthorized') ||
        message.includes('forbidden') ||
        this.extractHttpStatus(error) ||
        (code === 'ENOTFOUND' && message.includes('api')) ||
        (code === 'ETIMEDOUT' && message.includes('api'))) {
      return this.ERROR_TYPES.API;
    }

    // File system errors (enhanced detection)
    if (name === 'FileSystemError' ||
        code === 'ENOENT' ||
        code === 'EACCES' ||
        code === 'ENOSPC' ||
        code === 'EISDIR' ||
        code === 'ENOTDIR' ||
        code === 'EMFILE' ||
        code === 'ENFILE' ||
        message.includes('file') ||
        message.includes('directory') ||
        message.includes('permission') ||
        message.includes('path') ||
        message.includes('folder')) {
      return this.ERROR_TYPES.FILE_SYSTEM;
    }

    // User input errors (enhanced detection)
    if (name === 'ValidationError' ||
        name === 'UserInputError' ||
        name === 'SyntaxError' ||
        message.includes('invalid') ||
        message.includes('malformed') ||
        message.includes('unexpected') ||
        message.includes('parse') ||
        message.includes('syntax')) {
      return this.ERROR_TYPES.USER_INPUT;
    }

    // Network errors (enhanced detection)
    if (code === 'ECONNREFUSED' ||
        code === 'ECONNRESET' ||
        code === 'ENETUNREACH' ||
        code === 'EHOSTUNREACH' ||
        code === 'ENOTFOUND' ||
        code === 'ETIMEDOUT' ||
        message.includes('network') ||
        message.includes('connection') ||
        message.includes('dns') ||
        message.includes('host')) {
      return this.ERROR_TYPES.NETWORK;
    }

    // System errors (enhanced detection)
    if (name === 'SystemError' ||
        code === 'ENOMEM' ||
        code === 'EMFILE' ||
        code === 'ENFILE' ||
        message.includes('memory') ||
        message.includes('system') ||
        message.includes('resource') ||
        message.includes('limit')) {
      return this.ERROR_TYPES.SYSTEM;
    }

    return this.ERROR_TYPES.UNKNOWN;
  }

  /**
   * Get enhanced user-friendly error message with more context
   * @param {string} errorType - The error type
   * @param {Error} error - The error object
   * @returns {string} - User-friendly message
   */
  getUserMessage(errorType, error) {
    const code = error.code;
    const message = error.message;
    const httpStatus = this.extractHttpStatus(error);

    switch (errorType) {
      case this.ERROR_TYPES.API:
        if (message.includes('authentication') || httpStatus === '401') {
          return 'Your API key is invalid or has expired. Please update your configuration with a valid API key.';
        }
        if (message.includes('forbidden') || httpStatus === '403') {
          return 'Access forbidden. Your API key may not have permission for this operation.';
        }
        if (message.includes('rate limit') || httpStatus === '429') {
          return 'You have exceeded the API rate limit. Please wait before making more requests.';
        }
        if (httpStatus === '500' || httpStatus === '502' || httpStatus === '503') {
          return 'The AI service is currently experiencing issues. Please try again in a few minutes.';
        }
        if (code === 'ETIMEDOUT' || message.includes('timeout')) {
          return 'The request to the AI service timed out. This may be due to a slow connection or high server load.';
        }
        if (code === 'ENOTFOUND') {
          return 'Cannot connect to the AI service. Please check your internet connection and try again.';
        }
        if (message.includes('quota') || message.includes('limit')) {
          return 'You have reached your API usage quota. Please check your account limits or upgrade your plan.';
        }
        return 'Unable to communicate with the AI service. Please verify your connection and API configuration.';

      case this.ERROR_TYPES.FILE_SYSTEM:
        if (code === 'ENOENT') {
          const path = this.extractPath(error);
          return `The file or directory "${path}" does not exist. Please check the path and try again.`;
        }
        if (code === 'EACCES') {
          const path = this.extractPath(error);
          return `Permission denied when accessing "${path}". You may need to change file permissions or run with elevated privileges.`;
        }
        if (code === 'ENOSPC') {
          return 'Your disk is full. Please free up some space by deleting unnecessary files and try again.';
        }
        if (code === 'EISDIR') {
          return 'Cannot perform this operation on a directory. Please specify a file instead.';
        }
        if (code === 'ENOTDIR') {
          return 'Cannot perform this operation on a file. Please specify a directory instead.';
        }
        if (code === 'EMFILE' || code === 'ENFILE') {
          return 'Too many files are open. Please close some applications and try again.';
        }
        return `File system operation failed: ${message}. Please check file paths and permissions.`;

      case this.ERROR_TYPES.USER_INPUT:
        if (message.includes('mention') || message.includes('@')) {
          return 'Invalid file mention format. Use @filename to reference files in your project.';
        }
        if (message.includes('command')) {
          return 'Invalid command format. Use /help to see available commands.';
        }
        if (message.includes('json') || message.includes('parse')) {
          return 'Invalid JSON format in configuration or input. Please check the syntax.';
        }
        return `Invalid input provided: ${message}. Please check your input format and try again.`;

      case this.ERROR_TYPES.NETWORK:
        if (code === 'ECONNREFUSED') {
          return 'Connection refused by the server. The service may be down or your firewall may be blocking the connection.';
        }
        if (code === 'ECONNRESET') {
          return 'Connection was unexpectedly closed by the server. Please try again.';
        }
        if (code === 'ENETUNREACH' || code === 'EHOSTUNREACH') {
          return 'Network is unreachable. Please check your internet connection and network settings.';
        }
        if (code === 'ENOTFOUND') {
          return 'DNS lookup failed. Please check your internet connection and DNS settings.';
        }
        return 'Network connection failed. Please check your internet connection and try again.';

      case this.ERROR_TYPES.TOOL_EXECUTION:
        if (message.includes('permission')) {
          return 'Tool execution failed due to insufficient permissions. You may need to approve the operation or run with elevated privileges.';
        }
        if (message.includes('parameter')) {
          return 'Tool execution failed due to invalid parameters. Please check the input values.';
        }
        if (message.includes('timeout')) {
          return 'Tool execution timed out. The operation may be taking longer than expected.';
        }
        return `Tool execution failed: ${message}. Please check the operation parameters and try again.`;

      case this.ERROR_TYPES.CONFIGURATION:
        if (message.includes('api key')) {
          return 'API key is missing or invalid. Please run the setup wizard (/setup) to configure your API key.';
        }
        if (message.includes('config')) {
          return 'Configuration file is missing or invalid. Please run /setup to create a new configuration.';
        }
        return `Configuration error: ${message}. Please check your settings and run /setup if needed.`;

      case this.ERROR_TYPES.SYSTEM:
        if (code === 'ENOMEM') {
          return 'System is out of memory. Please close some applications and try again, or restart the application.';
        }
        if (message.includes('resource')) {
          return 'System resource limit reached. Please try again later or restart the application.';
        }
        return `System error occurred: ${message}. You may need to restart the application.`;

      default:
        return `An unexpected error occurred: ${message}. Please try again or contact support if the issue persists.`;
    }
  }

  /**
   * Check if error is recoverable
   * @param {string} errorType - The error type
   * @param {Error} error - The error object
   * @returns {boolean} - True if recoverable
   */
  isRecoverable(errorType, error) {
    const code = error.code;
    const message = error.message?.toLowerCase() || '';

    // Non-recoverable errors
    if (code === 'ENOMEM' || code === 'ENOSPC') {
      return false;
    }

    // Recoverable errors
    switch (errorType) {
      case this.ERROR_TYPES.API:
        // Rate limits and timeouts are recoverable
        return message.includes('rate limit') || 
               message.includes('timeout') ||
               code === 'ETIMEDOUT' ||
               code === 'ENOTFOUND';

      case this.ERROR_TYPES.FILE_SYSTEM:
        // File not found is recoverable (user can provide correct path)
        // Permission errors might be recoverable
        return code === 'ENOENT';

      case this.ERROR_TYPES.USER_INPUT:
        // User input errors are always recoverable
        return true;

      case this.ERROR_TYPES.NETWORK:
        // Most network errors are recoverable
        return true;

      case this.ERROR_TYPES.SYSTEM:
        // Most system errors are not recoverable
        return false;

      default:
        return false;
    }
  }

  /**
   * Check if error is critical (requires shutdown)
   * @param {string} errorType - The error type
   * @param {Error} error - The error object
   * @returns {boolean} - True if critical
   */
  isCritical(errorType, error) {
    const code = error.code;

    // Critical errors that require shutdown
    return code === 'ENOMEM' || 
           code === 'ENOSPC' ||
           errorType === this.ERROR_TYPES.SYSTEM;
  }

  /**
   * Get recovery strategy for recoverable errors
   * @param {string} errorType - The error type
   * @param {Error} error - The error object
   * @param {Object} context - Error context
   * @returns {Object} - Recovery strategy
   */
  getRecoveryStrategy(errorType, error, context) {
    const code = error.code;
    const message = error.message?.toLowerCase() || '';

    switch (errorType) {
      case this.ERROR_TYPES.API:
        if (message.includes('rate limit') || message.includes('429')) {
          return {
            action: 'retry',
            delay: 5000,
            maxRetries: 3,
            backoff: 'exponential'
          };
        }
        if (code === 'ETIMEDOUT' || message.includes('timeout')) {
          return {
            action: 'retry',
            delay: 2000,
            maxRetries: 3,
            backoff: 'exponential'
          };
        }
        if (code === 'ENOTFOUND') {
          return {
            action: 'retry',
            delay: 3000,
            maxRetries: 2,
            backoff: 'linear'
          };
        }
        return null;

      case this.ERROR_TYPES.FILE_SYSTEM:
        if (code === 'ENOENT') {
          return {
            action: 'prompt',
            message: 'Would you like to provide a different file path?'
          };
        }
        return null;

      case this.ERROR_TYPES.USER_INPUT:
        return {
          action: 'prompt',
          message: 'Please provide valid input.'
        };

      case this.ERROR_TYPES.NETWORK:
        return {
          action: 'retry',
          delay: 2000,
          maxRetries: 3,
          backoff: 'exponential'
        };

      default:
        return null;
    }
  }

  /**
   * Extract file path from error message
   * @param {Error} error - The error object
   * @returns {string} - Extracted path or empty string
   */
  extractPath(error) {
    const message = error.message || '';
    const pathMatch = message.match(/['"]([^'"]+)['"]/);
    return pathMatch ? pathMatch[1] : '';
  }

  /**
   * Execute recovery strategy
   * @param {Object} recovery - Recovery strategy
   * @param {Function} retryFn - Function to retry
   * @returns {Promise} - Recovery result
   */
  async recover(recovery, retryFn) {
    if (!recovery) {
      return null;
    }

    if (recovery.action === 'retry' && retryFn) {
      return this.retryWithBackoff(retryFn, recovery);
    }

    return null;
  }

  /**
   * Retry with backoff strategy
   * @param {Function} fn - Function to retry
   * @param {Object} options - Retry options
   * @returns {Promise} - Result of successful retry
   */
  async retryWithBackoff(fn, options = {}) {
    const {
      delay = 1000,
      maxRetries = 3,
      backoff = 'exponential'
    } = options;

    let lastError;
    
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;
        
        if (attempt < maxRetries - 1) {
          const waitTime = backoff === 'exponential' 
            ? delay * Math.pow(2, attempt)
            : delay * (attempt + 1);
          
          if (this.logger) {
            this.logger.warn(`Retry attempt ${attempt + 1}/${maxRetries} after ${waitTime}ms`);
          }
          
          await this.sleep(waitTime);
        }
      }
    }

    throw lastError;
  }

  /**
   * Get helpful suggestions based on error and context
   * @param {Error} error - The error object
   * @param {Object} context - Error context
   * @param {string} errorType - Error type
   * @returns {Array<string>} Array of suggestions
   */
  getSuggestions(error, context, errorType) {
    const suggestions = [];
    const code = error.code;
    const httpStatus = this.extractHttpStatus(error);

    // Get suggestions from known error patterns
    const knownError = this.errorSolutions.get(code) || this.errorSolutions.get(httpStatus);
    if (knownError && knownError.suggestions) {
      suggestions.push(...knownError.suggestions);
    }

    // Add context-specific suggestions
    if (context.operation) {
      switch (context.operation) {
        case 'file_read':
          suggestions.push('Verify the file exists and you have read permissions');
          break;
        case 'file_write':
          suggestions.push('Check if you have write permissions to the directory');
          break;
        case 'api_call':
          suggestions.push('Check your API key and internet connection');
          break;
        case 'tool_execution':
          suggestions.push('Verify tool parameters and required permissions');
          break;
      }
    }

    // Add error type specific suggestions
    switch (errorType) {
      case this.ERROR_TYPES.API:
        if (!suggestions.length) {
          suggestions.push(
            'Check your API key configuration',
            'Verify your internet connection',
            'Try again in a few moments'
          );
        }
        break;

      case this.ERROR_TYPES.FILE_SYSTEM:
        if (!suggestions.length) {
          suggestions.push(
            'Check if the file or directory exists',
            'Verify file permissions',
            'Use absolute paths if relative paths fail'
          );
        }
        break;

      case this.ERROR_TYPES.USER_INPUT:
        suggestions.push(
          'Check the input format',
          'Use /help to see available commands',
          'Try using autocomplete with Tab key'
        );
        break;

      case this.ERROR_TYPES.CONFIGURATION:
        suggestions.push(
          'Run setup wizard: /setup',
          'Check configuration file: /config',
          'Update specific settings: /set key value'
        );
        break;

      case this.ERROR_TYPES.TOOL_EXECUTION:
        suggestions.push(
          'Check tool parameters',
          'Verify required permissions',
          'Try different autonomy mode: /mode'
        );
        break;
    }

    // Remove duplicates and limit to 5 suggestions
    return [...new Set(suggestions)].slice(0, 5);
  }

  /**
   * Get contextual help based on the current context
   * @param {Object} context - Current context
   * @returns {Array<string>} Array of contextual help items
   */
  getContextualHelp(context) {
    const help = [];

    if (context.feature) {
      const contextHelp = this.contextualHelp.get(context.feature);
      if (contextHelp) {
        help.push(...contextHelp);
      }
    }

    // Add general help if no specific context
    if (!help.length) {
      help.push(
        'Use /help for available commands',
        'Use @filename to mention files',
        'Use /search to find code',
        'Use /config to check settings'
      );
    }

    return help.slice(0, 4);
  }

  /**
   * Format error for display with colors and structure
   * @param {Object} errorInfo - Error information from handle()
   * @param {Object} options - Display options
   * @returns {string} Formatted error message
   */
  formatErrorForDisplay(errorInfo, options = {}) {
    const { useColors = this.options.colorOutput, showAll = false } = options;
    let output = '';

    // Error title and message
    if (useColors) {
      output += `\n🚨 ${errorInfo.title}\n`;
      output += `💬 ${errorInfo.message}\n`;
    } else {
      output += `\nError: ${errorInfo.title}\n`;
      output += `${errorInfo.message}\n`;
    }

    // Suggestions
    if (errorInfo.suggestions && errorInfo.suggestions.length > 0) {
      if (useColors) {
        output += `\n💡 Suggestions:\n`;
      } else {
        output += `\nSuggestions:\n`;
      }
      errorInfo.suggestions.forEach((suggestion, index) => {
        output += `   ${index + 1}. ${suggestion}\n`;
      });
    }

    // Contextual help
    if (errorInfo.contextualHelp && errorInfo.contextualHelp.length > 0) {
      if (useColors) {
        output += `\n🔧 Quick Help:\n`;
      } else {
        output += `\nQuick Help:\n`;
      }
      errorInfo.contextualHelp.forEach(help => {
        output += `   • ${help}\n`;
      });
    }

    // Recovery information
    if (errorInfo.recoverable && errorInfo.recovery) {
      if (useColors) {
        output += `\n🔄 Recovery: `;
      } else {
        output += `\nRecovery: `;
      }
      
      if (errorInfo.recovery.action === 'retry') {
        output += `Will retry automatically in ${errorInfo.recovery.delay}ms\n`;
      } else if (errorInfo.recovery.action === 'prompt') {
        output += `${errorInfo.recovery.message}\n`;
      }
    }

    // Technical details (if enabled)
    if (showAll && errorInfo.technicalDetails) {
      output += `\nTechnical Details:\n`;
      if (errorInfo.technicalDetails.code) {
        output += `   Code: ${errorInfo.technicalDetails.code}\n`;
      }
      output += `   Original: ${errorInfo.technicalDetails.originalMessage}\n`;
    }

    return output;
  }

  /**
   * Create a user-friendly error report
   * @param {Object} errorInfo - Error information from handle()
   * @returns {Object} Error report for logging or support
   */
  createErrorReport(errorInfo) {
    return {
      timestamp: errorInfo.timestamp,
      type: errorInfo.type,
      title: errorInfo.title,
      message: errorInfo.message,
      code: errorInfo.technicalDetails?.code,
      recoverable: errorInfo.recoverable,
      critical: errorInfo.critical,
      suggestions: errorInfo.suggestions,
      originalMessage: errorInfo.originalError?.message,
      stack: errorInfo.originalError?.stack
    };
  }

  /**
   * Sleep utility
   * @param {number} ms - Milliseconds to sleep
   * @returns {Promise} - Promise that resolves after delay
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
