// codestral/src/utils/logger.js
/**
 * Logger - Winston-based logging configuration
 * 
 * Provides structured logging with file and console transports
 * Requirements: 9.2, 9.3, 9.4
 */

import winston from 'winston';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Ensure logs directory exists
const logsDir = path.join(process.cwd(), 'logs');
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir, { recursive: true });
}

// Custom format for console output
const consoleFormat = winston.format.combine(
  winston.format.colorize(),
  winston.format.timestamp({ format: 'HH:mm:ss' }),
  winston.format.printf(({ timestamp, level, message, ...meta }) => {
    let msg = `${timestamp} [${level}]: ${message}`;
    
    // Add metadata if present
    if (Object.keys(meta).length > 0) {
      msg += ` ${JSON.stringify(meta)}`;
    }
    
    return msg;
  })
);

// Custom format for file output
const fileFormat = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  winston.format.errors({ stack: true }),
  winston.format.json()
);

/**
 * Create logger instance
 * @param {Object} options - Logger options
 * @returns {winston.Logger} - Configured logger
 */
function createLogger(options = {}) {
  const {
    level = process.env.LOG_LEVEL || 'info',
    debugMode = process.env.DEBUG === 'true',
    silent = false
  } = options;

  // Determine log level
  const logLevel = debugMode ? 'debug' : level;

  // Create transports
  const transports = [];

  // Console transport
  if (!silent) {
    transports.push(
      new winston.transports.Console({
        level: logLevel,
        format: consoleFormat
      })
    );
  }

  // File transport for errors
  transports.push(
    new winston.transports.File({
      filename: path.join(logsDir, 'error.log'),
      level: 'error',
      format: fileFormat,
      maxsize: 5242880, // 5MB
      maxFiles: 5
    })
  );

  // File transport for all logs
  transports.push(
    new winston.transports.File({
      filename: path.join(logsDir, 'combined.log'),
      level: logLevel,
      format: fileFormat,
      maxsize: 5242880, // 5MB
      maxFiles: 5
    })
  );

  // Debug file transport (only in debug mode)
  if (debugMode) {
    transports.push(
      new winston.transports.File({
        filename: path.join(logsDir, 'debug.log'),
        level: 'debug',
        format: fileFormat,
        maxsize: 5242880, // 5MB
        maxFiles: 3
      })
    );
  }

  // Create logger
  const logger = winston.createLogger({
    level: logLevel,
    format: fileFormat,
    transports,
    exitOnError: false
  });

  // Add helper methods
  logger.setDebugMode = (enabled) => {
    const newLevel = enabled ? 'debug' : 'info';
    logger.level = newLevel;
    logger.transports.forEach(transport => {
      transport.level = newLevel;
    });
    
    // Add or remove debug file transport
    if (enabled && !logger.transports.some(t => t.filename?.includes('debug.log'))) {
      logger.add(
        new winston.transports.File({
          filename: path.join(logsDir, 'debug.log'),
          level: 'debug',
          format: fileFormat,
          maxsize: 5242880,
          maxFiles: 3
        })
      );
    }
  };

  logger.getLogLevel = () => {
    return logger.level;
  };

  logger.isDebugMode = () => {
    return logger.level === 'debug';
  };

  return logger;
}

// Create default logger instance
const logger = createLogger();

// Export both the logger and the factory function
export { createLogger };
export default logger;
