// codestral/src/setup/WebSetupServer.js
import http from 'http';
import { URL } from 'url';
import chalk from 'chalk';
import open from 'open';

/**
 * WebSetupServer - Web-based setup interface
 * Opens a browser for easy API key configuration
 */
class WebSetupServer {
  constructor(configManager) {
    this.configManager = configManager;
    this.server = null;
    this.port = 3737;
    this.setupComplete = false;
    this.setupData = null;
  }

  /**
   * Start the web setup server
   */
  async start() {
    return new Promise((resolve, reject) => {
      this.server = http.createServer((req, res) => {
        this.handleRequest(req, res, resolve);
      });

      this.server.listen(this.port, () => {
        const url = `http://localhost:${this.port}`;
        console.log(chalk.cyan('\n🌐 Opening setup page in your browser...'));
        console.log(chalk.gray(`If it doesn't open automatically, visit: ${url}\n`));
        
        // Open browser
        open(url).catch(() => {
          console.log(chalk.yellow('Could not open browser automatically.'));
          console.log(chalk.white(`Please open: ${chalk.cyan(url)}`));
        });
      });

      this.server.on('error', (error) => {
        if (error.code === 'EADDRINUSE') {
          this.port++;
          this.server.listen(this.port);
        } else {
          reject(error);
        }
      });
    });
  }

  /**
   * Handle HTTP requests
   */
  handleRequest(req, res, resolve) {
    const parsedUrl = new URL(req.url, `http://localhost:${this.port}`);
    const pathname = parsedUrl.pathname;

    if (pathname === '/' && req.method === 'GET') {
      this.serveSetupPage(res);
    } else if (pathname === '/api/setup' && req.method === 'POST') {
      this.handleSetupSubmit(req, res, resolve);
    } else if (pathname === '/api/test' && req.method === 'POST') {
      this.handleApiTest(req, res);
    } else {
      res.writeHead(404);
      res.end('Not Found');
    }
  }

  /**
   * Serve the setup HTML page
   */
  serveSetupPage(res) {
    const html = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Codestral Setup</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }

        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            display: flex;
            align-items: center;
            justify-content: center;
            padding: 20px;
        }

        .container {
            background: white;
            border-radius: 20px;
            box-shadow: 0 20px 60px rgba(0, 0, 0, 0.3);
            max-width: 600px;
            width: 100%;
            overflow: hidden;
        }

        .header {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            padding: 40px;
            text-align: center;
        }

        .header h1 {
            font-size: 2.5em;
            margin-bottom: 10px;
            font-weight: 700;
        }

        .header p {
            font-size: 1.1em;
            opacity: 0.9;
        }

        .content {
            padding: 40px;
        }

        .step {
            margin-bottom: 30px;
        }

        .step-number {
            display: inline-block;
            width: 30px;
            height: 30px;
            background: #667eea;
            color: white;
            border-radius: 50%;
            text-align: center;
            line-height: 30px;
            font-weight: bold;
            margin-right: 10px;
        }

        .step-title {
            font-size: 1.2em;
            font-weight: 600;
            color: #333;
            margin-bottom: 10px;
        }

        .step-description {
            color: #666;
            margin-bottom: 15px;
            line-height: 1.6;
        }

        .form-group {
            margin-bottom: 20px;
        }

        label {
            display: block;
            font-weight: 600;
            color: #333;
            margin-bottom: 8px;
        }

        input[type="password"],
        input[type="text"],
        select {
            width: 100%;
            padding: 12px 15px;
            border: 2px solid #e0e0e0;
            border-radius: 8px;
            font-size: 1em;
            transition: border-color 0.3s;
        }

        input[type="password"]:focus,
        input[type="text"]:focus,
        select:focus {
            outline: none;
            border-color: #667eea;
        }

        .info-box {
            background: #f0f4ff;
            border-left: 4px solid #667eea;
            padding: 15px;
            border-radius: 8px;
            margin-bottom: 20px;
        }

        .info-box a {
            color: #667eea;
            text-decoration: none;
            font-weight: 600;
        }

        .info-box a:hover {
            text-decoration: underline;
        }

        .btn {
            width: 100%;
            padding: 15px;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            border: none;
            border-radius: 8px;
            font-size: 1.1em;
            font-weight: 600;
            cursor: pointer;
            transition: transform 0.2s, box-shadow 0.2s;
        }

        .btn:hover {
            transform: translateY(-2px);
            box-shadow: 0 10px 20px rgba(102, 126, 234, 0.4);
        }

        .btn:active {
            transform: translateY(0);
        }

        .btn:disabled {
            opacity: 0.6;
            cursor: not-allowed;
            transform: none;
        }

        .btn-secondary {
            background: #6c757d;
            margin-top: 10px;
        }

        .success-message {
            background: #d4edda;
            border: 2px solid #28a745;
            color: #155724;
            padding: 20px;
            border-radius: 8px;
            text-align: center;
            margin-top: 20px;
        }

        .success-message h3 {
            margin-bottom: 10px;
        }

        .error-message {
            background: #f8d7da;
            border: 2px solid #dc3545;
            color: #721c24;
            padding: 15px;
            border-radius: 8px;
            margin-top: 15px;
            display: none;
        }

        .loading {
            display: none;
            text-align: center;
            padding: 20px;
        }

        .spinner {
            border: 4px solid #f3f3f3;
            border-top: 4px solid #667eea;
            border-radius: 50%;
            width: 40px;
            height: 40px;
            animation: spin 1s linear infinite;
            margin: 0 auto 15px;
        }

        @keyframes spin {
            0% { transform: rotate(0deg); }
            100% { transform: rotate(360deg); }
        }

        .option-description {
            font-size: 0.9em;
            color: #666;
            margin-top: 5px;
        }

        .badge {
            display: inline-block;
            padding: 3px 8px;
            border-radius: 4px;
            font-size: 0.8em;
            font-weight: 600;
            margin-left: 8px;
        }

        .badge-recommended {
            background: #28a745;
            color: white;
        }

        .badge-balanced {
            background: #ffc107;
            color: #333;
        }

        .badge-fast {
            background: #17a2b8;
            color: white;
        }

        .footer {
            text-align: center;
            padding: 20px;
            color: #666;
            font-size: 0.9em;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🚀 Codestral</h1>
            <p>AI Terminal Assistant Setup</p>
        </div>

        <div class="content">
            <form id="setupForm">
                <!-- Step 1: API Key -->
                <div class="step">
                    <div class="step-title">
                        <span class="step-number">1</span>
                        API Key
                    </div>
                    <div class="step-description">
                        Enter your Mistral AI API key to get started.
                    </div>
                    <div class="info-box">
                        📝 Don't have an API key? Get one at 
                        <a href="https://console.mistral.ai/" target="_blank">console.mistral.ai</a>
                    </div>
                    <div class="form-group">
                        <label for="apiKey">Mistral AI API Key *</label>
                        <input 
                            type="password" 
                            id="apiKey" 
                            name="apiKey" 
                            placeholder="Enter your API key"
                            required
                        >
                    </div>
                </div>

                <!-- Step 2: Model Selection -->
                <div class="step">
                    <div class="step-title">
                        <span class="step-number">2</span>
                        AI Model
                    </div>
                    <div class="step-description">
                        Choose the AI model that best fits your needs.
                    </div>
                    <div class="form-group">
                        <label for="model">Model</label>
                        <select id="model" name="model">
                            <option value="mistral-large-latest">
                                mistral-large-latest (Recommended)
                            </option>
                            <option value="mistral-medium-latest">
                                mistral-medium-latest (Balanced)
                            </option>
                            <option value="mistral-small-latest">
                                mistral-small-latest (Fast)
                            </option>
                        </select>
                        <div class="option-description" id="modelDescription">
                            Best performance for complex tasks
                        </div>
                    </div>
                </div>

                <!-- Step 3: Autonomy Mode -->
                <div class="step">
                    <div class="step-title">
                        <span class="step-number">3</span>
                        Autonomy Mode
                    </div>
                    <div class="step-description">
                        Control how much the AI can do automatically.
                    </div>
                    <div class="form-group">
                        <label for="autonomyMode">Autonomy Mode</label>
                        <select id="autonomyMode" name="autonomyMode">
                            <option value="safe">Safe (Asks before changes)</option>
                            <option value="autonomous">Autonomous (Auto-approve files)</option>
                            <option value="full">Full (Maximum automation)</option>
                        </select>
                        <div class="option-description" id="autonomyDescription">
                            AI asks permission before modifying files
                        </div>
                    </div>
                </div>

                <button type="submit" class="btn" id="submitBtn">
                    Complete Setup
                </button>

                <div class="error-message" id="errorMessage"></div>

                <div class="loading" id="loading">
                    <div class="spinner"></div>
                    <p>Saving configuration...</p>
                </div>
            </form>

            <div id="successMessage" style="display: none;">
                <div class="success-message">
                    <h3>✅ Setup Complete!</h3>
                    <p>Your configuration has been saved.</p>
                    <p style="margin-top: 10px;">You can close this window and return to your terminal.</p>
                </div>
            </div>
        </div>

        <div class="footer">
            Codestral AI Terminal Assistant
        </div>
    </div>

    <script>
        // Model descriptions
        const modelDescriptions = {
            'mistral-large-latest': 'Best performance for complex tasks',
            'mistral-medium-latest': 'Good balance of performance and speed',
            'mistral-small-latest': 'Fastest responses for simple tasks'
        };

        // Autonomy descriptions
        const autonomyDescriptions = {
            'safe': 'AI asks permission before modifying files',
            'autonomous': 'AI can read and write files automatically',
            'full': 'AI has full control (use with caution)'
        };

        // Update descriptions on change
        document.getElementById('model').addEventListener('change', (e) => {
            document.getElementById('modelDescription').textContent = modelDescriptions[e.target.value];
        });

        document.getElementById('autonomyMode').addEventListener('change', (e) => {
            document.getElementById('autonomyDescription').textContent = autonomyDescriptions[e.target.value];
        });

        // Handle form submission
        document.getElementById('setupForm').addEventListener('submit', async (e) => {
            e.preventDefault();

            const submitBtn = document.getElementById('submitBtn');
            const loading = document.getElementById('loading');
            const errorMessage = document.getElementById('errorMessage');
            const form = document.getElementById('setupForm');
            const successMessage = document.getElementById('successMessage');

            // Get form data
            const formData = {
                apiKey: document.getElementById('apiKey').value,
                model: document.getElementById('model').value,
                autonomyMode: document.getElementById('autonomyMode').value
            };

            // Validate
            if (!formData.apiKey || formData.apiKey.trim().length === 0) {
                errorMessage.textContent = 'Please enter your API key';
                errorMessage.style.display = 'block';
                return;
            }

            // Show loading
            submitBtn.disabled = true;
            loading.style.display = 'block';
            errorMessage.style.display = 'none';

            try {
                // Submit to server
                const response = await fetch('/api/setup', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(formData)
                });

                const result = await response.json();

                if (result.success) {
                    // Show success message
                    form.style.display = 'none';
                    loading.style.display = 'none';
                    successMessage.style.display = 'block';

                    // Close window after 3 seconds
                    setTimeout(() => {
                        window.close();
                    }, 3000);
                } else {
                    throw new Error(result.error || 'Setup failed');
                }
            } catch (error) {
                loading.style.display = 'none';
                submitBtn.disabled = false;
                errorMessage.textContent = error.message;
                errorMessage.style.display = 'block';
            }
        });
    </script>
</body>
</html>
    `;

    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(html);
  }

  /**
   * Handle setup form submission
   */
  async handleSetupSubmit(req, res, resolve) {
    let body = '';

    req.on('data', chunk => {
      body += chunk.toString();
    });

    req.on('end', async () => {
      try {
        const data = JSON.parse(body);

        // Validate data
        if (!data.apiKey || data.apiKey.trim().length === 0) {
          res.writeHead(400, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ success: false, error: 'API key is required' }));
          return;
        }

        // Save configuration
        this.setupData = {
          apiKey: data.apiKey.trim(),
          model: data.model || 'mistral-large-latest',
          autonomyMode: data.autonomyMode || 'safe'
        };

        this.setupComplete = true;

        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ success: true }));

        // Close server after a short delay
        setTimeout(() => {
          this.stop();
          resolve(this.setupData);
        }, 1000);

      } catch (error) {
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ success: false, error: error.message }));
      }
    });
  }

  /**
   * Handle API key test
   */
  async handleApiTest(req, res) {
    let body = '';

    req.on('data', chunk => {
      body += chunk.toString();
    });

    req.on('end', async () => {
      try {
        const data = JSON.parse(body);
        
        // Here you could add actual API validation
        // For now, just check if it's not empty
        const isValid = data.apiKey && data.apiKey.trim().length > 10;

        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ valid: isValid }));
      } catch (error) {
        res.writeHead(500, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ valid: false, error: error.message }));
      }
    });
  }

  /**
   * Stop the server
   */
  stop() {
    if (this.server) {
      this.server.close();
    }
  }
}

export default WebSetupServer;
